---
title: "Integraton Anchor Heatmaps"
author: "Stefan Boeing stefan.boeing@crick.ac.uk"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
abstract: "In this section, integration anchor heatmaps are created."
output: 
    html_document:
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 5
        toc_float: true
        css: ../Main_Analysis/src/style/style.css

always_allow_html: yes

---

```{css setup_css, echo=FALSE}


.table{
  width:auto;
  font-size: 10px;
}

```

```{r setup, include=FALSE, warning=FALSE, message=F, echo=F, results=F}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a

if (!require("remotes")){
  install.packages("remotes")
}

if (!require("renv")){
  remotes::install_github("rstudio/renv")
}
  
if (!file.exists("renv.lock")){
    renv::init()
} else {
    renv::restore(prompt = FALSE)
}

urlString <- "biologic.crick.ac.uk"
```



```{r set_directories, eval=T, echo=F, results=F}
## Setup plot collection object
library(knitr)
library(ggplot2)
library(ggpubr)
library(DT)
library(biologicSeqTools2)

addCorCatsToLabDb <- FALSE
figureCount <- 1
chnkVec <- as.vector(NULL, mode = "character")

VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}





## Heatmap setup ##
## Select Heatmap samples ##
FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")
dbTable <- read.delim(
  FN,
  header = F,
  sep = "\t",
  stringsAsFactors = F
)

db.pwd <- as.vector(dbTable[1,1])
#setwd(Obio@parameterList$localWorkDir)

# The active biologic data object is expected to be found in ../../../../data/biologic_active_object/

source("load.biologic.robj.R")

# ObioFN <- paste0("../",list.files("..")[grep(".bioLOGIC.Robj", list.files(".."))])
# 
# load(ObioFN)


# Obio <- setMountingPoint(Obio)
# Obio <- setAnalysisPaths(Obio)
# #Obio <- setCrickGenomeAndGeneNameTable(Obio)
# Obio <- createAnalysisFolders(
#     Obio
# )
# Obio <- setDataBaseParameters(Obio)


if (is.null(Obio@parameterList[["reportFigDir"]])){
    Obio@parameterList[["reportFigDir"]] <- paste0(Obio@parameterList$html_local, "report_figures/")
}



## Create outputfolders ##
if (!dir.exists(Obio@parameterList[["reportFigDir"]])){
    dir.create(Obio@parameterList[["reportFigDir"]])
}

figureCount <- 1

## Order design file ##
# dfDesign <- Obio@dfDesign
# orderVec <- unique(dfDesign$dataseries)
# orderVec <- c("WtFemale", "HetFemale", "HomFemale", "WtMale", "HomMale")
# 
# dfDesign$dataseries <- factor(dfDesign$dataseries, levels = orderVec)
# dfDesign <- dfDesign[order(dfDesign$dataseries),]
# Obio@dfDesign <- data.frame(NULL)
# Obio@dfDesign <- dfDesign
# source("save.biologic.robj.R")

```


<!-- Optional set data access parameters for R-users ------------------------->
```{r child = '02-B.1.2.set.data.access.parameters.for.R.data.import.Rmd', eval=TRUE}
```

## Custom Heatmap Creation
You can use the provided code to generate a custom heatmap. 

```{r alignment_dat, echo=TRUE, eval=TRUE, warning=FALSE, result="asis"}

library(Seurat)
FN <- "../../../../workdir/integration01.Seurat.Robj"
base::load(FN)

dfData <- data.frame(OsC[["RNA"]]@data)
###############################################################################
## Prepare annotation                                                        ##
## Annotate highlight genes
#dfAnno <- unique(dfMainData[,c(geneIDcolumn, alignmentGeneID)])

# dfAnnoHL <- dfAnno[dfAnno[,geneIDcolumn] %in% highlightGenes,]


## Done with annotation                                                      ##
###############################################################################

###############################################################################
## select heatmap genes                                                      ##

## Gene selection for the heatmap:

## Select all genes with a Foxl2 peak
# genSel <- as.vector(
#     dfMainData[dfMainData$,geneIDcolumn]
# )
# 
# 
# geneSel <- mostVarGenes

## Making sure that all genes are present in the dataset:

## First select relevant subset of the data

geneSel <- unique(Obio@dataTableList$referenceList[["sampleAnchors"]])

## Check genes used for the heatmap:
# sort(geneSel)

## Select TPM data for the heatmap
#selVec <- c(geneIDcolumn, alignmentGeneID, names(dfMainData)[grep("norm_counts", names(dfMainData))])
dfTPM <- unique(dfData[row.names(dfData) %in%  geneSel, ])
## Adjust column names
#names(dfTPM) <- gsub("norm_counts_", "", names(dfTPM))

## Temporary fix: Remove duplicates:
#dfTPM <- dfTPM[!duplicated(dfTPM$gene_id), ]

## Remove duplicated gene names, if they exists
#dup <- sum(duplicated(dfTPM$gene_id))



# if (dup > 0){
#     #print(paste0(dup, " gene names in the heatmap data are duplicated."))
#     dfTPM[duplicated(dfTPM$gene_id),"gene_id"] <- paste0(dfTPM[duplicated(dfTPM$gene_id),"gene_id"], "_", dfTPM[duplicated(dfTPM$gene_id), alignmentGeneID])
#     
# }


# row.names(dfTPM) <- dfTPM[,alignmentGeneID]
# dfTPM$gene_id <- NULL
# dfTPM$mgi_symbol <- NULL


## Done                                                                      ##
###############################################################################


## Step 2 Calculate correlation on most variable genes

df.data <- data.matrix(dfTPM)


## Do manual correlaton calculation ##
## Get clusters using hclust and dist ##

mHmBase <- df.data 
#heat.dat   <- t(apply(heat.dat,1,function(x){(x-mean(x))/sd(x)}))
mHmBase   <- t(apply(mHmBase,1,function(x){log2((x/mean(x)))}))
mHmBase[mHmBase == -Inf] <- 0
mHmBase <- round(mHmBase, 3)
mHmBase <- na.omit(mHmBase)



dfCol <- unique(OsC@meta.data[,c( "sampleName", "sampleColor")])
colVec <- dfCol$sampleColor
names(colVec) <- dfCol$sampleName

dfCol2 <- unique(OsC@meta.data[,c("clusterName", "clusterColor")])
colVec2 <- dfCol2$clusterColor
names(colVec2) <- dfCol2$clusterName

## Setting Heatmap Annotation and Colors
ht.anno <- ComplexHeatmap::HeatmapAnnotation(
    Samples = OsC@meta.data$sampleName,
    Clusters =  OsC@meta.data$clusterName,
    col = list(
        Clusters = colVec2,
        Samples = colVec
    )
)





## Setup Split By Chromosome
# selectedColumns2 <- c(
#   #geneIDcolumn, 
#   alignmentGeneID,
#   "add_venn_X_Metacore_Transcripton_Factor",
#   "add_venn_X_Foxl2_ChIP_Peak",
#   "add_venn_X_Foxl2_Chip_Peak_within_2kb"
# )
# 
# dfSplit <- unique(dfMainData[,selectedColumns2])

## Select all genes with Foxl2 ChIP peak
# dfSplit <- dfSplit[dfSplit$add_venn_X_Foxl2_ChIP_Peak == 1, ]
# dfSplit[["label"]] <- "Other Genes"
# dfSplit[dfSplit$add_venn_X_Metacore_Transcripton_Factor == 1, "label"] <- "TFs"
# 
# dfSplit <- dfSplit[dfSplit[,alignmentGeneID] %in% row.names(dfTPM), ]
# 
# geneGroupList <- list(
#     "TFs" = as.vector(dfSplit[dfSplit$label == "TFs", alignmentGeneID]),
#     "Other Genes" = as.vector(dfSplit[dfSplit$label == "Other Genes", alignmentGeneID])
# )

# rowSplitVec <- row.names(mHmBase)   
# for (i in 1:length(geneGroupList)){
#   rowSplitVec[rowSplitVec %in% geneGroupList[[i]]] <- names(geneGroupList)[i]
# }


## Set options
figureCount <- 1

ComplexHeatmap::ht_opt(
    legend_border = "black",
    heatmap_border = TRUE,
    annotation_border = TRUE
)

columnSplitVec <- OsC@meta.data$sampleName

tag <- "Custom_Heatmap"
plotListHM <- list()
chnkVec <- as.vector(NULL, mode="character")

## Create correlation heatmap and clustering
plotListHM[[tag]] <- ComplexHeatmap::Heatmap(
    mHmBase,
    #row_split               = rowSplitVec,
    name                    = "log2 row mean",
    
    show_column_names       = FALSE,
    show_row_names          = FALSE,
    
    column_title_gp         = grid::gpar(fontsize = 8),
    row_title_rot           = 0,
    column_split            = factor(columnSplitVec, levels=unique(columnSplitVec)),
    cluster_column_slices   = FALSE,
    cluster_columns         = FALSE,
    cluster_rows            = TRUE,
    column_names_side       = "bottom",
    show_column_dend        = TRUE,
    row_dend_width          = grid::unit(20, "mm"),
    show_heatmap_legend     = FALSE,
    column_names_max_height = grid::unit(8, "cm"),
    row_names_gp            = grid::gpar(fontsize = 6),
    top_annotation          = ht.anno,
    col                     = circlize::colorRamp2(c(-3, 0, 3),c("#3060cf", "#fffbbc","#c4463a")),
    column_names_rot        = 90,
    border                  = TRUE
) 

## To view the heatmap in an interactive sesssion run:

# print(plotListHM[[tag]])

```

```{r, echo=F, eval=TRUE, warning=FALSE, result="asis", include = FALSE}
## Create a file name
## Datestamp
VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")


FNbase <- paste0(tag, VersionPdfExt)

FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
FNrel <- paste0("report_figures/", FNbase)

pdf(FN)
    print(plotListHM[[tag]])
dev.off()


## Create R markdown chunk ##
figLegend <- paste0(
    "**Figure ",
    figureCount,
    ":** Heatmap showing all genes in this RNA-Seq Study that also feature a Foxl2 peak. ",
    "Download a pdf of this figure [here](", FNrel,"). "
)


figureCount <- figureCount + 1

NewChnk <- paste0(
    "### ", tag,
    "\n```{r ",
    ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
    figLegend,"', fig.asp = 1, fig.align = 'center'}\n",
    "\n",
    "\n print(plotListHM[['",tag,"']])",
    "\n cat(  '\n')",
    "\n\n\n```\n"
)


chnkVec <- unique(c(
    chnkVec,
    NewChnk
))

## Step 3 Plot by median timeseries

## Upload correlation class genes as gene cats
```

```{r knit_cor_plot_characterization_plot, echo=F, eval=TRUE, warning=FALSE, results="asis"}
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
```

## Documentation
```{r documentation, eval=TRUE, echo=F, results=T, error=F}
#renv::snapshot(prompt=FALSE)

print(paste0("Projectfolder: ", getwd()))
print(paste0("Project ID: ", Obio@parameterList$project_id))

sessionInfo()
```
