

processing file: sc_PartB_Analysis.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |.                                                                     |   2%
  ordinary text without R code

  |                                                                              |...                                                                   |   4%
label: setup (with options) 
List of 1
 $ include: logi FALSE

  |                                                                              |....                                                                  |   6%
  ordinary text without R code

  |                                                                              |.....                                                                 |   8%
label: unnamed-chunk-1 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B.1.set.parameters.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B.1.set.parameters.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..............                                                        |  20%
  ordinary text without R code

  |                                                                              |............................                                          |  40%
label: set_parameters_init (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................................                            |  60%
  ordinary text without R code

  |                                                                              |........................................................              |  80%
label: B_set_directories (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |.......                                                               |  10%
  ordinary text without R code

  |                                                                              |........                                                              |  12%
label: unnamed-chunk-2 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B.2.retrieve.reference.gene.sets.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B.2.retrieve.reference.gene.sets.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..............                                                        |  20%
  ordinary text without R code

  |                                                                              |............................                                          |  40%
label: partB_retrieve_reference_gene_sets_init (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................................                            |  60%
  ordinary text without R code

  |                                                                              |........................................................              |  80%
label: partB_create_sample_list (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Loading required package: DBI
  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |..........                                                            |  14%
  ordinary text without R code

  |                                                                              |...........                                                           |  16%
label: unnamed-chunk-3 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B.3.create.integrated.seurat.object.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B.3.create.integrated.seurat.object.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..............                                                        |  20%
  ordinary text without R code

  |                                                                              |............................                                          |  40%
label: partB_create_integrated_Seurat_object_init (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................................                            |  60%
  ordinary text without R code

  |                                                                              |........................................................              |  80%
label: B3_partB_create_sample_list (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Computing 2000 integration features
Scaling features for provided objects
Finding all pairwise anchors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 5496 anchors
Filtering anchors
	Retained 3969 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 7174 anchors
Filtering anchors
	Retained 5176 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 5775 anchors
Filtering anchors
	Retained 3311 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 6713 anchors
Filtering anchors
	Retained 4875 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 5201 anchors
Filtering anchors
	Retained 3234 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 7419 anchors
Filtering anchors
	Retained 5838 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 4288 anchors
Filtering anchors
	Retained 2326 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 4325 anchors
Filtering anchors
	Retained 1883 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 4704 anchors
Filtering anchors
	Retained 2076 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 4427 anchors
Filtering anchors
	Retained 2170 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 5993 anchors
Filtering anchors
	Retained 1947 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 5776 anchors
Filtering anchors
	Retained 2073 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 6331 anchors
Filtering anchors
	Retained 2340 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 5547 anchors
Filtering anchors
	Retained 2377 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 4467 anchors
Filtering anchors
	Retained 3437 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 1538 anchors
Filtering anchors
	Retained 1371 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 1630 anchors
Filtering anchors
	Retained 1399 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 1487 anchors
Filtering anchors
	Retained 1382 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 1493 anchors
Filtering anchors
	Retained 1355 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 1443 anchors
Filtering anchors
	Retained 1378 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 1643 anchors
Filtering anchors
	Retained 1609 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 7680 anchors
Filtering anchors
	Retained 1816 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 7356 anchors
Filtering anchors
	Retained 1567 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 8137 anchors
Filtering anchors
	Retained 1937 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 6328 anchors
Filtering anchors
	Retained 2588 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 4536 anchors
Filtering anchors
	Retained 2886 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 7466 anchors
Filtering anchors
	Retained 5258 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 1669 anchors
Filtering anchors
	Retained 1033 anchors
Extracting within-dataset neighbors
Merging dataset 7 into 6
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 5 into 6 7
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 4 into 3
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 8 into 6 7 5
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 2 into 1
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 1 2 into 3 4
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 6 7 5 8 into 3 4 1 2
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Warning: Adding a command log without an assay associated with it
Warning: The following features are not present in the object: Atad2, Blm, Brip1, Casp8ap2, Cdc45, Cdc6, Cdca7, Chaf1b, Clspn, Dscc1, Dtl, E2f8, Exo1, Fen1, Gins2, Gmnn, Hells, Mcm2, Mcm4, Mcm5, Mcm6, Msh2, Nasp, Pcna, Pola1, Pold3, Prim1, Rad51, Rad51ap1, Rfc2, Rpa2, Rrm1, Slbp, Tipin, Tyms, Ubr7, Uhrf1, Ung, Usp1, Wdr76, not searching for symbol synonyms
Warning: The following features are not present in the object: Anln, Aurkb, Bub1, Cbx5, Cdc25c, Cdca2, Ckap2, Ckap2l, Ckap5, Ctcf, Dlgap5, Fam64a, G2e3, Gtse1, Hjurp, Hn1, Kif11, Kif20b, Kif2c, Lbr, Ncapd2, Ndc80, Nek2, Nuf2, Rangap1, Smc4, Tmpo, Ttk, Tubb4b, not searching for symbol synonyms
  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |............                                                          |  18%
  ordinary text without R code

  |                                                                              |..............                                                        |  20%
label: unnamed-chunk-4 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B.4.create.variation.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B.4.create.variation.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..........                                                            |  14%
  ordinary text without R code

  |                                                                              |....................                                                  |  29%
label: partB_create_variation_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..............................                                        |  43%
  ordinary text without R code

  |                                                                              |........................................                              |  57%
label: partB_create-individual-varFeatures (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster all
Finished averaging integrated for cluster all
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   6%  |                                                                              |=========                                                             |  12%  |                                                                              |=============                                                         |  19%  |                                                                              |==================                                                    |  25%  |                                                                              |======================                                                |  31%  |                                                                              |==========================                                            |  38%  |                                                                              |===============================                                       |  44%  |                                                                              |===================================                                   |  50%  |                                                                              |=======================================                               |  56%  |                                                                              |============================================                          |  62%  |                                                                              |================================================                      |  69%  |                                                                              |====================================================                  |  75%  |                                                                              |=========================================================             |  81%  |                                                                              |=============================================================         |  88%  |                                                                              |==================================================================    |  94%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster KPC1viable
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |==========                                                            |  14%  |                                                                              |===============                                                       |  21%  |                                                                              |====================                                                  |  29%  |                                                                              |=========================                                             |  36%  |                                                                              |==============================                                        |  43%  |                                                                              |===================================                                   |  50%  |                                                                              |========================================                              |  57%  |                                                                              |=============================================                         |  64%  |                                                                              |==================================================                    |  71%  |                                                                              |=======================================================               |  79%  |                                                                              |============================================================          |  86%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster KPC2viable
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |==========                                                            |  14%  |                                                                              |===============                                                       |  21%  |                                                                              |====================                                                  |  29%  |                                                                              |=========================                                             |  36%  |                                                                              |==============================                                        |  43%  |                                                                              |===================================                                   |  50%  |                                                                              |========================================                              |  57%  |                                                                              |=============================================                         |  64%  |                                                                              |==================================================                    |  71%  |                                                                              |=======================================================               |  79%  |                                                                              |============================================================          |  86%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster KPC3viable
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |==========                                                            |  14%  |                                                                              |===============                                                       |  21%  |                                                                              |====================                                                  |  29%  |                                                                              |=========================                                             |  36%  |                                                                              |==============================                                        |  43%  |                                                                              |===================================                                   |  50%  |                                                                              |========================================                              |  57%  |                                                                              |=============================================                         |  64%  |                                                                              |==================================================                    |  71%  |                                                                              |=======================================================               |  79%  |                                                                              |============================================================          |  86%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster KPC4viable
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |==========                                                            |  14%  |                                                                              |===============                                                       |  21%  |                                                                              |====================                                                  |  29%  |                                                                              |=========================                                             |  36%  |                                                                              |==============================                                        |  43%  |                                                                              |===================================                                   |  50%  |                                                                              |========================================                              |  57%  |                                                                              |=============================================                         |  64%  |                                                                              |==================================================                    |  71%  |                                                                              |=======================================================               |  79%  |                                                                              |============================================================          |  86%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster KPC1fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |=========                                                             |  13%  |                                                                              |==============                                                        |  20%  |                                                                              |===================                                                   |  27%  |                                                                              |=======================                                               |  33%  |                                                                              |============================                                          |  40%  |                                                                              |=================================                                     |  47%  |                                                                              |=====================================                                 |  53%  |                                                                              |==========================================                            |  60%  |                                                                              |===============================================                       |  67%  |                                                                              |===================================================                   |  73%  |                                                                              |========================================================              |  80%  |                                                                              |=============================================================         |  87%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster KPC2fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |=========                                                             |  13%  |                                                                              |==============                                                        |  20%  |                                                                              |===================                                                   |  27%  |                                                                              |=======================                                               |  33%  |                                                                              |============================                                          |  40%  |                                                                              |=================================                                     |  47%  |                                                                              |=====================================                                 |  53%  |                                                                              |==========================================                            |  60%  |                                                                              |===============================================                       |  67%  |                                                                              |===================================================                   |  73%  |                                                                              |========================================================              |  80%  |                                                                              |=============================================================         |  87%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster KPC3fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |==========                                                            |  14%  |                                                                              |===============                                                       |  21%  |                                                                              |====================                                                  |  29%  |                                                                              |=========================                                             |  36%  |                                                                              |==============================                                        |  43%  |                                                                              |===================================                                   |  50%  |                                                                              |========================================                              |  57%  |                                                                              |=============================================                         |  64%  |                                                                              |==================================================                    |  71%  |                                                                              |=======================================================               |  79%  |                                                                              |============================================================          |  86%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster KPC4fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |=========                                                             |  13%  |                                                                              |==============                                                        |  20%  |                                                                              |===================                                                   |  27%  |                                                                              |=======================                                               |  33%  |                                                                              |============================                                          |  40%  |                                                                              |=================================                                     |  47%  |                                                                              |=====================================                                 |  53%  |                                                                              |==========================================                            |  60%  |                                                                              |===============================================                       |  67%  |                                                                              |===================================================                   |  73%  |                                                                              |========================================================              |  80%  |                                                                              |=============================================================         |  87%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%
  |                                                                              |..................................................                    |  71%
   inline R code fragments

  |                                                                              |............................................................          |  86%
label: create-var-feature-plot-II (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |...............                                                       |  22%
  ordinary text without R code

  |                                                                              |................                                                      |  24%
label: unnamed-chunk-5 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B.5.create.pca.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B.5.create.pca.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |.........                                                             |  12%
  ordinary text without R code

  |                                                                              |..................                                                    |  25%
label: partB_pca_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................                                            |  38%
  ordinary text without R code

  |                                                                              |...................................                                   |  50%
label: partB_create_PCA_plots (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session
23:35:37 UMAP embedding parameters a = 0.9922 b = 1.112
23:35:37 Read 19648 rows and found 30 numeric columns
23:35:37 Using Annoy for neighbor search, n_neighbors = 30
23:35:37 Building Annoy index with metric = cosine, n_trees = 50
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
23:35:41 Writing NN index file to temp file /tmp/slurm_25480163/RtmpxQTLxA/file45211b5e08e0
23:35:41 Searching Annoy index using 1 thread, search_k = 3000
23:35:47 Annoy recall = 100%
23:35:48 Commencing smooth kNN distance calibration using 1 thread
23:35:51 Initializing from normalized Laplacian + noise
23:35:52 Commencing optimization for 200 epochs, with 887862 positive edges
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
23:36:18 Optimization finished
Computing nearest neighbor graph
Computing SNN
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|

Attaching package: 'scales'

The following object is masked from 'package:purrr':

    discard

The following object is masked from 'package:readr':

    col_factor

Calculating cluster C15
Calculating cluster C3
Calculating cluster C1
Calculating cluster C0
Calculating cluster C10
Calculating cluster C7
Calculating cluster C5
Calculating cluster C13
Calculating cluster C6
Calculating cluster C9
Calculating cluster C2
Calculating cluster C12
Calculating cluster C8
Calculating cluster C14
Calculating cluster C4
Calculating cluster C16
Calculating cluster C17
Calculating cluster C11
  |                                                                              |............................................                          |  62%
  ordinary text without R code

  |                                                                              |....................................................                  |  75%
label: data_prep_sample_UMAP_rendering (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE

Registered S3 method overwritten by 'treeio':
  method     from
  root.phylo ape 
ggtree v1.16.6  For help: https://yulab-smu.github.io/treedata-book/

If you use ggtree in published research, please cite the most appropriate paper(s):

[36m-[39m Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods for mapping and visualizing associated data on phylogeny using ggtree. Molecular Biology and Evolution 2018, 35(12):3041-3043. doi: 10.1093/molbev/msy194
[36m-[39m Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution 2017, 8(1):28-36, doi:10.1111/2041-210X.12628



Attaching package: 'ggtree'

The following object is masked from 'package:IRanges':

    collapse

The following object is masked from 'package:S4Vectors':

    expand

The following object is masked from 'package:tidyr':

    expand

Finished averaging RNA for cluster KPC1viable
Finished averaging RNA for cluster KPC2viable
Finished averaging RNA for cluster KPC3viable
Finished averaging RNA for cluster KPC4viable
Finished averaging RNA for cluster KPC1fib
Finished averaging RNA for cluster KPC2fib
Finished averaging RNA for cluster KPC3fib
Finished averaging RNA for cluster KPC4fib
  |                                                                              |.............................................................         |  88%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Plot_tsne_data_plotting (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |..................                                                    |  25%
  ordinary text without R code

  |                                                                              |...................                                                   |  27%
label: unnamed-chunk-6 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.calculate.average.expression.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.calculate.average.expression.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..................                                                    |  25%
  ordinary text without R code

  |                                                                              |...................................                                   |  50%
label: partB_calculate_avg_expr (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |....................................................                  |  75%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: Identify differential expressed genes across conditions (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

Finished averaging RNA for cluster all
Finished averaging integrated for cluster all
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   6%  |                                                                              |=========                                                             |  12%  |                                                                              |=============                                                         |  19%  |                                                                              |==================                                                    |  25%  |                                                                              |======================                                                |  31%  |                                                                              |==========================                                            |  38%  |                                                                              |===============================                                       |  44%  |                                                                              |===================================                                   |  50%  |                                                                              |=======================================                               |  56%  |                                                                              |============================================                          |  62%  |                                                                              |================================================                      |  69%  |                                                                              |====================================================                  |  75%  |                                                                              |=========================================================             |  81%  |                                                                              |=============================================================         |  88%  |                                                                              |==================================================================    |  94%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster C15_KPC1viable
Finished averaging RNA for cluster C3_KPC1viable
Finished averaging RNA for cluster C1_KPC1viable
Finished averaging RNA for cluster C0_KPC1viable
Finished averaging RNA for cluster C10_KPC1viable
Finished averaging RNA for cluster C7_KPC1viable
Finished averaging RNA for cluster C5_KPC1viable
Finished averaging RNA for cluster C13_KPC1viable
Finished averaging RNA for cluster C6_KPC1viable
Finished averaging RNA for cluster C9_KPC1viable
Finished averaging RNA for cluster C2_KPC1viable
Finished averaging RNA for cluster C12_KPC1viable
Finished averaging RNA for cluster C8_KPC1viable
Finished averaging RNA for cluster C14_KPC1viable
Finished averaging RNA for cluster C4_KPC1viable
Finished averaging RNA for cluster C16_KPC1viable
Finished averaging RNA for cluster C17_KPC1viable
Finished averaging RNA for cluster C7_KPC2viable
Finished averaging RNA for cluster C14_KPC2viable
Finished averaging RNA for cluster C10_KPC2viable
Finished averaging RNA for cluster C9_KPC2viable
Finished averaging RNA for cluster C0_KPC2viable
Finished averaging RNA for cluster C1_KPC2viable
Finished averaging RNA for cluster C15_KPC2viable
Finished averaging RNA for cluster C3_KPC2viable
Finished averaging RNA for cluster C2_KPC2viable
Finished averaging RNA for cluster C13_KPC2viable
Finished averaging RNA for cluster C16_KPC2viable
Finished averaging RNA for cluster C4_KPC2viable
Finished averaging RNA for cluster C6_KPC2viable
Finished averaging RNA for cluster C8_KPC2viable
Finished averaging RNA for cluster C17_KPC2viable
Finished averaging RNA for cluster C12_KPC2viable
Finished averaging RNA for cluster C5_KPC2viable
Finished averaging RNA for cluster C3_KPC3viable
Finished averaging RNA for cluster C0_KPC3viable
Finished averaging RNA for cluster C6_KPC3viable
Finished averaging RNA for cluster C12_KPC3viable
Finished averaging RNA for cluster C13_KPC3viable
Finished averaging RNA for cluster C1_KPC3viable
Finished averaging RNA for cluster C14_KPC3viable
Finished averaging RNA for cluster C2_KPC3viable
Finished averaging RNA for cluster C9_KPC3viable
Finished averaging RNA for cluster C10_KPC3viable
Finished averaging RNA for cluster C5_KPC3viable
Finished averaging RNA for cluster C17_KPC3viable
Finished averaging RNA for cluster C8_KPC3viable
Finished averaging RNA for cluster C15_KPC3viable
Finished averaging RNA for cluster C7_KPC3viable
Finished averaging RNA for cluster C16_KPC3viable
Finished averaging RNA for cluster C11_KPC3viable
Finished averaging RNA for cluster C4_KPC3viable
Finished averaging RNA for cluster C9_KPC4viable
Finished averaging RNA for cluster C0_KPC4viable
Finished averaging RNA for cluster C1_KPC4viable
Finished averaging RNA for cluster C12_KPC4viable
Finished averaging RNA for cluster C3_KPC4viable
Finished averaging RNA for cluster C2_KPC4viable
Finished averaging RNA for cluster C10_KPC4viable
Finished averaging RNA for cluster C4_KPC4viable
Finished averaging RNA for cluster C6_KPC4viable
Finished averaging RNA for cluster C5_KPC4viable
Finished averaging RNA for cluster C13_KPC4viable
Finished averaging RNA for cluster C8_KPC4viable
Finished averaging RNA for cluster C7_KPC4viable
Finished averaging RNA for cluster C14_KPC4viable
Finished averaging RNA for cluster C16_KPC4viable
Finished averaging RNA for cluster C11_KPC4viable
Finished averaging RNA for cluster C15_KPC4viable
Finished averaging RNA for cluster C17_KPC4viable
Finished averaging RNA for cluster C2_KPC1fib
Finished averaging RNA for cluster C4_KPC1fib
Finished averaging RNA for cluster C0_KPC1fib
Finished averaging RNA for cluster C12_KPC1fib
Finished averaging RNA for cluster C5_KPC1fib
Finished averaging RNA for cluster C9_KPC1fib
Finished averaging RNA for cluster C7_KPC1fib
Finished averaging RNA for cluster C8_KPC1fib
Finished averaging RNA for cluster C1_KPC1fib
Finished averaging RNA for cluster C6_KPC1fib
Finished averaging RNA for cluster C11_KPC1fib
Finished averaging RNA for cluster C17_KPC1fib
Finished averaging RNA for cluster C3_KPC1fib
Finished averaging RNA for cluster C13_KPC1fib
Finished averaging RNA for cluster C14_KPC1fib
Finished averaging RNA for cluster C15_KPC1fib
Finished averaging RNA for cluster C10_KPC1fib
Finished averaging RNA for cluster C16_KPC1fib
Finished averaging RNA for cluster C8_KPC2fib
Finished averaging RNA for cluster C7_KPC2fib
Finished averaging RNA for cluster C6_KPC2fib
Finished averaging RNA for cluster C2_KPC2fib
Finished averaging RNA for cluster C1_KPC2fib
Finished averaging RNA for cluster C12_KPC2fib
Finished averaging RNA for cluster C5_KPC2fib
Finished averaging RNA for cluster C0_KPC2fib
Finished averaging RNA for cluster C4_KPC2fib
Finished averaging RNA for cluster C3_KPC2fib
Finished averaging RNA for cluster C10_KPC2fib
Finished averaging RNA for cluster C11_KPC2fib
Finished averaging RNA for cluster C14_KPC2fib
Finished averaging RNA for cluster C9_KPC2fib
Finished averaging RNA for cluster C17_KPC2fib
Finished averaging RNA for cluster C16_KPC2fib
Finished averaging RNA for cluster C13_KPC2fib
Finished averaging RNA for cluster C2_KPC3fib
Finished averaging RNA for cluster C5_KPC3fib
Finished averaging RNA for cluster C9_KPC3fib
Finished averaging RNA for cluster C6_KPC3fib
Finished averaging RNA for cluster C11_KPC3fib
Finished averaging RNA for cluster C4_KPC3fib
Finished averaging RNA for cluster C0_KPC3fib
Finished averaging RNA for cluster C3_KPC3fib
Finished averaging RNA for cluster C1_KPC3fib
Finished averaging RNA for cluster C14_KPC3fib
Finished averaging RNA for cluster C8_KPC3fib
Finished averaging RNA for cluster C10_KPC3fib
Finished averaging RNA for cluster C12_KPC3fib
Finished averaging RNA for cluster C7_KPC3fib
Finished averaging RNA for cluster C15_KPC3fib
Finished averaging RNA for cluster C16_KPC3fib
Finished averaging RNA for cluster C2_KPC4fib
Finished averaging RNA for cluster C4_KPC4fib
Finished averaging RNA for cluster C11_KPC4fib
Finished averaging RNA for cluster C8_KPC4fib
Finished averaging RNA for cluster C3_KPC4fib
Finished averaging RNA for cluster C9_KPC4fib
Finished averaging RNA for cluster C6_KPC4fib
Finished averaging RNA for cluster C5_KPC4fib
Finished averaging RNA for cluster C7_KPC4fib
Finished averaging RNA for cluster C14_KPC4fib
Finished averaging RNA for cluster C12_KPC4fib
Finished averaging RNA for cluster C1_KPC4fib
Finished averaging RNA for cluster C0_KPC4fib
Finished averaging RNA for cluster C15_KPC4fib
Finished averaging RNA for cluster C17_KPC4fib
Finished averaging RNA for cluster C13_KPC4fib
Finished averaging RNA for cluster C10_KPC4fib
Finished averaging integrated for cluster C15_KPC1viable
Finished averaging integrated for cluster C3_KPC1viable
Finished averaging integrated for cluster C1_KPC1viable
Finished averaging integrated for cluster C0_KPC1viable
Finished averaging integrated for cluster C10_KPC1viable
Finished averaging integrated for cluster C7_KPC1viable
Finished averaging integrated for cluster C5_KPC1viable
Finished averaging integrated for cluster C13_KPC1viable
Finished averaging integrated for cluster C6_KPC1viable
Finished averaging integrated for cluster C9_KPC1viable
Finished averaging integrated for cluster C2_KPC1viable
Finished averaging integrated for cluster C12_KPC1viable
Finished averaging integrated for cluster C8_KPC1viable
Finished averaging integrated for cluster C14_KPC1viable
Finished averaging integrated for cluster C4_KPC1viable
Finished averaging integrated for cluster C16_KPC1viable
Finished averaging integrated for cluster C17_KPC1viable
Finished averaging integrated for cluster C7_KPC2viable
Finished averaging integrated for cluster C14_KPC2viable
Finished averaging integrated for cluster C10_KPC2viable
Finished averaging integrated for cluster C9_KPC2viable
Finished averaging integrated for cluster C0_KPC2viable
Finished averaging integrated for cluster C1_KPC2viable
Finished averaging integrated for cluster C15_KPC2viable
Finished averaging integrated for cluster C3_KPC2viable
Finished averaging integrated for cluster C2_KPC2viable
Finished averaging integrated for cluster C13_KPC2viable
Finished averaging integrated for cluster C16_KPC2viable
Finished averaging integrated for cluster C4_KPC2viable
Finished averaging integrated for cluster C6_KPC2viable
Finished averaging integrated for cluster C8_KPC2viable
Finished averaging integrated for cluster C17_KPC2viable
Finished averaging integrated for cluster C12_KPC2viable
Finished averaging integrated for cluster C5_KPC2viable
Finished averaging integrated for cluster C3_KPC3viable
Finished averaging integrated for cluster C0_KPC3viable
Finished averaging integrated for cluster C6_KPC3viable
Finished averaging integrated for cluster C12_KPC3viable
Finished averaging integrated for cluster C13_KPC3viable
Finished averaging integrated for cluster C1_KPC3viable
Finished averaging integrated for cluster C14_KPC3viable
Finished averaging integrated for cluster C2_KPC3viable
Finished averaging integrated for cluster C9_KPC3viable
Finished averaging integrated for cluster C10_KPC3viable
Finished averaging integrated for cluster C5_KPC3viable
Finished averaging integrated for cluster C17_KPC3viable
Finished averaging integrated for cluster C8_KPC3viable
Finished averaging integrated for cluster C15_KPC3viable
Finished averaging integrated for cluster C7_KPC3viable
Finished averaging integrated for cluster C16_KPC3viable
Finished averaging integrated for cluster C11_KPC3viable
Finished averaging integrated for cluster C4_KPC3viable
Finished averaging integrated for cluster C9_KPC4viable
Finished averaging integrated for cluster C0_KPC4viable
Finished averaging integrated for cluster C1_KPC4viable
Finished averaging integrated for cluster C12_KPC4viable
Finished averaging integrated for cluster C3_KPC4viable
Finished averaging integrated for cluster C2_KPC4viable
Finished averaging integrated for cluster C10_KPC4viable
Finished averaging integrated for cluster C4_KPC4viable
Finished averaging integrated for cluster C6_KPC4viable
Finished averaging integrated for cluster C5_KPC4viable
Finished averaging integrated for cluster C13_KPC4viable
Finished averaging integrated for cluster C8_KPC4viable
Finished averaging integrated for cluster C7_KPC4viable
Finished averaging integrated for cluster C14_KPC4viable
Finished averaging integrated for cluster C16_KPC4viable
Finished averaging integrated for cluster C11_KPC4viable
Finished averaging integrated for cluster C15_KPC4viable
Finished averaging integrated for cluster C17_KPC4viable
Finished averaging integrated for cluster C2_KPC1fib
Finished averaging integrated for cluster C4_KPC1fib
Finished averaging integrated for cluster C0_KPC1fib
Finished averaging integrated for cluster C12_KPC1fib
Finished averaging integrated for cluster C5_KPC1fib
Finished averaging integrated for cluster C9_KPC1fib
Finished averaging integrated for cluster C7_KPC1fib
Finished averaging integrated for cluster C8_KPC1fib
Finished averaging integrated for cluster C1_KPC1fib
Finished averaging integrated for cluster C6_KPC1fib
Finished averaging integrated for cluster C11_KPC1fib
Finished averaging integrated for cluster C17_KPC1fib
Finished averaging integrated for cluster C3_KPC1fib
Finished averaging integrated for cluster C13_KPC1fib
Finished averaging integrated for cluster C14_KPC1fib
Finished averaging integrated for cluster C15_KPC1fib
Finished averaging integrated for cluster C10_KPC1fib
Finished averaging integrated for cluster C16_KPC1fib
Finished averaging integrated for cluster C8_KPC2fib
Finished averaging integrated for cluster C7_KPC2fib
Finished averaging integrated for cluster C6_KPC2fib
Finished averaging integrated for cluster C2_KPC2fib
Finished averaging integrated for cluster C1_KPC2fib
Finished averaging integrated for cluster C12_KPC2fib
Finished averaging integrated for cluster C5_KPC2fib
Finished averaging integrated for cluster C0_KPC2fib
Finished averaging integrated for cluster C4_KPC2fib
Finished averaging integrated for cluster C3_KPC2fib
Finished averaging integrated for cluster C10_KPC2fib
Finished averaging integrated for cluster C11_KPC2fib
Finished averaging integrated for cluster C14_KPC2fib
Finished averaging integrated for cluster C9_KPC2fib
Finished averaging integrated for cluster C17_KPC2fib
Finished averaging integrated for cluster C16_KPC2fib
Finished averaging integrated for cluster C13_KPC2fib
Finished averaging integrated for cluster C2_KPC3fib
Finished averaging integrated for cluster C5_KPC3fib
Finished averaging integrated for cluster C9_KPC3fib
Finished averaging integrated for cluster C6_KPC3fib
Finished averaging integrated for cluster C11_KPC3fib
Finished averaging integrated for cluster C4_KPC3fib
Finished averaging integrated for cluster C0_KPC3fib
Finished averaging integrated for cluster C3_KPC3fib
Finished averaging integrated for cluster C1_KPC3fib
Finished averaging integrated for cluster C14_KPC3fib
Finished averaging integrated for cluster C8_KPC3fib
Finished averaging integrated for cluster C10_KPC3fib
Finished averaging integrated for cluster C12_KPC3fib
Finished averaging integrated for cluster C7_KPC3fib
Finished averaging integrated for cluster C15_KPC3fib
Finished averaging integrated for cluster C16_KPC3fib
Finished averaging integrated for cluster C2_KPC4fib
Finished averaging integrated for cluster C4_KPC4fib
Finished averaging integrated for cluster C11_KPC4fib
Finished averaging integrated for cluster C8_KPC4fib
Finished averaging integrated for cluster C3_KPC4fib
Finished averaging integrated for cluster C9_KPC4fib
Finished averaging integrated for cluster C6_KPC4fib
Finished averaging integrated for cluster C5_KPC4fib
Finished averaging integrated for cluster C7_KPC4fib
Finished averaging integrated for cluster C14_KPC4fib
Finished averaging integrated for cluster C12_KPC4fib
Finished averaging integrated for cluster C1_KPC4fib
Finished averaging integrated for cluster C0_KPC4fib
Finished averaging integrated for cluster C15_KPC4fib
Finished averaging integrated for cluster C17_KPC4fib
Finished averaging integrated for cluster C13_KPC4fib
Finished averaging integrated for cluster C10_KPC4fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   6%  |                                                                              |=========                                                             |  12%  |                                                                              |=============                                                         |  19%  |                                                                              |==================                                                    |  25%  |                                                                              |======================                                                |  31%  |                                                                              |==========================                                            |  38%  |                                                                              |===============================                                       |  44%  |                                                                              |===================================                                   |  50%  |                                                                              |=======================================                               |  56%  |                                                                              |============================================                          |  62%  |                                                                              |================================================                      |  69%  |                                                                              |====================================================                  |  75%  |                                                                              |=========================================================             |  81%  |                                                                              |=============================================================         |  88%  |                                                                              |==================================================================    |  94%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster C15
Finished averaging RNA for cluster C3
Finished averaging RNA for cluster C1
Finished averaging RNA for cluster C0
Finished averaging RNA for cluster C10
Finished averaging RNA for cluster C7
Finished averaging RNA for cluster C5
Finished averaging RNA for cluster C13
Finished averaging RNA for cluster C6
Finished averaging RNA for cluster C9
Finished averaging RNA for cluster C2
Finished averaging RNA for cluster C12
Finished averaging RNA for cluster C8
Finished averaging RNA for cluster C14
Finished averaging RNA for cluster C4
Finished averaging RNA for cluster C16
Finished averaging RNA for cluster C17
Finished averaging RNA for cluster C11
Finished averaging integrated for cluster C15
Finished averaging integrated for cluster C3
Finished averaging integrated for cluster C1
Finished averaging integrated for cluster C0
Finished averaging integrated for cluster C10
Finished averaging integrated for cluster C7
Finished averaging integrated for cluster C5
Finished averaging integrated for cluster C13
Finished averaging integrated for cluster C6
Finished averaging integrated for cluster C9
Finished averaging integrated for cluster C2
Finished averaging integrated for cluster C12
Finished averaging integrated for cluster C8
Finished averaging integrated for cluster C14
Finished averaging integrated for cluster C4
Finished averaging integrated for cluster C16
Finished averaging integrated for cluster C17
Finished averaging integrated for cluster C11
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   6%  |                                                                              |=========                                                             |  12%  |                                                                              |=============                                                         |  19%  |                                                                              |==================                                                    |  25%  |                                                                              |======================                                                |  31%  |                                                                              |==========================                                            |  38%  |                                                                              |===============================                                       |  44%  |                                                                              |===================================                                   |  50%  |                                                                              |=======================================                               |  56%  |                                                                              |============================================                          |  62%  |                                                                              |================================================                      |  69%  |                                                                              |====================================================                  |  75%  |                                                                              |=========================================================             |  81%  |                                                                              |=============================================================         |  88%  |                                                                              |==================================================================    |  94%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster KPC1viable
Finished averaging RNA for cluster KPC2viable
Finished averaging RNA for cluster KPC3viable
Finished averaging RNA for cluster KPC4viable
Finished averaging RNA for cluster KPC1fib
Finished averaging RNA for cluster KPC2fib
Finished averaging RNA for cluster KPC3fib
Finished averaging RNA for cluster KPC4fib
Finished averaging integrated for cluster KPC1viable
Finished averaging integrated for cluster KPC2viable
Finished averaging integrated for cluster KPC3viable
Finished averaging integrated for cluster KPC4viable
Finished averaging integrated for cluster KPC1fib
Finished averaging integrated for cluster KPC2fib
Finished averaging integrated for cluster KPC3fib
Finished averaging integrated for cluster KPC4fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   6%  |                                                                              |=========                                                             |  12%  |                                                                              |=============                                                         |  19%  |                                                                              |==================                                                    |  25%  |                                                                              |======================                                                |  31%  |                                                                              |==========================                                            |  38%  |                                                                              |===============================                                       |  44%  |                                                                              |===================================                                   |  50%  |                                                                              |=======================================                               |  56%  |                                                                              |============================================                          |  62%  |                                                                              |================================================                      |  69%  |                                                                              |====================================================                  |  75%  |                                                                              |=========================================================             |  81%  |                                                                              |=============================================================         |  88%  |                                                                              |==================================================================    |  94%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster KPC1viable
Finished averaging RNA for cluster KPC2viable
Finished averaging RNA for cluster KPC3viable
Finished averaging RNA for cluster KPC4viable
Finished averaging RNA for cluster KPC1fib
Finished averaging RNA for cluster KPC2fib
Finished averaging RNA for cluster KPC3fib
Finished averaging RNA for cluster KPC4fib

  |                                                                              |.....................                                                 |  29%
  ordinary text without R code

  |                                                                              |......................                                                |  31%
label: unnamed-chunk-7 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.plot.pca.and.tsne.results.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.plot.pca.and.tsne.results.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_pca_tsne_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: Plot_dimred_by_cluster_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE

Finished averaging RNA for cluster KPC1viable
Finished averaging RNA for cluster KPC2viable
Finished averaging RNA for cluster KPC3viable
Finished averaging RNA for cluster KPC4viable
Finished averaging RNA for cluster KPC1fib
Finished averaging RNA for cluster KPC2fib
Finished averaging RNA for cluster KPC3fib
Finished averaging RNA for cluster KPC4fib
Finished averaging RNA for cluster C15
Finished averaging RNA for cluster C3
Finished averaging RNA for cluster C1
Finished averaging RNA for cluster C0
Finished averaging RNA for cluster C10
Finished averaging RNA for cluster C7
Finished averaging RNA for cluster C5
Finished averaging RNA for cluster C13
Finished averaging RNA for cluster C6
Finished averaging RNA for cluster C9
Finished averaging RNA for cluster C2
Finished averaging RNA for cluster C12
Finished averaging RNA for cluster C8
Finished averaging RNA for cluster C14
Finished averaging RNA for cluster C4
Finished averaging RNA for cluster C16
Finished averaging RNA for cluster C17
Finished averaging RNA for cluster C11
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Plot_dim_red_by_cluster (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.......................                                               |  33%
  ordinary text without R code

  |                                                                              |.........................                                             |  35%
label: unnamed-chunk-8 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.euclidian.distances.figure.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.euclidian.distances.figure.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..............                                                        |  20%
  ordinary text without R code

  |                                                                              |............................                                          |  40%
label: partB_euclidian_distances (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................................                            |  60%
  ordinary text without R code

  |                                                                              |........................................................              |  80%
label: Plot_umap_sleepwalk (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

Estimating 'maxdist' for feature matrix 1
  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |..........................                                            |  37%
  ordinary text without R code

  |                                                                              |...........................                                           |  39%
label: unnamed-chunk-9 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.doublet.estimation.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.doublet.estimation.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_doublet_estimates (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: doublet-data (with options) 
List of 5
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE
 $ message: logi FALSE

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: doublet-plot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.............................                                         |  41%
  ordinary text without R code

  |                                                                              |..............................                                        |  43%
label: unnamed-chunk-10 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/cell.cycle.phase.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/cell.cycle.phase.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_cell_cycle_phase (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: Plot_dimred_by_cell_cycle_phase_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE

Warning: The following features are not present in the object: MCM5, PCNA, TYMS, FEN1, MCM2, MCM4, RRM1, UNG, GINS2, MCM6, CDCA7, DTL, PRIM1, UHRF1, MLF1IP, HELLS, RFC2, RPA2, NASP, RAD51AP1, GMNN, WDR76, SLBP, CCNE2, UBR7, POLD3, MSH2, ATAD2, RAD51, RRM2, CDC45, CDC6, EXO1, TIPIN, DSCC1, BLM, CASP8AP2, USP1, CLSPN, POLA1, CHAF1B, BRIP1, E2F8, not searching for symbol synonyms
Warning: The following features are not present in the object: HMGB2, CDK1, NUSAP1, UBE2C, BIRC5, TPX2, TOP2A, NDC80, CKS2, NUF2, CKS1B, MKI67, TMPO, CENPF, TACC3, FAM64A, SMC4, CCNB2, CKAP2L, CKAP2, AURKB, BUB1, KIF11, ANP32E, TUBB4B, GTSE1, KIF20B, HJURP, CDCA3, HN1, CDC20, TTK, CDC25C, KIF2C, RANGAP1, NCAPD2, DLGAP5, CDCA2, CDCA8, ECT2, KIF23, HMMR, AURKA, PSRC1, ANLN, LBR, CKAP5, CENPE, CTCF, NEK2, G2E3, GAS2L3, CBX5, CENPA, not searching for symbol synonyms
Finished averaging RNA for cluster KPC1viable
Finished averaging RNA for cluster KPC2viable
Finished averaging RNA for cluster KPC3viable
Finished averaging RNA for cluster KPC4viable
Finished averaging RNA for cluster KPC1fib
Finished averaging RNA for cluster KPC2fib
Finished averaging RNA for cluster KPC3fib
Finished averaging RNA for cluster KPC4fib
Finished averaging RNA for cluster C15
Finished averaging RNA for cluster C3
Finished averaging RNA for cluster C1
Finished averaging RNA for cluster C0
Finished averaging RNA for cluster C10
Finished averaging RNA for cluster C7
Finished averaging RNA for cluster C5
Finished averaging RNA for cluster C13
Finished averaging RNA for cluster C6
Finished averaging RNA for cluster C9
Finished averaging RNA for cluster C2
Finished averaging RNA for cluster C12
Finished averaging RNA for cluster C8
Finished averaging RNA for cluster C14
Finished averaging RNA for cluster C4
Finished averaging RNA for cluster C16
Finished averaging RNA for cluster C17
Finished averaging RNA for cluster C11
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Plot_dim_red_by_cell_cycle_phase (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |................................                                      |  45%
  ordinary text without R code

  |                                                                              |.................................                                     |  47%
label: unnamed-chunk-11 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.Ncell.barchart.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.Ncell.barchart.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_N_cell_barchart (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: make_percent_N_cells_barchart (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: plot_barchart (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |..................................                                    |  49%
  ordinary text without R code

  |                                                                              |....................................                                  |  51%
label: unnamed-chunk-12 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.AUC.cat.enrichments.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.AUC.cat.enrichments.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_AUC_cat_enrichment (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: AUC_prep_from_file (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Quantiles for the number of genes detected by cell: 
(Non-detected genes are shuffled at the end of the ranking. Keep it in mind when choosing the threshold for calculating the AUC).
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: enrichment_plots (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.....................................                                 |  53%
  ordinary text without R code

  |                                                                              |......................................                                |  55%
label: unnamed-chunk-13 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.linear.regression.data.prep.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.linear.regression.data.prep.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_linear_regression (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: data_pre_linear_regression (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: plot_linear_regression (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |........................................                              |  57%
  ordinary text without R code

  |                                                                              |.........................................                             |  59%
label: unnamed-chunk-14 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.heatmaps.and.dotplots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.heatmaps.and.dotplots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_heatmaps_and_dotplots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: Cluster_overview_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Cluster_overview (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |...........................................                           |  61%
  ordinary text without R code

  |                                                                              |............................................                          |  63%
label: unnamed-chunk-15 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.make.cluster.table.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.make.cluster.table.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_make_cluster_table (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: ClusterTable_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


Attaching package: 'DT'

The following object is masked from 'package:Seurat':

    JS

  |                                                                              |..........................................................            |  83%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: render_ClusterTable (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.............................................                         |  65%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: unnamed-chunk-16 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.cell.type.estimate.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.cell.type.estimate.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_cell_type_estimate (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: CellTypePlot_dataprep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F


Registered S3 method overwritten by 'enrichplot':
  method               from
  fortify.enrichResult DOSE
clusterProfiler v3.12.0  For help: https://guangchuangyu.github.io/software/clusterProfiler

If you use clusterProfiler in published research, please cite:
Guangchuang Yu, Li-Gen Wang, Yanyan Han, Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology. 2012, 16(5):284-287.

Attaching package: 'clusterProfiler'

The following object is masked from 'package:DelayedArray':

    simplify

The following object is masked from 'package:purrr':

    simplify

Loading required package: GSEABase
Loading required package: annotate
Loading required package: AnnotationDbi

Attaching package: 'AnnotationDbi'

The following object is masked from 'package:dplyr':

    select

Loading required package: XML
Loading required package: graph

Attaching package: 'graph'

The following object is masked from 'package:XML':

    addNode

The following object is masked from 'package:stringr':

    boundary

--> No gene can be mapped....
--> Expected input gene ID: Fcgr2b,Fzd2,Ccl22,Bank1,Cd38,Ly86
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Rab27b,Il18r1,Fcer1a,H2-Aa,Slamf8,Ggt5
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Tcl1,Dapk2,P2rx1,Ccl19,Cxcl13,Emr1
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Clec10a,Slc15a3,Ms4a2,Emr1,Paqr5,Ccl2
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Il1b,Ccr7,Myb,Tnfrsf4,Mak,Ido1
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Dcstamp,Sell,Irf8,Blk,Ebi3,Gpr18
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Cd86,Sept8,Cd69,Frk,Trem2,Cxcl2
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Cd22,Ccl2,H2-Ob,Il2ra,Ccl22,Igsf6
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Il18r1,Rgs13,Cyp27b1,Ggt5,Cd4,Gm2023
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Cd40,Faim3,Hoxa1,Smpd3,Ncf2,Mmp9
--> return NULL...
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: CellTypePlot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |................................................                      |  69%
  ordinary text without R code

  |                                                                              |.................................................                     |  71%
label: unnamed-chunk-17 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.gene.set.heatmaps.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.gene.set.heatmaps.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_gene_set_heatmaps (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: MakeHM (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |..........................................................            |  83%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: calculate-markdown-heatmap-chunks-dynamically (with options) 
List of 4
 $ echo   : symbol F
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |...................................................                   |  73%
  ordinary text without R code

  |                                                                              |....................................................                  |  75%
label: unnamed-chunk-18 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.example.feature.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.example.feature.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_example_feature_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: ExampleFeaturePlot_dataprep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |..........................................................            |  83%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: ExampleFeaturePlot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |......................................................                |  76%
  ordinary text without R code

  |                                                                              |.......................................................               |  78%
label: unnamed-chunk-19 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.pca.cell.and.loading.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.pca.cell.and.loading.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |....                                                                  |   6%
  ordinary text without R code

  |                                                                              |.........                                                             |  12%
label: partB_pca_loadings (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |.............                                                         |  19%
  ordinary text without R code

  |                                                                              |..................                                                    |  25%
label: elbow_plot (with options) 
List of 3
 $ fig.height: num 6
 $ fig.width : num 10
 $ fig.cap   : language paste0("**Figure, ", figureCount, ": Elbowplot** Variance explained per PCA dimension")

  |                                                                              |......................                                                |  31%
  ordinary text without R code

  |                                                                              |..........................                                            |  38%
label: PCA_dataprep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F


Attaching package: 'ggpubr'

The following object is masked from 'package:ggtree':

    rotate


Attaching package: 'gridExtra'

The following object is masked from 'package:Biobase':

    combine

The following object is masked from 'package:BiocGenerics':

    combine

The following object is masked from 'package:dplyr':

    combine

  |                                                                              |...............................                                       |  44%
   inline R code fragments

  |                                                                              |...................................                                   |  50%
label: plot_PCADimPlots (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |.......................................                               |  56%
  ordinary text without R code

  |                                                                              |............................................                          |  62%
label: pca_boxplot_dataprep (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

  |                                                                              |................................................                      |  69%
   inline R code fragments

  |                                                                              |....................................................                  |  75%
label: pca_boxplot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |.........................................................             |  81%
  ordinary text without R code

  |                                                                              |.............................................................         |  88%
label: create-pca-enrichment-data (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

--> No gene can be mapped....
--> Expected input gene ID: Gm13309,Ccl2,Ccl4,Fcer1a,Cd69,Hdc
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Dach1,Tlr7,Faim3,Hesx1,H2-Ob,Hpgds
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Fcer1a,H2-Aa,Hoxa1,Emr1,Ptprcap,Smpdl3b
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Lamp3,Rgs13,Ms4a1,P2rx5,Aif1,Alox5
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Fcgr2b,Frmd4a,Socs1,Cxcr5,Dapk2,P2ry10
--> return NULL...
  |                                                                              |..................................................................    |  94%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: create-pca-enrichment-plot (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |........................................................              |  80%
  ordinary text without R code

  |                                                                              |..........................................................            |  82%
label: unnamed-chunk-20 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.pseudotime.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.pseudotime.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_pseudotime (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: pseudotime-enrichment-data (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: pseudotime-plot (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |...........................................................           |  84%
  ordinary text without R code

  |                                                                              |............................................................          |  86%
label: unnamed-chunk-21 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.add.extra.tables.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.add.extra.tables.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..................                                                    |  25%
  ordinary text without R code

  |                                                                              |...................................                                   |  50%
label: partB_add_extra_tables (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |....................................................                  |  75%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: Additional tables (with options) 
List of 4
 $ echo   : symbol F
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F


  |                                                                              |..............................................................        |  88%
  ordinary text without R code

  |                                                                              |...............................................................       |  90%
label: saveobject (with options) 
List of 3
 $ eval   : logi TRUE
 $ echo   : symbol T
 $ results: symbol F

  |                                                                              |.................................................................     |  92%
  ordinary text without R code

  |                                                                              |..................................................................    |  94%
label: create_report_params (with options) 
List of 2
 $ eval   : symbol T
 $ results: chr "asis"

Project Database could not be reached or has no entry in Obio@parameterList$lims.id for this analysis.
  |                                                                              |...................................................................   |  96%
  ordinary text without R code

  |                                                                              |..................................................................... |  98%
label: documentation (with options) 
List of 3
 $ eval   : logi TRUE
 $ echo   : symbol T
 $ results: symbol T

  |                                                                              |......................................................................| 100%
   inline R code fragments


output file: sc_PartB_Analysis.knit.md

/camp/apps/eb/dev/software/Anaconda2/2019.10/bin/pandoc +RTS -K512m -RTS sc_PartB_Analysis.utf8.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /camp/stp/babs/working/boeings/Projects/hillc/danielle.park/439_scRNAseq_KPC_PDAC_GSE129455/workdir/html_local/sc_PartB_Analysis.html --lua-filter /camp/apps/misc/stp/babs/manual/software/r/r-site-library/bioc-3.9/rmarkdown/rmarkdown/lua/pagebreak.lua --lua-filter /camp/apps/misc/stp/babs/manual/software/r/r-site-library/bioc-3.9/rmarkdown/rmarkdown/lua/latex-div.lua --self-contained --variable bs3=TRUE --standalone --section-divs --table-of-contents --toc-depth 5 --variable toc_float=1 --variable toc_selectors=h1,h2,h3,h4,h5 --variable toc_collapsed=1 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /camp/apps/misc/stp/babs/manual/software/r/r-site-library/bioc-3.9/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --css src/assets/style/style.css --variable theme=paper --include-in-header /tmp/slurm_25480163/RtmpxQTLxA/rmarkdown-str45212af031f7.html --mathjax --variable 'mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --variable code_folding=hide --variable code_menu=1 

Output created: /camp/stp/babs/working/boeings/Projects/hillc/danielle.park/439_scRNAseq_KPC_PDAC_GSE129455/workdir/html_local/sc_PartB_Analysis.html
There were 45 warnings (use warnings() to see them)
