

processing file: sc_PartB_Analysis.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |.                                                                     |   2%
  ordinary text without R code

  |                                                                              |...                                                                   |   4%
label: setup (with options) 
List of 1
 $ include: logi FALSE

  |                                                                              |....                                                                  |   6%
  ordinary text without R code

  |                                                                              |.....                                                                 |   8%
label: unnamed-chunk-1 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B.1.set.parameters.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B.1.set.parameters.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..............                                                        |  20%
  ordinary text without R code

  |                                                                              |............................                                          |  40%
label: set_parameters_init (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................................                            |  60%
  ordinary text without R code

  |                                                                              |........................................................              |  80%
label: B_set_directories (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |.......                                                               |  10%
  ordinary text without R code

  |                                                                              |........                                                              |  12%
label: unnamed-chunk-2 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B.2.retrieve.reference.gene.sets.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B.2.retrieve.reference.gene.sets.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..............                                                        |  20%
  ordinary text without R code

  |                                                                              |............................                                          |  40%
label: partB_retrieve_reference_gene_sets_init (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................................                            |  60%
  ordinary text without R code

  |                                                                              |........................................................              |  80%
label: partB_create_sample_list (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Loading required package: DBI
  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |..........                                                            |  14%
  ordinary text without R code

  |                                                                              |...........                                                           |  16%
label: unnamed-chunk-3 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B.3.create.integrated.seurat.object.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B.3.create.integrated.seurat.object.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_create_integrated_Seurat_object_init (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: B3_partB_create_sample_list (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')
Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')
Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')
Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')
Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')
Computing 2000 integration features
Scaling features for provided objects
Finding all pairwise anchors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 4866 anchors
Filtering anchors
	Retained 2485 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 4800 anchors
Filtering anchors
	Retained 3377 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 6559 anchors
Filtering anchors
	Retained 4358 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 4561 anchors
Filtering anchors
	Retained 2727 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 8005 anchors
Filtering anchors
	Retained 6443 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 7408 anchors
Filtering anchors
	Retained 5498 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 5673 anchors
Filtering anchors
	Retained 2295 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 9067 anchors
Filtering anchors
	Retained 7083 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 7619 anchors
Filtering anchors
	Retained 4191 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 9431 anchors
Filtering anchors
	Retained 6742 anchors
Extracting within-dataset neighbors
Merging dataset 2 into 5
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 4 into 5 2
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 1 into 3
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 3 1 into 5 2 4
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Warning: Adding a command log without an assay associated with it
  |                                                                              |..........................................................            |  83%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: partB_create_PCA_plots (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session
16:33:22 UMAP embedding parameters a = 0.9922 b = 1.112
16:33:22 Read 21131 rows and found 30 numeric columns
16:33:23 Using Annoy for neighbor search, n_neighbors = 30
16:33:23 Building Annoy index with metric = cosine, n_trees = 50
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
16:33:27 Writing NN index file to temp file /tmp/RtmpSLSyWu/file8ad023bebd9a
16:33:27 Searching Annoy index using 1 thread, search_k = 3000
16:33:36 Annoy recall = 100%
16:33:37 Commencing smooth kNN distance calibration using 1 thread
16:33:40 Initializing from normalized Laplacian + noise
16:33:43 Commencing optimization for 200 epochs, with 932568 positive edges
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
16:34:17 Optimization finished
Computing nearest neighbor graph
Computing SNN
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|

Attaching package: 'scales'

The following object is masked from 'package:purrr':

    discard

The following object is masked from 'package:readr':

    col_factor

Calculating cluster C20
Calculating cluster C6
Calculating cluster C10
Calculating cluster C5
Calculating cluster C15
Calculating cluster C0
Calculating cluster C21
Calculating cluster C3
Calculating cluster C19
Calculating cluster C22
Calculating cluster C8
Calculating cluster C2
Calculating cluster C7
Calculating cluster C4
Calculating cluster C9
Calculating cluster C11
Calculating cluster C16
Calculating cluster C14
Calculating cluster C1
Calculating cluster C17
Calculating cluster C18
Calculating cluster C12
Calculating cluster C13

  |                                                                              |............                                                          |  18%
  ordinary text without R code

  |                                                                              |..............                                                        |  20%
label: unnamed-chunk-4 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B.4.create.variation.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B.4.create.variation.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..........                                                            |  14%
  ordinary text without R code

  |                                                                              |....................                                                  |  29%
label: partB_create_variation_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..............................                                        |  43%
  ordinary text without R code

  |                                                                              |........................................                              |  57%
label: partB_create-individual-varFeatures (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster all
Finished averaging integrated for cluster all
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   6%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |===========                                                           |  16%  |                                                                              |==============                                                        |  19%  |                                                                              |================                                                      |  23%  |                                                                              |==================                                                    |  26%  |                                                                              |====================                                                  |  29%  |                                                                              |=======================                                               |  32%  |                                                                              |=========================                                             |  35%  |                                                                              |===========================                                           |  39%  |                                                                              |=============================                                         |  42%  |                                                                              |================================                                      |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |======================================                                |  55%  |                                                                              |=========================================                             |  58%  |                                                                              |===========================================                           |  61%  |                                                                              |=============================================                         |  65%  |                                                                              |===============================================                       |  68%  |                                                                              |==================================================                    |  71%  |                                                                              |====================================================                  |  74%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  81%  |                                                                              |===========================================================           |  84%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  94%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster LiverNormalM1
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   6%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |===========                                                           |  16%  |                                                                              |==============                                                        |  19%  |                                                                              |================                                                      |  23%  |                                                                              |==================                                                    |  26%  |                                                                              |====================                                                  |  29%  |                                                                              |=======================                                               |  32%  |                                                                              |=========================                                             |  35%  |                                                                              |===========================                                           |  39%  |                                                                              |=============================                                         |  42%  |                                                                              |================================                                      |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |======================================                                |  55%  |                                                                              |=========================================                             |  58%  |                                                                              |===========================================                           |  61%  |                                                                              |=============================================                         |  65%  |                                                                              |===============================================                       |  68%  |                                                                              |==================================================                    |  71%  |                                                                              |====================================================                  |  74%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  81%  |                                                                              |===========================================================           |  84%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  94%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster SubcutNoLiverMetM2
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   6%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |===========                                                           |  16%  |                                                                              |==============                                                        |  19%  |                                                                              |================                                                      |  23%  |                                                                              |==================                                                    |  26%  |                                                                              |====================                                                  |  29%  |                                                                              |=======================                                               |  32%  |                                                                              |=========================                                             |  35%  |                                                                              |===========================                                           |  39%  |                                                                              |=============================                                         |  42%  |                                                                              |================================                                      |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |======================================                                |  55%  |                                                                              |=========================================                             |  58%  |                                                                              |===========================================                           |  61%  |                                                                              |=============================================                         |  65%  |                                                                              |===============================================                       |  68%  |                                                                              |==================================================                    |  71%  |                                                                              |====================================================                  |  74%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  81%  |                                                                              |===========================================================           |  84%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  94%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster BackgroundLiverM3
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   6%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |===========                                                           |  16%  |                                                                              |==============                                                        |  19%  |                                                                              |================                                                      |  23%  |                                                                              |==================                                                    |  26%  |                                                                              |====================                                                  |  29%  |                                                                              |=======================                                               |  32%  |                                                                              |=========================                                             |  35%  |                                                                              |===========================                                           |  39%  |                                                                              |=============================                                         |  42%  |                                                                              |================================                                      |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |======================================                                |  55%  |                                                                              |=========================================                             |  58%  |                                                                              |===========================================                           |  61%  |                                                                              |=============================================                         |  65%  |                                                                              |===============================================                       |  68%  |                                                                              |==================================================                    |  71%  |                                                                              |====================================================                  |  74%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  81%  |                                                                              |===========================================================           |  84%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  94%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster LiverMetM3
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   6%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |===========                                                           |  16%  |                                                                              |==============                                                        |  19%  |                                                                              |================                                                      |  23%  |                                                                              |==================                                                    |  26%  |                                                                              |====================                                                  |  29%  |                                                                              |=======================                                               |  32%  |                                                                              |=========================                                             |  35%  |                                                                              |===========================                                           |  39%  |                                                                              |=============================                                         |  42%  |                                                                              |================================                                      |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |======================================                                |  55%  |                                                                              |=========================================                             |  58%  |                                                                              |===========================================                           |  61%  |                                                                              |=============================================                         |  65%  |                                                                              |===============================================                       |  68%  |                                                                              |==================================================                    |  71%  |                                                                              |====================================================                  |  74%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  81%  |                                                                              |===========================================================           |  84%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  94%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster SubcutM3
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   6%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |===========                                                           |  16%  |                                                                              |==============                                                        |  19%  |                                                                              |================                                                      |  23%  |                                                                              |==================                                                    |  26%  |                                                                              |====================                                                  |  29%  |                                                                              |=======================                                               |  32%  |                                                                              |=========================                                             |  35%  |                                                                              |===========================                                           |  39%  |                                                                              |=============================                                         |  42%  |                                                                              |================================                                      |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |======================================                                |  55%  |                                                                              |=========================================                             |  58%  |                                                                              |===========================================                           |  61%  |                                                                              |=============================================                         |  65%  |                                                                              |===============================================                       |  68%  |                                                                              |==================================================                    |  71%  |                                                                              |====================================================                  |  74%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  81%  |                                                                              |===========================================================           |  84%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  94%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%
  |                                                                              |..................................................                    |  71%
   inline R code fragments

  |                                                                              |............................................................          |  86%
label: create-var-feature-plot-II (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |...............                                                       |  22%
  ordinary text without R code

  |                                                                              |................                                                      |  24%
label: unnamed-chunk-5 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B.6.create.pca.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B.6.create.pca.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |.........                                                             |  12%
  ordinary text without R code

  |                                                                              |..................                                                    |  25%
label: partB_pcaB6_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................                                            |  38%
  ordinary text without R code

  |                                                                              |...................................                                   |  50%
label: partB_upload_markers (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

  |                                                                              |............................................                          |  62%
  ordinary text without R code

  |                                                                              |....................................................                  |  75%
label: data_prep_sample_UMAP_rendering (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE

Registered S3 method overwritten by 'treeio':
  method     from
  root.phylo ape 
ggtree v1.16.6  For help: https://yulab-smu.github.io/treedata-book/

If you use ggtree in published research, please cite the most appropriate paper(s):

[36m-[39m Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods for mapping and visualizing associated data on phylogeny using ggtree. Molecular Biology and Evolution 2018, 35(12):3041-3043. doi: 10.1093/molbev/msy194
[36m-[39m Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution 2017, 8(1):28-36, doi:10.1111/2041-210X.12628



Attaching package: 'ggtree'

The following object is masked from 'package:tidyr':

    expand

Finished averaging RNA for cluster LiverNormalM1
Finished averaging RNA for cluster SubcutNoLiverMetM2
Finished averaging RNA for cluster BackgroundLiverM3
Finished averaging RNA for cluster LiverMetM3
Finished averaging RNA for cluster SubcutM3
  |                                                                              |.............................................................         |  88%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Plot_tsne_data_plotting (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |..................                                                    |  25%
  ordinary text without R code

  |                                                                              |...................                                                   |  27%
label: unnamed-chunk-6 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B7.calculate.average.expression.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B7.calculate.average.expression.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..................                                                    |  25%
  ordinary text without R code

  |                                                                              |...................................                                   |  50%
label: partB_calculate_avg_expr (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |....................................................                  |  75%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: Identify differential expressed genes across conditions (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

Finished averaging RNA for cluster all
Finished averaging integrated for cluster all
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   6%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |===========                                                           |  16%  |                                                                              |==============                                                        |  19%  |                                                                              |================                                                      |  23%  |                                                                              |==================                                                    |  26%  |                                                                              |====================                                                  |  29%  |                                                                              |=======================                                               |  32%  |                                                                              |=========================                                             |  35%  |                                                                              |===========================                                           |  39%  |                                                                              |=============================                                         |  42%  |                                                                              |================================                                      |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |======================================                                |  55%  |                                                                              |=========================================                             |  58%  |                                                                              |===========================================                           |  61%  |                                                                              |=============================================                         |  65%  |                                                                              |===============================================                       |  68%  |                                                                              |==================================================                    |  71%  |                                                                              |====================================================                  |  74%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  81%  |                                                                              |===========================================================           |  84%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  94%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster C20_LiverNormalM1
Finished averaging RNA for cluster C6_LiverNormalM1
Finished averaging RNA for cluster C10_LiverNormalM1
Finished averaging RNA for cluster C5_LiverNormalM1
Finished averaging RNA for cluster C15_LiverNormalM1
Finished averaging RNA for cluster C0_LiverNormalM1
Finished averaging RNA for cluster C21_LiverNormalM1
Finished averaging RNA for cluster C3_LiverNormalM1
Finished averaging RNA for cluster C19_LiverNormalM1
Finished averaging RNA for cluster C22_LiverNormalM1
Finished averaging RNA for cluster C8_LiverNormalM1
Finished averaging RNA for cluster C2_LiverNormalM1
Finished averaging RNA for cluster C7_LiverNormalM1
Finished averaging RNA for cluster C4_LiverNormalM1
Finished averaging RNA for cluster C9_LiverNormalM1
Finished averaging RNA for cluster C11_LiverNormalM1
Finished averaging RNA for cluster C16_LiverNormalM1
Finished averaging RNA for cluster C14_LiverNormalM1
Finished averaging RNA for cluster C1_LiverNormalM1
Finished averaging RNA for cluster C17_LiverNormalM1
Finished averaging RNA for cluster C18_LiverNormalM1
Finished averaging RNA for cluster C12_LiverNormalM1
Finished averaging RNA for cluster C13_LiverNormalM1
Finished averaging RNA for cluster C12_SubcutNoLiverMetM2
Finished averaging RNA for cluster C0_SubcutNoLiverMetM2
Finished averaging RNA for cluster C3_SubcutNoLiverMetM2
Finished averaging RNA for cluster C11_SubcutNoLiverMetM2
Finished averaging RNA for cluster C9_SubcutNoLiverMetM2
Finished averaging RNA for cluster C4_SubcutNoLiverMetM2
Finished averaging RNA for cluster C21_SubcutNoLiverMetM2
Finished averaging RNA for cluster C7_SubcutNoLiverMetM2
Finished averaging RNA for cluster C13_SubcutNoLiverMetM2
Finished averaging RNA for cluster C1_SubcutNoLiverMetM2
Finished averaging RNA for cluster C14_SubcutNoLiverMetM2
Finished averaging RNA for cluster C8_SubcutNoLiverMetM2
Finished averaging RNA for cluster C18_SubcutNoLiverMetM2
Finished averaging RNA for cluster C2_SubcutNoLiverMetM2
Finished averaging RNA for cluster C20_SubcutNoLiverMetM2
Finished averaging RNA for cluster C15_SubcutNoLiverMetM2
Finished averaging RNA for cluster C6_SubcutNoLiverMetM2
Finished averaging RNA for cluster C16_SubcutNoLiverMetM2
Finished averaging RNA for cluster C22_SubcutNoLiverMetM2
Finished averaging RNA for cluster C17_SubcutNoLiverMetM2
Finished averaging RNA for cluster C10_SubcutNoLiverMetM2
Finished averaging RNA for cluster C5_SubcutNoLiverMetM2
Finished averaging RNA for cluster C19_SubcutNoLiverMetM2
Finished averaging RNA for cluster C5_BackgroundLiverM3
Finished averaging RNA for cluster C1_BackgroundLiverM3
Finished averaging RNA for cluster C22_BackgroundLiverM3
Finished averaging RNA for cluster C3_BackgroundLiverM3
Finished averaging RNA for cluster C0_BackgroundLiverM3
Finished averaging RNA for cluster C11_BackgroundLiverM3
Finished averaging RNA for cluster C10_BackgroundLiverM3
Finished averaging RNA for cluster C4_BackgroundLiverM3
Finished averaging RNA for cluster C13_BackgroundLiverM3
Finished averaging RNA for cluster C15_BackgroundLiverM3
Finished averaging RNA for cluster C8_BackgroundLiverM3
Finished averaging RNA for cluster C12_BackgroundLiverM3
Finished averaging RNA for cluster C6_BackgroundLiverM3
Finished averaging RNA for cluster C16_BackgroundLiverM3
Finished averaging RNA for cluster C7_BackgroundLiverM3
Finished averaging RNA for cluster C18_BackgroundLiverM3
Finished averaging RNA for cluster C19_BackgroundLiverM3
Finished averaging RNA for cluster C9_BackgroundLiverM3
Finished averaging RNA for cluster C14_BackgroundLiverM3
Finished averaging RNA for cluster C21_BackgroundLiverM3
Finished averaging RNA for cluster C17_BackgroundLiverM3
Finished averaging RNA for cluster C20_BackgroundLiverM3
Finished averaging RNA for cluster C4_LiverMetM3
Finished averaging RNA for cluster C1_LiverMetM3
Finished averaging RNA for cluster C0_LiverMetM3
Finished averaging RNA for cluster C7_LiverMetM3
Finished averaging RNA for cluster C17_LiverMetM3
Finished averaging RNA for cluster C16_LiverMetM3
Finished averaging RNA for cluster C8_LiverMetM3
Finished averaging RNA for cluster C14_LiverMetM3
Finished averaging RNA for cluster C6_LiverMetM3
Finished averaging RNA for cluster C3_LiverMetM3
Finished averaging RNA for cluster C10_LiverMetM3
Finished averaging RNA for cluster C15_LiverMetM3
Finished averaging RNA for cluster C12_LiverMetM3
Finished averaging RNA for cluster C13_LiverMetM3
Finished averaging RNA for cluster C19_LiverMetM3
Finished averaging RNA for cluster C5_LiverMetM3
Finished averaging RNA for cluster C9_LiverMetM3
Finished averaging RNA for cluster C2_LiverMetM3
Finished averaging RNA for cluster C20_LiverMetM3
Finished averaging RNA for cluster C11_LiverMetM3
Finished averaging RNA for cluster C21_LiverMetM3
Finished averaging RNA for cluster C18_LiverMetM3
Finished averaging RNA for cluster C22_LiverMetM3
Finished averaging RNA for cluster C0_SubcutM3
Finished averaging RNA for cluster C2_SubcutM3
Finished averaging RNA for cluster C9_SubcutM3
Finished averaging RNA for cluster C1_SubcutM3
Finished averaging RNA for cluster C3_SubcutM3
Finished averaging RNA for cluster C15_SubcutM3
Finished averaging RNA for cluster C14_SubcutM3
Finished averaging RNA for cluster C8_SubcutM3
Finished averaging RNA for cluster C12_SubcutM3
Finished averaging RNA for cluster C16_SubcutM3
Finished averaging RNA for cluster C6_SubcutM3
Finished averaging RNA for cluster C4_SubcutM3
Finished averaging RNA for cluster C7_SubcutM3
Finished averaging RNA for cluster C5_SubcutM3
Finished averaging RNA for cluster C10_SubcutM3
Finished averaging RNA for cluster C11_SubcutM3
Finished averaging RNA for cluster C20_SubcutM3
Finished averaging RNA for cluster C19_SubcutM3
Finished averaging RNA for cluster C13_SubcutM3
Finished averaging RNA for cluster C21_SubcutM3
Finished averaging RNA for cluster C22_SubcutM3
Finished averaging RNA for cluster C17_SubcutM3
Finished averaging RNA for cluster C18_SubcutM3
Finished averaging integrated for cluster C20_LiverNormalM1
Finished averaging integrated for cluster C6_LiverNormalM1
Finished averaging integrated for cluster C10_LiverNormalM1
Finished averaging integrated for cluster C5_LiverNormalM1
Finished averaging integrated for cluster C15_LiverNormalM1
Finished averaging integrated for cluster C0_LiverNormalM1
Finished averaging integrated for cluster C21_LiverNormalM1
Finished averaging integrated for cluster C3_LiverNormalM1
Finished averaging integrated for cluster C19_LiverNormalM1
Finished averaging integrated for cluster C22_LiverNormalM1
Finished averaging integrated for cluster C8_LiverNormalM1
Finished averaging integrated for cluster C2_LiverNormalM1
Finished averaging integrated for cluster C7_LiverNormalM1
Finished averaging integrated for cluster C4_LiverNormalM1
Finished averaging integrated for cluster C9_LiverNormalM1
Finished averaging integrated for cluster C11_LiverNormalM1
Finished averaging integrated for cluster C16_LiverNormalM1
Finished averaging integrated for cluster C14_LiverNormalM1
Finished averaging integrated for cluster C1_LiverNormalM1
Finished averaging integrated for cluster C17_LiverNormalM1
Finished averaging integrated for cluster C18_LiverNormalM1
Finished averaging integrated for cluster C12_LiverNormalM1
Finished averaging integrated for cluster C13_LiverNormalM1
Finished averaging integrated for cluster C12_SubcutNoLiverMetM2
Finished averaging integrated for cluster C0_SubcutNoLiverMetM2
Finished averaging integrated for cluster C3_SubcutNoLiverMetM2
Finished averaging integrated for cluster C11_SubcutNoLiverMetM2
Finished averaging integrated for cluster C9_SubcutNoLiverMetM2
Finished averaging integrated for cluster C4_SubcutNoLiverMetM2
Finished averaging integrated for cluster C21_SubcutNoLiverMetM2
Finished averaging integrated for cluster C7_SubcutNoLiverMetM2
Finished averaging integrated for cluster C13_SubcutNoLiverMetM2
Finished averaging integrated for cluster C1_SubcutNoLiverMetM2
Finished averaging integrated for cluster C14_SubcutNoLiverMetM2
Finished averaging integrated for cluster C8_SubcutNoLiverMetM2
Finished averaging integrated for cluster C18_SubcutNoLiverMetM2
Finished averaging integrated for cluster C2_SubcutNoLiverMetM2
Finished averaging integrated for cluster C20_SubcutNoLiverMetM2
Finished averaging integrated for cluster C15_SubcutNoLiverMetM2
Finished averaging integrated for cluster C6_SubcutNoLiverMetM2
Finished averaging integrated for cluster C16_SubcutNoLiverMetM2
Finished averaging integrated for cluster C22_SubcutNoLiverMetM2
Finished averaging integrated for cluster C17_SubcutNoLiverMetM2
Finished averaging integrated for cluster C10_SubcutNoLiverMetM2
Finished averaging integrated for cluster C5_SubcutNoLiverMetM2
Finished averaging integrated for cluster C19_SubcutNoLiverMetM2
Finished averaging integrated for cluster C5_BackgroundLiverM3
Finished averaging integrated for cluster C1_BackgroundLiverM3
Finished averaging integrated for cluster C22_BackgroundLiverM3
Finished averaging integrated for cluster C3_BackgroundLiverM3
Finished averaging integrated for cluster C0_BackgroundLiverM3
Finished averaging integrated for cluster C11_BackgroundLiverM3
Finished averaging integrated for cluster C10_BackgroundLiverM3
Finished averaging integrated for cluster C4_BackgroundLiverM3
Finished averaging integrated for cluster C13_BackgroundLiverM3
Finished averaging integrated for cluster C15_BackgroundLiverM3
Finished averaging integrated for cluster C8_BackgroundLiverM3
Finished averaging integrated for cluster C12_BackgroundLiverM3
Finished averaging integrated for cluster C6_BackgroundLiverM3
Finished averaging integrated for cluster C16_BackgroundLiverM3
Finished averaging integrated for cluster C7_BackgroundLiverM3
Finished averaging integrated for cluster C18_BackgroundLiverM3
Finished averaging integrated for cluster C19_BackgroundLiverM3
Finished averaging integrated for cluster C9_BackgroundLiverM3
Finished averaging integrated for cluster C14_BackgroundLiverM3
Finished averaging integrated for cluster C21_BackgroundLiverM3
Finished averaging integrated for cluster C17_BackgroundLiverM3
Finished averaging integrated for cluster C20_BackgroundLiverM3
Finished averaging integrated for cluster C4_LiverMetM3
Finished averaging integrated for cluster C1_LiverMetM3
Finished averaging integrated for cluster C0_LiverMetM3
Finished averaging integrated for cluster C7_LiverMetM3
Finished averaging integrated for cluster C17_LiverMetM3
Finished averaging integrated for cluster C16_LiverMetM3
Finished averaging integrated for cluster C8_LiverMetM3
Finished averaging integrated for cluster C14_LiverMetM3
Finished averaging integrated for cluster C6_LiverMetM3
Finished averaging integrated for cluster C3_LiverMetM3
Finished averaging integrated for cluster C10_LiverMetM3
Finished averaging integrated for cluster C15_LiverMetM3
Finished averaging integrated for cluster C12_LiverMetM3
Finished averaging integrated for cluster C13_LiverMetM3
Finished averaging integrated for cluster C19_LiverMetM3
Finished averaging integrated for cluster C5_LiverMetM3
Finished averaging integrated for cluster C9_LiverMetM3
Finished averaging integrated for cluster C2_LiverMetM3
Finished averaging integrated for cluster C20_LiverMetM3
Finished averaging integrated for cluster C11_LiverMetM3
Finished averaging integrated for cluster C21_LiverMetM3
Finished averaging integrated for cluster C18_LiverMetM3
Finished averaging integrated for cluster C22_LiverMetM3
Finished averaging integrated for cluster C0_SubcutM3
Finished averaging integrated for cluster C2_SubcutM3
Finished averaging integrated for cluster C9_SubcutM3
Finished averaging integrated for cluster C1_SubcutM3
Finished averaging integrated for cluster C3_SubcutM3
Finished averaging integrated for cluster C15_SubcutM3
Finished averaging integrated for cluster C14_SubcutM3
Finished averaging integrated for cluster C8_SubcutM3
Finished averaging integrated for cluster C12_SubcutM3
Finished averaging integrated for cluster C16_SubcutM3
Finished averaging integrated for cluster C6_SubcutM3
Finished averaging integrated for cluster C4_SubcutM3
Finished averaging integrated for cluster C7_SubcutM3
Finished averaging integrated for cluster C5_SubcutM3
Finished averaging integrated for cluster C10_SubcutM3
Finished averaging integrated for cluster C11_SubcutM3
Finished averaging integrated for cluster C20_SubcutM3
Finished averaging integrated for cluster C19_SubcutM3
Finished averaging integrated for cluster C13_SubcutM3
Finished averaging integrated for cluster C21_SubcutM3
Finished averaging integrated for cluster C22_SubcutM3
Finished averaging integrated for cluster C17_SubcutM3
Finished averaging integrated for cluster C18_SubcutM3
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   6%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |===========                                                           |  16%  |                                                                              |==============                                                        |  19%  |                                                                              |================                                                      |  23%  |                                                                              |==================                                                    |  26%  |                                                                              |====================                                                  |  29%  |                                                                              |=======================                                               |  32%  |                                                                              |=========================                                             |  35%  |                                                                              |===========================                                           |  39%  |                                                                              |=============================                                         |  42%  |                                                                              |================================                                      |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |======================================                                |  55%  |                                                                              |=========================================                             |  58%  |                                                                              |===========================================                           |  61%  |                                                                              |=============================================                         |  65%  |                                                                              |===============================================                       |  68%  |                                                                              |==================================================                    |  71%  |                                                                              |====================================================                  |  74%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  81%  |                                                                              |===========================================================           |  84%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  94%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster C20
Finished averaging RNA for cluster C6
Finished averaging RNA for cluster C10
Finished averaging RNA for cluster C5
Finished averaging RNA for cluster C15
Finished averaging RNA for cluster C0
Finished averaging RNA for cluster C21
Finished averaging RNA for cluster C3
Finished averaging RNA for cluster C19
Finished averaging RNA for cluster C22
Finished averaging RNA for cluster C8
Finished averaging RNA for cluster C2
Finished averaging RNA for cluster C7
Finished averaging RNA for cluster C4
Finished averaging RNA for cluster C9
Finished averaging RNA for cluster C11
Finished averaging RNA for cluster C16
Finished averaging RNA for cluster C14
Finished averaging RNA for cluster C1
Finished averaging RNA for cluster C17
Finished averaging RNA for cluster C18
Finished averaging RNA for cluster C12
Finished averaging RNA for cluster C13
Finished averaging integrated for cluster C20
Finished averaging integrated for cluster C6
Finished averaging integrated for cluster C10
Finished averaging integrated for cluster C5
Finished averaging integrated for cluster C15
Finished averaging integrated for cluster C0
Finished averaging integrated for cluster C21
Finished averaging integrated for cluster C3
Finished averaging integrated for cluster C19
Finished averaging integrated for cluster C22
Finished averaging integrated for cluster C8
Finished averaging integrated for cluster C2
Finished averaging integrated for cluster C7
Finished averaging integrated for cluster C4
Finished averaging integrated for cluster C9
Finished averaging integrated for cluster C11
Finished averaging integrated for cluster C16
Finished averaging integrated for cluster C14
Finished averaging integrated for cluster C1
Finished averaging integrated for cluster C17
Finished averaging integrated for cluster C18
Finished averaging integrated for cluster C12
Finished averaging integrated for cluster C13
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   6%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |===========                                                           |  16%  |                                                                              |==============                                                        |  19%  |                                                                              |================                                                      |  23%  |                                                                              |==================                                                    |  26%  |                                                                              |====================                                                  |  29%  |                                                                              |=======================                                               |  32%  |                                                                              |=========================                                             |  35%  |                                                                              |===========================                                           |  39%  |                                                                              |=============================                                         |  42%  |                                                                              |================================                                      |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |======================================                                |  55%  |                                                                              |=========================================                             |  58%  |                                                                              |===========================================                           |  61%  |                                                                              |=============================================                         |  65%  |                                                                              |===============================================                       |  68%  |                                                                              |==================================================                    |  71%  |                                                                              |====================================================                  |  74%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  81%  |                                                                              |===========================================================           |  84%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  94%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster LiverNormalM1
Finished averaging RNA for cluster SubcutNoLiverMetM2
Finished averaging RNA for cluster BackgroundLiverM3
Finished averaging RNA for cluster LiverMetM3
Finished averaging RNA for cluster SubcutM3
Finished averaging integrated for cluster LiverNormalM1
Finished averaging integrated for cluster SubcutNoLiverMetM2
Finished averaging integrated for cluster BackgroundLiverM3
Finished averaging integrated for cluster LiverMetM3
Finished averaging integrated for cluster SubcutM3
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   6%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |===========                                                           |  16%  |                                                                              |==============                                                        |  19%  |                                                                              |================                                                      |  23%  |                                                                              |==================                                                    |  26%  |                                                                              |====================                                                  |  29%  |                                                                              |=======================                                               |  32%  |                                                                              |=========================                                             |  35%  |                                                                              |===========================                                           |  39%  |                                                                              |=============================                                         |  42%  |                                                                              |================================                                      |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |======================================                                |  55%  |                                                                              |=========================================                             |  58%  |                                                                              |===========================================                           |  61%  |                                                                              |=============================================                         |  65%  |                                                                              |===============================================                       |  68%  |                                                                              |==================================================                    |  71%  |                                                                              |====================================================                  |  74%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  81%  |                                                                              |===========================================================           |  84%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  94%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster LiverNormalM1
Finished averaging RNA for cluster SubcutNoLiverMetM2
Finished averaging RNA for cluster BackgroundLiverM3
Finished averaging RNA for cluster LiverMetM3
Finished averaging RNA for cluster SubcutM3

  |                                                                              |.....................                                                 |  29%
  ordinary text without R code

  |                                                                              |......................                                                |  31%
label: unnamed-chunk-7 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B8.plot.pca.and.tsne.results.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B8.plot.pca.and.tsne.results.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_pca_tsne_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: Plot_dimred_by_cluster_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE

Finished averaging RNA for cluster LiverNormalM1
Finished averaging RNA for cluster SubcutNoLiverMetM2
Finished averaging RNA for cluster BackgroundLiverM3
Finished averaging RNA for cluster LiverMetM3
Finished averaging RNA for cluster SubcutM3
Finished averaging RNA for cluster C20
Finished averaging RNA for cluster C6
Finished averaging RNA for cluster C10
Finished averaging RNA for cluster C5
Finished averaging RNA for cluster C15
Finished averaging RNA for cluster C0
Finished averaging RNA for cluster C21
Finished averaging RNA for cluster C3
Finished averaging RNA for cluster C19
Finished averaging RNA for cluster C22
Finished averaging RNA for cluster C8
Finished averaging RNA for cluster C2
Finished averaging RNA for cluster C7
Finished averaging RNA for cluster C4
Finished averaging RNA for cluster C9
Finished averaging RNA for cluster C11
Finished averaging RNA for cluster C16
Finished averaging RNA for cluster C14
Finished averaging RNA for cluster C1
Finished averaging RNA for cluster C17
Finished averaging RNA for cluster C18
Finished averaging RNA for cluster C12
Finished averaging RNA for cluster C13
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Plot_dim_red_by_cluster (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.......................                                               |  33%
  ordinary text without R code

  |                                                                              |.........................                                             |  35%
label: unnamed-chunk-8 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B9.euclidian.distances.figure.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B9.euclidian.distances.figure.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..............                                                        |  20%
  ordinary text without R code

  |                                                                              |............................                                          |  40%
label: partB_euclidian_distances (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................................                            |  60%
  ordinary text without R code

  |                                                                              |........................................................              |  80%
label: Plot_umap_sleepwalk (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

Estimating 'maxdist' for feature matrix 1
  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |..........................                                            |  37%
  ordinary text without R code

  |                                                                              |...........................                                           |  39%
label: unnamed-chunk-9 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B10.doublet.estimation.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B10.doublet.estimation.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_doublet_estimates (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: doublet-data (with options) 
List of 5
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE
 $ message: logi FALSE

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: doublet-plot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.............................                                         |  41%
  ordinary text without R code

  |                                                                              |..............................                                        |  43%
label: unnamed-chunk-10 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B11.cell.cycle.phase.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B11.cell.cycle.phase.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_cell_cycle_phase (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: Plot_dimred_by_cell_cycle_phase_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE

Warning: The following features are not present in the object: MCM5, PCNA, TYMS, FEN1, MCM2, MCM4, RRM1, UNG, GINS2, MCM6, CDCA7, DTL, PRIM1, UHRF1, MLF1IP, HELLS, RFC2, RPA2, NASP, RAD51AP1, GMNN, WDR76, SLBP, CCNE2, UBR7, POLD3, MSH2, ATAD2, RAD51, RRM2, CDC45, CDC6, EXO1, TIPIN, DSCC1, BLM, CASP8AP2, USP1, CLSPN, POLA1, CHAF1B, BRIP1, E2F8, not searching for symbol synonyms
Warning: The following features are not present in the object: HMGB2, CDK1, NUSAP1, UBE2C, BIRC5, TPX2, TOP2A, NDC80, CKS2, NUF2, CKS1B, MKI67, TMPO, CENPF, TACC3, FAM64A, SMC4, CCNB2, CKAP2L, CKAP2, AURKB, BUB1, KIF11, ANP32E, TUBB4B, GTSE1, KIF20B, HJURP, CDCA3, HN1, CDC20, TTK, CDC25C, KIF2C, RANGAP1, NCAPD2, DLGAP5, CDCA2, CDCA8, ECT2, KIF23, HMMR, AURKA, PSRC1, ANLN, LBR, CKAP5, CENPE, CTCF, NEK2, G2E3, GAS2L3, CBX5, CENPA, not searching for symbol synonyms
Finished averaging RNA for cluster LiverNormalM1
Finished averaging RNA for cluster SubcutNoLiverMetM2
Finished averaging RNA for cluster BackgroundLiverM3
Finished averaging RNA for cluster LiverMetM3
Finished averaging RNA for cluster SubcutM3
Finished averaging RNA for cluster C20
Finished averaging RNA for cluster C6
Finished averaging RNA for cluster C10
Finished averaging RNA for cluster C5
Finished averaging RNA for cluster C15
Finished averaging RNA for cluster C0
Finished averaging RNA for cluster C21
Finished averaging RNA for cluster C3
Finished averaging RNA for cluster C19
Finished averaging RNA for cluster C22
Finished averaging RNA for cluster C8
Finished averaging RNA for cluster C2
Finished averaging RNA for cluster C7
Finished averaging RNA for cluster C4
Finished averaging RNA for cluster C9
Finished averaging RNA for cluster C11
Finished averaging RNA for cluster C16
Finished averaging RNA for cluster C14
Finished averaging RNA for cluster C1
Finished averaging RNA for cluster C17
Finished averaging RNA for cluster C18
Finished averaging RNA for cluster C12
Finished averaging RNA for cluster C13
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Plot_dim_red_by_cell_cycle_phase (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |................................                                      |  45%
  ordinary text without R code

  |                                                                              |.................................                                     |  47%
label: unnamed-chunk-11 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B12.Ncell.barchart.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B12.Ncell.barchart.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_N_cell_barchart (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: make_percent_N_cells_barchart (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: plot_barchart (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |..................................                                    |  49%
  ordinary text without R code

  |                                                                              |....................................                                  |  51%
label: unnamed-chunk-12 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B13.AUC.cat.enrichments.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B13.AUC.cat.enrichments.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_AUC_cat_enrichment (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: AUC_prep_from_file (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Quantiles for the number of genes detected by cell: 
(Non-detected genes are shuffled at the end of the ranking. Keep it in mind when choosing the threshold for calculating the AUC).
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: enrichment_plots (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.....................................                                 |  53%
  ordinary text without R code

  |                                                                              |......................................                                |  55%
label: unnamed-chunk-13 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B14.linear.regression.data.prep.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B14.linear.regression.data.prep.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_linear_regression (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: data_pre_linear_regression (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: plot_linear_regression (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |........................................                              |  57%
  ordinary text without R code

  |                                                                              |.........................................                             |  59%
label: unnamed-chunk-14 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B15.heatmaps.and.dotplots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B15.heatmaps.and.dotplots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_heatmaps_and_dotplots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: Cluster_overview_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Cluster_overview (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |...........................................                           |  61%
  ordinary text without R code

  |                                                                              |............................................                          |  63%
label: unnamed-chunk-15 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B16.make.cluster.table.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B16.make.cluster.table.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_make_cluster_table (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: ClusterTable_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


Attaching package: 'DT'

The following object is masked from 'package:Seurat':

    JS

  |                                                                              |..........................................................            |  83%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: render_ClusterTable (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.............................................                         |  65%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: unnamed-chunk-16 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B17.cell.type.estimate.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B17.cell.type.estimate.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_cell_type_estimate (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: CellTypePlot_dataprep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F


Registered S3 method overwritten by 'enrichplot':
  method               from
  fortify.enrichResult DOSE
clusterProfiler v3.12.0  For help: https://guangchuangyu.github.io/software/clusterProfiler

If you use clusterProfiler in published research, please cite:
Guangchuang Yu, Li-Gen Wang, Yanyan Han, Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology. 2012, 16(5):284-287.

Attaching package: 'clusterProfiler'

The following object is masked from 'package:purrr':

    simplify

Loading required package: GSEABase
Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: 'BiocGenerics'

The following objects are masked from 'package:parallel':

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from 'package:dplyr':

    combine, intersect, setdiff, union

The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs

The following objects are masked from 'package:base':

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which, which.max, which.min

Loading required package: Biobase
Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.

Loading required package: annotate
Loading required package: AnnotationDbi
Loading required package: stats4
Loading required package: IRanges
Loading required package: S4Vectors

Attaching package: 'S4Vectors'

The following object is masked from 'package:ggtree':

    expand

The following objects are masked from 'package:dplyr':

    first, rename

The following object is masked from 'package:tidyr':

    expand

The following object is masked from 'package:base':

    expand.grid


Attaching package: 'IRanges'

The following object is masked from 'package:ggtree':

    collapse

The following objects are masked from 'package:dplyr':

    collapse, desc, slice

The following object is masked from 'package:purrr':

    reduce


Attaching package: 'AnnotationDbi'

The following object is masked from 'package:dplyr':

    select

Loading required package: XML
Loading required package: graph

Attaching package: 'graph'

The following object is masked from 'package:XML':

    addNode

The following object is masked from 'package:stringr':

    boundary

--> No gene can be mapped....
--> Expected input gene ID: Mfsd7c,Cd69,Mgl2,Cxcl10,Hhex,LOC100043921
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Hdc,Fes,Cd38,Tmem156,Ntrk1,Etv3
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Paqr5,Hdc,Gm2457,Il18r1,Map3k13,St8sia1
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Cd37,Ccl20,Il18r1,Pla2g7,Slc2a6,Ms4a1
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Treml2,Fcgr2b,Ly86,Bank1,Igsf6,Adam28
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Ear11,Fcer1a,Ncf2,Il1rl1,Tec,Tmem255a
--> return NULL...
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: CellTypePlot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |................................................                      |  69%
  ordinary text without R code

  |                                                                              |.................................................                     |  71%
label: unnamed-chunk-17 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B18.gene.set.heatmaps.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B18.gene.set.heatmaps.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_gene_set_heatmaps (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: MakeHM (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |..........................................................            |  83%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: calculate-markdown-heatmap-chunks-dynamically (with options) 
List of 4
 $ echo   : symbol F
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |...................................................                   |  73%
  ordinary text without R code

  |                                                                              |....................................................                  |  75%
label: unnamed-chunk-18 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B19.example.feature.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B19.example.feature.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_example_feature_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: ExampleFeaturePlot_dataprep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |..........................................................            |  83%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: ExampleFeaturePlot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |......................................................                |  76%
  ordinary text without R code

  |                                                                              |.......................................................               |  78%
label: unnamed-chunk-19 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B20.pca.cell.and.loading.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B20.pca.cell.and.loading.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |....                                                                  |   6%
  ordinary text without R code

  |                                                                              |.........                                                             |  12%
label: partB_pca_loadings (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |.............                                                         |  19%
  ordinary text without R code

  |                                                                              |..................                                                    |  25%
label: elbow_plot (with options) 
List of 3
 $ fig.height: num 6
 $ fig.width : num 10
 $ fig.cap   : language paste0("**Figure, ", figureCount, ": Elbowplot** Variance explained per PCA dimension")

  |                                                                              |......................                                                |  31%
  ordinary text without R code

  |                                                                              |..........................                                            |  38%
label: PCA_dataprep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F


Attaching package: 'ggpubr'

The following object is masked from 'package:ggtree':

    rotate


Attaching package: 'gridExtra'

The following object is masked from 'package:Biobase':

    combine

The following object is masked from 'package:BiocGenerics':

    combine

The following object is masked from 'package:dplyr':

    combine

  |                                                                              |...............................                                       |  44%
   inline R code fragments

  |                                                                              |...................................                                   |  50%
label: plot_PCADimPlots (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |.......................................                               |  56%
  ordinary text without R code

  |                                                                              |............................................                          |  62%
label: pca_boxplot_dataprep (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

  |                                                                              |................................................                      |  69%
   inline R code fragments

  |                                                                              |....................................................                  |  75%
label: pca_boxplot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |.........................................................             |  81%
  ordinary text without R code

  |                                                                              |.............................................................         |  88%
label: create-pca-enrichment-data (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

--> No gene can be mapped....
--> Expected input gene ID: Prg2,Aim2,Stap1,Pkd2l2,Gm2457,Ccl2
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: P2ry14,St3gal6,Gpr18,Bcl7a,Treml2,Siglec1
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Ntrk1,Ccl17,Ccl17,Cpa3,Il1rl1,Hpgds
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Eaf2,Il5ra,Ptgir,Ccl22,C3ar1,P2rx5
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Treml2,Rab27b,Ccl5,Rgs13,Rgs1,Ugt1a10
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: Ms4a1,Hoxa1,Il5,Mmp12,Tlr8,Cxcr5
--> return NULL...
  |                                                                              |..................................................................    |  94%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: create-pca-enrichment-plot (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |........................................................              |  80%
  ordinary text without R code

  |                                                                              |..........................................................            |  82%
label: unnamed-chunk-20 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B21.pseudotime.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B21.pseudotime.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_pseudotime (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: pseudotime-enrichment-data (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: pseudotime-plot (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |...........................................................           |  84%
  ordinary text without R code

  |                                                                              |............................................................          |  86%
label: unnamed-chunk-21 (with options) 
List of 2
 $ child: chr "src/modules/section_B/B22.add.extra.tables.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/section_B/B22.add.extra.tables.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..................                                                    |  25%
  ordinary text without R code

  |                                                                              |...................................                                   |  50%
label: partB_add_extra_tables (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |....................................................                  |  75%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: Additional tables (with options) 
List of 4
 $ echo   : symbol F
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F


  |                                                                              |..............................................................        |  88%
  ordinary text without R code

  |                                                                              |...............................................................       |  90%
label: saveobject (with options) 
List of 3
 $ eval   : logi TRUE
 $ echo   : symbol T
 $ results: symbol F

  |                                                                              |.................................................................     |  92%
  ordinary text without R code

  |                                                                              |..................................................................    |  94%
label: create_report_params (with options) 
List of 2
 $ eval   : symbol T
 $ results: chr "asis"

  |                                                                              |...................................................................   |  96%
  ordinary text without R code

  |                                                                              |..................................................................... |  98%
label: documentation (with options) 
List of 3
 $ eval   : logi TRUE
 $ echo   : symbol T
 $ results: symbol T

  |                                                                              |......................................................................| 100%
   inline R code fragments


output file: sc_PartB_Analysis.knit.md

/camp/apps/eb/dev/software/Anaconda2/2019.10/bin/pandoc +RTS -K512m -RTS sc_PartB_Analysis.utf8.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /camp/stp/babs/working/boeings/Projects/sahaie/rebecca.lee/461_mouse_liver_single_cell_SC20154/workdir/html_local/sc_PartB_Analysis.html --lua-filter /camp/apps/misc/stp/babs/manual/software/r/r-site-library/bioc-3.9/rmarkdown/rmarkdown/lua/pagebreak.lua --lua-filter /camp/apps/misc/stp/babs/manual/software/r/r-site-library/bioc-3.9/rmarkdown/rmarkdown/lua/latex-div.lua --self-contained --variable bs3=TRUE --standalone --section-divs --table-of-contents --toc-depth 5 --variable toc_float=1 --variable toc_selectors=h1,h2,h3,h4,h5 --variable toc_collapsed=1 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /camp/apps/misc/stp/babs/manual/software/r/r-site-library/bioc-3.9/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --css src/assets/style/style.css --variable theme=paper --include-in-header /tmp/RtmpSLSyWu/rmarkdown-str8ad03ab5ac32.html --mathjax --variable 'mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --variable code_folding=hide --variable code_menu=1 

Output created: /camp/stp/babs/working/boeings/Projects/sahaie/rebecca.lee/461_mouse_liver_single_cell_SC20154/workdir/html_local/sc_PartB_Analysis.html
There were 24 warnings (use warnings() to see them)
