

processing file: sc_PartB_Analysis.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |.                                                                     |   2%
  ordinary text without R code

  |                                                                              |...                                                                   |   4%
label: setup (with options) 
List of 1
 $ include: logi FALSE

  |                                                                              |....                                                                  |   6%
  ordinary text without R code

  |                                                                              |......                                                                |   8%
label: unnamed-chunk-1 (with options) 
List of 2
 $ child: chr "src/modules/settings/partB.set.parameters.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/settings/partB.set.parameters.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..........                                                            |  14%
  ordinary text without R code

  |                                                                              |....................                                                  |  29%
label: set_parameters_init (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..............................                                        |  43%
  ordinary text without R code

  |                                                                              |........................................                              |  57%
label: B_set_directories (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |..................................................                    |  71%
  ordinary text without R code

  |                                                                              |............................................................          |  86%
label: next_block (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

Loading required package: DBI
  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |.......                                                               |  10%
  ordinary text without R code

  |                                                                              |.........                                                             |  12%
label: unnamed-chunk-2 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.create.integrated.seurat.object.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.create.integrated.seurat.object.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..............                                                        |  20%
  ordinary text without R code

  |                                                                              |............................                                          |  40%
label: partB_create_integrated_Seurat_object_init (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................................                            |  60%
  ordinary text without R code

  |                                                                              |........................................................              |  80%
label: partB_create_sample_list (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Computing 2000 integration features
Scaling features for provided objects
Finding all pairwise anchors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 250 anchors
Filtering anchors
	Retained 163 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 205 anchors
Filtering anchors
	Retained 203 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 220 anchors
Filtering anchors
	Retained 220 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 250 anchors
Filtering anchors
	Retained 100 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 563 anchors
Filtering anchors
	Retained 212 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 220 anchors
Filtering anchors
	Retained 101 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 250 anchors
Filtering anchors
	Retained 137 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 524 anchors
Filtering anchors
	Retained 274 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 220 anchors
Filtering anchors
	Retained 125 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 704 anchors
Filtering anchors
	Retained 344 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 250 anchors
Filtering anchors
	Retained 139 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 539 anchors
Filtering anchors
	Retained 259 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 220 anchors
Filtering anchors
	Retained 133 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 806 anchors
Filtering anchors
	Retained 398 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 643 anchors
Filtering anchors
	Retained 335 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 250 anchors
Filtering anchors
	Retained 111 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 548 anchors
Filtering anchors
	Retained 221 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 220 anchors
Filtering anchors
	Retained 104 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 923 anchors
Filtering anchors
	Retained 366 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 678 anchors
Filtering anchors
	Retained 261 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 749 anchors
Filtering anchors
	Retained 303 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 250 anchors
Filtering anchors
	Retained 148 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 492 anchors
Filtering anchors
	Retained 281 anchors
Extracting within-dataset neighbors
Running CCA
Warning in irlba(A = mat3, nv = num.cc) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Merging objects
Finding neighborhoods
Finding anchors
	Found 220 anchors
Filtering anchors
	Retained 141 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 637 anchors
Filtering anchors
	Retained 354 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 561 anchors
Filtering anchors
	Retained 318 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 582 anchors
Filtering anchors
	Retained 332 anchors
Extracting within-dataset neighbors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 624 anchors
Filtering anchors
	Retained 362 anchors
Extracting within-dataset neighbors
Merging dataset 3 into 2
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 1 into 2 3
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 8 into 7
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 5 into 2 3 1
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 6 into 4
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 2 3 1 5 into 4 6
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Merging dataset 7 8 into 4 6 2 3 1 5
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
Warning: Adding a command log without an assay associated with it
Warning: The following features are not present in the object: ATAD2, BLM, BRIP1, CASP8AP2, CCNE2, CDC45, CDC6, CDCA7, CHAF1B, CLSPN, DSCC1, DTL, E2F8, EXO1, FEN1, GINS2, GMNN, HELLS, MCM2, MCM4, MCM5, MCM6, MLF1IP, MSH2, POLA1, POLD3, PRIM1, RAD51, RAD51AP1, RFC2, RPA2, RRM1, RRM2, SLBP, TIPIN, TYMS, UBR7, UHRF1, UNG, WDR76, not searching for symbol synonyms
Warning: The following features are not present in the object: ANLN, AURKA, AURKB, BIRC5, BUB1, CCNB2, CDC25C, CDCA2, CDCA3, CDCA8, CDK1, CENPA, CENPE, CENPF, CKAP2, CKAP2L, CKAP5, CTCF, DLGAP5, ECT2, FAM64A, G2E3, GAS2L3, GTSE1, HJURP, HMMR, KIF11, KIF20B, KIF23, KIF2C, LBR, MKI67, NCAPD2, NDC80, NEK2, NUF2, NUSAP1, PSRC1, RANGAP1, TACC3, TMPO, TOP2A, TPX2, TTK, UBE2C, not searching for symbol synonyms
  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |..........                                                            |  14%
  ordinary text without R code

  |                                                                              |...........                                                           |  16%
label: unnamed-chunk-3 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.create.variation.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.create.variation.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..........                                                            |  14%
  ordinary text without R code

  |                                                                              |....................                                                  |  29%
label: partB_create_variation_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..............................                                        |  43%
  ordinary text without R code

  |                                                                              |........................................                              |  57%
label: partB_create-individual-varFeatures (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster all
Finished averaging integrated for cluster all
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   5%  |                                                                              |=======                                                               |  10%  |                                                                              |==========                                                            |  15%  |                                                                              |==============                                                        |  20%  |                                                                              |==================                                                    |  25%  |                                                                              |=====================                                                 |  30%  |                                                                              |========================                                              |  35%  |                                                                              |============================                                          |  40%  |                                                                              |================================                                      |  45%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================                                |  55%  |                                                                              |==========================================                            |  60%  |                                                                              |==============================================                        |  65%  |                                                                              |=================================================                     |  70%  |                                                                              |====================================================                  |  75%  |                                                                              |========================================================              |  80%  |                                                                              |============================================================          |  85%  |                                                                              |===============================================================       |  90%  |                                                                              |==================================================================    |  95%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster P01Fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |======                                                                |   8%  |                                                                              |============                                                          |  17%  |                                                                              |==================                                                    |  25%  |                                                                              |=======================                                               |  33%  |                                                                              |=============================                                         |  42%  |                                                                              |===================================                                   |  50%  |                                                                              |=========================================                             |  58%  |                                                                              |===============================================                       |  67%  |                                                                              |====================================================                  |  75%  |                                                                              |==========================================================            |  83%  |                                                                              |================================================================      |  92%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster P03Fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |=========                                                             |  13%  |                                                                              |==============                                                        |  20%  |                                                                              |===================                                                   |  27%  |                                                                              |=======================                                               |  33%  |                                                                              |============================                                          |  40%  |                                                                              |=================================                                     |  47%  |                                                                              |=====================================                                 |  53%  |                                                                              |==========================================                            |  60%  |                                                                              |===============================================                       |  67%  |                                                                              |===================================================                   |  73%  |                                                                              |========================================================              |  80%  |                                                                              |=============================================================         |  87%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster P05Fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   8%  |                                                                              |===========                                                           |  15%  |                                                                              |================                                                      |  23%  |                                                                              |======================                                                |  31%  |                                                                              |===========================                                           |  38%  |                                                                              |================================                                      |  46%  |                                                                              |======================================                                |  54%  |                                                                              |===========================================                           |  62%  |                                                                              |================================================                      |  69%  |                                                                              |======================================================                |  77%  |                                                                              |===========================================================           |  85%  |                                                                              |=================================================================     |  92%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster P06Fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   6%  |                                                                              |========                                                              |  11%  |                                                                              |============                                                          |  17%  |                                                                              |================                                                      |  22%  |                                                                              |===================                                                   |  28%  |                                                                              |=======================                                               |  33%  |                                                                              |===========================                                           |  39%  |                                                                              |===============================                                       |  44%  |                                                                              |===================================                                   |  50%  |                                                                              |=======================================                               |  56%  |                                                                              |===========================================                           |  61%  |                                                                              |===============================================                       |  67%  |                                                                              |===================================================                   |  72%  |                                                                              |======================================================                |  78%  |                                                                              |==========================================================            |  83%  |                                                                              |==============================================================        |  89%  |                                                                              |==================================================================    |  94%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster P07Fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   6%  |                                                                              |=========                                                             |  12%  |                                                                              |=============                                                         |  19%  |                                                                              |==================                                                    |  25%  |                                                                              |======================                                                |  31%  |                                                                              |==========================                                            |  38%  |                                                                              |===============================                                       |  44%  |                                                                              |===================================                                   |  50%  |                                                                              |=======================================                               |  56%  |                                                                              |============================================                          |  62%  |                                                                              |================================================                      |  69%  |                                                                              |====================================================                  |  75%  |                                                                              |=========================================================             |  81%  |                                                                              |=============================================================         |  88%  |                                                                              |==================================================================    |  94%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster P08Fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   6%  |                                                                              |========                                                              |  12%  |                                                                              |============                                                          |  18%  |                                                                              |================                                                      |  24%  |                                                                              |=====================                                                 |  29%  |                                                                              |=========================                                             |  35%  |                                                                              |=============================                                         |  41%  |                                                                              |=================================                                     |  47%  |                                                                              |=====================================                                 |  53%  |                                                                              |=========================================                             |  59%  |                                                                              |=============================================                         |  65%  |                                                                              |=================================================                     |  71%  |                                                                              |======================================================                |  76%  |                                                                              |==========================================================            |  82%  |                                                                              |==============================================================        |  88%  |                                                                              |==================================================================    |  94%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster P09Fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   6%  |                                                                              |=========                                                             |  12%  |                                                                              |=============                                                         |  19%  |                                                                              |==================                                                    |  25%  |                                                                              |======================                                                |  31%  |                                                                              |==========================                                            |  38%  |                                                                              |===============================                                       |  44%  |                                                                              |===================================                                   |  50%  |                                                                              |=======================================                               |  56%  |                                                                              |============================================                          |  62%  |                                                                              |================================================                      |  69%  |                                                                              |====================================================                  |  75%  |                                                                              |=========================================================             |  81%  |                                                                              |=============================================================         |  88%  |                                                                              |==================================================================    |  94%  |                                                                              |======================================================================| 100%
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finished averaging RNA for cluster P10Fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |=========                                                             |  13%  |                                                                              |==============                                                        |  20%  |                                                                              |===================                                                   |  27%  |                                                                              |=======================                                               |  33%  |                                                                              |============================                                          |  40%  |                                                                              |=================================                                     |  47%  |                                                                              |=====================================                                 |  53%  |                                                                              |==========================================                            |  60%  |                                                                              |===============================================                       |  67%  |                                                                              |===================================================                   |  73%  |                                                                              |========================================================              |  80%  |                                                                              |=============================================================         |  87%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%
  |                                                                              |..................................................                    |  71%
   inline R code fragments

  |                                                                              |............................................................          |  86%
label: create-var-feature-plot-II (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |.............                                                         |  18%
  ordinary text without R code

  |                                                                              |..............                                                        |  20%
label: unnamed-chunk-4 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.create.pca.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.create.pca.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |.........                                                             |  12%
  ordinary text without R code

  |                                                                              |..................                                                    |  25%
label: partB_pca_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................                                            |  38%
  ordinary text without R code

  |                                                                              |...................................                                   |  50%
label: partB_create_PCA_plots (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session
11:59:22 UMAP embedding parameters a = 0.9922 b = 1.112
11:59:22 Read 1133 rows and found 30 numeric columns
11:59:22 Using Annoy for neighbor search, n_neighbors = 30
11:59:22 Building Annoy index with metric = cosine, n_trees = 50
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
11:59:22 Writing NN index file to temp file /tmp/slurm_25137365/RtmpEt7lA6/filea3fd49382c18
11:59:22 Searching Annoy index using 1 thread, search_k = 3000
11:59:22 Annoy recall = 100%
11:59:23 Commencing smooth kNN distance calibration using 1 thread
11:59:25 Initializing from normalized Laplacian + noise
11:59:25 Commencing optimization for 500 epochs, with 52346 positive edges
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
11:59:30 Optimization finished
Computing nearest neighbor graph
Computing SNN
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|

Attaching package: 'scales'

The following object is masked from 'package:purrr':

    discard

The following object is masked from 'package:readr':

    col_factor

Calculating cluster C0
Calculating cluster C1
Calculating cluster C2
  |                                                                              |............................................                          |  62%
  ordinary text without R code

  |                                                                              |....................................................                  |  75%
label: data_prep_sample_UMAP_rendering (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE

Registered S3 method overwritten by 'treeio':
  method     from
  root.phylo ape 
ggtree v1.16.6  For help: https://yulab-smu.github.io/treedata-book/

If you use ggtree in published research, please cite the most appropriate paper(s):

[36m-[39m Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods for mapping and visualizing associated data on phylogeny using ggtree. Molecular Biology and Evolution 2018, 35(12):3041-3043. doi: 10.1093/molbev/msy194
[36m-[39m Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution 2017, 8(1):28-36, doi:10.1111/2041-210X.12628



Attaching package: 'ggtree'

The following object is masked from 'package:IRanges':

    collapse

The following object is masked from 'package:S4Vectors':

    expand

The following object is masked from 'package:tidyr':

    expand

Finished averaging RNA for cluster P01Fib
Finished averaging RNA for cluster P03Fib
Finished averaging RNA for cluster P05Fib
Finished averaging RNA for cluster P06Fib
Finished averaging RNA for cluster P07Fib
Finished averaging RNA for cluster P08Fib
Finished averaging RNA for cluster P09Fib
Finished averaging RNA for cluster P10Fib
  |                                                                              |.............................................................         |  88%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Plot_tsne_data_plotting (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |................                                                      |  22%
  ordinary text without R code

  |                                                                              |.................                                                     |  24%
label: unnamed-chunk-5 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.calculate.average.expression.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.calculate.average.expression.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..................                                                    |  25%
  ordinary text without R code

  |                                                                              |...................................                                   |  50%
label: partB_calculate_avg_expr (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |....................................................                  |  75%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: Identify differential expressed genes across conditions (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

Finished averaging RNA for cluster all
Finished averaging integrated for cluster all
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   5%  |                                                                              |=======                                                               |  10%  |                                                                              |==========                                                            |  15%  |                                                                              |==============                                                        |  20%  |                                                                              |==================                                                    |  25%  |                                                                              |=====================                                                 |  30%  |                                                                              |========================                                              |  35%  |                                                                              |============================                                          |  40%  |                                                                              |================================                                      |  45%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================                                |  55%  |                                                                              |==========================================                            |  60%  |                                                                              |==============================================                        |  65%  |                                                                              |=================================================                     |  70%  |                                                                              |====================================================                  |  75%  |                                                                              |========================================================              |  80%  |                                                                              |============================================================          |  85%  |                                                                              |===============================================================       |  90%  |                                                                              |==================================================================    |  95%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster C0_P01Fib
Finished averaging RNA for cluster C1_P01Fib
Finished averaging RNA for cluster C2_P01Fib
Finished averaging RNA for cluster C1_P03Fib
Finished averaging RNA for cluster C0_P03Fib
Finished averaging RNA for cluster C2_P03Fib
Finished averaging RNA for cluster C0_P05Fib
Finished averaging RNA for cluster C1_P05Fib
Finished averaging RNA for cluster C2_P05Fib
Finished averaging RNA for cluster C0_P06Fib
Finished averaging RNA for cluster C1_P06Fib
Finished averaging RNA for cluster C2_P06Fib
Finished averaging RNA for cluster C0_P07Fib
Finished averaging RNA for cluster C1_P07Fib
Finished averaging RNA for cluster C2_P07Fib
Finished averaging RNA for cluster C1_P08Fib
Finished averaging RNA for cluster C0_P08Fib
Finished averaging RNA for cluster C2_P08Fib
Finished averaging RNA for cluster C0_P09Fib
Finished averaging RNA for cluster C1_P09Fib
Finished averaging RNA for cluster C2_P09Fib
Finished averaging RNA for cluster C0_P10Fib
Finished averaging RNA for cluster C1_P10Fib
Finished averaging RNA for cluster C2_P10Fib
Finished averaging integrated for cluster C0_P01Fib
Finished averaging integrated for cluster C1_P01Fib
Finished averaging integrated for cluster C2_P01Fib
Finished averaging integrated for cluster C1_P03Fib
Finished averaging integrated for cluster C0_P03Fib
Finished averaging integrated for cluster C2_P03Fib
Finished averaging integrated for cluster C0_P05Fib
Finished averaging integrated for cluster C1_P05Fib
Finished averaging integrated for cluster C2_P05Fib
Finished averaging integrated for cluster C0_P06Fib
Finished averaging integrated for cluster C1_P06Fib
Finished averaging integrated for cluster C2_P06Fib
Finished averaging integrated for cluster C0_P07Fib
Finished averaging integrated for cluster C1_P07Fib
Finished averaging integrated for cluster C2_P07Fib
Finished averaging integrated for cluster C1_P08Fib
Finished averaging integrated for cluster C0_P08Fib
Finished averaging integrated for cluster C2_P08Fib
Finished averaging integrated for cluster C0_P09Fib
Finished averaging integrated for cluster C1_P09Fib
Finished averaging integrated for cluster C2_P09Fib
Finished averaging integrated for cluster C0_P10Fib
Finished averaging integrated for cluster C1_P10Fib
Finished averaging integrated for cluster C2_P10Fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   5%  |                                                                              |=======                                                               |  10%  |                                                                              |==========                                                            |  15%  |                                                                              |==============                                                        |  20%  |                                                                              |==================                                                    |  25%  |                                                                              |=====================                                                 |  30%  |                                                                              |========================                                              |  35%  |                                                                              |============================                                          |  40%  |                                                                              |================================                                      |  45%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================                                |  55%  |                                                                              |==========================================                            |  60%  |                                                                              |==============================================                        |  65%  |                                                                              |=================================================                     |  70%  |                                                                              |====================================================                  |  75%  |                                                                              |========================================================              |  80%  |                                                                              |============================================================          |  85%  |                                                                              |===============================================================       |  90%  |                                                                              |==================================================================    |  95%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster C0
Finished averaging RNA for cluster C1
Finished averaging RNA for cluster C2
Finished averaging integrated for cluster C0
Finished averaging integrated for cluster C1
Finished averaging integrated for cluster C2
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   5%  |                                                                              |=======                                                               |  10%  |                                                                              |==========                                                            |  15%  |                                                                              |==============                                                        |  20%  |                                                                              |==================                                                    |  25%  |                                                                              |=====================                                                 |  30%  |                                                                              |========================                                              |  35%  |                                                                              |============================                                          |  40%  |                                                                              |================================                                      |  45%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================                                |  55%  |                                                                              |==========================================                            |  60%  |                                                                              |==============================================                        |  65%  |                                                                              |=================================================                     |  70%  |                                                                              |====================================================                  |  75%  |                                                                              |========================================================              |  80%  |                                                                              |============================================================          |  85%  |                                                                              |===============================================================       |  90%  |                                                                              |==================================================================    |  95%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster P01Fib
Finished averaging RNA for cluster P03Fib
Finished averaging RNA for cluster P05Fib
Finished averaging RNA for cluster P06Fib
Finished averaging RNA for cluster P07Fib
Finished averaging RNA for cluster P08Fib
Finished averaging RNA for cluster P09Fib
Finished averaging RNA for cluster P10Fib
Finished averaging integrated for cluster P01Fib
Finished averaging integrated for cluster P03Fib
Finished averaging integrated for cluster P05Fib
Finished averaging integrated for cluster P06Fib
Finished averaging integrated for cluster P07Fib
Finished averaging integrated for cluster P08Fib
Finished averaging integrated for cluster P09Fib
Finished averaging integrated for cluster P10Fib
Performing log-normalization
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Centering and scaling data matrix
  |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   5%  |                                                                              |=======                                                               |  10%  |                                                                              |==========                                                            |  15%  |                                                                              |==============                                                        |  20%  |                                                                              |==================                                                    |  25%  |                                                                              |=====================                                                 |  30%  |                                                                              |========================                                              |  35%  |                                                                              |============================                                          |  40%  |                                                                              |================================                                      |  45%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================                                |  55%  |                                                                              |==========================================                            |  60%  |                                                                              |==============================================                        |  65%  |                                                                              |=================================================                     |  70%  |                                                                              |====================================================                  |  75%  |                                                                              |========================================================              |  80%  |                                                                              |============================================================          |  85%  |                                                                              |===============================================================       |  90%  |                                                                              |==================================================================    |  95%  |                                                                              |======================================================================| 100%
Finished averaging RNA for cluster P01Fib
Finished averaging RNA for cluster P03Fib
Finished averaging RNA for cluster P05Fib
Finished averaging RNA for cluster P06Fib
Finished averaging RNA for cluster P07Fib
Finished averaging RNA for cluster P08Fib
Finished averaging RNA for cluster P09Fib
Finished averaging RNA for cluster P10Fib

  |                                                                              |...................                                                   |  27%
  ordinary text without R code

  |                                                                              |....................                                                  |  29%
label: unnamed-chunk-6 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.plot.pca.and.tsne.results.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.plot.pca.and.tsne.results.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_pca_tsne_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: Plot_dimred_by_cluster_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE

Finished averaging RNA for cluster P01Fib
Finished averaging RNA for cluster P03Fib
Finished averaging RNA for cluster P05Fib
Finished averaging RNA for cluster P06Fib
Finished averaging RNA for cluster P07Fib
Finished averaging RNA for cluster P08Fib
Finished averaging RNA for cluster P09Fib
Finished averaging RNA for cluster P10Fib
Finished averaging RNA for cluster C0
Finished averaging RNA for cluster C1
Finished averaging RNA for cluster C2
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Plot_dim_red_by_cluster (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.....................                                                 |  31%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: unnamed-chunk-7 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.euclidian.distances.figure.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.euclidian.distances.figure.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..............                                                        |  20%
  ordinary text without R code

  |                                                                              |............................                                          |  40%
label: partB_euclidian_distances (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |..........................................                            |  60%
  ordinary text without R code

  |                                                                              |........................................................              |  80%
label: Plot_umap_sleepwalk (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

Estimating 'maxdist' for feature matrix 1
  |                                                                              |......................................................................| 100%
  ordinary text without R code


  |                                                                              |........................                                              |  35%
  ordinary text without R code

  |                                                                              |..........................                                            |  37%
label: unnamed-chunk-8 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.doublet.estimation.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.doublet.estimation.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_doublet_estimates (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: doublet-data (with options) 
List of 5
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE
 $ message: logi FALSE

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: doublet-plot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |...........................                                           |  39%
  ordinary text without R code

  |                                                                              |.............................                                         |  41%
label: unnamed-chunk-9 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/cell.cycle.phase.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/cell.cycle.phase.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_cell_cycle_phase (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: Plot_dimred_by_cell_cycle_phase_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: logi FALSE

Warning: The following features are not present in the object: UHRF1, MLF1IP, CASP8AP2, not searching for symbol synonyms
Warning: The following features are not present in the object: FAM64A, HJURP, CDC25C, DLGAP5, CDCA2, not searching for symbol synonyms
Finished averaging RNA for cluster P01Fib
Finished averaging RNA for cluster P03Fib
Finished averaging RNA for cluster P05Fib
Finished averaging RNA for cluster P06Fib
Finished averaging RNA for cluster P07Fib
Finished averaging RNA for cluster P08Fib
Finished averaging RNA for cluster P09Fib
Finished averaging RNA for cluster P10Fib
Finished averaging RNA for cluster C0
Finished averaging RNA for cluster C1
Finished averaging RNA for cluster C2
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Plot_dim_red_by_cell_cycle_phase (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |..............................                                        |  43%
  ordinary text without R code

  |                                                                              |...............................                                       |  45%
label: unnamed-chunk-10 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.Ncell.barchart.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.Ncell.barchart.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_N_cell_barchart (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: make_percent_N_cells_barchart (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: plot_barchart (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.................................                                     |  47%
  ordinary text without R code

  |                                                                              |..................................                                    |  49%
label: unnamed-chunk-11 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.AUC.cat.enrichments.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.AUC.cat.enrichments.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_AUC_cat_enrichment (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: AUC_prep_from_file (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Quantiles for the number of genes detected by cell: 
(Non-detected genes are shuffled at the end of the ranking. Keep it in mind when choosing the threshold for calculating the AUC).
Warning in AUCell_exploreThresholds(cells_AUC, plotHist = TRUE, nCores = 1,  :
  Skipping genesets with all AUC 0: Mk
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: enrichment_plots (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |....................................                                  |  51%
  ordinary text without R code

  |                                                                              |.....................................                                 |  53%
label: unnamed-chunk-12 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.linear.regression.data.prep.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.linear.regression.data.prep.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_linear_regression (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: data_pre_linear_regression (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: plot_linear_regression (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.......................................                               |  55%
  ordinary text without R code

  |                                                                              |........................................                              |  57%
label: unnamed-chunk-13 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.heatmaps.and.dotplots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.heatmaps.and.dotplots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_heatmaps_and_dotplots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: Cluster_overview_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: Cluster_overview (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.........................................                             |  59%
  ordinary text without R code

  |                                                                              |...........................................                           |  61%
label: unnamed-chunk-14 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.make.cluster.table.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.make.cluster.table.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_make_cluster_table (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: ClusterTable_data_prep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


Attaching package: 'DT'

The following object is masked from 'package:Seurat':

    JS

  |                                                                              |..........................................................            |  83%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: render_ClusterTable (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |............................................                          |  63%
  ordinary text without R code

  |                                                                              |..............................................                        |  65%
label: unnamed-chunk-15 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.cell.type.estimate.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.cell.type.estimate.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_cell_type_estimate (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: CellTypePlot_dataprep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F


Registered S3 method overwritten by 'enrichplot':
  method               from
  fortify.enrichResult DOSE
clusterProfiler v3.12.0  For help: https://guangchuangyu.github.io/software/clusterProfiler

If you use clusterProfiler in published research, please cite:
Guangchuang Yu, Li-Gen Wang, Yanyan Han, Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology. 2012, 16(5):284-287.

Attaching package: 'clusterProfiler'

The following object is masked from 'package:DelayedArray':

    simplify

The following object is masked from 'package:purrr':

    simplify

Loading required package: GSEABase
Loading required package: annotate
Loading required package: AnnotationDbi

Attaching package: 'AnnotationDbi'

The following object is masked from 'package:dplyr':

    select

Loading required package: XML
Loading required package: graph

Attaching package: 'graph'

The following object is masked from 'package:XML':

    addNode

The following object is masked from 'package:stringr':

    boundary

--> No gene can be mapped....
--> Expected input gene ID: MMP9,HHEX,PLA2G7,CCL22,BEND5,MGAM
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: GATA3,B2M,PEX26,CD151,MMP15,IFITM3
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: CCL22,SCN9A,CD22,NIPSNAP3B,MSC,BACH2
--> return NULL...
  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: CellTypePlot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |...............................................                       |  67%
  ordinary text without R code

  |                                                                              |.................................................                     |  69%
label: unnamed-chunk-16 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.gene.set.heatmaps.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.gene.set.heatmaps.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_gene_set_heatmaps (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: MakeHM (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |..........................................................            |  83%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: calculate-markdown-heatmap-chunks-dynamically (with options) 
List of 4
 $ echo   : symbol F
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |..................................................                    |  71%
  ordinary text without R code

  |                                                                              |...................................................                   |  73%
label: unnamed-chunk-17 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.example.feature.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.example.feature.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_example_feature_plots (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: ExampleFeaturePlot_dataprep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |..........................................................            |  83%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: ExampleFeaturePlot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |.....................................................                 |  76%
  ordinary text without R code

  |                                                                              |......................................................                |  78%
label: unnamed-chunk-18 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.pca.cell.and.loading.plots.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.pca.cell.and.loading.plots.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |....                                                                  |   6%
  ordinary text without R code

  |                                                                              |.........                                                             |  12%
label: partB_pca_loadings (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |.............                                                         |  19%
  ordinary text without R code

  |                                                                              |..................                                                    |  25%
label: elbow_plot (with options) 
List of 3
 $ fig.height: num 6
 $ fig.width : num 10
 $ fig.cap   : language paste0("**Figure, ", figureCount, ": Elbowplot** Variance explained per PCA dimension")

  |                                                                              |......................                                                |  31%
  ordinary text without R code

  |                                                                              |..........................                                            |  38%
label: PCA_dataprep (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F


Attaching package: 'ggpubr'

The following object is masked from 'package:ggtree':

    rotate


Attaching package: 'gridExtra'

The following object is masked from 'package:Biobase':

    combine

The following object is masked from 'package:BiocGenerics':

    combine

The following object is masked from 'package:dplyr':

    combine

  |                                                                              |...............................                                       |  44%
   inline R code fragments

  |                                                                              |...................................                                   |  50%
label: plot_PCADimPlots (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |.......................................                               |  56%
  ordinary text without R code

  |                                                                              |............................................                          |  62%
label: pca_boxplot_dataprep (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

  |                                                                              |................................................                      |  69%
   inline R code fragments

  |                                                                              |....................................................                  |  75%
label: pca_boxplot (with options) 
List of 4
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"

  |                                                                              |.........................................................             |  81%
  ordinary text without R code

  |                                                                              |.............................................................         |  88%
label: create-pca-enrichment-data (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

--> No gene can be mapped....
--> Expected input gene ID: PDE6C,SLC15A3,NTRK1,CCL19,CLIC2,IL18RAP
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: CRISP3,PLA1A,IRF8,HHEX,PLA1A,GPR183
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: IL5RA,CLIC2,BPI,LTC4S,PLA1A,FZD2
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: RAB27B,ALOX15,RASSF4,ABCB4,STAP1,CXCL3
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: CCL20,LRMP,APOL6,FAIM3,C3AR1,SLC12A8
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: CD33,NMBR,SLAMF1,CD69,ZNF222,EMR1
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: MAP3K13,CLIC2,CCL13,SLC2A6,CCR7,P2RY14
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: CCL22,TEC,CCL13,CCL18,LAG3,CLC
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: BACH2,NPL,CD244,KYNU,CPA3,IL7
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: IDO1,CD209,RNASE6,ALOX5,IFI44L,FCER1A
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: CCL1,STXBP6,FCER2,EMR3,LINC00921,CMA1
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: MARCO,IL5,MMP9,EPN2,LINC00597,TMEM255A
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: PTGIR,GUSBP11,MS4A3,CD38,GSTT1,CD209
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: TNFRSF13B,P2RY14,BACH2,CD244,TCL1A,DACH1
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: RASGRP2,CD69,PLA2G7,MMP12,CXCL10,LRMP
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: MAK,STXBP6,PDCD1LG2,FRK,P2RX1,NMBR
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: MS4A1,KYNU,STXBP6,CCL20,BLK,FPR3
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: CCL8,HLA-DOB,AQP9,CD1C,CR2,CCL19
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: NTRK1,PTGIR,EMR3,SMPDL3B,IL5,PNOC
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: DCSTAMP,TNFSF14,ST3GAL6,CD1E,MICAL3,TNFRSF13B
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: CD38,IFNA10,ZNF286A,TREM2,HLA-DOB,CD209
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: CLEC4A,STAP1,PNOC,PTPRCAP,RGS13,RNASE2
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: SMPDL3B,BPI,ZBTB32,P2RX5,NME8,IL1RL1
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: ACHE,NMBR,IGHD,CCDC102B,IL1RL1,BACH2
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: ADAMDEC1,ATP8B4,CCL4,CD68,RGS1,SLAMF8
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: TREM2,VPREB3,CD40,DAPK2,PPFIBP1,GGT5
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: SLC2A6,HHEX,BCL7A,LINC00597,VNN1,GPR97
--> return NULL...
--> No gene can be mapped....
--> Expected input gene ID: PLA2G7,DACH1,FFAR2,TREM2,GGT5,CD68
--> return NULL...
  |                                                                              |..................................................................    |  94%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: create-pca-enrichment-plot (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |........................................................              |  80%
  ordinary text without R code

  |                                                                              |.........................................................             |  82%
label: unnamed-chunk-19 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.pseudotime.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.pseudotime.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |............                                                          |  17%
  ordinary text without R code

  |                                                                              |.......................                                               |  33%
label: partB_pseudotime (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |...................................                                   |  50%
  ordinary text without R code

  |                                                                              |...............................................                       |  67%
label: pseudotime-enrichment-data (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F

  |                                                                              |..........................................................            |  83%
   inline R code fragments

  |                                                                              |......................................................................| 100%
label: pseudotime-plot (with options) 
List of 4
 $ echo   : symbol T
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: chr "asis"


  |                                                                              |...........................................................           |  84%
  ordinary text without R code

  |                                                                              |............................................................          |  86%
label: unnamed-chunk-20 (with options) 
List of 2
 $ child: chr "src/modules/partB_main_figures/partB.add.extra.tables.Rmd"
 $ eval : logi TRUE



processing file: ./src/modules/partB_main_figures/partB.add.extra.tables.Rmd
  |                                                                              |                                                                      |   0%  |                                                                              |..................                                                    |  25%
  ordinary text without R code

  |                                                                              |...................................                                   |  50%
label: partB_add_extra_tables (with options) 
List of 3
 $ echo   : logi TRUE
 $ eval   : logi TRUE
 $ warning: logi FALSE

  |                                                                              |....................................................                  |  75%
  ordinary text without R code

  |                                                                              |......................................................................| 100%
label: Additional tables (with options) 
List of 4
 $ echo   : symbol F
 $ eval   : logi TRUE
 $ warning: logi FALSE
 $ results: symbol F


  |                                                                              |.............................................................         |  88%
  ordinary text without R code

  |                                                                              |...............................................................       |  90%
label: saveobject (with options) 
List of 3
 $ eval   : logi TRUE
 $ echo   : symbol T
 $ results: symbol F

  |                                                                              |................................................................      |  92%
  ordinary text without R code

  |                                                                              |..................................................................    |  94%
label: create_report_params (with options) 
List of 2
 $ eval   : symbol T
 $ results: chr "asis"

Project Database could not be reached or has no entry in Obio@parameterList$lims.id for this analysis.
  |                                                                              |...................................................................   |  96%
  ordinary text without R code

  |                                                                              |..................................................................... |  98%
label: documentation (with options) 
List of 3
 $ eval   : logi TRUE
 $ echo   : symbol T
 $ results: symbol T

  |                                                                              |......................................................................| 100%
   inline R code fragments


output file: sc_PartB_Analysis.knit.md

/camp/apps/eb/dev/software/Anaconda2/2019.10/bin/pandoc +RTS -K512m -RTS sc_PartB_Analysis.utf8.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /camp/stp/babs/working/boeings/Projects/hillc/danielle.park/437subC_fibroblast_GSE154778_primTumor_C4/workdir/html_local/sc_PartB_Analysis.html --lua-filter /camp/apps/misc/stp/babs/manual/software/r/r-site-library/bioc-3.9/rmarkdown/rmarkdown/lua/pagebreak.lua --lua-filter /camp/apps/misc/stp/babs/manual/software/r/r-site-library/bioc-3.9/rmarkdown/rmarkdown/lua/latex-div.lua --self-contained --variable bs3=TRUE --standalone --section-divs --table-of-contents --toc-depth 5 --variable toc_float=1 --variable toc_selectors=h1,h2,h3,h4,h5 --variable toc_collapsed=1 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /camp/apps/misc/stp/babs/manual/software/r/r-site-library/bioc-3.9/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --css src/assets/style/style.css --variable theme=paper --include-in-header /tmp/slurm_25137365/RtmpEt7lA6/rmarkdown-stra3fd3ab1176b.html --mathjax --variable 'mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --variable code_folding=hide --variable code_menu=1 

Output created: /camp/stp/babs/working/boeings/Projects/hillc/danielle.park/437subC_fibroblast_GSE154778_primTumor_C4/workdir/html_local/sc_PartB_Analysis.html
There were 50 or more warnings (use warnings() to see the first 50)
