---

output: 
    html_document:
        highlight: default
        theme: paper
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 3
        toc_float: true
        css: /camp/stp/babs/working/boeings/Stefan/protocol_files/github/boeings/templates/style/style.css

---
    
    

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```


```{r hpc_notes, include=FALSE}

## Get interactive session ##
#  srun --time=08:00:00 --mem=40G -p int --pty bash

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;R;

# sbatch --time=12:00:00 --wrap "module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;Rscript runA.r" --job-name="rA"  --mem=100G -o rA.slurm >> commands.txt

# sbatch --time=24:00:00 --wrap "module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;Rscript runDGE.r" --job-name="rDGE" -p hmem --mem=300G -o rDGE.slurm >> commands.txt

# --mem-per-cpu=14G -p hmem --pty bash

```

```{r populate_meta_data_database, eval=TRUE, echo=F, results=F}
#Create the environment and load a suitable version of R, e.g. so:

myPaths <- .libPaths()
myNewPaths <- c("/camp/stp/babs/working/boeings/Projects/pachnisv/song.chng/330_10xscRNAseq_DIV4_DIV11_DIV20_SC19069/basedata", myPaths)
.libPaths(myNewPaths)
library(glmGamPoi)
.libPaths(myPaths)

library(Seurat)
library(ggplot2)
library(tidyverse)
library(knitr)

VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}


FN <- paste0(hpc.mount, "Projects/reference_data/documentation/BC.parameters.txt")
dbTable <- read.delim(
    FN, 
    sep = "\t",
    stringsAsFactors = F
)

db.pwd <- as.vector(dbTable[1,1])

figureCount <- 1

source("assets/R/SBwebtools.pckg.r")

if (length(.libPaths()) > 2){
    .libPaths(.libPaths()[2:3])
}
## Create biologic Object for visualization ##

ObioFN <- paste0("../", list.files("..")[grep(".bioLOGIC.Robj", list.files(".."))])

load(ObioFN)

checkFile = paste0(
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
)


Obio <- setMountingPoint(Obio)
Obio <- setAnalysisPaths(Obio)
Obio <- setCrickGenomeAndGeneNameTable(Obio)
Obio <- createAnalysisFolders(
    Obio
)
Obio <- setDataBaseParameters(Obio)

## Upload metadata table > p315_PCA
# Obio@parameterList$host <- "10.152.22.193"
# Obio@parameterList$db.user <- "boeingS"
# db.pwd <- "5+3f4nB040420"

## Create url string
if (Obio@parameterList$host == "10.27.241.234"){
    urlString <- "biologic.thecrick.org"
} else {
    urlString <- "biologic.crick.ac.uk"
}

legendDotSize <- 5


shinyURL <- paste0(
    "https://shiny-bioinformatics.crick.ac.uk/shiny/boeings/",
    Obio@parameterList$project_id,
    "_app/"
)            
    

## Set file paths ##
baseFN <- paste0(
   Obio@parameterList$project_id, 
   ".DGE.table.xlsx"
)


outPutFN <- paste0(
     Obio@parameterList$reportTableDir,
     baseFN
)
  
 
FNrel <- paste0("report_tables/", baseFN)
 

## Create table link string ##

tableLink <- paste0('<a href="https://',urlString,'/mdata/',Obio@parameterList$project_id, '/html/', FNrel,' target="_blank">here</a>')  

tableString <- paste0('An Excel table with the DGE results can be downloaded ',
                      tableLink
)


load(paste0(
    Obio@parameterList$localWorkDir,
    Obio@parameterList$project_id,
    ".Seurat.Robj"
)
)






###############################################################################
## Do DGEsc                                                                  ##
setGeneric(
  name="doDGEsc",
  def=function(
    obj,
    DGEselCol = "sub_clusters_ExNeurons",
    colName = "DGE_name",
    contrastTag = "contrast_1_",
    DGEsampleList = list(
      "M1" = c(5),
      "M2" = c(1)
    )
    
    ) {
    
    colName <- unlist(colName)
    DGEselCol <- unlist(DGEselCol)

    myPaths <- .libPaths()
    myNewPaths <- c("/camp/stp/babs/working/boeings/Projects/pachnisv/song.chng/330_10xscRNAseq_DIV4_DIV11_DIV20_SC19069/basedata", myPaths)
    .libPaths(myNewPaths)
    library(glmGamPoi)
    .libPaths(myPaths)

    library(Seurat)
    pos <- grep(colName, names(obj@meta.data))
    if (length(pos) > 0){
      obj@meta.data <- obj@meta.data[,-pos]
    }

    mColName <- paste0("meta_", colName)
    obj@meta.data[[mColName]] <- "Rest" 
    ## Add DGEsamples to meta.data
    for (i in 1:length(DGEsampleList)){
        obj@meta.data[obj@meta.data[,DGEselCol] %in% DGEsampleList[[i]], mColName] <- names(DGEsampleList)[i]
    }
    
    dfMeta <- obj@meta.data
    dfMeta <- dfMeta[dfMeta[,mColName] != "Rest",]
    #dfMeta$cellID <- row.names(dfMeta)


    cellVec <- dfMeta$cellID
    group <-  as.vector(dfMeta[,mColName])
    
    dfMatrix <- OsC[["RNA"]]@counts
    dfMatrix <- data.matrix(dfMatrix[,cellVec])
    dfMatrix <- dfMatrix[rowSums(dfMatrix) != 0, ]



    ## LRT ##
    # fit <- glm_gp(dfMatrix, design = group)
    # res <- test_de(fit, reduced_design = ~ 1)
    
    ## Pseudobulk ##
    #sample_labels <- rep(paste0("sample_", 1:6), length = ncol(pbmcs_subset))
    #cell_type_labels <- sample(c("T-cells", "B-cells", "Macrophages"), ncol(pbmcs_subset), replace = TRUE)
    
    
    #group <-  as.vector(dfMeta$Glia_vs_Neuro_all)
    sample_labels <- group
    sample_labels[sample_labels == names(DGEsampleList)[1]] <- paste0(sample_labels[sample_labels == names(DGEsampleList[1])], "_", 1:length(sample_labels[sample_labels == names(DGEsampleList)[1]]))
    
    sample_labels[sample_labels == names(DGEsampleList)[2]] <- paste0(sample_labels[sample_labels == names(DGEsampleList[2])], "_", 1:length(sample_labels[sample_labels == names(DGEsampleList)[2]]))
    
    cell_type_lables <- group
    unique(group)
    
    fit <- glm_gp(dfMatrix, design = group)
    comparison <- names(DGEsampleList)
    
    contrastString <- (paste0(comparison[1], " - ", comparison[2]))
    
    res <- test_de(fit, contrast = contrastString,
            pseudobulk_by = sample_labels, 
            #subset_to = cell_type_labels == "T-cells",
            #n_max = 4, 
            sort_by = pval, 
            decreasing = FALSE
    )
    
    dfRes <- res[order(res$lfc),]
    dfRes <- dfRes[dfRes$lfc > -100 & dfRes$lfc < 100, ]
    dfRes[["padj"]] <- dfRes$adj_pval
    minP <- min(dfRes$padj[dfRes$padj != 0])
    dfRes[dfRes$padj == 0, "padj"] <- minP
    dfRes[["lg10p"]] <- -1 * log10(dfRes$padj)
    
    
    comp_1 <- dfRes
    comp_1[["gene"]] <- dfRes$name
    names(comp_1) <- gsub("lfc", paste0(contrastTag, "logFC_" ,colName), names(comp_1))
    
    names(comp_1) <- gsub("padj", paste0(contrastTag, "padj_",colName), names(comp_1))
    names(comp_1) <- gsub("lg10p", paste0(contrastTag, "lg10p_",colName), names(comp_1))
    
    selVec <- c(
        "gene",
        names(comp_1)[grep(contrastTag, names(comp_1))]
    )
    
    comp_1 <- unique(comp_1[,selVec])
    
    pos <- grep(paste0("DGE_", DGEselCol), names(Obio@dataTableList))
    
    returnList <- list(
      "OsC" = obj,
      DGEtable = comp_1
    )
    
    return(returnList)
})


## End doDGEsc Function                                                      ##
###############################################################################


DGEtagVec <- as.vector(NULL, mode = "character")


DGEinputList <- list(
  "comp_1" = list(
      "name"    = "DGE_C5_vs_C7",
      "DGEselCol" = "seurat_clusters",
      "DGEsampleList" = list(
          "C5"   = c(5),
          "C7" = c(7)
      )
  ) 
)







###############################################################################
## DGE Analysis                                                              ##

plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

for (i in 1:length(DGEinputList)){
    tag <- paste0(unlist(DGEinputList[[i]]["name"]))
    
    
    
    resList <- doDGEsc(
        obj = OsC,
        DGEselCol = unlist(DGEinputList[[i]]["DGEselCol"]),
        colName = unlist(DGEinputList[[i]]["name"]),
        contrastTag = paste0("contrast_",i,"_"),
        DGEsampleList = DGEinputList[[i]][["DGEsampleList"]]
    )
    
    OsC <- resList$OsC
    
    ###########################################################################
    ## Add average intensity for MA-Plot to DGE table                        ##
    
    OsC_DGE <- OsC
    OsC_DGE@meta.data[["DGE_sel"]] <- 0
    
    OsC_DGE@meta.data[OsC_DGE@meta.data[,unlist(DGEinputList[[i]]["DGEselCol"])] != "", "DGE_sel"] <- 1
    
    row.names(OsC_DGE@meta.data) <- OsC_DGE@meta.data[,"cellID"]
    
    OsC_DGE <- subset(x = OsC_DGE, subset = DGE_sel == 1)
    
    
    Idents(OsC_DGE) <- "DGE_sel"
    
    cluster.averages <- AverageExpression(
        OsC_DGE, 
        return.seurat = TRUE
    )
    
    dfAvgExpr <- data.frame(cluster.averages[["RNA"]]@data)
    dfAvgExpr[["gene"]] <- row.names(dfAvgExpr)
    row.names(dfAvgExpr) <- NULL
    
    nameTag <- paste0("add_MA_cts_MA_Avg_", unlist(DGEinputList[[i]]))
    names(dfAvgExpr) <- gsub("X1", nameTag, names(dfAvgExpr))
    
    dfDataTable <- resList$DGEtable
    
    dfDataTable <- merge(
        dfDataTable, 
        dfAvgExpr, 
        by.x = "gene",
        by.y = "gene",
        all =TRUE
    )
    
    dfDataTable[is.na(dfDataTable)] <- 0
    
    ## Done adding avg intensity for MA-plot                                 ##
    ###########################################################################
    
    Obio@dataTableList[[tag]] <- data.frame(NULL)
    Obio@dataTableList[[tag]] <- dfDataTable
    
    DGEtagVec <- c(
      DGEtagVec,
      tag
    )
    
    ## Done with DGE Analysis                                                ##
    ###########################################################################
    
    ###########################################################################
    ## Create DGE Selection plot                                             ##
    dfPlot <- OsC@meta.data
    
    
    
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }

    pos <- grep("cellID", names(dfPlot))
      if (length(pos) == 0){
      dfPlot[["cellID"]] <- row.names(dfPlot)
    }

    dfPlot$UMAP_1 <- NULL
    dfPlot$UMAP_2 <- NULL
            
    ## Get UMAP coordinates ##
    coord <- data.frame(OsC@reductions$umap@cell.embeddings)
    coord[["cellID"]] <- row.names(coord)
    coord <-coord[coord$cellID %in% dfPlot$cellID, ]
            
    dfPlot <- merge(dfPlot, coord, by.x = "cellID", by.y="cellID", all=T)
    dfPlot[is.na(dfPlot)] <- 0
    dfPlot <- dfPlot[dfPlot$UMAP_1 != 0 & dfPlot$UMAP_2 != 0,]
            
            
    ## Add cluster colors ##
    dfPlot[["Cluster"]] <- dfPlot[,paste0("meta_", unlist(DGEinputList[[i]]["name"]))]
    dfPlot$Cluster[is.na(dfPlot$Cluster)] <- ""
    dfPlot[dfPlot$Cluster == "", "Cluster"] <- "Rest"
    
    label <- as.vector(gsub("DGE_", "", unlist(DGEinputList[[i]]["name"])))
    
    
    redGroup <- names(DGEinputList[[i]]$DGEsampleList)[1] #as.vector(unlist(strsplit(label, "_vs_"))[1])
    blueGroup <- names(DGEinputList[[i]]$DGEsampleList)[2] #as.vector(unlist(strsplit(label, "_vs_"))[2])
    greyGroup <- "Rest"
    
    ## Assigning colors ##
    dfCol <- data.frame(name = unique(dfPlot$Cluster), col = rep("", length(unique(dfPlot$Cluster))), stringsAsFactors = F)
    dfCol[dfCol$name == redGroup, "col"] <- "red"
    dfCol[dfCol$name == blueGroup, "col"] <- "blue"
    dfCol[dfCol$name == greyGroup, "col"] <- "grey"
    colVec <- dfCol$col
    names(colVec) <- dfCol$name
    
    dfColExtra <- dfCol[dfCol$col == "",]
    if (nrow(dfColExtra) > 0){
        eColVec <- dfColExtra$name
        library(scales)
        eCols = hue_pal()(length(eColVec))
        for (l in 1:length(eColVec)){
            dfCol[dfCol$name == eColVec[l], "col"] <- eCols[l]
        }
    }
    
    maxX <- 1.1*max(dfPlot$UMAP_1, na.rm = T)
    minX <- 1.1*min(dfPlot$UMAP_1, na.rm = T)
    maxY <- 1.1*max(dfPlot$UMAP_2, na.rm = T)
    minY <- 1.1*min(dfPlot$UMAP_2, na.rm = T)            

    #colVec <- dfPair$sampleColor
    #names(colVec) <- dfPair$sampleName
    levels <- as.vector(dfCol$name)
    levelsRest <- levels[grep("Rest", levels)]
    levels <- levels[levels != "Rest"]
    levels <- c(levels, levelsRest)
    
    dfPlot$Cluster <- factor(dfPlot$Cluster, levels = levels)

    dotsize  = 1
    if (nrow(dfPlot) > 10000){
      dotsize  = 0.5
    } else if (nrow(dfPlot) > 20000){
      dotsize = 0.25
    } else if (nrow(dfPlot) > 50000){
      dotsize = 0.1
    }

    plotList[[tag]] <- ggplot(
        data=dfPlot[dfPlot$included == "+",], 
        aes(UMAP_1, UMAP_2, color=Cluster)
      ) + geom_point( shape=16, size = as.numeric(dotsize)
      ) + xlab("UMAP1") + ylab("UMAP2"
      ) + theme_bw(
      )  +  theme(
          axis.text.y   = element_text(size=8),
          axis.text.x   = element_text(size=8),
          axis.title.y  = element_text(size=8),
          axis.title.x  = element_text(size=8),
          axis.line = element_line(colour = "black"),
          panel.border = element_rect(colour = "black", fill=NA, size=1),
          plot.title = element_text(hjust = 0.5, size = 12),
          legend.title = element_blank()
      ) + guides(col = guide_legend(override.aes = list(shape = 16, size = legendDotSize))
      ) + ggtitle(paste0("Cell Selection for ", gsub("_", " ", unlist(DGEinputList[[i]]["name"])))
      ) + xlim(minX, maxX) + ylim(minY, maxY
      ) + coord_fixed(ratio=1
      ) + scale_colour_manual("DGE Groups" ,values = colVec
      ) + guides(col = guide_legend(override.aes = list(shape = 16, size = legendDotSize))
      )


      ## Add colors if specified ##

            

      FNbase <- paste0(tag, VersionPdfExt)
      FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
      FNrel <- paste0("report_figures/", FNbase)
            
      pdf(FN)
          print(plotList[[tag]])
      dev.off()
            
      # link <- paste0('<a href="https://',urlString,'/',Obio@parameterList$project_id,'/pca?x_axis=UMAP_1&y_axis=UMAP_2" target="_blank">here</a>')  
      
      if (exists("shinyURL") & !is.null(shinyURL)){
          link <- paste0(
                  'An interactive version of this figure with additional viewing options can be found <a href="',shinyURL,'?_inputs_&y_axis=%22UMAP_2%22&x_axis=%22UMAP_1%22&colorBy=%22',paste0("meta_", tag),'%22&splitByColumn=%22all%22" target="_blank">here</a>. '
          )
                
      } else {
          link <- ""
      }
            
      figLegend <- paste0(
      '**Figure ', 
      figureCount, 
                ':** ',
                ' UMAP showing all cells from all samples together. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. ',
                'An interactive version of this figure can be found ', link
            )
            
      figureCount <- figureCount + 1
            
      NewChnk <- paste0(
          "#### ", tag,
          "\n```{r Sample_DGE_",
          tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
          figLegend,"'}\n",
          "\n",
          "\n print(plotList[['",tag,"']])",
          "\n cat(  '\n')",
          "\n\n\n```\n"   
      )
            
      chnkVec <- c(
          chnkVec,
          NewChnk
      )

      ## Done creating DGE selection plot                                      ##
      ###########################################################################
    
    
      print(paste0("Done with ", tag))
}

## Done DGE                                                                  ##
###############################################################################

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
``` 

### Differential Gene Expression Comparions Cell Selections {`r tabVar`}
In this section the cells that were used for the DGE comparisons are highlighted. 

```{r CellTypePlot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))

``` 


```{r data_DGE_summary, echo=F, eval=T, warning=FALSE, results="asis"}

###############################################################################
## Assemble DGE summary table                                                ##
#dfTemp <- Obio@dataTableList[["DGE_table"]] 

library(openxlsx)
wb <- createWorkbook()

hs1 <- createStyle(
    fontColour = "#ffffff",
    fgFill = "#000000",
    halign = "CENTER",
    textDecoration = "Bold"
)

for (i in 1:length(DGEtagVec)){
  ## Add to Excel workbook ##
  dfOutput <-   Obio@dataTableList[[DGEtagVec[i]]]
  orderVec <- names(dfOutput)
  dfOutput[["gene_FeatureViewLink"]] <- paste0(
      "https://shiny-bioinformatics.crick.ac.uk/shiny/boeings/",
      Obio@parameterList$project_id,
      "_app/?gene=",
      dfOutput$gene
  )
  orderVec <- orderVec[orderVec != "gene"]
  orderVec <- c("gene", "gene_FeatureViewLink", orderVec)
  dfOutput <- dfOutput[, orderVec]
  class(dfOutput$gene_FeatureViewLink) <- c(class(dfOutput$gene_FeatureViewLink), "hyperlink")
  
  sheetName <- substr(DGEtagVec[i],1,30)
  
  addWorksheet(
      wb, 
      sheetName = sheetName
  )
  
  
  freezePane(wb, sheetName ,  firstActiveRow = 2)
  writeData(wb, i, dfOutput, startRow = 1, startCol = 1, headerStyle = hs1)
  ## Done adding to Excel workbook ##
  if (i ==1){
    dfTemp <- Obio@dataTableList[[DGEtagVec[i]]]
    
  } else {
    dfTemp <- merge(
      dfTemp, 
      Obio@dataTableList[[DGEtagVec[i]]],
      by.x = "gene",
      by.y = "gene",
      all =TRUE
    )
    dfTemp[is.na(dfTemp)] <- 0
  
  }
}

## Save workbook ##
baseFN <- paste0(
   Obio@parameterList$project_id, 
   ".DGE.table.xlsx"
)


outPutFN <- paste0(
     Obio@parameterList$reportTableDir,
     baseFN
)

saveWorkbook(
        wb,
        outPutFN ,
        overwrite = TRUE
)

Obio@dataTableList[["DGE_table"]] <- NULL
Obio@dataTableList[["DGE_table"]] <- dfTemp

## Save DGE table to file ##
# tempDir <- paste0(Obio@parameterList$localWorkDir,"temp")
#             if(!dir.exists(tempDir)){
#                 dir.create(tempDir)
#             }
# 
# FN <- paste0(Obio@parameterList$localWorkDir, "/temp/DGE.result.table.txt")
# write.table(
#     dfTemp,
#     FN,
#     row.names = F, 
#     sep = "\t"
# )

## Done adding DGE table                                                     ##
###############################################################################
```


```{r data_MA_plots, echo=F, eval=T, warning=FALSE, results="asis"}


###############################################################################
## Make MA-plots                                                             ##
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

dfAllPlots <- Obio@dataTableList[["DGE_table"]] 

for (i in 1:length(DGEtagVec)){
    tag <- paste0("MA_Plot_", DGEtagVec[i])  
    label <- gsub("_", " ", DGEtagVec[i])
    
    selVec <- c(
        "gene",
        names(dfAllPlots)[grep(paste0(DGEtagVec[i], "$"), names(dfAllPlots))]
    )
    
    dfPlot <- dfAllPlots[,selVec]
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }
    
    logFCtwoSD <- 2*sd(dfPlot[,grep("_logFC_", names(dfPlot))])
    
    dfPlot[["DGE_Status"]] <- "Unchanged"
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] > logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Up"
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] < -1 *logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Down"
    
    colVec <- c("red", "blue", "black")
    names(colVec) <- c("Up", "Down", "Unchanged")
  
    
    dfPlot[["X"]] <- dfPlot[,grep("add_MA_cts_MA_Avg", names(dfPlot))]
    dfPlot[["Y"]] <- dfPlot[,grep("_logFC", names(dfPlot))]
    
    dotsize  = 1.25
    if (nrow(dfPlot) > 10000){
      dotsize  = 1
    } else if (nrow(dfPlot) > 20000){
      dotsize = 0.75
    } else if (nrow(dfPlot) > 50000){
      dotsize = 0.5
    }

    logFClims <- ceiling(max(abs(dfPlot$Y)))
    
    plotList[[tag]] <- ggplot(
        data=dfPlot[dfPlot$included == "+",], 
        aes(X, Y, color=DGE_Status,)
      ) + geom_hline(yintercept = c(logFCtwoSD, -1 * logFCtwoSD), col = "darkgrey", lty=2
      ) + geom_point( shape=16, size = as.numeric(dotsize),alpha = 0.7
      ) + xlab(paste0("log10 Avg Intensity ", gsub("_", " ",label))) + ylab(paste0("log2-Fold Change ", gsub("_", " ",label))
      ) + theme_bw(
      )  +  theme(
          axis.text.y   = element_text(size=8),
          axis.text.x   = element_text(size=8),
          axis.title.y  = element_text(size=8),
          axis.title.x  = element_text(size=8),
          axis.line = element_line(colour = "black"),
          panel.border = element_rect(colour = "black", fill=NA, size=1),
          plot.title = element_text(hjust = 0.5, size = 12)
          #legend.title = element_blank()
      ) + guides(col = guide_legend(override.aes = list(shape = 16, size = legendDotSize))
      ) + ggtitle(paste0("MA-stype Plot for DGE Comparison ", gsub("_", " ", label))
      ) + ylim(logFClims, -1 * logFClims
      ) + scale_color_manual("DGE Status" ,values = colVec
      ) + scale_y_continuous(breaks=c(seq(-20,20,2))
      )


      ## Add colors if specified ##

            

      FNbase <- paste0(tag, VersionPdfExt)
      FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
      FNrel <- paste0("report_figures/", FNbase)
            
      pdf(FN)
          print(plotList[[tag]])
      dev.off()
      
      xLabel <- names(dfPlot[grep("MA_Avg", names(dfPlot))])      
      yLabel <- names(dfPlot[grep("_logFC_", names(dfPlot))])  
      link <- paste0('<a href="https://',urlString,'/',Obio@parameterList$project_id,'/scatterplot?x_axis=',xLabel,'&y_axis=',yLabel,'" target="_blank">here</a>')  
            
      figLegend <- paste0(
      '**Figure ', 
      figureCount, 
                ':** ',
                ' MA-style plot for differential gene expression comparison ',gsub("_", " ", DGEtagVec[i]),'. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. ',
                'An interactive version of this figure can be found ', link, '. ',
                tableString
            )
            
      figureCount <- figureCount + 1
            
      NewChnk <- paste0(
          "#### ", tag,
          "\n```{r MA_DGE_",
          tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
          figLegend,"'}\n",
          "\n",
          "\n print(plotList[['",tag,"']])",
          "\n cat(  '\n')",
          "\n\n\n```\n"   
      )
            
      chnkVec <- c(
          chnkVec,
          NewChnk
      )

      ## Done creating DGE selection plot                                      ##
      ###########################################################################
    
    
    
    
}


## Done MA- plots                                                            ##
###############################################################################

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
``` 

### DGE MA-style Plots {`r tabVar`}

```{r DGE_MA_plot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))

``` 


```{r data_Volcano_plots, echo=F, eval=T, warning=FALSE, results="asis"}
###############################################################################
## Make Volcano plots
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

dfAllPlots <- Obio@dataTableList[["DGE_table"]] 

for (i in 1:length(DGEtagVec)){
    tag <- paste0("Volcano_Plot_", DGEtagVec[i])  
    label <- gsub("_", " ", DGEtagVec[i])
    
    selVec <- c(
        "gene",
        names(dfAllPlots)[grep(DGEtagVec[i], names(dfAllPlots))]
    )
    
    dfPlot <- dfAllPlots[,selVec]
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }
    
    logFCtwoSD <- 2*sd(dfPlot[,grep("_logFC_", names(dfPlot))])
    
    dfPlot[["DGE_Status"]] <- "Unchanged"
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] > logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Up"
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] < -1 *logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Down"
    
    colVec <- c("red", "blue", "black")
    names(colVec) <- c("Up", "Down", "Unchanged")
  
    
    dfPlot[["X"]] <- dfPlot[,grep("_logFC", names(dfPlot))]
    dfPlot[["Y"]] <- dfPlot[,grep("_lg10p_", names(dfPlot))]
    
    dotsize  = 1.25
    if (nrow(dfPlot) > 10000){
      dotsize  = 1
    } else if (nrow(dfPlot) > 20000){
      dotsize = 0.75
    } else if (nrow(dfPlot) > 50000){
      dotsize = 0.5
    }

    logFClims <- ceiling(max(abs(dfPlot$X)))
    
    plotList[[tag]] <- ggplot(
        data=dfPlot[dfPlot$included == "+",], 
        aes(X, Y, color=DGE_Status,)
      ) + geom_vline(xintercept = c(logFCtwoSD, -1 * logFCtwoSD), col = "darkgrey", lty=2
      ) + geom_vline(xintercept = 0, col = "darkgrey"
      ) + geom_hline(yintercept = 0, col = "darkgrey"
      ) + geom_point( shape=16, size = as.numeric(dotsize),alpha = 0.7
      ) + xlab(paste0("log2-Fold Change ", gsub("_", " ",label))) + ylab(paste0("-log10(adj p-value) ", gsub("_", " ",label))
      ) + theme_bw(
      )  +  theme(
          axis.text.y   = element_text(size=8),
          axis.text.x   = element_text(size=8),
          axis.title.y  = element_text(size=8),
          axis.title.x  = element_text(size=8),
          axis.line = element_line(colour = "black"),
          panel.border = element_rect(colour = "black", fill=NA, size=1),
          plot.title = element_text(hjust = 0.5, size = 12)
          #legend.title = element_blank()
      ) + guides(col = guide_legend(override.aes = list(shape = 16, size = legendDotSize))
      ) + ggtitle(paste0("Volcano Plot for DGE Comparison ", gsub("_", " ", label))
      ) + ylim(logFClims, -1 * logFClims
      ) + scale_color_manual("DGE Status" ,values = colVec
      ) + scale_y_continuous(breaks=c(seq(0,400,50))
      ) + scale_x_continuous(breaks=c(seq(-400,400,2))
      )


      ## Add colors if specified ##

            

      FNbase <- paste0(tag, VersionPdfExt)
      FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
      FNrel <- paste0("report_figures/", FNbase)
            
      pdf(FN)
          print(plotList[[tag]])
      dev.off()
      
      xLabel <- names(dfPlot[grep("_logFC_", names(dfPlot))])      
      yLabel <- names(dfPlot[grep("_lg10p_", names(dfPlot))])  
      link <- paste0('<a href="https://',urlString,'/',Obio@parameterList$project_id,'/scatterplot?x_axis=',xLabel,'&y_axis=',yLabel,'" target="_blank">here</a>')  
            
      figLegend <- paste0(
      '**Figure ', 
      figureCount, 
                ':** ',
                ' Volcano plot for differential gene expression comparison ',gsub("_", " ", DGEtagVec[i]),'. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. ',
                'An interactive version of this figure can be found ', link, '. ',
                tableString
            )
            
      figureCount <- figureCount + 1
            
      NewChnk <- paste0(
          "#### ", tag,
          "\n```{r Volcano_DGE_",
          tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
          figLegend,"'}\n",
          "\n",
          "\n print(plotList[['",tag,"']])",
          "\n cat(  '\n')",
          "\n\n\n```\n"   
      )
            
      chnkVec <- c(
          chnkVec,
          NewChnk
      )

      ## Done creating DGE selection plot                                      ##
      ###########################################################################
    
    
    
    
}


## Done Volcano plots
###############################################################################

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
``` 

### DGE Volcano Plots {`r tabVar`}

```{r DGE_volcano_plot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))

``` 


```{r create-pca-enrichment-data, echo=T, eval=TRUE, warning=FALSE, results=F}

## Create enriched genes list ##
EnrichedGenesList <- list()
dfAllPlots <- Obio@dataTableList[["DGE_table"]] 

if (Obio@parameterList$geneIDcolumn != "mgi_symbol" & Obio@parameterList$geneIDcolumn != "hgnc_symbol") {
            queryGS <- "hgnc_symbol" 
        } else {
            queryGS <- Obio@parameterList$geneIDcolumn
        }

for (i in 1:length(DGEtagVec)){
    tag <- paste0("MA_Plot_", DGEtagVec[i])  
    tagGLpos <- paste0(DGEtagVec[i], "_pos") 
    tagGLneg <- paste0(DGEtagVec[i], "_neg") 
    
    selVec <- c(
        "gene",
        names(dfAllPlots)[grep(DGEtagVec[i], names(dfAllPlots))]
    )
    
    dfPlot <- dfAllPlots[,selVec]
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }
    
    logFCtwoSD <- 2*sd(dfPlot[,grep("_logFC_", names(dfPlot))])
    
    dfPlot[["DGE_Status"]] <- "Unchanged"
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] > logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Up"
    EnrichedGenesList[[tagGLpos]] <- unique(dfPlot[dfPlot$DGE_Status == "Up", "gene"])
    
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] < -1 *logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Down"
    EnrichedGenesList[[tagGLneg]] <- unique(dfPlot[dfPlot$DGE_Status == "Down", "gene"])
} 

library(knitr)
library(ggplot2)

#save.image("temp.RData")
library(clusterProfiler)

    gmtList <- list()
    pos <- grep("clusterSigEnrichmentList", slotNames(Obio))
    
    if (length(pos) > 0){
    if (is.null(Obio@clusterSigEnrichmentList)){
      dbtableList <- list(
          "Cell Type Signatures" = "mysigdb_sc_sig",
          "Cell Type Signatures" = "cibersort_L22",
          "GO-MF" = "mysigdb_c5_MF",
          "Pathways" = "mysigdb_c2_1329_canonical_pathways",
          "Allen_Brain_Atlas" = "Allen_Brain_Atlas"
      )
    } else {
        dbtableList <- Obio@clusterSigEnrichmentList
    }
    } else {
      dbtableList <- list(
          "Cell Type Signatures" = "mysigdb_sc_sig",
          "Cell Type Signatures" = "cibersort_L22",
          "GO-MF" = "mysigdb_c5_MF",
          "Pathways" = "mysigdb_c2_1329_canonical_pathways",
          "Allen_Brain_Atlas" = "Allen_Brain_Atlas"
      )
    }
    
    
    rmVec <- grep("Cell Type Signatures", names(dbtableList))
    if (length(rmVec) > 0){
        dbtableList <- dbtableList[-rmVec]
    }
    
    for (i in 1:length(dbtableList)){
        
        dfTemp <- unique(import.db.table.from.db(
            host = Obio@dbDetailList$host,
            dbname = Obio@dbDetailList$ref.cat.db,
            dbtable = dbtableList [[i]],
            password = db.pwd,
            user = Obio@dbDetailList$db.user
        ))
        
        ## Remove duplicated entries ##
        dfTemp <- dfTemp[!(duplicated(dfTemp$cat_name)),]
        
        rmVec <- grep("temp_", dfTemp$cat_type)
        if (length(rmVec) > 0){
            dfTemp <- dfTemp[-rmVec, ]
        }
        
        dfTemp <- unique(dbcat2gmt(
            dfTemp, # As downloaded from reference_categories_db_new database
            gene.id.column = queryGS
        ))
        
        dfTemp <- unique(dfTemp[!duplicated(as.vector(dfTemp[,1])), ])
        
        write.table(
            dfTemp,
            "temp.gmt.txt",
            row.names = F, 
            sep = "\t",
            col.names = F,
            quote = F
        )
        
        CPgmt <- read.gmt("temp.gmt.txt")
        unlink("temp.gmt.txt")
        CPgmt <- unique(CPgmt[CPgmt$gene != "", ])
        
        gmtList[[dbtableList[[i]]]] <- CPgmt
    }
    
    ## Edit collection names for plot
    names(gmtList) <- gsub("mysigdb_", "",names(gmtList))
    names(gmtList) <- gsub("c2_1329_canonical_p", "P",names(gmtList))
    names(gmtList) <- gsub("sc_sig", "CellSig",names(gmtList))
    names(gmtList) <- gsub("cibersort_L22", "CellSig",names(gmtList))
    names(gmtList) <- gsub("c5_", "GO_",names(gmtList))
    names(gmtList) <- gsub("networkcategories", "Complexes",names(gmtList))
    ## Done creating gmt list
    ###########################
    
    ## Select colors ##
    library(scales)
    enrCols <- hue_pal()(length(gmtList))
    names(enrCols) <- names(gmtList)



plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")



for (j in 1:length(DGEtagVec)){
    posTestGeneSet <- as.vector(
        unique(
            EnrichedGenesList[[paste0(DGEtagVec[j], "_pos")]]
        )
    )
    
    
    negTestGeneSet <- as.vector(
        unique(
            EnrichedGenesList[[paste0(DGEtagVec[j], "_neg")]]
        )
    )
   
    
    ###########################################################################
    ## Create GMT file for category enrichment                               ##
    
    ###########################
    ## Create gmt list
    ## Retrieve gmt files from database
    ## Add custom gmt files
    
    
    
    
    ## Done                                                                  ##
    ###########################################################################
     
    library(clusterProfiler)
    library(ggplot2)
    library(tidyr)
        
        if (Obio@parameterList$geneIDcolumn != "mgi_symbol" & Obio@parameterList$geneIDcolumn != "hgnc_symbol") {
            queryGS <- "hgnc_symbol" 
        } else {
            queryGS <- Obio@parameterList$geneIDcolumn
        }
        
        if (Obio@parameterList$host == "10.27.241.234"){
            urlString <- "biologic.thecrick.org"
        } else {
            urlString <- "biologic.crick.ac.uk"
        }
    
    colVec <- c("red", "blue")
    pvalueCutoff <- 0.5
    topMaxCat <- 10
    
    ## Get background gene set ##
    #backgroundGeneVec <- row.names(OsC[["RNA"]]@counts)
    if ((length(posTestGeneSet) >= 3) | (length(negTestGeneSet) >= 3)){
        ## Do enrichment ##
        first <- TRUE
        if (length(posTestGeneSet) >= 3){
            for (k in 1:length(gmtList)){
                    egmt <- data.frame(
                        enricher(
                            negTestGeneSet, 
                            TERM2GENE=gmtList[[k]],
                            pvalueCutoff = pvalueCutoff
                        )
                    )
                    if (!is.null(egmt)){
                        if (nrow(egmt) > 0){
                            egmt[["Collection"]] <- substr(names(gmtList)[k], 1,10)
                        }
                        if (first){
                            dfTempEnriched <- egmt    
                            first <- FALSE
                        } else {
                            dfTempEnriched <- rbind(
                                dfTempEnriched, 
                                egmt
                            )    
                        }
                        
                    }
            }
            if (nrow(dfTempEnriched) > 0){
                dfTempEnriched[["direction"]] <- "positive"
                dfTempEnriched[["log10FDR"]] <- log10(dfTempEnriched$p.adjust)
                dfTempEnriched <- dfTempEnriched[order(dfTempEnriched$log10FDR, decreasing = F),]
                dfTempEnriched <- na.omit(dfTempEnriched)
                
                if (nrow(dfTempEnriched) > topMaxCat){
                    dfTempEnriched <- dfTempEnriched[1:topMaxCat, ]
                }
            }
          
            
        } # end positive
            
            ## Now the negative side ##
            if (length(negTestGeneSet) >= 3){
            first <- TRUE
            for (k in 1:length(gmtList)){
                    egmt <- data.frame(
                        enricher(
                            posTestGeneSet, 
                            TERM2GENE=gmtList[[k]],
                            pvalueCutoff = pvalueCutoff
                        )
                    )
                    if (!is.null(egmt)){
                        if (nrow(egmt) > 0){
                            egmt[["Collection"]] <- substr(names(gmtList)[k], 1,10)
                        }
                        if (first){
                            dfTempEnrichedNeg <- egmt    
                            first <- FALSE
                        } else {
                            dfTempEnrichedNeg <- rbind(
                                dfTempEnrichedNeg, 
                                egmt
                            )    
                        }
                        
                    } 
            }
            if (nrow(dfTempEnrichedNeg) > 0){
                dfTempEnrichedNeg[["direction"]] <- "negative"
                dfTempEnrichedNeg[["log10FDR"]] <- -1*log10(dfTempEnrichedNeg$p.adjust)
                dfTempEnrichedNeg <- dfTempEnrichedNeg[order(dfTempEnrichedNeg$log10FDR, decreasing = T),]
                dfTempEnrichedNeg <- na.omit(dfTempEnrichedNeg)
                
                if (nrow(dfTempEnrichedNeg) > topMaxCat){
                    dfTempEnrichedNeg <- dfTempEnrichedNeg[1:topMaxCat, ]
                }
            }
            } # end negative
        
            
            
            ## Make plot 
            if ((nrow(dfTempEnriched) > 0) | (nrow(dfTempEnrichedNeg) > 0)){
            
            
            
            
            dfSel <- rbind(
                dfTempEnriched,
                dfTempEnrichedNeg
            )
            
            dfSel <- na.omit(dfSel)
            dfSel <- dfSel[order(dfSel$log10FDR),]
            dfSel$log10FDR <- round(dfSel$log10FDR, 2)
            
            dfSel[["Category"]] <- ""
            dfSel[dfSel$log10FDR >= 0, "Category"] <- "Enr."
            dfSel[dfSel$log10FDR < 0, "Category"] <- "Depl."
            
            for (l in 1:nrow(dfSel)){
                if (nchar(dfSel[l, "ID"]) > 30){
                    part1 <- substr(dfSel[l, "ID"], 1, 30)
                    part2 <- substr(dfSel[l, "ID"], 31, 60)
                    dfSel[l, "ID"] <- paste0(part1, " \\n", part2)
                  
                }
            }
            
            
            #dfSel$Term <- gsub("\\(GO", "\\\n\\(GO", dfSel$Term)
            
            dfSel$ID <- factor(dfSel$ID, levels = unique(dfSel$ID))
            
            
            
            plotList[[paste0("PCA_ENR_", j)]] <- ggplot(
                data=dfSel, aes(x= ID, y=log10FDR, fill=Collection, order=log10FDR)
            ) + geom_bar(stat="identity", colour="black"
            ) + coord_flip() +scale_fill_manual(values=enrCols
            ) + theme_bw(
            )  +  theme(
                axis.text.y   = element_text(size=8),
                axis.text.x   = element_text(size=8),
                axis.title.y  = element_text(size=8),
                axis.title.x  = element_text(size=8),
                axis.line = element_line(colour = "black"),
                panel.border = element_rect(colour = "black", fill=NA, size=1),
                plot.title = element_text(hjust = 0.5, size = 12)
            )  + labs(title = paste0("Comparison ", DGEtagVec[j]," enriched genes") ,y = "-log10(FDR)", x = ""
            ) + geom_hline(yintercept = c(-log10(0.05), log10(0.05)), color = "grey", size=0.5, lty=2
            ) + geom_hline(yintercept = 0, color = "black", size=0.5
            ) 
            cat("  \n")
            
            
            
            ## Save to file ##
            FNbase <- paste0("DGE_comparison_", DGEtagVec[j],".enriched.genes", VersionPdfExt)
            FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
            FNrel <- paste0("report_figures/", FNbase)
            
           
            pdf(FN)
            print(plotList[[paste0("PCA_ENR_", j)]])
            dev.off()
            
            link <- paste0(
                '<a href="https://', urlString, '/',
                Obio@parameterList$project_id,
                '/category-view?category_type=GO-BP" target="_blank">CategoryView</a>'
            )
            
            ## Create R markdown chunk ##
            figLegend <- paste0(
                '**Figure ', 
                figureCount, 
                '**: Category enrichment analysis for the top genes that have  <font color = "',colVec[2],'"> the most positive </font> and <font color = "',colVec[1],'">the most negative</font> PCA loading values in dimension ', 
               DGEtagVec[j],
                ' associated with them. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. To view these gene sets in the context of your data, go to ',link,' and find these categories using the search box.'
            )
            figureCount <- figureCount + 1 
            
            NewChnk <- paste0(
                "#### ", DGEtagVec[j],
                "\n```{r enrichr_",
                j,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                figLegend,"'}\n",
                "\n",
                "\n print(plotList[['",paste0("PCA_ENR_", j),"']])",
                "\n cat(  '\n')",
                "\n\n\n```\n"   
            )
            
            chnkVec <- c(
                chnkVec,
                NewChnk
            )
        }
            
            
            ## done with plot 
            
    } ## Done with per dimension loops
}        
      
 

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
```

### Category Enrichments Differential Gene Expression Results {`r tabVar`} 
```{r create-cat-enrichment-plot, echo=T, eval=TRUE, warning=FALSE, results='asis'}
###############################################################################
## Do category enrichment on clusters                                        ##
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
## Done doing enrichment on clusters                                         ##
###############################################################################


```


<!-- Essential N: Upload meta data table -->
```{r child = 'src/modules/db_tools/upload.meta.data.table.to.DB.Rmd', eval=TRUE}

```


```{r documentation_saving, echo=F, eval=TRUE, warning=FALSE, results=F} 

save(Obio, 
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
     )
)

print("Obio Object saved.")

save(OsC,
    file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
        ".Seurat.Robj"
     )
)

###############################################################################
## Create Excel output files                                                 ##

## Add ref-gene column ##
#dfTemp[["ref_gene"]] <- ""
#dfTemp[dfTemp$gene %in% refGenes, "ref_gene"] <- "+"

library(openxlsx)

dfTemp <- Obio@dataTableList[["DGE_table"]] 

if (!dir.exists( paste0(
   Obio@parameterList$html_local,
   "report_tables/"))){
  dir.create(paste0(
   Obio@parameterList$html_local,
   "report_tables/"))
}

baseFN <- paste0(
   Obio@parameterList$project_id, 
   ".DGE.table.xlsx"
)


outPutFN <- paste0(
     Obio@parameterList$html_local,
      "report_tables/",
     baseFN
)
  
 
FNrel <- paste0("report_table/", baseFN)
    
wb <- createWorkbook()
    addWorksheet(wb, paste0(Obio@parameterList$project_id, "_subcluster_DGE"))
    freezePane(wb, paste0(Obio@parameterList$project_id, "_subcluster_DGE") ,  firstActiveRow = 2)
    
    ## Filter is inactivated, as it does not appear to be compatible with the current version of Excel
    #addFilter(wb, 1, row = 1, cols = 1:ncol(dfOutput))
    
## Style headers ##
hs1 <- createStyle(
  fontColour = "#ffffff",
  fgFill = "#000000", 
  halign = "CENTER", 
  textDecoration = "Bold"
)
    
writeData(wb, 1,dfTemp, startRow = 1, startCol = 1, headerStyle = hs1)
    
saveWorkbook(
  wb, 
  outPutFN , 
  overwrite = TRUE
)

## Done creating Excel output files                                          ##
###############################################################################


############
## Add to main dbTable
database.table <- import.db.table.from.db(
    dbtable= Obio@parameterList$rnaseqdbTableName,   
    dbname= Obio@dbDetailList$primDataDB,   
    user= Obio@dbDetailList$db.user,     
    password= db.pwd, 
    host= Obio@parameterList$host
)

database.table$row_names <- NULL

dfTemp <- Obio@dataTableList[["DGE_table"]]
dfTemp$ref_gene <- NULL

#names(dfTemp) <- gsub("logFC", "logFC_glmG", names(dfTemp))
#names(dfTemp) <- gsub("padj", "padj_glmG", names(dfTemp))
#names(dfTemp) <- gsub("lg10p", "lg10p_glmG", names(dfTemp))

selVec <- !(names(database.table) %in% names(dfTemp))
database.table <- database.table[!(names(database.table) %in% names(dfTemp))]


database.table <- merge(
    database.table, 
    dfTemp,
    by.x = Obio@parameterList$geneIDcolumn,
    by.y = "gene",
    all = TRUE
)

database.table[is.na(database.table)] <- 0

Obio@databaseTable <- data.frame(NULL)
Obio@databaseTable <- database.table

Obio@databaseTable$Gene_description <- NULL


cmd.vec <- upload.datatable.to.database(
    host = Obio@dbDetailList$host, 
    user = Obio@dbDetailList$db.user,
    password = db.pwd,
    prim.data.db = Obio@dbDetailList$primDataDB,
    dbTableName = Obio@parameterList$rnaseqdbTableName,
    df.data = Obio@databaseTable, #[Obio@databaseTable$count_cut_off > 1, ],
    db.col.parameter.list = list(
        "VARCHAR(255) CHARACTER SET latin1 COLLATE latin1_swedish_ci" = c("Gene_name","gene_description"),
        "VARCHAR(50) CHARACTER SET latin1 COLLATE latin1_swedish_ci" = c("ENSDARG", "dr_symbol","clone_based_gene_ID","Gene_description","Gene_type","ENSG", "ENSMUSG", "hgnc_symbol", "mgi_symbol", "uniprot", "entrezgene","display_ptm", "^sequence_window", "p_site_env","for_GSEA_gene_chip","associated_gene_name", "gene_type"),
        "VARCHAR(1) CHARACTER SET latin1 COLLATE latin1_swedish_ci" = c("ppos", "amino_acid", "charge","known_site"),
        "BIGINT(8) NULL DEFAULT NULL" = c("row_names"),
        "INT(8) NULL DEFAULT NULL" = c("DEseq2Normalized","CoVarOrder","row_id", "cluster_order","cluster_id", "count_cut_off", "^position$", "raw_counts"),
        "DECIMAL(6,3) NULL DEFAULT NULL" = c("Integrated_Samples_","_avg_diff_","Ind_var","add_counts_",  "lg2BaseMean","lg2baseMean","contrast_P_PCA_estimatePCA","^CoVar$","NES", "logFC", "lg2_avg","norm_counts", "intensity", "^int", "iBAQ","^localization_prob$", "stat", "lg10p","contrast_x_", "add_MA_cts_MA"),
        "DECIMAL(6,5) NULL DEFAULT NULL" = c("r2_PCA","padj", "pvalue","^pep$", "p_value_PCA")
    ),
    new.table = TRUE,
    cols2Index = c(Obio@parameterList$geneIDcolumn)
)

killDbConnections()

###############################################################################
# Create microwebsite                                                         #
###############################################################################

## Color samples by sample group ##
# Get sample order from database.table
## oder samples according to PC1
# df.pca <- df.pca[order(df.pca$pca1, decreasing = FALSE),]
# sample.order <- paste0("norm_counts_", df.pca$sample_id)


sample.order <- (names(Obio@databaseTable)[grep("norm_counts_", names(Obio@databaseTable))])


lg2Avg.order <- sort(names(Obio@databaseTable)[grep("lg2_avg", names(Obio@databaseTable))])

# sample.order <- c(
#     'norm_counts_Avg_log10_Expr_All_Ctrl','norm_counts_Avg_log10_Expr_All_Prad','norm_counts_Avg_log10_Expr_Cl_0_C','norm_counts_Avg_log10_Expr_Cl_0_P','norm_counts_Avg_log10_Expr_Cl_10_C','norm_counts_Avg_log10_Expr_Cl_10_P','norm_counts_Avg_log10_Expr_Cl_1_C','norm_counts_Avg_log10_Expr_Cl_1_P','norm_counts_Avg_log10_Expr_Cl_2_C','norm_counts_Avg_log10_Expr_Cl_2_P','norm_counts_Avg_log10_Expr_Cl_3_C','norm_counts_Avg_log10_Expr_Cl_3_P','norm_counts_Avg_log10_Expr_Cl_4_C','norm_counts_Avg_log10_Expr_Cl_4_P','norm_counts_Avg_log10_Expr_Cl_5_C','norm_counts_Avg_log10_Expr_Cl_5_P','norm_counts_Avg_log10_Expr_Cl_6_C','norm_counts_Avg_log10_Expr_Cl_6_P','norm_counts_Avg_log10_Expr_Cl_7_C','norm_counts_Avg_log10_Expr_Cl_7_P','norm_counts_Avg_log10_Expr_Cl_8_C','norm_counts_Avg_log10_Expr_Cl_8_P','norm_counts_Avg_log10_Expr_Cl_9_C','norm_counts_Avg_log10_Expr_Cl_9_P'
# )

## Sort database.table ##

#lg2Order <- gsub("norm_counts_", "lg2_avg_", sample.order)

orderVec <- names(Obio@databaseTable)
orderVec <- orderVec[!(orderVec %in% sample.order)]
orderVec <- orderVec[!(orderVec %in% lg2Avg.order)]

orderVec <- c(
    orderVec,
    sample.order,
    lg2Avg.order
)

Obio@databaseTable <- Obio@databaseTable[,orderVec]

## Get relevant colors ##
sampleIDs <- paste0("_", unique(OsC@meta.data$sampleID))

sample.colors <- sample.order
for (i in 1:length(sampleIDs)){
    sample.colors <- gsub(as.vector(sampleIDs[i]), "", sample.colors)    
}


library(scales)
color.groups <- unique(sample.colors)
#color.groups <- hue_pal()(length(color.groups))

#dfDesign[["sample_color"]] <- "orange"
nSampleGroups <- length(unique(color.groups))

# library(RColorBrewer)
# selcol <- colorRampPalette(brewer.pal(12,"Set3"))

color.sel = hue_pal()(length(color.groups))

for (i in 1:length(color.groups)){
    sample.colors[grep(color.groups[i], sample.colors)] <- color.sel[i]
}



if (!exists("gsea.cat.lines")){
    gsea.cat.lines <- ""
    downloadCatEnrichmentFNxlsx <- ""
} else {
    downloadCatEnrichmentFNxlsx <- paste0("outputs/", Obio@parameterList$project.code, ".enriched.categories.txt")
}

if (!exists("timecourse.cat.lines")){
    timecourse.cat.lines <- NA
}

#library(SBwebtools)
setwd(Obio@parameterList$localWorkDir)

#library(SBwebtools)
webSiteDir <- Obio@parameterList$localWorkDir

project.code <- gsub(substr(Obio@parameterList$project_id, 1, 3), "p", Obio@parameterList$project_id)


###############################################################################
## Set default X colum                                                       ##

pos <- grep(
  "^add_counts_Avg_log10_Expr_all$", 
  names(Obio@databaseTable)
)

if (length(pos) == 1){
  defaultX <- "add_counts_Avg_log10_Expr_all"
} else {
  defaultX <- "contrast_X_logFC_Expressed_in_N_Percent_Cells"
}

pos <- grep(
  "^add_counts_Avg_log10_Expr_C_0$",
  names(Obio@databaseTable)
)

if (length(pos) == 1){
  defaultY <- "add_counts_Avg_log10_Expr_C_0"
} else {
  defaultY <- sample.order[1]
}
## d

###############################################################################
## Create default plot selection                                             ##

plotSelection <- c(
    names(Obio@databaseTable)[grep("_logFC_",names(Obio@databaseTable))],
    
    names(Obio@databaseTable)[grep("_lg10p_",names(Obio@databaseTable))],

    names(Obio@databaseTable)[grep("lg2BaseMean",names(Obio@databaseTable))],
    names(Obio@databaseTable)[grep("corCoef",names(Obio@databaseTable))],
    names(Obio@databaseTable)[grep("add_counts",names(Obio@databaseTable))],
    names(Obio@databaseTable)[grep("norm_counts",names(Obio@databaseTable))],
    names(Obio@databaseTable)[grep("contrast_P_PCA_estimatePCA",names(Obio@databaseTable))],
    names(Obio@databaseTable)[grep("add_MA_cts_MA",names(Obio@databaseTable))]
)

chunk1 <- plotSelection[grep("_logFC_", plotSelection)]
chunkh2 <- plotSelection[grep("_lg10p_", plotSelection)]
chunk2a <-chunkh2[grep("_lg10p_PCA", chunkh2)]
chunk2b <- chunkh2[!(chunkh2  %in% chunk2a)]

chnk2c <- plotSelection[grep("_corCoef", plotSelection)]

chunk3 <- plotSelection[grep("_PCA_est", plotSelection)]
chunk3a <- plotSelection[grep("add_MA_cts_MA", plotSelection)]
chunk4 <- plotSelection[grep("_lg2BaseMean_", plotSelection)]

chunk5 <- sort(plotSelection[grep("Avg_log10_Expr", plotSelection)])
chunk6 <- sort(plotSelection[grep("add_counts_PCA", plotSelection)])

# rmVec <- grep("_LRT_", chunk4)
# if (length(rmVec) > 0){
#     chunk4 <- chunk4[-rmVec]
# }

plotSelectionNew <- c(
    chunk1,
    chunk2b,
    chnk2c,
    chunk2a,
    chunk4,
    chunk5,
    chunk6,
    chunk3,
    chunk3a,
    chunk4
)

plotAdd <- plotSelection[!(plotSelection %in% plotSelectionNew)]

plotSelection <- c(
    defaultX,
    defaultY,
    plotSelectionNew,
    plotAdd
)


## Done creating default plot selection                                      ##
###############################################################################

###############################################################################
## Do default Venn selection                                                 ##
defaultVennSel <- names(Obio@databaseTable)
defaultVennSel <- c(
    defaultVennSel[grep("contrast_", defaultVennSel)],
    defaultVennSel[grep("add_venn_", defaultVennSel)],
    defaultVennSel[grep("_TsCor_", defaultVennSel)],
    defaultVennSel[grep("_Residuals", defaultVennSel)]

)

rmVec <- c(
    grep("contrast_P_", defaultVennSel),
    grep("_lg2BaseMean_", defaultVennSel),
    grep("_padj_LRT_", defaultVennSel),
    grep("_lg10p_", defaultVennSel),
    grep("_stat_", defaultVennSel),
    grep("pValueCor_", defaultVennSel)
)

if (length(rmVec) > 0){
    defaultVennSel <- defaultVennSel[-rmVec]
}

## Add lrt
defaultVennSel <- c(
    defaultVennSel,
    names(Obio@databaseTable)[grep("contrast_L_lg10p", names(Obio@databaseTable))]
)

## Done with LRT selection                                                   ##
###############################################################################


###############################################################################
## Deploy next generation website                                            ##
createSettingsFile(
    obj = Obio,
    df.data = Obio@databaseTable,
    defaultXcolName = defaultX,
    defaultYcolName = defaultY,
    primDataTable = Obio@parameterList$rnaseqdbTableName,
    pcaDbTable = Obio@parameterList$PCAdbTableName, 
    sample.order = sample.order, #set to "" to go with default sorting
    heatmapSampleOrder = "",
    sample.names = "", # default is sample.order
    count.sample.colors = sample.colors,
    ptm.colum = "",
    count.table.headline = Obio@parameterList$count.table.headline,
    count.table.sidelabel = Obio@parameterList$count.table.sidelabel,
    venn.slider.selector.strings = defaultVennSel,
    plot.selection.strings = plotSelection,
    webSiteDir = paste0(hpc.mount, "Stefan/protocol_files/github/biologic/src/experiments/"),
    upper_heatmap_limit = 3, 
    lower_heatmap_limit = -3,
    heamap.headline.text = Obio@parameterList$heamap.headline.text,
    project_id = Obio@parameterList$project_id
)

setwd(Obio@parameterList$localWorkDir)
## Done deploying next generation website                                    ##
###############################################################################

save(Obio, 
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
     )
)

print("Obio Object saved.")

save(OsC,
    file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
        ".Seurat.Robj"
     )
)

```




```{r create_report_params, eval=T, results="asis"}
documentationParams <- list(

    "title" = "Pseudobulk Differential Gene Expression Analysis",
    "subtitle" =  "",
    "abstract" = ""

)

Obio@parameterList$lims.id <- "SC19235"

## Try to retrieve project data from db ##
library(RMySQL)
db.pwd2 <- "_asf_"
db.user2 <- "asf"
host2 <- "ms1.thecrick.org"
projectParams <- documentationParams

tryCatch({
    dbDB = dbConnect(drv = RMySQL::MySQL(), user = db.user2, password = db.pwd2, host = host2, dbname = "asf");
dfProposal =  dbGetQuery(dbDB, paste0("SELECT * FROM asf_proposals WHERE project_name ='",Obio@parameterList$lims.id,"'"));
dbDisconnect(dbDB)
  }, error = function(x) {
    message("Project Database could not be reached or has no entry in Obio@parameterList$lims.id for this analysis.")
   
})

if (exists("dfProposal")){
  if (nrow(dfProposal) == 1){
      if (!is.na(dfProposal[1,"ProjectAlias"]) & dfProposal[1,"ProjectAlias"] != ""){
          projectParams[["title"]] = paste0(dfProposal[1,"ProjectAlias"], " - ", dfProposal[1,"project_name"])
      }
      
      if (!is.na(dfProposal[1,"project_user"]) & dfProposal[1,"project_user"] != ""){
          projectParams[["subtitle"]] = paste0(dfProposal[1,"user_lab"], " Lab - ", dfProposal[1,"project_user"])
          projectParams[["subtitle"]] <- gsub("^ Lab - ", "", projectParams[["subtitle"]])
          
      }
      
      if (!is.na(dfProposal[1,"proposal_text"]) & dfProposal[1,"proposal_text"] != ""){
          projectParams[["abstract"]] = dfProposal[1,"proposal_text"]
         
          
      }
  }
}
   
## Escape all special characters
projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("([.|()/\\^{}+$*?]|\\[|\\])", " ", x)
) 

projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("\\\n", " ", x)
) 


#projectParams$title <- "Title"
# projectParams$abstract <- "This is the QC section."
#projectParams$subtitle <- "Abstract"

```



## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
sessionInfo()
```

---
title: "`r projectParams$title`"
subtitle:  Pseudobulk Differential Gene Expression Analyses on Single-cell Results
author:
    - Bioinformatics: Stefan Boeing^[The Francis Crick Institute, stefan.boeing@crick.ac.uk]
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

abstract: |
    In this section pseudo-bulk differential gene expression analyses for this project are presented. The method used here was developed recently by the Huber lab and is desribed in detail [here](https://www.biorxiv.org/content/10.1101/2020.08.13.249623v2).

---