---

output: 
    html_document:
        highlight: default
        theme: paper
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 3
        toc_float: true
        css: src/assets/style/style.css

---
    
    

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```


```{r, include=FALSE}


## Recommended (Crick) R-setup: 
# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a

```

```{r, eval=TRUE, echo=F, results=F}
###############################################################################
##                                                                           ##
# if (!requireNamespace("remotes")){
#     install.packages("remotes")
# }
# 
# remotes::install_github("rstudio/renv")

if (file.exists("renv.lock")){
    renv::restore(prompt = FALSE)
} else {
    renv::init()
}

## Done                                                                      ##
###############################################################################


```

```{r, eval=TRUE, echo=T, results=F}
###############################################################################
##  Load packages                                                            ##
library(Seurat)
library(biologicSeqTools2)
library(dplyr)

VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

## Load bioLOGIC object
ObioFN <- paste0("../", list.files("..")[grep(".bioLOGIC.Robj", list.files(".."))])
load(ObioFN)

## Make sure paths are set correctly in the Obio object
Obio <- Obio %>% 
    biologicSeqTools2::setMountingPoint()  %>% 
    biologicSeqTools2::setAnalysisPaths() %>% 
    biologicSeqTools2::setCrickGenomeAndGeneNameTable() %>% 
    biologicSeqTools2::createAnalysisFolders() %>% 
    biologicSeqTools2::setDataBaseParameters()

## Load Seurat object ##
SeuratFN <- paste0("../", list.files("..")[grep(".Seurat.Robj", list.files(".."))])

load(SeuratFN)

##                                                                           ##
###############################################################################

###############################################################################
## Retrieve database password                                                ##

FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")
    dbTable <- read.delim(
      FN,
      header = F,
      sep = "\t",
      stringsAsFactors = F
    )
#}
db.pwd <- as.vector(dbTable[1,1])

## Done                                                                      ##
###############################################################################

###############################################################################
##  Set variables                                                            ##

# Plotting parameters
figureCount <- 1
legendDotSize <- 5

## Define project ID ##
project_id <- Obio@parameterList$project_id


## Set directory for report tables
reportTableDir <- Obio@parameterList$reportTableDir
reportFigDir <- Obio@parameterList$reportFigDir

localWorkDir <- Obio@parameterList$localWorkDir

## Create url string
if (Obio@parameterList$host == "10.27.241.234"){
    urlString <- "biologic.thecrick.org"
} else {
    urlString <- "biologic.crick.ac.uk"
}

shinyBaseUrl <- "https://bioinformatics.crick.ac.uk/shiny/users/boeings/"

shinyURL <- paste0(
    shinyBaseUrl,
    project_id,
    "_app/"
)            
   
## Set file paths ##
baseFN <- paste0(
   project_id, 
   ".DGE.table.xlsx"
)

FNrel <- paste0("report_tables/", baseFN)

outPutFN <- paste0(
     reportTableDir,
     baseFN
)

tableLink <- paste0(
    'https://',
    urlString,
    '/mdata/',project_id, '/html/', 
    FNrel
)

tableString <- paste0(
    'An Excel table with the DGE results can be downloaded ',
    tableLink
)

tableString <- paste0(
   ' An Excel table with the DGE results can be downloaded <a href="',tableLink,'" target="_blank">here</a>. '
)
  

##  Set variables                                                            ##
###############################################################################



###############################################################################
## Function doDGEsc                                                          ##

#' @title doDGEsc
#' @description Function to do glmgampoi differential gene expression. When using this function, make sure the glmGamPoi R-package is installed on your system: renv::install("renv::install("bioc::glmGamPoi") 
#' @param obj Seurat object for the DGE analysis
#' @param countSlot Seurat object assay to use to retrieve count data.
#' @param DGEselCol Meta data column in the above Seurat object to use for DGE sample group definitions. 
#' @param colName Name for the DGE comparison in the meta data
#' @param contrastTag contrastTag
#' @param DGEsampleList List specifying the DGE comparisons. 
#' @import Seurat
#' @export


setGeneric(
    name="doDGEsc",
    def=function(
        obj,
        countSlot = "RNA",
        DGEselCol = "sub_clusters_ExNeurons",
        colName = "DGE_name",
        contrastTag = "contrast_1_",
        DGEsampleList = list(
            "M1" = c(5),
            "M2" = c(1)
        )
    ) {
    
    ############################################################################
    ## Add DGE cell selection to meta-data                                    ##
    colName <- unlist(colName)
    DGEselCol <- unlist(DGEselCol)

    pos <- grep(colName, names(obj@meta.data))
    if (length(pos) > 0){
        obj@meta.data <- obj@meta.data[,-pos]
    }

    mColName <- substr(paste0("meta_", colName), 1, 63)
    obj@meta.data[[mColName]] <- "Rest" 
    
    for (i in 1:length(DGEsampleList)){
        obj@meta.data[obj@meta.data[,DGEselCol] %in% DGEsampleList[[i]], mColName] <- names(DGEsampleList)[i]
    }
    
    dfMeta <- obj@meta.data
    dfMeta <- dfMeta[dfMeta[,mColName] != "Rest",]
    
    ##                                                                        ##
    ############################################################################
    
    ############################################################################
    ## Get counts for DGE                                                     ##
    cellVec <- dfMeta$cellID
    group <-  as.vector(dfMeta[,mColName])
    
    dfMatrix <- obj[[countSlot]]@counts
    dfMatrix <- data.matrix(dfMatrix[,cellVec])
    dfMatrix <- dfMatrix[rowSums(dfMatrix) != 0, ]
    
    ## Pseudobulk DGE
    
    ## LRT ##
    # fit <- glm_gp(dfMatrix, design = group)
    # res <- test_de(fit, reduced_design = ~ 1)
    
    sample_labels <- group
    
    sample_labels[sample_labels == names(DGEsampleList)[1]] <- paste0(sample_labels[sample_labels == names(DGEsampleList[1])], "_", 1:length(sample_labels[sample_labels == names(DGEsampleList)[1]]))
    
    sample_labels[sample_labels == names(DGEsampleList)[2]] <- paste0(sample_labels[sample_labels == names(DGEsampleList[2])], "_", 1:length(sample_labels[sample_labels == names(DGEsampleList)[2]]))
    
    cell_type_lables <- group
    unique(group)
    
    fit <- glmGamPoi::glm_gp(
        dfMatrix, 
        design = group
    )
    
    comparison <- names(DGEsampleList)
    
    contrastString <- (paste0(comparison[1], " - ", comparison[2]))
    
    res <- glmGamPoi::test_de(fit, contrast = contrastString,
            pseudobulk_by = sample_labels, 
            #subset_to = cell_type_labels == "T-cells",
            #n_max = 4, 
            sort_by = pval, 
            decreasing = FALSE
    )
    
    dfRes <- res[order(res$lfc),]
    
    ## Remove extreme outliers ##
    dfRes <- dfRes[dfRes$lfc > -100 & dfRes$lfc < 100, ]
    dfRes[["padj"]] <- dfRes$adj_pval
    minP <- min(dfRes$padj[dfRes$padj != 0])
    dfRes[dfRes$padj == 0, "padj"] <- minP
    dfRes[["lg10p"]] <- -1 * log10(dfRes$padj)
    
    
    comp_1 <- dfRes
    comp_1[["gene"]] <- dfRes$name
    names(comp_1) <- gsub("lfc", paste0(contrastTag, "logFC_" ,colName), names(comp_1))
    
    names(comp_1) <- gsub("padj", paste0(contrastTag, "padj_",colName), names(comp_1))
    names(comp_1) <- gsub("lg10p", paste0(contrastTag, "lg10p_",colName), names(comp_1))
    
    selVec <- c(
        "gene",
        names(comp_1)[grep(contrastTag, names(comp_1))]
    )
    
    comp_1 <- unique(comp_1[,selVec])
    
    ## Cut all strings in comp_1 to a maximum of 63 charcters 
    names(comp_1) <- sapply(names(comp_1), function(x) substr(x, 1, 63))
    
    
    
    ##                                                                        ##
    ############################################################################
    
    
    
    returnList <- list(
      "OsC" = obj,
      DGEtable = comp_1
    )
    
    return(returnList)
})


## End doDGEsc Function                                                      ##
###############################################################################

###############################################################################
## Define differential gene expression comparisons in a list                 ##
## One list entry per comparison                                             ##


DGEtagVec <- as.vector(NULL, mode = "character")

DGEinputList <- list(
  "comp_1" = list(
      "name"    = "DGE_in_vitro_vs_in_vivo",
      "DGEselCol" = "meta_in_vitro_in_vivo",
      "DGEsampleList" = list(
          "in_vitro"   = c("inVitro"),
          "in_vivo"   = c("inVivo")
      )
  ),
  "comp_2" = list(
      "name"    = "DGE_in_vitro_C3_vs_in_vivo_C3",
      "DGEselCol" = "cluster_sample",
      "DGEsampleList" = list(
          "in_vitro_C3"   = c("C3_HuTEC135",  "C3_HuTEC136"),
          "in_vivo_C3"   = c("C3_Epip10sub", "C3_Epip14sub", "C3_Epip6sub",  "C3_Epip7sub")
      )
  )
  
  
)




###############################################################################
## Function do scDGE                                                         ##

doScDGEonList <- function(
    OsC,
    tag,
    DGEinputList,
    DGEtagVec = as.vector(NULL, mode = "character")
    
){
    resList <- doDGEsc(
        obj = OsC,
        DGEselCol = unlist(DGEinputList[[i]]["DGEselCol"]),
        colName = unlist(DGEinputList[[i]]["name"]),
        contrastTag = paste0("contrast_",i,"_"),
        DGEsampleList = DGEinputList[[i]][["DGEsampleList"]]
    )
    
    OsC <- resList$OsC
    
    ###########################################################################
    ## Add average intensity for MA-Plot to DGE table                        ##
    
    OsC_DGE <- OsC
    OsC_DGE@meta.data[["DGE_sel"]] <- 0
    
    OsC_DGE@meta.data[OsC_DGE@meta.data[,unlist(DGEinputList[[i]]["DGEselCol"])] != "", "DGE_sel"] <- 1
    
    #row.names(OsC_DGE@meta.data) <- OsC_DGE@meta.data[,"cellID"]
    
    OsC_DGE <- subset(x = OsC_DGE, subset = DGE_sel == 1)
    
    
    Idents(OsC_DGE) <- "DGE_sel"
    
    cluster.averages <- Seurat::AverageExpression(
        OsC_DGE, 
        return.seurat = TRUE
    )
    
    dfAvgExpr <- data.frame(cluster.averages[["RNA"]]@data)
    dfAvgExpr[["gene"]] <- row.names(dfAvgExpr)
    row.names(dfAvgExpr) <- NULL
    
    nameTag <- paste0("add_MA_cts_MA_Avg_", unlist(DGEinputList[[i]]$name))
    names(dfAvgExpr) <- gsub("all", nameTag, names(dfAvgExpr))
    
    dfDataTable <- resList$DGEtable
    
    dfDataTable <- merge(
        dfDataTable, 
        dfAvgExpr, 
        by.x = "gene",
        by.y = "gene",
        all =TRUE
    )
    
    dfDataTable[is.na(dfDataTable)] <- 0
    names(dfDataTable) <- sapply(names(dfDataTable), function(x) substr(x, 1, 63))
    
    DGEtagVec <- c(
      DGEtagVec,
      tag
    )
    
    resultList <- list(
        "dfDataTable" = dfDataTable,
        "DGEtagVec" = DGEtagVec,
        "OsC" = OsC
    )
    
    return(resultList)
    
}

## Done do scDGE                                                             ##
###############################################################################

###############################################################################
## DGE Analysis                                                              ##

DGEtagVec = as.vector(NULL, mode = "character")
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

## This does the DGE and creates an overview UMAP figure ##
for (i in 1:length(DGEinputList)){
    
    tag <- paste0(unlist(DGEinputList[[i]]["name"]))
    
    resList <- doScDGEonList(
        OsC = OsC,
        tag = tag,
        DGEinputList = DGEinputList,
        DGEtagVec = DGEtagVec
    )
  
    dfDataTable <- resList[["dfDataTable"]]
    DGEtagVec <- resList[["DGEtagVec"]]
    OsC <- resList[["OsC"]]
    
    ## Add results to biologic object 
    Obio@dataTableList[[tag]] <- data.frame(NULL)
    Obio@dataTableList[[tag]] <- dfDataTable
    
    ## Done with DGE Analysis                                                ##
    ###########################################################################
    
    ###########################################################################
    ## Create DGE Selection plot                                             ##
    dfPlot <- OsC@meta.data
    
    
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }

    pos <- grep("cellID", names(dfPlot))
      if (length(pos) == 0){
      dfPlot[["cellID"]] <- row.names(dfPlot)
    }

    dfPlot$UMAP_1 <- NULL
    dfPlot$UMAP_2 <- NULL
            
    ## Get UMAP coordinates ##
    coord <- data.frame(OsC@reductions$umap@cell.embeddings)
    coord[["cellID"]] <- row.names(coord)
    coord <-coord[coord$cellID %in% dfPlot$cellID, ]
            
    dfPlot <- merge(dfPlot, coord, by.x = "cellID", by.y="cellID", all=T)
    dfPlot[is.na(dfPlot)] <- 0
    dfPlot <- dfPlot[dfPlot$UMAP_1 != 0 & dfPlot$UMAP_2 != 0,]
            
            
    ## Add cluster colors ##
    dfPlot[["Cluster"]] <- dfPlot[,paste0("meta_", unlist(DGEinputList[[i]]["name"]))]
    dfPlot$Cluster[is.na(dfPlot$Cluster)] <- ""
    dfPlot[dfPlot$Cluster == "", "Cluster"] <- "Rest"
    
    label <- as.vector(gsub("DGE_", "", unlist(DGEinputList[[i]]["name"])))
    
    
    redGroup <- names(DGEinputList[[i]]$DGEsampleList)[1] #as.vector(unlist(strsplit(label, "_vs_"))[1])
    blueGroup <- names(DGEinputList[[i]]$DGEsampleList)[2] #as.vector(unlist(strsplit(label, "_vs_"))[2])
    greyGroup <- "Rest"
    
    ## Assigning colors ##
    dfCol <- data.frame(name = unique(dfPlot$Cluster), col = rep("", length(unique(dfPlot$Cluster))), stringsAsFactors = F)
    dfCol[dfCol$name == redGroup, "col"] <- "#000080"
    dfCol[dfCol$name == blueGroup, "col"] <- "#009900"
    dfCol[dfCol$name == greyGroup, "col"] <- "grey"
    colVec <- dfCol$col
    names(colVec) <- dfCol$name
    
    dfColExtra <- dfCol[dfCol$col == "",]
    if (nrow(dfColExtra) > 0){
        eColVec <- dfColExtra$name
        library(scales)
        eCols = hue_pal()(length(eColVec))
        for (l in 1:length(eColVec)){
            dfCol[dfCol$name == eColVec[l], "col"] <- eCols[l]
        }
    }
    
    maxX <- 1.1*max(dfPlot$UMAP_1, na.rm = T)
    minX <- 1.1*min(dfPlot$UMAP_1, na.rm = T)
    maxY <- 1.1*max(dfPlot$UMAP_2, na.rm = T)
    minY <- 1.1*min(dfPlot$UMAP_2, na.rm = T)            

    #colVec <- dfPair$sampleColor
    #names(colVec) <- dfPair$sampleName
    levels <- as.vector(dfCol$name)
    levelsRest <- levels[grep("Rest", levels)]
    levels <- levels[levels != "Rest"]
    levels <- c(levels, levelsRest)
    
    dfPlot$Cluster <- factor(dfPlot$Cluster, levels = levels)

    dotsize  = 1
    if (nrow(dfPlot) > 10000){
        dotsize  = 0.5
    } else if (nrow(dfPlot) > 20000){
        dotsize = 0.25
    } else if (nrow(dfPlot) > 50000){
        dotsize = 0.1
    }

    ## make plot ##
    plotList[[tag]] <- ggplot2::ggplot(
        data=dfPlot[dfPlot$included == "+",], 
        ggplot2::aes(UMAP_1, UMAP_2, color=Cluster)) + 
        ggplot2::geom_point( shape=16, size = as.numeric(dotsize)) + 
        ggplot2::xlab("UMAP1") + 
        ggplot2::ylab("UMAP2") + 
        ggplot2::theme_bw()  +  
        ggplot2::theme(
          axis.text.y   = ggplot2::element_text(size=8),
          axis.text.x   = ggplot2::element_text(size=8),
          axis.title.y  = ggplot2::element_text(size=8),
          axis.title.x  = ggplot2::element_text(size=8),
          axis.line = ggplot2::element_line(colour = "black"),
          panel.border = ggplot2::element_rect(colour = "black", fill=NA, size=1),
          plot.title = ggplot2::element_text(hjust = 0.5, size = 12),
          legend.title = ggplot2::element_blank()
        ) + 
        ggplot2::guides(col = ggplot2::guide_legend(override.aes = list(shape = 16, size = legendDotSize))) +     
        ggplot2::ggtitle(paste0("Cell Selection for ", gsub("_", " ", unlist(DGEinputList[[i]]["name"])))) + 
        ggplot2::xlim(minX, maxX) + 
        ggplot2::ylim(minY, maxY) + 
        ggplot2::coord_fixed(ratio=1) + 
        ggplot2::scale_colour_manual("DGE Groups" ,values = colVec) + 
        ggplot2::guides(col = ggplot2::guide_legend(override.aes = list(shape = 16, size = legendDotSize)))


      ## Add colors if specified ##

            
      ## Make plot figure ##
      FNbase <- paste0(tag, VersionPdfExt)
      FN <- paste0(reportFigDir, FNbase)
      FNrel <- paste0("report_figures/", FNbase)
              
      pdf(FN)
          print(plotList[[tag]])
      dev.off()
              
        # link <- paste0('<a href="https://',urlString,'/',Obio@parameterList$project_id,'/pca?x_axis=UMAP_1&y_axis=UMAP_2" target="_blank">here</a>')  
        
      if (exists("shinyURL") & !is.null(shinyURL)){
          link <- paste0(
                  'An interactive version of this figure with additional viewing options can be found <a href="',shinyURL,'?_inputs_&y_axis=%22UMAP_2%22&x_axis=%22UMAP_1%22&colorBy=%22',paste0("meta_", tag),'%22&splitByColumn=%22all%22" target="_blank">here</a>. '
          )
                
      } else {
          link <- ""
      }
              
      figLegend <- paste0(
      '**Figure ', 
      figureCount, 
                ':** ',
                ' UMAP showing all cells from all samples together. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. ',
                'An interactive version of this figure can be found ', link
            )
            
      figureCount <- figureCount + 1
            
      NewChnk <- paste0(
          "#### ", tag,
          "\n```{r Sample_DGE_",
          tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
          figLegend,"'}\n",
          "\n",
          "\n print(plotList[['",tag,"']])",
          "\n cat(  '\n')",
          "\n\n\n```\n"   
      )
            
      chnkVec <- c(
          chnkVec,
          NewChnk
      )
  
      ## Done creating DGE selection plot                                      ##
      ###########################################################################
      
      
      print(paste0("Done with ", tag))
}

## Done DGE                                                                  ##
###############################################################################

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
``` 

### Differential Gene Expression Comparions Cell Selections {`r tabVar`}
In this section the cells that were used for the DGE comparisons are highlighted. 

```{r CellTypePlot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))

``` 


```{r data_DGE_summary, echo=F, eval=T, warning=FALSE, results="asis"}

###############################################################################
## Assemble DGE summary table                                                ##
#dfTemp <- Obio@dataTableList[["DGE_table"]] 

wb <- openxlsx::createWorkbook()

hs1 <- openxlsx::createStyle(
    fontColour = "#ffffff",
    fgFill = "#000000",
    halign = "CENTER",
    textDecoration = "Bold"
)

###############################################################################
## Add Create DGE result Excel Spreadsheet                                   ##

for (i in 1:length(DGEtagVec)){
  
  ## Add to Excel workbook ##
  dfOutput <-   Obio@dataTableList[[DGEtagVec[i]]]
  orderVec <- names(dfOutput)
  dfOutput[["gene_FeatureViewLink"]] <- paste0(
      shinyURL,
      "?gene=",
      dfOutput$gene
  )
  orderVec <- orderVec[orderVec != "gene"]
  orderVec <- c("gene", "gene_FeatureViewLink", orderVec)
  dfOutput <- dfOutput[, orderVec]
  class(dfOutput$gene_FeatureViewLink) <- c(class(dfOutput$gene_FeatureViewLink), "hyperlink")
  
  sheetName <- substr(DGEtagVec[i],1,30)
  
  openxlsx::addWorksheet(
      wb, 
      sheetName = sheetName
  )
  
  
  openxlsx::freezePane(wb, sheetName ,  firstActiveRow = 2)
  openxlsx::writeData(wb, i, dfOutput, startRow = 1, startCol = 1, headerStyle = hs1)
  ## Done adding to Excel workbook ##
  
  if (i ==1){
      dfTemp <- Obio@dataTableList[[DGEtagVec[i]]]
  } else {
      dfTemp <- merge(
        dfTemp, 
        Obio@dataTableList[[DGEtagVec[i]]],
        by.x = "gene",
        by.y = "gene",
        all =TRUE
      )
      dfTemp[is.na(dfTemp)] <- 0
  
  }
}

## Save workbook ##
baseFN <- paste0(
   project_id, 
   ".DGE.table.xlsx"
)


outPutFN <- paste0(
     reportTableDir,
     baseFN
)

if (!dir.exists(reportTableDir)){
    dir.create(reportTableDir, recursive=T)
}

openxlsx::saveWorkbook(
        wb,
        outPutFN ,
        overwrite = TRUE
)

Obio@dataTableList[["DGE_table"]] <- NULL
Obio@dataTableList[["DGE_table"]] <- dfTemp


## Done creating xlsx spreadsheet                                            ##
###############################################################################
```


```{r, echo=F, eval=T, warning=FALSE, results="asis"}


###############################################################################
## Make scDGE MA-plots                                                       ##
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

dfAllPlots <- Obio@dataTableList[["DGE_table"]] 

for (i in 1:length(DGEtagVec)){
    tag <- paste0("MA_Plot_", DGEtagVec[i])  
    label <- gsub("_", " ", DGEtagVec[i])
    
    selVec <- c(
        "gene",
        names(dfAllPlots)[grep(paste0(DGEtagVec[i], "$"), names(dfAllPlots))]
    )
    
    dfPlot <- dfAllPlots[,selVec]
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }
    
    logFCtwoSD <- 2*sd(dfPlot[,grep("_logFC_", names(dfPlot))])
    
    dfPlot[["DGE_Status"]] <- "Unchanged"
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] > logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Up"
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] < -1 *logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Down"
    
    colVec <- c("red", "blue", "black")
    names(colVec) <- c("Up", "Down", "Unchanged")
  
    
    dfPlot[["X"]] <- dfPlot[,grep("add_MA_cts_MA_Avg", names(dfPlot))]
    dfPlot[["Y"]] <- dfPlot[,grep("_logFC", names(dfPlot))]
    
    dotsize  = 1.25
    if (nrow(dfPlot) > 10000){
      dotsize  = 1
    } else if (nrow(dfPlot) > 20000){
      dotsize = 0.75
    } else if (nrow(dfPlot) > 50000){
      dotsize = 0.5
    }

    logFClims <- ceiling(max(abs(dfPlot$Y)))
    
    ## Add to-regulated gene names ##
    topNgenes <- 10
    dfPlot <- dfPlot[order(dfPlot$Y, decreasing = T), ]
    topGenes <- as.vector(dfPlot[1:topNgenes,"gene"])
    
    dfPlot <- dfPlot[order(dfPlot$Y, decreasing = F), ]
    bottomGenes <- as.vector(dfPlot[1:topNgenes,"gene"])
    
    dfPlot[["label"]] <- ""
    dfPlot[dfPlot[,"gene"] %in% c(topGenes, bottomGenes), "label"] <- dfPlot[dfPlot[,"gene"] %in% c(topGenes, bottomGenes), "gene"]
    
    
    
    plotList[[tag]] <- ggplot2::ggplot(data=dfPlot[dfPlot$included == "+",], 
        ggplot2::aes(X, Y, color=DGE_Status, label = "label")) + 
        ggplot2::geom_hline(yintercept = c(logFCtwoSD, -1 * logFCtwoSD), col = "darkgrey", lty=2) + 
        ggplot2::geom_point( shape=16, size = as.numeric(dotsize),alpha = 0.7) + 
        ggplot2::xlab(paste0("log10 Avg Intensity ", gsub("_", " ",label))) + 
        ggplot2::ylab(paste0("log2-Fold Change ", gsub("_", " ",label))) + 
        ggplot2::theme_bw()  +  
        ggplot2::theme(
            axis.text.y   = ggplot2::element_text(size=8),
            axis.text.x   = ggplot2::element_text(size=8),
            axis.title.y  = ggplot2::element_text(size=8),
            axis.title.x  = ggplot2::element_text(size=8),
            axis.line = ggplot2::element_line(colour = "black"),
            panel.border = ggplot2::element_rect(colour = "black", fill=NA, size=1),
            plot.title = ggplot2::element_text(hjust = 0.5, size = 12)) + 
        ggplot2::guides(col = ggplot2::guide_legend(override.aes = list(shape = 16, size = legendDotSize))) + 
        ggplot2::ggtitle(paste0("MA-stype Plot for DGE Comparison ", gsub("_", " ", label))) + 
      ggplot2::ylim(logFClims, -1 * logFClims) + 
      ggplot2::scale_color_manual("DGE Status" ,values = colVec) + 
      ggplot2::scale_y_continuous(breaks=c(seq(-20,20,2)))


      ## Add labels 
      # plotList[[tag]] <- plotList[[tag]] + ggrepel::geom_text_repel(size = 3)

            

      FNbase <- paste0(tag, VersionPdfExt)
      FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
      FNrel <- paste0("report_figures/", FNbase)
            
      pdf(FN)
          print(plotList[[tag]])
      dev.off()
      
      xLabel <- names(dfPlot[grep("MA_Avg", names(dfPlot))])      
      yLabel <- names(dfPlot[grep("_logFC_", names(dfPlot))])  
      link <- paste0('<a href="https://',urlString,'/',Obio@parameterList$project_id,'/scatterplot?x_axis=',xLabel,'&y_axis=',yLabel,'" target="_blank">here</a>')  
            
      figLegend <- paste0(
      '**Figure ', 
      figureCount, 
                ':** ',
                ' MA-style plot for differential gene expression comparison ',gsub("_", " ", DGEtagVec[i]),'. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. ',
                'An interactive version of this figure can be found ', link, '. ',
                tableString
            )
            
      figureCount <- figureCount + 1
            
      NewChnk <- paste0(
          "#### ", tag,
          "\n```{r MA_DGE_",
          tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
          figLegend,"'}\n",
          "\n",
          "\n print(plotList[['",tag,"']])",
          "\n cat(  '\n')",
          "\n\n\n```\n"   
      )
            
      chnkVec <- c(
          chnkVec,
          NewChnk
      )

      ## Done creating DGE selection plot                                      ##
      ###########################################################################
    
    
    
    
}


## Done MA- plots                                                            ##
###############################################################################

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
``` 

### DGE MA-style Plots {`r tabVar`}

```{r DGE_MA_plot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))

``` 


```{r data_Volcano_plots, echo=F, eval=T, warning=FALSE, results="asis"}
###############################################################################
## Make Volcano plots
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

dfAllPlots <- Obio@dataTableList[["DGE_table"]] 

for (i in 1:length(DGEtagVec)){
    tag <- paste0("Volcano_Plot_", DGEtagVec[i])  
    label <- gsub("_", " ", DGEtagVec[i])
    
    selVec <- c(
        "gene",
        names(dfAllPlots)[grep(paste0(DGEtagVec[i], "$"), names(dfAllPlots))]
    )
    
    dfPlot <- dfAllPlots[,selVec]
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }
    
    logFCtwoSD <- 2*sd(dfPlot[,grep("_logFC_", names(dfPlot))])
    
    dfPlot[["DGE_Status"]] <- "Unchanged"
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] > logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Up"
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] < -1 *logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Down"
    
    colVec <- c("red", "blue", "black")
    names(colVec) <- c("Up", "Down", "Unchanged")
  
    
    dfPlot[["X"]] <- dfPlot[,grep("_logFC", names(dfPlot))]
    dfPlot[["Y"]] <- dfPlot[,grep("_lg10p_", names(dfPlot))]
    
    dotsize  = 1.25
    if (nrow(dfPlot) > 10000){
      dotsize  = 1
    } else if (nrow(dfPlot) > 20000){
      dotsize = 0.75
    } else if (nrow(dfPlot) > 50000){
      dotsize = 0.5
    }

    logFClims <- ceiling(max(abs(dfPlot$X)))
    
    ## Add to-regulated gene names ##
    topNgenes <- 10
    dfPlot <- dfPlot[order(dfPlot$Y, decreasing = T), ]
    topGenes <- as.vector(dfPlot[1:topNgenes,"gene"])
    
    dfPlot <- dfPlot[order(dfPlot$Y, decreasing = F), ]
    bottomGenes <- as.vector(dfPlot[1:topNgenes,"gene"])
    
    dfPlot[["label"]] <- ""
    dfPlot[dfPlot[,"gene"] %in% c(topGenes, bottomGenes), "label"] <- dfPlot[dfPlot[,"gene"] %in% c(topGenes, bottomGenes), "gene"]
    
    
    plotList[[tag]] <- ggplot2::ggplot(
        data=dfPlot[dfPlot$included == "+",], 
        ggplot2::aes(X, Y, color=DGE_Status, label = "label")) + 
        ggplot2::geom_vline(xintercept = c(logFCtwoSD, -1 * logFCtwoSD), col = "darkgrey", lty=2) +   
        ggplot2::geom_vline(xintercept = 0, col = "darkgrey") + 
        ggplot2::geom_hline(yintercept = 0, col = "darkgrey") + 
        ggplot2::geom_point( shape=16, size = as.numeric(dotsize),alpha = 0.7) + 
        ggplot2::xlab(paste0("log2-Fold Change ", gsub("_", " ",label))) + 
        ggplot2::ylab(paste0("-log10(adj p-value) ", gsub("_", " ",label))) + 
        ggplot2::theme_bw()  +  ggplot2::theme(
          axis.text.y   = ggplot2::element_text(size=8),
          axis.text.x   = ggplot2::element_text(size=8),
          axis.title.y  = ggplot2::element_text(size=8),
          axis.title.x  = ggplot2::element_text(size=8),
          axis.line = ggplot2::element_line(colour = "black"),
          panel.border = ggplot2::element_rect(colour = "black", fill=NA, size=1),
          plot.title = ggplot2::element_text(hjust = 0.5, size = 12)) + 
      ggplot2::guides(col = ggplot2::guide_legend(override.aes = list(shape = 16, size = legendDotSize))) + 
      ggplot2::ggtitle(paste0("Volcano Plot for DGE Comparison ", gsub("_", " ", label))) + 
      ggplot2::ylim(logFClims, -1 * logFClims) + 
      ggplot2::scale_color_manual("DGE Status" ,values = colVec) + 
      ggplot2::scale_y_continuous(breaks=c(seq(0,400,50))) + 
      ggplot2::scale_x_continuous(breaks=c(seq(-400,400,2)))


      ## Add labels 
      #plotList[[tag]] <- plotList[[tag]] + ggrepel::geom_text_repel(size = 3)
            

      FNbase <- paste0(tag, VersionPdfExt)
      FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
      FNrel <- paste0("report_figures/", FNbase)
            
      pdf(FN)
          print(plotList[[tag]])
      dev.off()
      
      xLabel <- names(dfPlot[grep("_logFC_", names(dfPlot))])      
      yLabel <- names(dfPlot[grep("_lg10p_", names(dfPlot))])  
      link <- paste0('<a href="https://',urlString,'/',Obio@parameterList$project_id,'/scatterplot?x_axis=',xLabel,'&y_axis=',yLabel,'" target="_blank">here</a>')  
            
      
      
      figLegend <- paste0(
      '**Figure ', 
      figureCount, 
                ':** ',
                ' Volcano plot for differential gene expression comparison ',gsub("_", " ", DGEtagVec[i]),'. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. ',
                'An interactive version of this figure can be found ', link, '. ',
                tableString
            )
            
      figureCount <- figureCount + 1
            
      NewChnk <- paste0(
          "#### ", tag,
          "\n```{r Volcano_DGE_",
          tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
          figLegend,"'}\n",
          "\n",
          "\n print(plotList[['",tag,"']])",
          "\n cat(  '\n')",
          "\n\n\n```\n"   
      )
            
      chnkVec <- c(
          chnkVec,
          NewChnk
      )

      ## Done creating DGE selection plot                                      ##
      ###########################################################################
    
    
    
    
}


## Done Volcano plots
###############################################################################

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
``` 

### DGE Volcano Plots {`r tabVar`}

```{r DGE_volcano_plot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))

``` 


```{r create-pca-enrichment-data, echo=T, eval=TRUE, warning=FALSE, results=F, error=F}
Obio@dbDetailList$db.user <- "babs"
## Create enriched genes list ##
EnrichedGenesList <- list()
dfAllPlots <- Obio@dataTableList[["DGE_table"]] 

if (Obio@parameterList$geneIDcolumn != "mgi_symbol" & Obio@parameterList$geneIDcolumn != "hgnc_symbol") {
            queryGS <- "hgnc_symbol" 
        } else {
            queryGS <- Obio@parameterList$geneIDcolumn
        }

for (i in 1:length(DGEtagVec)){
    tag <- paste0("MA_Plot_", DGEtagVec[i])  
    tagGLpos <- paste0(DGEtagVec[i], "_pos") 
    tagGLneg <- paste0(DGEtagVec[i], "_neg") 
    
    selVec <- c(
        "gene",
        names(dfAllPlots)[grep(paste0(DGEtagVec[i], "$"), names(dfAllPlots))]
    )
    
    dfPlot <- dfAllPlots[,selVec]
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }
    
    logFCtwoSD <- 2*sd(dfPlot[,grep("_logFC_", names(dfPlot))])
    
    dfPlot[["DGE_Status"]] <- "Unchanged"
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] > logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Up"
    EnrichedGenesList[[tagGLpos]] <- unique(dfPlot[dfPlot$DGE_Status == "Up", "gene"])
    
    dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] < -1 *logFCtwoSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "DGE_Status"] <- "Down"
    EnrichedGenesList[[tagGLneg]] <- unique(dfPlot[dfPlot$DGE_Status == "Down", "gene"])
} 

#library(knitr::)
#library(ggplot2)

#save.image("temp.RData")
#library(clusterProfiler)

    gmtList <- list()
    pos <- grep("clusterSigEnrichmentList", slotNames(Obio))
    
    if (length(pos) > 0){
    if (is.null(Obio@clusterSigEnrichmentList)){
      dbtableList <- list(
          "Cell Type Signatures" = "mysigdb_sc_sig",
          "Cell Type Signatures" = "cibersort_L22",
          "GO-MF" = "mysigdb_c5_MF",
          "Pathways" = "mysigdb_c2_1329_canonical_pathways",
          "Protein Complexes" = "networkcategories"
      )
    } else {
        dbtableList <- Obio@clusterSigEnrichmentList
    }
    } else {
      dbtableList <- list(
          "Cell Type Signatures" = "mysigdb_sc_sig",
          "Cell Type Signatures" = "cibersort_L22",
          "GO-MF" = "mysigdb_c5_MF",
          "Pathways" = "mysigdb_c2_1329_canonical_pathways",
          "Protein Complexes" = "networkcategories"
      )
    }
    
    
    rmVec <- grep("Cell Type Signatures", names(dbtableList))
    if (length(rmVec) > 0){
        dbtableList <- dbtableList[-rmVec]
    }
    
    for (i in 1:length(dbtableList)){
        
        dfTemp <- unique(import.db.table.from.db(
            host = Obio@dbDetailList$host,
            dbname = Obio@dbDetailList$ref.cat.db,
            dbtable = dbtableList [[i]],
            password = db.pwd,
            user = Obio@dbDetailList$db.user
        ))
        
        ## Remove duplicated entries ##
        dfTemp <- dfTemp[!(duplicated(dfTemp$cat_name)),]
        
        rmVec <- grep("temp_", dfTemp$cat_type)
        if (length(rmVec) > 0){
            dfTemp <- dfTemp[-rmVec, ]
        }
        
        dfTemp <- unique(dbcat2gmt(
            dfTemp, # As downloaded from reference_categories_db_new database
            gene.id.column = queryGS
        ))
        
        dfTemp <- unique(dfTemp[!duplicated(as.vector(dfTemp[,1])), ])
        
        write.table(
            dfTemp,
            "temp.gmt.txt",
            row.names = F, 
            sep = "\t",
            col.names = F,
            quote = F
        )
        
        CPgmt <- clusterProfiler::read.gmt("temp.gmt.txt")
        unlink("temp.gmt.txt")
        CPgmt <- unique(CPgmt[CPgmt$gene != "", ])
        
        gmtList[[dbtableList[[i]]]] <- CPgmt
    }
    
    ## Edit collection names for plot
    names(gmtList) <- gsub("mysigdb_", "",names(gmtList))
    names(gmtList) <- gsub("c2_1329_canonical_p", "P",names(gmtList))
    names(gmtList) <- gsub("sc_sig", "CellSig",names(gmtList))
    names(gmtList) <- gsub("cibersort_L22", "CellSig",names(gmtList))
    names(gmtList) <- gsub("c5_", "GO_",names(gmtList))
    names(gmtList) <- gsub("networkcategories", "Complexes",names(gmtList))
    ## Done creating gmt list
    ###########################
    
    ## Select colors ##
    #library(scales)
    enrCols <- scales::hue_pal()(length(gmtList))
    names(enrCols) <- substr(names(gmtList),1,10)



plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")



for (j in 1:length(DGEtagVec)){
    posTestGeneSet <- as.vector(
        unique(
            EnrichedGenesList[[paste0(DGEtagVec[j], "_pos")]]
        )
    )
    
    
    negTestGeneSet <- as.vector(
        unique(
            EnrichedGenesList[[paste0(DGEtagVec[j], "_neg")]]
        )
    )
   
    
    ###########################################################################
    ## Create GMT file for category enrichment                               ##
    
    ###########################
    ## Create gmt list
    ## Retrieve gmt files from database
    ## Add custom gmt files
    
    
    
    
    ## Done                                                                  ##
    ###########################################################################
     
    #library(clusterProfiler)
    #library(ggplot2)
    #library(tidyr)
        
        if (Obio@parameterList$geneIDcolumn != "mgi_symbol" & Obio@parameterList$geneIDcolumn != "hgnc_symbol") {
            queryGS <- "hgnc_symbol" 
        } else {
            queryGS <- Obio@parameterList$geneIDcolumn
        }
        
        if (Obio@parameterList$host == "10.27.241.234"){
            urlString <- "biologic.thecrick.org"
        } else {
            urlString <- "biologic.crick.ac.uk"
        }
    
    colVec <- c("#009900", "#000080")
    pvalueCutoff <- 0.5
    topMaxCat <- 10
    
    ## Get background gene set ##
    #backgroundGeneVec <- row.names(OsC[["RNA"]]@counts)
    if ((length(posTestGeneSet) >= 3) | (length(negTestGeneSet) >= 3)){
        ## Do enrichment ##
        first <- TRUE
        if (length(posTestGeneSet) >= 3){
            for (k in 1:length(gmtList)){
                    egmt <- data.frame(
                        clusterProfiler::enricher(
                            negTestGeneSet, 
                            TERM2GENE=gmtList[[k]],
                            pvalueCutoff = pvalueCutoff
                        )
                    )
                    if (!is.null(egmt)){
                        if (nrow(egmt) > 0){
                            egmt[["Collection"]] <- substr(names(gmtList)[k], 1,10)
                        }
                        if (first){
                            dfTempEnriched <- egmt    
                            first <- FALSE
                        } else {
                            dfTempEnriched <- rbind(
                                dfTempEnriched, 
                                egmt
                            )    
                        }
                        
                    }
            }
           
            if (exists("dfTempEnriched") && nrow(dfTempEnriched) > 0){
                dfTempEnriched[["direction"]] <- "positive"
                dfTempEnriched[["log10FDR"]] <- log10(dfTempEnriched$p.adjust)
                dfTempEnriched <- dfTempEnriched[order(dfTempEnriched$log10FDR, decreasing = F),]
                dfTempEnriched <- na.omit(dfTempEnriched)
                dfEnriched <- dfTempEnriched
                
                if (nrow(dfTempEnriched) > topMaxCat){
                    dfTempEnriched <- dfTempEnriched[1:topMaxCat, ]
                }
            }
          
            
        } # end positive
            
            ## Now the negative side ##
            if (length(negTestGeneSet) >= 3){
            first <- TRUE
            for (k in 1:length(gmtList)){
                    egmt <- data.frame(
                        clusterProfiler::enricher(
                            posTestGeneSet, 
                            TERM2GENE=gmtList[[k]],
                            pvalueCutoff = pvalueCutoff
                        )
                    )
                    if (!is.null(egmt)){
                        if (nrow(egmt) > 0){
                            egmt[["Collection"]] <- substr(names(gmtList)[k], 1,10)
                        }
                        if (first){
                            dfTempEnrichedNeg <- egmt    
                            first <- FALSE
                        } else {
                            dfTempEnrichedNeg <- rbind(
                                dfTempEnrichedNeg, 
                                egmt
                            )    
                        }
                        
                    } 
            }
            if (exists("dfTempEnrichedNeg") && nrow(dfTempEnrichedNeg) > 0){
                dfTempEnrichedNeg[["direction"]] <- "negative"
                dfTempEnrichedNeg[["log10FDR"]] <- -1*log10(dfTempEnrichedNeg$p.adjust)
                dfTempEnrichedNeg <- dfTempEnrichedNeg[order(dfTempEnrichedNeg$log10FDR, decreasing = T),]
                dfTempEnrichedNeg <- na.omit(dfTempEnrichedNeg)
                dfEnrichedNeg <- dfTempEnrichedNeg
                
                if (nrow(dfTempEnrichedNeg) > topMaxCat){
                    dfTempEnrichedNeg <- dfTempEnrichedNeg[1:topMaxCat, ]
                }
            }
            } # end negative
        
            
            
            ## Make plot 
            if ((nrow(dfTempEnriched) > 0) | (nrow(dfTempEnrichedNeg) > 0)){
            
            
            
            
            dfSel <- rbind(
                dfTempEnriched,
                dfTempEnrichedNeg
            )
            
            dfSel <- na.omit(dfSel)
            dfSel <- dfSel[order(dfSel$log10FDR),]
            dfSel$log10FDR <- round(dfSel$log10FDR, 2)
            
            dfSel[["Category"]] <- ""
            dfSel[dfSel$log10FDR >= 0, "Category"] <- "Enr."
            dfSel[dfSel$log10FDR < 0, "Category"] <- "Depl."
            
            for (l in 1:nrow(dfSel)){
                if (nchar(dfSel[l, "ID"]) > 30){
                    part1 <- substr(dfSel[l, "ID"], 1, 30)
                    part2 <- substr(dfSel[l, "ID"], 31, 60)
                    dfSel[l, "ID"] <- paste0(part1, " \n ", part2)
                  
                }
            }
            
            
            #dfSel$Term <- gsub("\\(GO", "\\\n\\(GO", dfSel$Term)
            
            dfSel$ID <- factor(dfSel$ID, levels = unique(dfSel$ID))
            
            
            
            plotList[[paste0("PCA_ENR_", j)]] <- ggplot2::ggplot(
                data=dfSel, ggplot2::aes(x= ID, y=log10FDR, fill=Collection, order=log10FDR)
            ) + ggplot2::geom_bar(stat="identity", colour="black"
            ) + ggplot2::coord_flip() + ggplot2::scale_fill_manual(values=enrCols
            ) + ggplot2::theme_bw(
            )  +  ggplot2::theme(
                axis.text.y   = ggplot2::element_text(size=8),
                axis.text.x   = ggplot2::element_text(size=8),
                axis.title.y  = ggplot2::element_text(size=8),
                axis.title.x  = ggplot2::element_text(size=8),
                axis.line = ggplot2::element_line(colour = "black"),
                panel.border = ggplot2::element_rect(colour = "black", fill=NA, size=1),
                plot.title = ggplot2::element_text(hjust = 0.5, size = 12)
            )  + ggplot2::labs(title = paste0("Comparison ", DGEtagVec[j]," enriched genes") ,y = "-log10(FDR)", x = ""
            ) + ggplot2::geom_hline(yintercept = c(-log10(0.05), log10(0.05)), color = "grey", size=0.5, lty=2
            ) + ggplot2::geom_hline(yintercept = 0, color = "black", size=0.5
            ) 
            cat("  \n")
            
            
            
            ## Save to file ##
            FNbase <- paste0("DGE_comparison_", DGEtagVec[j],".enriched.genes", VersionPdfExt)
            FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
            FNrel <- paste0("report_figures/", FNbase)
            
           
            pdf(FN)
            print(plotList[[paste0("PCA_ENR_", j)]])
            dev.off()
            
            link <- paste0(
                '<a href="https://', urlString, '/',
                Obio@parameterList$project_id,
                '/category-view?category_type=GO-BP" target="_blank">CategoryView</a>'
            )
            
            ## Create R markdown chunk ##
            figLegend <- paste0(
                '**Figure ', 
                figureCount, 
                '**: Category enrichment analysis for the top genes that have  <font color = "',colVec[2],'"> the most positive </font> and <font color = "',colVec[1],'">the most negative</font> PCA loading values in dimension ', 
               DGEtagVec[j],
                ' associated with them. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. To view these gene sets in the context of your data, go to ',link,' and find these categories using the search box.'
            )
            figureCount <- figureCount + 1 
            
            NewChnk <- paste0(
                "#### ", DGEtagVec[j],
                "\n```{r enrichr_",
                j,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                figLegend,"'}\n",
                "\n",
                "\n print(plotList[['",paste0("PCA_ENR_", j),"']])",
                "\n cat(  '\n')",
                "\n\n\n```\n"   
            )
            
            chnkVec <- c(
                chnkVec,
                NewChnk
            )
            
            
            ####################################################################
            ## Create full output table                                       ##
            dfResTemp <- rbind(
                dfEnriched,
                dfEnrichedNeg
            )
            
            dfResTemp[["comparison"]] <- tag
            
            if (j==1){
                dfResTable <- dfResTemp
            } else {
                dfResTable <- rbind(
                    dfResTable, 
                    dfResTemp
                )
            }
            ## Done                                                           ##
            ####################################################################
        }
            
            
            ## done with plot 
            
    } ## Done with per dimension loops
}        
     
 
 

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
```

### Category Enrichments Differential Gene Expression Results {`r tabVar`} 
```{r create-cat-enrichment-plot, echo=T, eval=TRUE, warning=FALSE, results='asis'}
###############################################################################
## Do category enrichment on clusters                                        ##
cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))
## Done doing enrichment on clusters                                         ##
###############################################################################


```


<!-- Essential N: Upload meta data table -->
```{r child = 'src/modules/db_tools/upload.meta.data.table.to.DB.Rmd', eval=FALSE, error = FALSE}

```


```{r documentation_saving, echo=F, eval=TRUE, warning=FALSE, results=F, error = FALSE} 

save(Obio,
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
     )
)

print("Obio Object saved.")

save(OsC,
    file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
        ".Seurat.Robj"
     )
)

###############################################################################
## Rerunning of part C is required                                           ##

###############################################################################
## Create Excel output files                                                 ##

## Add ref-gene column ##
#dfTemp[["ref_gene"]] <- ""
#dfTemp[dfTemp$gene %in% refGenes, "ref_gene"] <- "+"

library(openxlsx)

dfTemp <- Obio@dataTableList[["DGE_table"]] 

if (!dir.exists( paste0(
   Obio@parameterList$html_local,
   "report_tables/"))){
  dir.create(paste0(
   Obio@parameterList$html_local,
   "report_tables/"))
}

baseFN <- paste0(
   Obio@parameterList$project_id, 
   ".DGE.table.xlsx"
)


outPutFN <- paste0(
     Obio@parameterList$html_local,
      "report_tables/",
     baseFN
)
  
 
FNrel <- paste0("report_table/", baseFN)
    
sheetName <- substr(paste0(Obio@parameterList$project_id, "_subcluster_DGE"),1,30)

wb <- createWorkbook()
    addWorksheet(wb, sheetName)
    freezePane(wb, sheetName ,  firstActiveRow = 2)
    
    ## Filter is inactivated, as it does not appear to be compatible with the current version of Excel
    #addFilter(wb, 1, row = 1, cols = 1:ncol(dfOutput))
    
## Style headers ##
hs1 <- createStyle(
  fontColour = "#ffffff",
  fgFill = "#000000", 
  halign = "CENTER", 
  textDecoration = "Bold"
)
    
writeData(wb, 1,dfTemp, startRow = 1, startCol = 1, headerStyle = hs1)
    
saveWorkbook(
  wb, 
  outPutFN , 
  overwrite = TRUE
)

## Done creating Excel output files                                          ##
###############################################################################

###############################################################################
## Create cat enrichment table                                               ##

## Add ref-gene column ##
#dfTemp[["ref_gene"]] <- ""
#dfTemp[dfTemp$gene %in% refGenes, "ref_gene"] <- "+"

library(openxlsx)

## Enrichment result table is called dfResTable

if (!dir.exists( paste0(
   Obio@parameterList$html_local,
   "report_tables/"))){
  dir.create(paste0(
   Obio@parameterList$html_local,
   "report_tables/"))
}

baseFN <- paste0(
   Obio@parameterList$project_id, 
   ".DGE.cat.enrichment.table.xlsx"
)


outPutFN <- paste0(
     Obio@parameterList$html_local,
      "report_tables/",
     baseFN
)
  
 
FNrel <- paste0("report_table/", baseFN)
    
sheetName <- substr(paste0(Obio@parameterList$project_id, "_cat_enrichment_DGE"),1,30)

wb <- openxlsx::createWorkbook()
    openxlsx::addWorksheet(wb, sheetName)
    openxlsx::freezePane(wb, sheetName ,  firstActiveRow = 2)
    
    ## Filter is inactivated, as it does not appear to be compatible with the current version of Excel
    #addFilter(wb, 1, row = 1, cols = 1:ncol(dfOutput))
    
## Style headers ##
hs1 <- openxlsx::createStyle(
  fontColour = "#ffffff",
  fgFill = "#000000", 
  halign = "CENTER", 
  textDecoration = "Bold"
)
    
openxlsx::writeData(wb, 1,dfTemp, startRow = 1, startCol = 1, headerStyle = hs1)
    
saveWorkbook(
  wb, 
  outPutFN , 
  overwrite = TRUE
)

## Done creating Excel output files                                          ##
###############################################################################

```




```{r create_report_params, eval=T, results="asis"}
documentationParams <- list(

    "title" = "Pseudobulk Differential Gene Expression Analysis",
    "subtitle" =  "",
    "abstract" = ""

)

Obio@parameterList$lims.id <- "SC19235"

## Try to retrieve project data from db ##
library(RMySQL)
db.pwd2 <- "_asf_"
db.user2 <- "asf"
host2 <- "ms1.thecrick.org"
projectParams <- documentationParams

tryCatch({
    dbDB = dbConnect(drv = RMySQL::MySQL(), user = db.user2, password = db.pwd2, host = host2, dbname = "asf");
dfProposal =  dbGetQuery(dbDB, paste0("SELECT * FROM asf_proposals WHERE project_name ='",Obio@parameterList$lims.id,"'"));
dbDisconnect(dbDB)
  }, error = function(x) {
    message("Project Database could not be reached or has no entry in Obio@parameterList$lims.id for this analysis.")
   
})

if (exists("dfProposal")){
  if (nrow(dfProposal) == 1){
      if (!is.na(dfProposal[1,"ProjectAlias"]) & dfProposal[1,"ProjectAlias"] != ""){
          projectParams[["title"]] = paste0(dfProposal[1,"ProjectAlias"], " - ", dfProposal[1,"project_name"])
      }
      
      if (!is.na(dfProposal[1,"project_user"]) & dfProposal[1,"project_user"] != ""){
          projectParams[["subtitle"]] = paste0(dfProposal[1,"user_lab"], " Lab - ", dfProposal[1,"project_user"])
          projectParams[["subtitle"]] <- gsub("^ Lab - ", "", projectParams[["subtitle"]])
          
      }
      
      if (!is.na(dfProposal[1,"proposal_text"]) & dfProposal[1,"proposal_text"] != ""){
          projectParams[["abstract"]] = dfProposal[1,"proposal_text"]
         
          
      }
  }
}
   
## Escape all special characters
projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("([.|()/\\^{}+$*?]|\\[|\\])", " ", x)
) 

projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("\\\n", " ", x)
) 


#projectParams$title <- "Title"
# projectParams$abstract <- "This is the QC section."
#projectParams$subtitle <- "Abstract"

```



## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
sessionInfo()
```

---
title: "`r projectParams$title`"
subtitle:  Pseudobulk Differential Gene Expression Analyses on Single-cell Results
author:
    - Bioinformatics: Stefan Boeing^[The Francis Crick Institute, stefan.boeing@crick.ac.uk]
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

abstract: |
    In this section pseudo-bulk differential gene expression analyses for this project are presented. The method used here was developed recently by the Huber lab and is desribed in detail [here](https://www.biorxiv.org/content/10.1101/2020.08.13.249623v2).

---
