---
title: "Label Transfer"
author: "Stefan Boeing"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

output: 
    html_document:
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 3
        toc_float: true
        css:
    
always_allow_html: yes

---
    
    
```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```


## Part B Database


```{r hpc_notes, include=FALSE}

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;R;

## Get interactive session ##
#  srun --time=08:00:00 --mem=40G -p int --pty bash

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;R;

# sbatch --time=12:00:00 --wrap "module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;Rscript runD.r" --job-name="rD" -p hmem --mem=300G -o rD.slurm >> commands.txt

# --mem-per-cpu=14G -p hmem --pty bash

```


```{r populate_meta_data_database, eval=TRUE, echo=F, results=F}
## libraries ##
library(tidyverse)
library(Seurat)

VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

#Create the environment and load a suitable version of R, e.g. so:
FN <- paste0(hpc.mount, "Projects/reference_data/documentation/BC.parameters.txt")
dbTable <- read.delim(
    FN, 
    sep = "\t",
    stringsAsFactors = F
)

db.pwd <- as.vector(dbTable[1,1])






# source(
#     paste0(
#         hpc.mount,
#         "Stefan/protocol_files/github/boeings/packages/packageSourceCode/SBwebtools.pckg.r"
#     )
# )

# source(
#     paste0(
#         hpc.mount,
#         "Stefan/protocol_files/github/boeings/packages/scTools/scTools.r"
#     )
# )

source("assets/R/scTools.r")
source("assets/R/SBwebtools.pckg.r")


if (length(.libPaths()) > 2){
    .libPaths(.libPaths()[2:3])
}

ObioFN <- paste0("../", list.files("..")[grep(".bioLOGIC.Robj", list.files(".."))])

if (file.exists(ObioFN)){
    load(paste0(ObioFN))
    print(paste0("Obio object ", Obio@parameterList$localWorkDir, ObioFN, " exists and is loaded."))
} else {
    exit()
}

## Reset paths to local environment


Obio <- setMountingPoint(Obio)
Obio <- setAnalysisPaths(Obio)
Obio <- setCrickGenomeAndGeneNameTable(Obio)
Obio <- createAnalysisFolders(
    Obio
)
Obio <- setDataBaseParameters(Obio)

## Load Seurat object
## Goal: create a table, that can be added via the 'gene' column to the main 
## database table

ObioFN <- paste0("../", list.files("..")[grep(".Seurat.Robj", list.files(".."))])

load(ObioFN)

```



```{r populate_expr_database, eval=TRUE, echo=F, results=F}
###############################################################################
## Otional add extra data set                                                ##

#Create transfer list
transfer.list <- SplitObject(
    OsC, 
    split.by = "meta_Zeppelli_or_Tepe"
)

# for (i in 1:length(transfer.list)) {
#     transfer.list[[i]] <- NormalizeData(transfer.list[[i]], verbose = FALSE)
#     transfer.list[[i]] <- FindVariableFeatures(transfer.list[[i]], selection.method = "vst", 
#         nfeatures = 2000, verbose = FALSE)
# }

# Query = destination
transfer.query <- transfer.list[["Tepe"]]



    # pcaproject: Project the PCA from the reference onto the query. We recommend using PCA when reference and query datasets are from scRNA-seq
    # 
    # lsiproject: Project the LSI from the reference onto the query. We recommend using LSI when reference and query datasets are from scATAC-seq. This requires that LSI has been computed for the reference dataset, and the same features (eg, peaks or genome bins) are present in both the reference and query. See RunTFIDF and RunSVD
    # 
    # rpca: Project the PCA from the reference onto the query, and the PCA from the query onto the reference (reciprocal PCA projection).
    # 
    # cca: Run a CCA on the reference and query


transfer.anchors <- Seurat::FindTransferAnchors(
    reference = transfer.list[["Zeppelli"]], 
    query = transfer.query, 
    reduction = "cca",
    dims = 1:Obio@parameterList$singleCellSeuratNpcs4PCA
)

#transfer.anchors <- Obio@dataTableList$referenceList[["sampleAnchors"]]

# Oref <- transfer.list[["Zeppelli"]]
predictions <- Seurat::TransferData(
    anchorset = transfer.anchors, 
    refdata = transfer.list[["Zeppelli"]]$meta_Zeppilli_subclusters, 
    weight.reduction = "cca",
    dims = 1:Obio@parameterList$singleCellSeuratNpcs4PCA
)
transfer.query <- AddMetaData(transfer.query, metadata = predictions)

dfAdd <- transfer.query@meta.data[,c("predicted.id", "cellID")]
names(dfAdd) <- gsub("predicted.id", "meta_Zeppilli_subclusters", names(dfAdd))

## Get other half ##
dfAdd2 <- transfer.list[["Zeppelli"]]@meta.data[,c("meta_Zeppilli_subclusters", "cellID")]

dfAddFinal <- rbind(dfAdd, dfAdd2)
names(dfAddFinal) <- gsub("meta_Zeppilli_subclusters","meta_LT_Zeppilli_subclusters", names(dfAddFinal))
row.names(dfAddFinal) <- dfAddFinal$cellID
dfAddFinal$cellID <- NULL



OsC <- addDf2seuratMetaData(
    OsC,
    dfAdd = dfAddFinal
) 
   
save(OsC, 
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".Seurat.Robj"
     )
)


## Upload new metadata table ##
#```{r child = 'src/modules/db_tools/upload.meta.data.table.to.DB.Rmd', eval=TRUE}
#
#```




#database.table[is.na(database.table)] <- 0
## Done adding extra dataset                                                 ##
###############################################################################

###############################################################################
## Otional add extra data set                                                ##

#Create transfer list
transfer.list <- SplitObject(
    OsC, 
    split.by = "meta_Zeppelli_or_Tepe"
)

# for (i in 1:length(transfer.list)) {
#     transfer.list[[i]] <- NormalizeData(transfer.list[[i]], verbose = FALSE)
#     transfer.list[[i]] <- FindVariableFeatures(transfer.list[[i]], selection.method = "vst", 
#         nfeatures = 2000, verbose = FALSE)
# }

# Query = destination
transfer.query <- transfer.list[["Zeppelli"]]



    # pcaproject: Project the PCA from the reference onto the query. We recommend using PCA when reference and query datasets are from scRNA-seq
    # 
    # lsiproject: Project the LSI from the reference onto the query. We recommend using LSI when reference and query datasets are from scATAC-seq. This requires that LSI has been computed for the reference dataset, and the same features (eg, peaks or genome bins) are present in both the reference and query. See RunTFIDF and RunSVD
    # 
    # rpca: Project the PCA from the reference onto the query, and the PCA from the query onto the reference (reciprocal PCA projection).
    # 
    # cca: Run a CCA on the reference and query


transfer.anchors <- Seurat::FindTransferAnchors(
    reference = transfer.list[["Tepe"]], 
    query = transfer.query, 
    reduction = "cca",
    dims = 1:Obio@parameterList$singleCellSeuratNpcs4PCA
)

#transfer.anchors <- Obio@dataTableList$referenceList[["sampleAnchors"]]

# Oref <- transfer.list[["Zeppelli"]]
predictions <- Seurat::TransferData(
    anchorset = transfer.anchors, 
    refdata = transfer.list[["Tepe"]]$meta_ClusterName_Tepe, 
    weight.reduction = "cca",
    dims = 1:Obio@parameterList$singleCellSeuratNpcs4PCA
)
transfer.query <- AddMetaData(transfer.query, metadata = predictions)

dfAdd <- transfer.query@meta.data[,c("predicted.id", "cellID")]
names(dfAdd) <- gsub("predicted.id", "meta_ClusterName_Tepe", names(dfAdd))

## Get other half ##
dfAdd2 <- transfer.list[["Tepe"]]@meta.data[,c("meta_ClusterName_Tepe", "cellID")]

dfAddFinal <- rbind(dfAdd, dfAdd2)
names(dfAddFinal) <- gsub("meta_ClusterName_Tepe","meta_LT_ClusterName_Tepe", names(dfAddFinal))
row.names(dfAddFinal) <- dfAddFinal$cellID
dfAddFinal$cellID <- NULL



OsC <- addDf2seuratMetaData(
    OsC,
    dfAdd = dfAddFinal
) 
   
save(OsC, 
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".Seurat.Robj"
     )
)


## Upload new metadata table ##
#```{r child = 'src/modules/db_tools/upload.meta.data.table.to.DB.Rmd', eval=TRUE}
#
#```




#database.table[is.na(database.table)] <- 0
## Done adding extra dataset                                                 ##
###############################################################################
```


```{r make_figures, eval=TRUE, echo=F, results=F}


```

```{r saveobject, eval=TRUE, echo=T, results=F}
### Will save Obio object here, so it can be re-used with different parameters
save(Obio, 
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
     )
)

print("Obio Object saved.")

# save(OsC,
#     file = paste0(
#          Obio@parameterList$localWorkDir,
#          Obio@parameterList$project_id,
#         ".Seurat.Robj"
#      )
# )

```

## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
sessionInfo()
```