---
title: "Add additional datasets"
author: "Stefan Boeing"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

output: 
    html_document:
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 3
        toc_float: true
        css:
    
always_allow_html: yes

---
    
    
```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```


## Part B Database


```{r hpc_notes, include=FALSE}

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a;R

```

```{r, eval=TRUE, echo=F, results=F}
VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")


#Create the environment and load a suitable version of R, e.g. so:
library(dplyr)
library(Seurat)
library(tidyverse)
library(biologicSeqTools)
library(biologicToolsSC)

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}


FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")

db.pwd <- FN %>% 
    readr::read_delim(., delim = "\t", col_names = F) %>% 
    dplyr::pull(., var = 1) %>% 
    dplyr::first()



###############################################################################
##                                                                           ##

if (file.exists("renv.lock")){
    renv::restore()
} else {
    renv::init()
}

## Done                                                                      ##
###############################################################################

#install.packages("remotes")
# renv::install("decusInLabore/biologicSeqTools")
# renv::install("decusInLabore/biologicToolsSC")
# renv::install("decusinlabore/biologicViewerSC")
# source("assets/R/scTools.r")
# source("assets/R/SBwebtools.pckg.r")


# if (length(.libPaths()) > 2){
#     .libPaths(.libPaths()[2:3])
# }

ObioFN <- paste0("../", list.files("..")[grep(".bioLOGIC.Robj", list.files(".."))])

if (file.exists(ObioFN)){
    load(paste0(ObioFN))
    print(paste0("Obio object ", Obio@parameterList$localWorkDir, ObioFN, " exists and is loaded."))
} else {
    exit()
}

## Reset paths to local environment
Obio <- Obio %>% 
    biologicSeqTools2::setMountingPoint()  %>% 
    biologicSeqTools2::setAnalysisPaths() %>% 
    biologicSeqTools2::setCrickGenomeAndGeneNameTable() %>% 
    biologicSeqTools2::createAnalysisFolders() %>% 
    biologicSeqTools2::setDataBaseParameters()


## Load Seurat object
SeuratFN <- paste0(Obio@parameterList$localWorkDir,list.files(Obio@parameterList$localWorkDir)[grep(".Seurat.Robj", list.files(Obio@parameterList$localWorkDir))])

if (file.exists(SeuratFN)){
    load(SeuratFN)
    print(paste0("Obio object ", Obio@parameterList$localWorkDir,SeuratFN, " exists and is loaded."))
    
} else {
    exit()
}

row.names(OsC@meta.data) <- OsC@meta.data$cellID

if (Obio@parameterList$scIntegrationMethod == "SCT"){
    resultSlot = "SCT"
} else {
    resultSlot = "RNA"
}

```



```{r, eval=TRUE, echo=F, results=F}
###############################################################################
## Optional add extra data set                                                ##

## Goal: create a table, that can be added via the 'gene' column to the main 
## database table


FN <- "/camp/stp/babs/working/boeings/Projects/hillc/danielle.park/437_fibroblast_GSE154778/auxiliarydata/WT_v_Inhba_Inbbb_null_combined.20210429.txt"

dfAddCounts <- FN %>% 
    readr::read_delim(., delim = "\t") 


###############################################################################
## Load reference file from database                                         ##

dfAddCounts <- biologicSeqTools2::import.db.table.from.db(
    dbname = "chl_data",
    dbtable = "RN19081_rna_seq_table",
    host = Obio@dbDetailList$host,
    user = Obio@dbDetailList$db.user,
    password = db.pwd
)

selVec <- c(
    #"ENSMUSG",
    "hgnc_symbol",
    #"mgi_symbol",
    names(dfAddCounts)[grep("contrast_4", names(dfAddCounts))]
)

dfAddCounts <- unique(dfAddCounts[,selVec])
names(dfAddCounts) <- gsub("contrast_4_", "", names(dfAddCounts))

dfAddCounts$stat_Activin_Nullcomb_vs_WT <- NULL

## Use only those with a human homologue
dfAddCounts <- dfAddCounts[dfAddCounts$hgnc_symbol != "",]

## Done                                                                      ##
###############################################################################


dfAddCounts <- dfAddCounts %>% 
    dplyr::rename(contrast_A_logFC_RN19081_logFC_Activin_Nullcomb_vs_WT = logFC_Activin_Nullcomb_vs_WT) %>% 
    dplyr::rename(contrast_A_padj_RN19081_Activin_Nullcomb_vs_WT = padj_Activin_Nullcomb_vs_WT) %>% 
    dplyr::rename(contrast_A_lg10p_RN19081_Activin_Nullcomb_vs_WT = lg10p_Activin_Nullcomb_vs_WT) %>% 
    dplyr::rename(contrast_A_lg2BaseMean_RN19081_Activin_Nullcomb_vs_WT = lg2BaseMean_Activin_Nullcomb_vs_WT ) 



dfAddCounts[is.na(dfAddCounts)] <- 0



## Add Inhba correlation from experiment SC21185 ##

dfAdd <- biologicSeqTools2::import.db.table.from.db(
    dbname = "chl_data",
    dbtable = "SC21185_sc_rna_seq_table ",
    host = Obio@dbDetailList$host,
    user = Obio@dbDetailList$db.user,
    password = db.pwd
)


selVec <- c("hgnc_symbol", "contrast_C_Inhba_Correlation_Coef") 
dfAdd <- unique(dfAdd[,selVec])
dfAdd <- dfAdd[dfAdd$hgnc_symbol != "", ]


dfAddCounts <- merge(
    dfAddCounts, 
    dfAdd, 
    by.x = "hgnc_symbol",
    by.y = "hgnc_symbol",
    all = TRUE
)

dfAddCounts[is.na(dfAddCounts)] <- 0

names(dfAddCounts) <- gsub("corCoef_Inhba", "contrast_C_SC21185_Inhba_Cor_Coef", names(dfAddCounts))



dfRefDatasets <- dfAddCounts


## Done
######################

Obio@dataTableList[["dfRefDatasets"]] <- data.frame(NULL)
Obio@dataTableList[["dfRefDatasets"]] <- dfRefDatasets

#database.table[is.na(database.table)] <- 0
## Done adding extra dataset                                                 ##
###############################################################################
```

```{r saveobject, eval=TRUE, echo=T, results=F}
### Will save Obio object here, so it can be re-used with different parameters
save(Obio, 
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
     )
)

print("Obio Object saved.")

# save(OsC,
#     file = paste0(
#          Obio@parameterList$localWorkDir,
#          Obio@parameterList$project_id,
#         ".Seurat.Robj"
#      )
# )

```

## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
sessionInfo()
```