---
title: "Add additional datasets"
author: "Stefan Boeing"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

output: 
    html_document:
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 3
        toc_float: true
        css:
    
always_allow_html: yes

---
    
    
```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```


## Part B Database


```{r hpc_notes, include=FALSE}

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;R;

## Get interactive session ##
#  srun --time=08:00:00 --mem=40G -p int --pty bash

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;R;

# sbatch --time=12:00:00 --wrap "module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;Rscript runD.r" --job-name="rD" -p hmem --mem=300G -o rD.slurm >> commands.txt

# --mem-per-cpu=14G -p hmem --pty bash

```


```{r populate_meta_data_database, eval=TRUE, echo=F, results=F}
## libraries ##
library(tidyverse)
library(Seurat)

VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

#Create the environment and load a suitable version of R, e.g. so:
FN <- paste0(hpc.mount, "Projects/reference_data/documentation/BC.parameters.txt")
dbTable <- read.delim(
    FN, 
    sep = "\t",
    stringsAsFactors = F
)

db.pwd <- as.vector(dbTable[1,1])






# source(
#     paste0(
#         hpc.mount,
#         "Stefan/protocol_files/github/boeings/packages/packageSourceCode/SBwebtools.pckg.r"
#     )
# )

# source(
#     paste0(
#         hpc.mount,
#         "Stefan/protocol_files/github/boeings/packages/scTools/scTools.r"
#     )
# )

source("assets/scTools.r")
source("assets/SBwebtools.pckg.r")


if (length(.libPaths()) > 2){
    .libPaths(.libPaths()[2:3])
}

ObioFN <- paste0("../", list.files("..")[grep(".bioLOGIC.Robj", list.files(".."))])

if (file.exists(ObioFN)){
    load(paste0(ObioFN))
    print(paste0("Obio object ", Obio@parameterList$localWorkDir, ObioFN, " exists and is loaded."))
} else {
    exit()
}

## Reset paths to local environment
Obio <- setMountingPoint(Obio)
Obio <- setAnalysisPaths(Obio)
Obio <- setCrickGenomeAndGeneNameTable(Obio)
Obio <- createAnalysisFolders(
    Obio,
    baseDir="/camp/stp/babs/working/boeings/Projects/",
    localBaseDir = paste0(hpc.mount, "Projects/")
)
Obio <- setDataBaseParameters(Obio)
```



```{r populate_expr_database, eval=TRUE, echo=F, results=F}
###############################################################################
## Otional add extra data set                                                ##

## Goal: create a table, that can be added via the 'gene' column to the main 
## database table

dfAddCounts <- import.db.table.from.db(
    dbtable = "vpl369A_rna_seq_table",
    dbname = "vpl_data",
    user = Obio@dbDetailList$db.user,
    password = db.pwd,
    host = Obio@dbDetailList$host
)
# # 
selVec <- c(
    Obio@parameterList$geneIDcolumn,
    names(dfAddCounts)[grep("contrast_1_logFC", names(dfAddCounts))],
    names(dfAddCounts)[grep("contrast_1_lg10p", names(dfAddCounts))],
    names(dfAddCounts)[grep("contrast_1_padj", names(dfAddCounts))],
    names(dfAddCounts)[grep("contrast_2_logFC", names(dfAddCounts))],
    names(dfAddCounts)[grep("contrast_2_lg10p", names(dfAddCounts))],
    names(dfAddCounts)[grep("contrast_2_padj", names(dfAddCounts))],
    names(dfAddCounts)[grep("contrast_3_logFC", names(dfAddCounts))],
    names(dfAddCounts)[grep("contrast_3_lg10p", names(dfAddCounts))],
    names(dfAddCounts)[grep("contrast_3_padj", names(dfAddCounts))]
)
# # 
dfAddCounts <- unique(dfAddCounts[,selVec])
dfAddCounts <- dfAddCounts[dfAddCounts$mgi_symbol != "",]

#dfNormCounts <- spread(data = dfNormCounts, key=condition, value=measurement)
#names(dfNormCounts) <- paste0("norm_counts_", names(dfNormCounts))
#names(dfNormCounts) <- gsub("norm_counts_gene", "gene", names(dfNormCounts))

names(dfAddCounts) <- gsub("contrast_1_logFC_", "contrast_A_logFC_bulkRNASeq_vpl369A_", names(dfAddCounts) )
names(dfAddCounts) <- gsub("contrast_1_padj_", "contrast_A_padj_bulkRNASeq_vpl369A_", names(dfAddCounts) )
names(dfAddCounts) <- gsub("contrast_1_lg10p_", "contrast_A_lg10p_bulkRNASeq_vpl369A_", names(dfAddCounts) )

names(dfAddCounts) <- gsub("contrast_2_logFC_", "contrast_A_logFC_bulkRNASeq_vpl369A_", names(dfAddCounts) )
names(dfAddCounts) <- gsub("contrast_2_padj_", "contrast_A_padj_bulkRNASeq_vpl369A_", names(dfAddCounts) )
names(dfAddCounts) <- gsub("contrast_2_lg10p_", "contrast_A_lg10p_bulkRNASeq_vpl369A_", names(dfAddCounts) )

names(dfAddCounts) <- gsub("contrast_3_logFC_", "contrast_A_logFC_bulkRNASeq_vpl369A_", names(dfAddCounts) )
names(dfAddCounts) <- gsub("contrast_3_padj_", "contrast_A_padj_bulkRNASeq_vpl369A_", names(dfAddCounts) )
names(dfAddCounts) <- gsub("contrast_3_lg10p_", "contrast_A_lg10p_bulkRNASeq_vpl369A_", names(dfAddCounts) )


dfRefDatasets <- dfAddCounts

####################3
## Add vpl367
dfAddCounts <- import.db.table.from.db(
    dbtable = "vpl367_rna_seq_table",
    dbname = "vpl_data",
    user = Obio@dbDetailList$db.user,
    password = db.pwd,
    host = Obio@dbDetailList$host
)
# # 
selVec <- c(
    Obio@parameterList$geneIDcolumn,
    names(dfAddCounts)[grep("contrast_1_logFC", names(dfAddCounts))],
    names(dfAddCounts)[grep("contrast_1_lg10p", names(dfAddCounts))],
    names(dfAddCounts)[grep("contrast_1_padj", names(dfAddCounts))]
)
# # 
dfAddCounts <- unique(dfAddCounts[,selVec])
dfAddCounts <- dfAddCounts[Obio@parameterList$geneIDcolumn != "",]

#dfNormCounts <- spread(data = dfNormCounts, key=condition, value=measurement)
#names(dfNormCounts) <- paste0("norm_counts_", names(dfNormCounts))
#names(dfNormCounts) <- gsub("norm_counts_gene", "gene", names(dfNormCounts))

names(dfAddCounts) <- gsub("contrast_1_logFC_", "contrast_A_logFC_bulkRNASeq_vpl367_", names(dfAddCounts) )
names(dfAddCounts) <- gsub("contrast_1_padj_", "contrast_A_padj_bulkRNASeq_vpl367_", names(dfAddCounts) )
names(dfAddCounts) <- gsub("contrast_1_lg10p_", "contrast_A_lg10p_bulkRNASeq_vpl367_", names(dfAddCounts) )

dfRefDatasets <- merge(
    dfRefDatasets, 
    dfAddCounts, 
    by.x = Obio@parameterList$geneIDcolumn, 
    by.y = Obio@parameterList$geneIDcolumn,
    all =TRUE
)

dfRefDatasets[is.na(dfRefDatasets)] <- 0
## Done
######################

Obio@dataTableList[["dfRefDatasets"]] <- data.frame(NULL)
Obio@dataTableList[["dfRefDatasets"]] <- dfRefDatasets

#database.table[is.na(database.table)] <- 0
## Done adding extra dataset                                                 ##
###############################################################################
```

```{r saveobject, eval=TRUE, echo=T, results=F}
### Will save Obio object here, so it can be re-used with different parameters
save(Obio, 
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
     )
)

print("Obio Object saved.")

# save(OsC,
#     file = paste0(
#          Obio@parameterList$localWorkDir,
#          Obio@parameterList$project_id,
#         ".Seurat.Robj"
#      )
# )

```

## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
sessionInfo()
```