---
title: "Sub Clustering"
author: "Stefan Boeing \n stefan.boeing@crick.ac.uk"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

output: 
    html_document:
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 5
        toc_float: true
        css:
    
always_allow_html: yes

---
    
    
```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```


```{r populate_meta_data_database, eval=TRUE, echo=F, results=F}

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;R;

## libraries ##
library(tidyverse)
library(Seurat)
library(knitr)

VersionPdfExt <- paste0("Subclustering.Exitatory.V", gsub("-", "", Sys.Date()), ".pdf")

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

#Create the environment and load a suitable version of R, e.g. so:
FN <- paste0(hpc.mount, "Projects/reference_data/documentation/BC.parameters.txt")
dbTable <- read.delim(
    FN, 
    sep = "\t",
    stringsAsFactors = F
)

db.pwd <- as.vector(dbTable[1,1])


source(
    paste0(
        hpc.mount,
        "Stefan/protocol_files/github/boeings/packages/packageSourceCode/SBwebtools.pckg.r"
    )
)

figureCount <- 1

## Load R module load R/3.5.1-foss-2018b ##
#setwd(Obio@parameterList$localWorkDir)

if (length(.libPaths()) > 2){
    .libPaths(.libPaths()[2:3])
}


ObioFN <- paste0("../", list.files("..")[grep(".bioLOGIC.Robj", list.files(".."))])

load(ObioFN)


# if (file.exists(ObioFN)){
#     load(paste0(ObioFN))
#     print(paste0("Obio object ", Obio@parameterList$localWorkDir, ObioFN, " exists and is loaded."))
# } else {
#     exit()
# }

## Reset paths to local environment
Obio <- setMountingPoint(Obio)
Obio <- setAnalysisPaths(Obio)
Obio <- setCrickGenomeAndGeneNameTable(Obio)
Obio <- createAnalysisFolders(
    Obio,
    baseDir="/camp/stp/babs/working/boeings/Projects/",
    localBaseDir = paste0(hpc.mount, "Projects/")
)
Obio <- setDataBaseParameters(Obio)

if (Obio@dbDetailList$host == "10.27.241.234"){
      urlString <- "biologic.thecrick.org"
    } else {
      urlString <- "biologic.crick.ac.uk"
    }

## Load Seurat object
SeuratFN <- paste0("../", list.files("..")[grep(".Seurat.Robj", list.files(".."))])

load(SeuratFN)

if (file.exists(paste0(Obio@parameterList$localWorkDir,SeuratFN))){
    print(paste0("Obio object ", Obio@parameterList$localWorkDir,SeuratFN, " exists and is loaded."))
    load(paste0(Obio@parameterList$localWorkDir, SeuratFN))
} else {
    exit()
}
```


## Overview Plot
This plot depicts the cells from the overall experiment that were subjected to subclustering
```{r Overviewplot, eval=TRUE, echo=F, results=F}

## Set parameters ##

## Select for subclustering ##
dfMeta <- OsC@meta.data
dfMeta[["subclustering"]] <- "" 

dfMeta[dfMeta[,Obio@parameterList$singleCellClusterString] %in% c(2), "subclustering"] <- "Subcluster Selection"

dfMeta[!(dfMeta[,Obio@parameterList$singleCellClusterString] %in% c(2)), "subclustering"] <- "Rest"

OsC@meta.data[,Obio@parameterList$singleCellClusterString] <- NULL
OsC@meta.data[,"seurat_clusters"] <- NULL

Obio@parameterList$singleCellClusterParameter <- 0.3
Obio@parameterList$singleCellClusterString <- paste0("integrated_snn_res.", Obio@parameterList$singleCellClusterParameter)
clusterTag <- "sub_clusters_Glia"

cluster_letters <- as.vector(dfMeta$subclustering)
names(cluster_letters) <- rownames(dfMeta)
OsC <- AddMetaData(
  object = OsC,
  metadata = cluster_letters,
  col.name = 'subclustering'
)

plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

## First make variation plot for integrated samples, than for all individual samples separately
tag <- "Overview_plot"

#dotsize <- round(7500/nrow(dfTemp),1)
dotsize <- 0.3
dfTemp <- OsC@meta.data

plotList[[tag]] <- ggplot(dfTemp, aes(UMAP_1, UMAP_2, color=subclustering)
        )+ geom_point( 
            shape = 16,
            size = as.numeric(dotsize)
        ) + xlab("UMAP1") + ylab("UMAP2")  +  theme(
            axis.text.y   = element_text(size=8),
            axis.text.x   = element_text(size=8),
            axis.title.y  = element_text(size=8),
            axis.title.x  = element_text(size=8),
            axis.line = element_line(colour = "black"),
            panel.border = element_rect(colour = "black", fill=NA, size=1),
            plot.title = element_text(hjust = 0.5, size = 12),
            panel.background = element_rect(fill = "lightgrey")
        ) + ggtitle("Selected for Subclustering"
        ) + scale_color_manual(values=c("#AAAAAA","#FF0000")
        )

#+ xlim(minX, maxX) + ylim(minY, maxY)  
    
    
    
            ## Save to file ##
            FNbase <- paste0("SubClusterScatterPlot", VersionPdfExt)
            FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
            FNrel <- paste0("report_figures/", FNbase)
            
           
            pdf(FN)
            print(plotList[[tag]])
            dev.off()
            
            
            
            ## Create R markdown chunk ##
            figLegend <- paste0(
                "**Figure ", 
                figureCount, 
                "**: Figure depicting the selection for subclustering. Download a pdf of this figure [here](", FNrel, "). "
            )
            
            figureCount <- figureCount + 1 
            
            NewChnk <- paste0(
                #"\n #### Subclustering Selection", 
                "\n```{r ", tag, ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                figLegend,"'}\n",
                "\n",
                "\n print(plotList[['",tag,"']])",
                "\n cat(  '\n')",
                "\n\n\n```\n"   
            )
            
            chnkVec <- c(
                chnkVec,
                NewChnk
            )

```

```{r overview_plot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
## plot list will be integrated in full figure ##
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
```


## Subclustering
This plot depicts the cells from the overall experiment that were subjected to subclustering
```{r Subclustering, eval=TRUE, echo=F, results=F}

## Add columns for subclustering

## Select cells for sublcustering and subset dataset ##
#Newcluster7 alone in the 4 epithelial populations (pbl359A)?
OsCsub <- subset(
    x = OsC, 
    subset = subclustering == "Subcluster Selection")
## Do dimensionality reduction on reduced cells ##
## Set to RNA

## Plot most variable genes ##

```


Start here with part B from Result Figures on

## Result Figures
### Quality Control Plots: Variable Features in the combined and all individual dataset {.tabset .tabset-fade .tabset-pills}
```{r create-individual-varFeatures, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

## First make variation plot for integrated samples, than for all individual samples separately
tag <- "Integrated_Samples"
DefaultAssay(OsCsub) <- "RNA"

    OsCsub <- FindVariableFeatures(object = OsCsub)

    OsCsub <- FindVariableFeatures(
        object = OsCsub,
        selection.method = 'vst', 
        nfeatures = 2000
    )

    # Identify the 10 most highly variable genes
    label2000 <- paste0("integrated", "_", "top2000var")
    Obio@dataTableList$referenceList[[ label2000]]<- head(
        x = VariableFeatures(object = OsCsub), 
        2000
    )
    
    label30 <- paste0("integrated", "_", "top30var")
    Obio@dataTableList$referenceList[[ label30]]<- head(
        x = VariableFeatures(object = OsCsub), 
        30
    )


    ## slot for variable features OsCsub@assays$RNA@var.features

    dfVar <- OsCsub@assays$RNA@meta.features
    names(dfVar) <- gsub("vst.", "",names(dfVar))
    dfVar[["gene"]] <- row.names(dfVar)

    OsCsub@meta.data[["all"]] <- "all"
    Idents(OsCsub) <- "all"
    
    
    cluster.averages <- AverageExpression(
        OsCsub, 
        return.seurat = TRUE
    )
    
    Idents(OsCsub) <- "sampleID"
    
    dfAvgExpr <- data.frame(cluster.averages[["RNA"]]@data)
    dfAvgExpr[["gene"]] <- row.names(dfAvgExpr)
    names(dfAvgExpr)[1] <- "Avg.Expression"

    dfVar <- merge(
        dfVar, 
        dfAvgExpr, 
        by.x = "gene",
        by.y = "gene"
    )

    dfVar[["Type"]] <- "Standard"
    dfVar[dfVar$gene %in% OsCsub@assays$RNA@var.features, "Type"] <- "Most Variable"

    dfVar[["text"]] <- ""
    dfVar[dfVar$gene %in% as.vector(Obio@dataTableList$referenceList[[label30]]), "text"] <- Obio@dataTableList$referenceList[label30]

    dotsize <- 0.5

    library(ggrepel)

    plotList[[tag]] <- ggplot(
        data = dfVar, 
        aes(
            x=Avg.Expression, 
            y=variance.standardized, label = text, color = Type)
            ) + geom_point( shape=16, size = dotsize
            )  + xlab("Average Expression") + ylab("Variance Standarized")  +  theme(
                axis.text.y   = element_text(size=8),
                axis.text.x   = element_text(size=8),
                axis.title.y  = element_text(size=8),
                axis.title.x  = element_text(size=8),
                axis.line = element_line(colour = "black"),
                panel.border = element_rect(colour = "black", fill=NA, size=1),
                plot.title = element_text(hjust = 0.5, size = 12)
            )+ ggtitle(paste0("Variance vs. Expression in the Overall Experiment")
            ) + scale_color_manual(values=c("#FF0000", "#000000")
            ) + geom_text_repel()


    
    ###########################################################################
    ## Save plot to file                                                     ##
    FNbase <- paste0("variation.integrated.samples.", VersionPdfExt)
    FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    pdf(FN)
        print(plotList[[tag]])
    dev.off()
    ##                                                                       ##
    ###########################################################################
    
    
    
    
    figCap <- paste0(
        "**Figure ",
        figureCount,
        "C:** Variance versus averaged gene expression for overall sample.",
        "Download a pdf of this figure [here](", FNrel, "). "
    )
    
    NewChnk <- paste0(
        "#### ",tag,
        "\n```{r varplot_",tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",figCap,"'}\n",
        "\n",
        "\n print(plotList[['",tag,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
    )
     
    ## Histogram Part C done                                                 ##
    ###########################################################################
    
       
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
    
    figureCount <- figureCount + 1
    print(" All variation done.")

## Now the individual samples ##

xmax <- ceiling(max(dfVar$Avg.Expression))
ymax <- ceiling(max(dfVar$variance.standardized))

dfVarRes <- unique(
    dfVar[,c("gene", "Avg.Expression", "variance.standardized")]
)

names(dfVarRes) <- gsub(
    "Avg.Expression", paste0(tag, "_AvgExpr"), names(dfVarRes)
)

names(dfVarRes) <- gsub(
    "variance.standardized", paste0(tag, "_var_std"), names(dfVarRes)
)
  
SampleList <- list("Subclusters" = OsCsub)  
for (i in 1:length(SampleList)){
    tag <- paste0("Ind_var_", names(SampleList)[i])
    
    
    DefaultAssay(SampleList[[i]]) <- "RNA"

    SampleList[[i]] <- FindVariableFeatures(object = SampleList[[i]])

    SampleList[[i]] <- FindVariableFeatures(
        object = SampleList[[i]],
        selection.method = 'vst', 
        nfeatures = 2000
    )

    # Identify the 10 most highly variable genes
    label2000 <- paste0(names(SampleList)[i], "_", "top2000var")
    Obio@dataTableList$referenceList[[ label2000]]<- head(
        x = VariableFeatures(object = SampleList[[i]]), 
        2000
    )
    
    label30 <- paste0(names(SampleList)[i], "_", "top30var")
    Obio@dataTableList$referenceList[[ label30]]<- head(
        x = VariableFeatures(object = SampleList[[i]]), 
        30
    )


    ## slot for variable features OsCsub@assays$RNA@var.features

    dfVar <- SampleList[[i]]@assays$RNA@meta.features
    names(dfVar) <- gsub("vst.", "",names(dfVar))
    dfVar[["gene"]] <- row.names(dfVar)

    
    cluster.averages <- AverageExpression(
        SampleList[[i]], 
        return.seurat = TRUE
    )
    
    dfAvgExpr <- data.frame(cluster.averages[["RNA"]]@data)
    dfAvgExpr[["gene"]] <- row.names(dfAvgExpr)
    names(dfAvgExpr)[1] <- "Avg.Expression"

    dfVar <- merge(
        dfVar, 
        dfAvgExpr, 
        by.x = "gene",
        by.y = "gene"
    )

    dfVar[["Type"]] <- "Standard"
    dfVar[dfVar$gene %in% SampleList[[i]]@assays$RNA@var.features, "Type"] <- "Most Variable"

    dfVar[["text"]] <- ""
    dfVar[dfVar$gene %in% as.vector(Obio@dataTableList$referenceList[[label30]]), "text"] <- Obio@dataTableList$referenceList[label30]

    dotsize <- 0.5

    library(ggrepel)

    plotList[[tag]] <- ggplot(
        data = dfVar, 
        aes(
            x=Avg.Expression, 
            y=variance.standardized, label = text, color = Type)
            ) + geom_point( shape=16, size = dotsize
            )  + xlab("Average Expression") + ylab("Variance Standarized")  +  theme(
                axis.text.y   = element_text(size=8),
                axis.text.x   = element_text(size=8),
                axis.title.y  = element_text(size=8),
                axis.title.x  = element_text(size=8),
                axis.line = element_line(colour = "black"),
                panel.border = element_rect(colour = "black", fill=NA, size=1),
                plot.title = element_text(hjust = 0.5, size = 12)
            )+ ggtitle(paste0("Individual variance in the ", names(SampleList)[i],  " sample.")
            ) + scale_color_manual(values=c("#FF0000", "#000000")
            ) + geom_text_repel(
            ) +  xlim(0, xmax) + ylim(0, ymax)


    
    ###########################################################################
    ## Save plot to file                                                     ##
    FNbase <- paste0("Individual.var.features",names(SampleList)[i], VersionPdfExt)
    FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    pdf(FN)
        print(plotList[[tag]])
    dev.off()
    ##                                                                       ##
    ###########################################################################
    
    
    
    
    figCap <- paste0(
        "**Figure ",
        figureCount,
        "C:** Variance versus averaged gene expression for sample ", 
        names(SampleList)[i],
        ". ",
        "Download a pdf of this figure [here](", FNrel, "). "
    )
    
    NewChnk <- paste0(
        "#### ",tag,
        "\n```{r varplot_",tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",figCap,"'}\n",
        "\n",
        "\n print(plotList[['",tag,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
    )
     
    ## Histogram Part C done                                                 ##
    ###########################################################################
    
       
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
    
    figureCount <- figureCount + 1
    print(paste0(names(SampleList)[i], " individual variation done."))
    
    ## Add to result table ##
    dfVarTemp <- unique(
    dfVar[,c("gene", "Avg.Expression", "variance.standardized")]
    )
    
    names(dfVarTemp) <- gsub(
        "Avg.Expression", paste0(tag, "_AvgExpr"), names(dfVarTemp)
    )
    
    names(dfVarTemp) <- gsub(
        "variance.standardized", paste0(tag, "_var_std"), names(dfVarTemp)
    )
    
    dfVarRes <- merge(
        dfVarRes, 
        dfVarTemp, 
        by.x = "gene",
        by.y = "gene",
        all = TRUE
    )
    dfVarRes[is.na(dfVarRes)] <- 0
    
}

## Make sure summary goes first ##
Obio@dataTableList[["dfVariation"]] <- dfVarRes

```

```{r create-var-feature-plot-II, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
## plot list will be integrated in full figure ##
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
```



```{r PCA_UMAP_tSNE, echo=TRUE, eval=TRUE, warning=FALSE, results=FALSE} 
###############################################################################
## Perform integrated analysis                                               ##

#if (length(names(Obio@sampleDetailList)) > 1) {
    DefaultAssay(OsCsub) <- "integrated"
# } else {
#     Obio@parameterList$singleCellClusterString <- gsub("integrated", "RNA", Obio@parameterList$singleCellClusterString)
#}


# Run the standard workflow for visualization and clustering
## This will scale on the most variable features only
OsCsub <- ScaleData(OsCsub, verbose = FALSE)

OsCsub <- RunPCA(
    OsCsub, 
    npcs = Obio@parameterList$singleCellSeuratNpcs4PCA, verbose = FALSE
)
# t-SNE and Clustering

## Add PCA clusters to data collection ##



OsCsub <- RunUMAP(OsCsub, reduction = "pca", dims = 1:20)

OsCsub <- RunTSNE(OsCsub, reduction = "pca", dims = 1:20)

OsCsub <- FindNeighbors(OsCsub, reduction = "pca", dims = 1:20)

OsCsub <- FindClusters(OsCsub, resolution = Obio@parameterList$singleCellClusterParameter)

## Rational: Run PCA on variable features, then scale data for heatmaps and other applications

if (Obio@parameterList$scIntegrationMethod != "RNA"){
DefaultAssay(OsCsub) <- Obio@parameterList$scIntegrationMethod
    allGenes <- rownames(x = OsCsub@assays[[Obio@parameterList$scIntegrationMethod]])
    OsCsub <- ScaleData(OsCsub, verbose = FALSE, features=allGenes)
}

DefaultAssay(OsCsub) <- "RNA"
allGenes <- rownames(x = OsCsub@assays$RNA)
OsCsub <- ScaleData(OsCsub, verbose = FALSE, features=allGenes)



OsCsub@meta.data[["hmIdent"]] <- factor(OsCsub@meta.data$seurat_clusters, levels=sort(unique(OsCsub@meta.data$seurat_clusters)))
OsCsub@meta.data[["hmIdent2"]] <- factor(paste("C",OsCsub@meta.data$seurat_clusters), levels=sort(unique(paste("C",OsCsub@meta.data$seurat_clusters ))))
OsCsub@meta.data[["clust_Ident"]] <- factor(paste("C_",OsCsub@meta.data$seurat_clusters), levels=sort(unique(paste("C_",OsCsub@meta.data$seurat_clusters ))))

dfAdd <- OsCsub@meta.data



dfAdd[[clusterTag]] <-  dfAdd$seurat_clusters
dfAdd[[paste0(clusterTag, "_names")]] <-  paste0(clusterTag, "_1_Cluster_",dfAdd$seurat_clusters)
dfAdd <- data.frame(dfAdd[,c(clusterTag,paste0(clusterTag, "_names"))])
dfAdd[["cellID"]] <- row.names(dfAdd)

## Get and add umap ##
dfUMAP  <- data.frame(OsCsub$umap@cell.embeddings)
names(dfUMAP) <- paste0(names(dfUMAP), "_", clusterTag)
dfUMAP[["cellID"]] <- row.names(dfUMAP)
dfAdd <- merge(
    dfAdd, 
    dfUMAP, 
    by.x = "cellID",
    by.y = "cellID"
)
row.names(dfAdd) <- dfAdd$cellID
dfAdd$cellID <- NULL

dfMold <- OsC@meta.data
dfMold <- dfMold[!(row.names(dfMold) %in% row.names(dfAdd)), ]
dfFill <- data.frame(matrix(nrow=nrow(dfMold), ncol=4))
dfFill[is.na(dfFill)] <- ""
names(dfFill) <- names(dfAdd)
row.names(dfFill) <- row.names(dfMold)

dfAdd <- rbind(
    dfAdd, 
    dfFill
)

## Add subclusters to OsC ##
OsC <- addDf2seuratMetaData(
    OsC, 
    dfAdd
)

## Save Seurat Object ##
save(OsC, 
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".Seurat.Robj"
     )
)


```

### Dimensionality Reduction Plots by Sample {.tabset .tabset-fade .tabset-pills}
If you wish to get a bit of background on tSNE dimensionality reduction, take a look at 
[this youtube video](https://www.youtube.com/watch?v=NEaUSP4YerM) by Josh Starmer from the University of North Carolina.

If you wish to get a bit of background on UMAP (and other) dimensionality reduction algorithms, take a look at 
[this youtube video](https://www.youtube.com/watch?v=9iol3Lk6kyU) recaping a lecture at the PyData 2018 conference.   
```{r Plot_tsne_data_rendering, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 


## Get Coordinates ##
# dfCoord <- data.frame(OsCsub$umap@cell.embeddings)
# names(dfCoord) <- paste0(names(dfCoord), "_SubC_Ext")
# rmVec <- names(dfCoord)
# dfCoord[["cellID"]] <- row.names(dfCoord)
# #write.table(dfCoord, "dfCoord.exitiatory.txt", row.names = F, sep = "\t")
dfData <- import.db.table.from.db(
    dbtable = Obio@parameterList$PCAdbTableName,
    dbname =  Obio@dbDetailList$primDataDB,
    user = Obio@dbDetailList$db.user,
    password = db.pwd,
    host = Obio@dbDetailList$host
)

delVec <- c(names(dfAdd), "sub_clusters_1", "sub_clusters_1_names")
selVec <- names(dfData)[!(names(dfData) %in% delVec)]
dfData <- dfData[,selVec]

dfAdd[["cellID"]] <- row.names(dfAdd)
dfAdd <- dfAdd[dfAdd$cellID %in% dfData$cellID, ]

dfData <- merge(
    dfData, 
    dfAdd, 
    by.x = "cellID",
    by.y = "cellID",
    all =T
)
#dfData[is.na(dfData)] <- ""
# 
# dfData$row_names <- NULL
# 
# dfData[,rmVec[1]] <- NULL
# dfData[,rmVec[2]] <- NULL
# 
# dfData <- merge(
#     dfData,
#     dfCoord, 
#     by.x = "cellID",
#     by.y = "cellID",
#     all = TRUE
# )
#
#dfData[is.na(dfData)] <- 0
names(dfData) <- gsub("[.]", "_", names(dfData))

upload.datatable.to.database(
    host = Obio@dbDetailList$host,
    user = Obio@dbDetailList$db.user,
    password = db.pwd,
    prim.data.db = Obio@dbDetailList$primDataDB,
    dbTableName = Obio@parameterList$PCAdbTableName,
    df.data = dfData,
    db.col.parameter.list = list(
        "VARCHAR(100) CHARACTER SET utf8 COLLATE utf8_general_ci" = c("subclustering","sub_clusters","cellType","cellIdent","sample_","cellID","orig_ident","old_ident","clustIdent", "hmIdent","sampleID"),
       "VARCHAR(10) CHARACTER SET utf8 COLLATE utf8_general_ci" = c("all","included","Phase", gsub("[.]", "_", Obio@parameterList$singleCellClusterString), "seurat_clusters"),
        "BIGINT(8) NULL DEFAULT NULL" = c("row_names"),
        "INT(8) NULL DEFAULT NULL" = c("nCount_", "nFeature_","DF_Classification"),
        "DECIMAL(6,3) NULL DEFAULT NULL" = c("percent_mt","PC", "tSNE", "UMAP","^DC","DM_Pseudotime"),
        "DECIMAL(6,5) NULL DEFAULT NULL" = c("mAUC_","_Score","DF_pANN")
    ),
    new.table = TRUE
)

killDbConnections()

reductionVec <- c("umap", "tsne")

plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

for (i in 1:length(reductionVec)){
    tag <- paste0("Dimplot_by_sample_", i)
    
    plotList[[tag]] <- DimPlot(OsCsub, reduction = reductionVec[i], group.by = "sampleID")
    
    
    ## Save to file ##
    FNbase <- paste0("dimplot.by.sample.", reductionVec[i],".", VersionPdfExt)
    FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    pdf(FN)
        print(plotList[[tag]])
    dev.off()
    
    figLegend <- paste0(
        "**Figure ", 
        figureCount, 
        ":** ",
        reductionVec[i],
        " plot depicting all samples. Download a pdf of this figure [here](", FNrel,")."
    )
    
    NewChnk <- paste0(
        "#### ",reductionVec[i],
        "\n```{r Dimplot_by_sample_",
        i,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
        figLegend,"'}\n",
        "\n",
        "\n print(plotList[['",tag,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
    )
    
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
}

## Done integraed analysis                                                   ##
###############################################################################
```

```{r Plot_tsne_data_plotting, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))

```


### Dimensionality Reduction Plots by Cluster {.tabset .tabset-fade .tabset-pills}
```{r Plot_dimred_by_cluster_data_prep, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

reductionVec <- c("umap", "tsne")

plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")


for (i in 1:length(reductionVec)){
    tag <- paste0("Dimplot_by_cluster", i)
    
    plotList[[tag]] <- DimPlot(OsCsub, reduction = reductionVec[i], label = TRUE, split.by = "sampleID")
    
    
    ## Save to file ##
    FNbase <- paste0("dimplot.by.cluster.", reductionVec[i],".", VersionPdfExt)
    FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    pdf(FN)
        print(plotList[[tag]])
    dev.off()
    
    figLegend <- paste0(
        "**Figure ", 
        figureCount, 
        ":** ",
        reductionVec[i],
        " plot depicting all samples. Download a pdf of this figure [here](", FNrel,")."
    )
    
    NewChnk <- paste0(
        "#### ",reductionVec[i],
        "\n```{r Dimplot_bycluster_",
        i,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
        figLegend,"'}\n",
        "\n",
        "\n print(plotList[['",tag,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
    )
    
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
}


## Add cluster dendrogram ##
library(ggtree)
OsCsub <- BuildClusterTree(OsCsub)
    
tag <- paste0("Cluster_Dendrogram")
    
OsCsub@tools$BuildClusterTree$tip.label <- paste0("C", OsCsub@tools$BuildClusterTree$tip.label)
    
plotList[[tag]]  <- ggplot(OsCsub@tools$BuildClusterTree
    ) + geom_tree(
    ) + theme_tree(
    ) + geom_tiplab(
    ) + labs(title=tag
    ) + theme(
      panel.border = element_rect(colour = "black", fill=NA, size=1),
      axis.title.x=element_blank(),
      plot.title = element_text(hjust = 0.5, size = 12)
    ) 
    
    #+ xlim(-1,1.2*max(OsCsub@tools$BuildClusterTree$edge)) 
    
    
## Save to file ##
FNbase <- paste0(tag,".", VersionPdfExt)
FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
FNrel <- paste0("report_figures/", FNbase)
    
pdf(FN)
    print(plotList[[tag]])
dev.off()
    
figLegend <- paste0(
    "**Figure ", 
        figureCount, 
        ":** ",
        " Clusterplot dendrogram. ","A pdf of this figure can be downloaded [here](",FNrel,")."
    )
    
    
    NewChnk <- paste0(
        "#### Cluster Dendrogram",
        "\n```{r ", tag, "results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
        figLegend,"'}\n",
        "\n",
        "\n print(plotList[['",tag,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
    )
    
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
    
    
    figureCount <- figureCount + 1

## Done integraed analysis                                                   ##
###############################################################################


###############################################################################
## Find all markers                                                          ##
DefaultAssay(OsCsub) <- "RNA"
Idents(OsCsub) <- Obio@parameterList$singleCellClusterString

lgFCthreshold <- 0.25

dfGeneralMarkers <- FindAllMarkers(
    object = OsCsub, 
    only.pos = FALSE, 
    min.pct = 0.1, 
    logfc.threshold = lgFCthreshold,
    test.use = "roc",
    assay =  "RNA",
    slot = "data"
)

if (nrow(dfGeneralMarkers) == 0){
    lgFCthreshold <- 0.05
dfGeneralMarkers <- FindAllMarkers(
    object = OsCsub, 
    only.pos = FALSE, 
    min.pct = 0.1, 
    logfc.threshold = lgFCthreshold,
    test.use = "roc",
    assay =  "RNA",
    slot = "data"
)
}

if (nrow(dfGeneralMarkers) > 0){
    dfGeneralMarkers[["direction"]] <- ""
    dfGeneralMarkers[dfGeneralMarkers$avg_diff >= 0, "direction"] <- "positive"
    dfGeneralMarkers[dfGeneralMarkers$avg_diff < 0, "direction"] <- "negative"
    
    Obio@dataTableList[["dfGeneralMarkers"]] <- dfGeneralMarkers
    
    dfGeneralMarkersFilt <- dfGeneralMarkers[dfGeneralMarkers$avg_diff > lgFCthreshold | dfGeneralMarkers$avg_diff < -lgFCthreshold,]
    Obio@dataTableList[["dfGeneralMarkersFilt"]] <- dfGeneralMarkersFilt
    
    dfTop1 <- data.frame(dfGeneralMarkers %>% group_by(cluster) %>% top_n(1, avg_diff))
    dfTop5 <- data.frame(dfGeneralMarkers %>% group_by(cluster) %>% top_n(5, avg_diff))
    dfTop10 <- data.frame(dfGeneralMarkers %>% group_by(cluster) %>% top_n(10, avg_diff))
    
    Obio@dataTableList[["dfGeneralMarkersTop10"]] <- dfTop10
    
    Obio@dataTableList$referenceList[["Top10clusterMarkers"]] <- as.vector(
        unique(
            dfTop10$gene
        )
    )
} else {
    Obio@dataTableList[["dfGeneralMarkers"]] <- NULL
}

#############################################################

```

```{r Plot_dim_red_by_cluster, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))

```


### Check Accuracy UMAP Distances
The sleepwalk tool will provide the euclidean distances between individual cells. This will help you to determine clustering accuracy. Find more information [here]("https://anders-biostat.github.io/sleepwalk/"). 
<div align="center">
```{r Plot_umap_sleepwalk, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
library(sleepwalk)
sleepwalk( 
  OsCsub@reductions$umap@cell.embeddings, 
  OsCsub@reductions$pca@cell.embeddings,
  saveToFile=paste(Obio@parameterList$outputDir,"sleepwalk.UMAP.html",sep='')
)
htmltools::includeHTML(paste(Obio@parameterList$outputDir,"sleepwalk.UMAP.html",sep=''))
  
          
```
</div>




### Cell Cycle Markers
```{r Cell_cycle_markers, echo=TRUE, eval=TRUE, warning=FALSE, fig.cap=paste0("**Figure ",figureCount,":** tSNE colored by cell cycle phase", results="asis"), results="asis"}
figureCount <- figureCount + 1
###############################################################################
## Estimate cell cycle genes                                                 ##
exprFN <- paste0(hpc.mount, "Projects/reference_data/cell_cycle_vignette_files/nestorawa_forcellcycle_expressionMatrix.txt")

exp.mat <- read.table(file = exprFN, header = TRUE, 
                      as.is = TRUE, row.names = 1)


# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

print(paste0("Used as S-phase marker genes: ", sort(unique(paste(s.genes, collapse = ", ")))))
print(paste0("Used as G2M-phase marker genes: ", sort(unique(paste(g2m.genes, collapse = ", ")))))

# Create our Seurat object and complete the initalization steps
OsCsub <- CellCycleScoring(OsCsub, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# view cell cycle scores and phase assignments
# head(OsCsub[[]])

## Make tsne showing cell cycle ##
p1 <- DimPlot(
    OsCsub, 
    reduction = Obio@parameterList$primReduction,
    pt.size = 0.25,
    label =T,
    no.legend = FALSE,
    do.label = FALSE,
    group.by = "Phase",
    split.by = "sampleID"
)

print(p1)


## Save to file ##
FNbase <- paste0("tsne.cell.cycle.phase", VersionPdfExt)
FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
FNrel <- paste0("report_figures/", FNbase)

pdf(FN)
print(p1)
dev.off()


cat("Download a pdf of this figure [here](",FNrel,"). ")
cat("\n")


#RidgePlot(OsCsub, features = c("TOP2A", "PCNA"), ncol = 2, log=T, slot="data")

Obio@dataTableList[["dfCellCycleIndex"]] <- OsCsub[[]]

## Done estimating cell cycle phase                                          ##
###############################################################################
```

### Barchart N cells and Percent in Clusters {.tabset .tabset-fade .tabset-pills}
if you could prepare a bar-shape graph with the %of cells clusters representing our populations (like in the Nat Med)

```{r make_percent_N_cells_barchart, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
###############################################################################
## Create datatable for plotting                                             ##


## This plotting procedure requires three sets: the sets cellTypeIDs, clusterIDs, cellTypeIDs
## level1ID, level2ID, level3ID

sampleIDs <- unique(OsCsub@meta.data$sampleID)
#Obio@parameterList$singleCellSeuratMtCutoff <- 20

if (is.null(Obio@parameterList$singleCellSeuratMtCutoff)){
    Obio@parameterList$singleCellSeuratMtCutoff <- rep(10, length(sampleIDs))    
} else if (length(Obio@parameterList$singleCellSeuratMtCutoff) == 1){
    Obio@parameterList$singleCellSeuratMtCutoff <- rep(
        Obio@parameterList$singleCellSeuratMtCutoff, 
        length(sampleIDs)
    )
} else if (length(Obio@parameterList$singleCellSeuratMtCutoff) == length(sampleIDs)){
    Obio@parameterList$singleCellSeuratMtCutoff <- Obio@parameterList$singleCellSeuratMtCutoff
} else {
    print("Can't determine mitochondrial cut off. ")
    stop()
}

sampleIDs <- unique(OsCsub@meta.data$sampleID)
clusterIDs <- unique(OsCsub@meta.data[,Obio@parameterList$singleCellClusterString])

if (length(grep("cellIdent", names(OsCsub@meta.data))) == 0){
  OsCsub@meta.data[["cellIdent"]] <- "All"
}

cellTypeIDs <- unique(OsCsub@meta.data[,"cellIdent"])

dfTemp <- OsCsub@meta.data

if (length(grep("^cellIdent$", names(dfTemp))) == 0){
  dfTemp[["cellIdent"]] <- "All"
}

dfTemp <- dfTemp[dfTemp$percent.mt <= max(Obio@parameterList$singleCellSeuratMtCutoff), ]
dfTemp[["cellID"]] <- row.names(dfTemp)
dfTemp <- unique(dfTemp[,c("cellID", "sampleID", Obio@parameterList$singleCellClusterString, "cellIdent")])
names(dfTemp) <- gsub(Obio@parameterList$singleCellClusterString, "Cluster", names(dfTemp) )
names(dfTemp) <- gsub(Obio@parameterList$singleCellClusterString, "Cluster", names(dfTemp) )
 dfTemp <- unique(dfTemp[,c("cellID", "sampleID", "Cluster", "cellIdent")])
 
dfRes <- dfTemp
dfRes$cellID <- NULL
row.names(dfRes) <- NULL
dfRes <- unique(dfRes)
dfRes[["N_cells"]] <- 0


for (i in 1:nrow(dfRes)){
  dfRes[i, "N_cells"] <- nrow(dfTemp[dfTemp$sampleID == dfRes[i,"sampleID"] & dfTemp$Cluster == dfRes[i,"Cluster"] & dfTemp$cellIdent == dfRes[i,"cellIdent"], ])
}




## Calculate cluster percentages per celltypeID ##
dfRes[["Perc_cells"]] <- 0
for (i in 1:length(cellTypeIDs)){
  dfResTemp2 <- dfRes[dfRes$cellIdent == cellTypeIDs[i], ]
  tempCluster <- as.vector(unique(dfResTemp2$Cluster))
  
  for (j in 1:length(tempCluster)){
    dfResTemp3 <- dfResTemp2[dfResTemp2$Cluster == tempCluster[j],]
    NclusterTotal <- sum(dfResTemp3[, "N_cells"])
    dfResTemp3[,"Perc_cells"] <- round(dfResTemp3[,"N_cells"]/NclusterTotal, 4)*100
    
    
    if (j ==1){
      dfRes3 <- dfResTemp3
    } else {
      dfRes3 <- rbind(dfResTemp3, dfRes3)
    }
  }
  
  if (i ==1){
    dfRes4 <- dfRes3
  } else {
    dfRes4 <- rbind(dfRes3, dfRes4)
  }
  
}
 
dfRes <- dfRes4

plotListNcells <- list()
plotListPercent <- list()
chnkVec <- as.vector(NULL, mode = "character")

for (i in 1:length(cellTypeIDs)){
  #############################################################################
  ## Create cell number plot                                                 ##
  tag <- cellTypeIDs[i]
  dfResTemp <- dfRes[dfRes$cellIdent == cellTypeIDs[i], ]
  
  ## Calculate percentages for this subset ##
  
  plotListNcells[[tag]] <- ggplot(
    ) + geom_bar(aes(y = N_cells, x = Cluster, fill = sampleID), data = dfResTemp, stat="identity"
    ) + labs(title=tag, x="Cluster", y = "Cell Count"
    ) +  theme(
      panel.border = element_rect(colour = "black", fill=NA, size=1),
      axis.title.x=element_blank(),
      plot.title = element_text(hjust = 0.5, size = 12)
    ) + coord_flip()
  
  ###########################################################################
  ## Save plot to file                                                     ##
  FNbase <- paste0(tag,".Ncells", VersionPdfExt)
  FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
  FNrel <- paste0("report_figures/", FNbase)
    
  pdf(FN)
      print(plotListNcells[[tag]])
  dev.off()
  ##                                                                       ##
  ###########################################################################
    
  ###########################################################################
  ## Add to chunk                                                          ##
  figCap <- paste0(
      "**Figure ",
      figureCount,
      "A:** Cell Count in each cluster for ", 
      tag,
      "Download a pdf of this figure [here](", FNrel, "). "
  )
  
  NewChnk <- paste0(
    paste0("#### Barchart_ ", tag),
        "\n```{r Barchart-",tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",figCap,"'}\n",
        "\n",
        "\n print(plotListNcells[['",tag,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
  )
    
  chnkVec <- c(
      chnkVec,
      NewChnk
  )
  ## Done adding                                                             ##
  #############################################################################
  
  #############################################################################
  ## Add percentage plot                                                     ##
  plotListPercent[[tag]] <- ggplot(
    ) + geom_bar(aes(x = Cluster, y = Perc_cells, fill = sampleID), data = dfResTemp, stat="identity"
    ) + labs(title=tag, x="Cluster", y = "Percent Cells"
    ) +  theme(
      panel.border = element_rect(colour = "black", fill=NA, size=1),
      axis.title.x=element_blank(),
      plot.title = element_text(hjust = 0.5, size = 12)
    ) +  coord_flip()
  
  ###########################################################################
  ## Save plot to file                                                     ##
  FNbase <- paste0(tag, ".percent.cells",VersionPdfExt)
  FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
  FNrel <- paste0("report_figures/", FNbase)
    
  pdf(FN)
      print(plotListPercent[[tag]])
  dev.off()
  ##                                                                       ##
  ###########################################################################
    
  ###########################################################################
  ## Add to chunk                                                          ##
  figCap <- paste0(
      "**Figure ",
      figureCount,
      "B:** Cell percentages in each cluster for ", 
      tag,
      "Download a pdf of this figure [here](", FNrel, "). "
  )
  
  NewChnk <- paste0(
        "\n```{r Barchart-percent_",tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",figCap,"'}\n",
        "\n",
        "\n print(plotListPercent[['",tag,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
  )
    
  chnkVec <- c(
      chnkVec,
      NewChnk
  )
  ## Done adding percentage plot                                             ##
  #############################################################################
  
  
  figureCount <- figureCount + 1
}

## Done creating data table                                                  ##
###############################################################################


```

```{r plot_barchart, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
```


```{r Identify_cluster_marker, echo=TRUE, eval=TRUE, warning=FALSE, fig.cap="**Figure 1:** tSNE"}
###############################################################################
## Identify conserved markers                                                ##

#To identify canonical cell type marker genes that are conserved across conditions, we provide the FindConservedMarkers function. This function performs differential gene expression testing for each dataset/group and combines the p-values using meta-analysis methods from the MetaDE R package. For example, we can calculated the genes that are conserved markers irrespective of stimulation condition in cluster 6 (NK cells).

# DefaultAssay(immune.combined) <- "RNA"
# markers <- FindConservedMarkers(OsCsub, ident.1 = 6, grouping.var = "orig.ident", verbose = FALSE)
# head(markers)
# 
# markers <- FindConservedMarkers(OsCsub,  grouping.var = "orig.ident", verbose = FALSE)
# head(markers)

## Done identify conserved markers                                           ##
###############################################################################

##################
## Bespoke                                                                   ##
DefaultAssay(OsCsub) <- "RNA"
Idents(OsCsub) <- Obio@parameterList$singleCellClusterString

if (!is.null(Obio@parameterList$DGEbyCluster)){
    for (i in 1:length(Obio@parameterList$DGEbyCluster)){
        Obio@enrichmentList[[names(Obio@parameterList$DGEbyCluster)[i]]] <- FindMarkers(
            object = OsCsub, 
            ident.1 = as.vector(unlist(Obio@parameterList$DGEbyCluster[[i]][1])),
            ident.2 = as.vector(unlist(Obio@parameterList$DGEbyCluster[[i]][2])),
            min.pct = 0.05,
            logfc.threshold = 0
        )
    }
}
        



## Done bespoke                                                              ##
###############################################################################

## Done finding all markers                                                  ##
###############################################################################
```

### Category Enrichment Scatterplots {.tabset .tabset-fade .tabset-pills}

```{r AUC_prep_from_file, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}

# save(Obio, 
#      file = paste0(
#          Obio@parameterList$localWorkDir,
#          Obio@parameterList$project_id,
#          ".temp.bioLOGIC.Robj"
#      )
# )

#print("Obio Object saved.")

# save(OsCsub,
#     file = paste0(
#          Obio@parameterList$localWorkDir,
#          Obio@parameterList$project_id,
#         ".Seurat.Robj"
#      )
# )

library(AUCell)
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

# Defined in the section above #

## This needs to become a gmt file ##
if (is.null(Obio@parameterList$catRefFile)){
    FNcat <- paste0(hpc.mount, "Projects/schaefera/tobias.ackels/360_scRNAseq_mm_10X_1M_neurons_20k/basedata/asl320.referenceCats.txt")
} else {
    FNcat <- Obio@parameterList$catRefFile
}

if (length(grep(".gmt$", FNcat)) > 0){
    print("Load gmt file. To be implemented.")
    stop()
} else {
    dfHeatmapGenes <- read.delim(
      FNcat,
      header = T,
      sep = "\t",
      stringsAsFactors = F
      
    )
    
    if (is.null(Obio@parameterList[["cat2DotplotList"]])){
        Obio@parameterList[["cat2DotplotList"]] <- list()
    }
    
    if (is.null(Obio@parameterList[["cat2HMplotList"]])){
        Obio@parameterList[["cat2HMplotList"]] <- list()
    }
    
    
    

for (i in 1:ncol(dfHeatmapGenes)){
    genes <- as.vector(dfHeatmapGenes[2:nrow(dfHeatmapGenes),i])
    genes <- genes[genes %in% rownames(x = OsCsub@assays$RNA)]
    if (length(unique(genes)) < 61 |  (length(unique(genes)) > 0)){
        Obio@parameterList[["cat2DotplotList"]][[names(dfHeatmapGenes)[i]]] <- genes
    }
    
    if ((length(unique(genes)) < 501) |  (length(unique(genes)) > 2) ){
        Obio@parameterList[["cat2HMplotList"]] [[names(dfHeatmapGenes)[i]]] <- genes
    }
}
}


## Add transcription factors to dotplot ##
if (Obio@parameterList$geneIDcolumn != "mgi_symbol" & Obio@parameterList$geneIDcolumn != "hgnc_symbol") {
    queryGS <- "hgnc_symbol" 
} else {
    queryGS <- Obio@parameterList$geneIDcolumn
}


tempVec <- retrieve.gene.category.from.db(
    cat_id = "ag_lab_categories__10",
    password = db.pwd,
    gene.symbol = queryGS,
    user = Obio@dbDetailList$db.user,
    host = Obio@dbDetailList$host
)

###############################################################################
## If this is fish, translation is non-human or non-mouse, translation is necessary
if (queryGS != Obio@parameterList$geneIDcolumn){
    dfAnno <- Obio@dfGeneAnnotation
    dfAnno <- unique(dfAnno[,c("hgnc_symbol",Obio@parameterList$geneIDcolumn )])
    dfAnno <- dfAnno[dfAnno$hgnc_symbol != "", ]
    dfAnno <- dfAnno[dfAnno$hgnc_symbol %in% tempVec, ]
    tempVec <- unique(dfAnno[,Obio@parameterList$geneIDcolumn])
    tempVec <- tempVec[tempVec != ""]
}

dfHMG <- dfGeneralMarkers[dfGeneralMarkers$gene %in% tempVec, ]
dfHMGsel <- data.frame(dfHMG %>% group_by(cluster) %>% top_n(5, avg_diff))

Obio@parameterList[["cat2DotplotList"]][["Top5_TF_per_cluster_Markers"]] <- as.vector(unique(dfHMGsel$gene))

## Add cluster defining transcription factors to the collection ##
## For the dotplot ##


###############################################################################
## Get backdrop

exprMatrix <- as.matrix(OsCsub@assays$RNA@counts)
#logMat <- log10(exprMatrix+1)

# When using a Seurat object #
logMat <- data.frame(OsCsub[["RNA"]]@data)

## Load tSNE coordinates ##
cellsTsne <- data.frame(OsCsub@reductions$umap@cell.embeddings)

## done
FNbase <- paste0("CatScatter_Rankings", VersionPdfExt)
FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
FNrel <- paste0("report_figures/", FNbase)
    

pdf(FN)
    cells_rankings <- AUCell_buildRankings(exprMatrix)
dev.off()

geneSets <- Obio@parameterList$cat2DotplotList

cells_AUC <- AUCell_calcAUC(geneSets, cells_rankings, aucMaxRank=nrow(cells_rankings)*0.05)

## Select thresholds ##


FNbase <- paste0("CatScatterHist", VersionPdfExt)
FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
FNrel <- paste0("report_figures/", FNbase)
            
pdf(FN)
    set.seed(123)
    cells_assignment <- AUCell_exploreThresholds(
        cells_AUC, 
        plotHist=TRUE, 
        nCores=1, 
        assign=TRUE
    )
dev.off()


## Add data to dfExpr ##

## Plot CatScatters ##
for (i in 1:length(Obio@parameterList$cat2DotplotList)){
    HMname <- names(Obio@parameterList$cat2DotplotList)[i]
    tag <- gsub("[.]", "_", HMname)
    
    FNbase <- paste0("CatScatterHist_", HMname, VersionPdfExt)
    FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    
    selectedThresholds <-  cells_assignment[[i]]$aucThr$thresholds 
    
    if ("minimumDens" %in% rownames(selectedThresholds)) {
        pThr <- selectedThresholds["minimumDens", "threshold"]
    } else if ("Global_k1" %in% rownames(selectedThresholds)){
        pThr <- selectedThresholds["Global_k1", "threshold"]
    } else {
        pThr <- selectedThresholds[1, "threshold"]
    }
    
    if (nrow(cellsTsne) > 15000){
        cex = 0.25
    } else if (nrow(cellsTsne) > 1000){
        cex = 0.5 
    } else {
        cex = 1
    }
    
    
    ## Get AUC matrix ##
    tSNE.df <- data.frame(cellsTsne, cell=rownames(cellsTsne))
    mAUC <- getAUC(cells_AUC)[HMname,rownames(tSNE.df)]
    dfAUC <- data.frame(mAUC)
    dfAUC[["cellID"]] <- row.names(dfAUC)
    dfAUC <- merge(dfAUC, tSNE.df, by.x = "cellID", by.y = "cell")
    
    
    input <- list(
        "x_axis" = "UMAP1",
        "y_axis" = "UMAP2",
        "gene" = HMname
    )
    dotsize <- cex
    
    legendNote <- paste0(
            " The following genes of this dataset are represented in this figure: ",
            paste0(sort(Obio@parameterList$cat2DotplotList[[i]]), collapse = ", ")
        )
    
     plotList[[tag]] <- ggplot(data = dfAUC, aes(x=UMAP_1, y=UMAP_2, color = mAUC)
            )+ geom_point( shape=16, size = dotsize
            ) + scale_color_gradient2(low="grey", high="darkblue"
            ) + xlab(input$x_axis) + ylab(input$y_axis)  +  theme(
                axis.text.y   = element_text(size=8),
                axis.text.x   = element_text(size=8),
                axis.title.y  = element_text(size=8),
                axis.title.x  = element_text(size=8),
                axis.line = element_line(colour = "black"),
                panel.border = element_rect(colour = "black", fill=NA, size=1),
                plot.title = element_text(hjust = 0.5, size = 12)
            )+ ggtitle(paste0("Category: ", input$gene)
            ) + coord_fixed(ratio = 1) #+ theme(legend.position="none") 
     
    FNbase <- paste0("CatScatter", HMname, VersionPdfExt)
    FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    pdf(FN)
        print(plotList[[tag]])
    dev.off()
    ## Create R markdown chunk ##
    figLegend <- paste0(
        "**Figure ", 
        figureCount, 
        "A:** Category Scatter showing gene category ", 
        HMname, ". ", legendNote, 
        ". Download a pdf of this figure [here](", FNrel,"). "
    )
            
            
   
            
    NewChnk <- paste0(
        "#### Category Feature Plot ",HMname,
                "\n```{r CatFeatPlot1_",
                i,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                figLegend,"'}\n",
                "\n",
                "\n print(plotList[['",tag,"']])",
                "\n cat(  '\n')",
                "\n\n\n```\n"   
            )
          
        
    
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
    
    ###########################################################################
    ## Add part B - dotplot                                                  ##
    DefaultAssay(OsCsub) <- "RNA"

    OsCsub@meta.data[["hmIdent2"]] <- paste0("C", OsCsub@meta.data[,Obio@parameterList$singleCellClusterString])
    
    levels <- paste0(
      "C",
      sort(unique(OsCsub@meta.data[,Obio@parameterList$singleCellClusterString]))
      )
    
    OsCsub@meta.data$hmIdent2 <- factor(OsCsub@meta.data$hmIdent2, levels=levels)
    
    Idents(OsCsub) <- "hmIdent2"
    
    
    
    
   
        HMname <- paste0("Dotplot_", names(Obio@parameterList$cat2DotplotList)[i])
        tag <- gsub("[.]", "_", HMname)
        
        dpGenes <- unique(Obio@parameterList$cat2DotplotList[[i]])
        legendNote <- paste0("The following genes were found in this category and the single-cell dataset: ", paste0(dpGenes, collapse=", "))
        
        plotList[[tag]] <- DotPlotSB(
            object = OsCsub,
            features = dpGenes,
            #cols = cols,
            group.by = NULL,
            split.by = NULL,
            dot.scale = 4,
            col.min = 0,
            col.max = 5,
            #assay = "RNA"
        ) + ggtitle(gsub("_", "", tag)) + coord_fixed() + coord_flip()
        
        FNbase <- paste0(HMname, VersionPdfExt)
        FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
        FNrel <- paste0("report_figures/", FNbase)
        
        pdf(FN)
            print(plotList[[tag]])
        dev.off()
        ## Create R markdown chunk ##
        figLegend <- paste0(
            "**Figure ", 
            figureCount, 
            "B:** Dotplot showing gene category ", 
            HMname, ". ", legendNote, 
            ". Download a pdf of this figure [here](", FNrel,"). "
        )
                
                
        figureCount <- figureCount + 1 
                
        NewChnk <- paste0(
                    "\n```{r ",tag,
                    ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                    figLegend,"'}\n",
                    "\n",
                    "\n print(plotList[['",tag,"']])",
                    "\n cat(  '\n')",
                    "\n\n\n```\n"   
                )
              
            
        
        chnkVec <- c(
            chnkVec,
            NewChnk
        )
    
    
    ## Done adding dotplot                                                   ##
    ###########################################################################
    
    
}


```

```{r enrichment_plots, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
```



### Heatmap and Dotplot for the top5 cluster defining genes
```{r Cluster_overview_data_prep, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
#plan("sequential")

plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

###############################################################################
## Make Heatmap                                                              ##

OsCsub@meta.data[["hmIdent2"]] <- paste0("C", OsCsub@meta.data[,Obio@parameterList$singleCellClusterString])

Idents(OsCsub) <- "hmIdent2"

levels <- paste0(
  "C",
  sort(unique(OsCsub@meta.data[,Obio@parameterList$singleCellClusterString]))
  )

levels(OsCsub) <- levels

## Deal with more than 5000 cells ##
if (nrow(OsCsub@meta.data) > 5000){
    set.seed(127)
    n.cells <- 5000
    OsCsub_HM <- SubsetData(OsCsub, cells.use = sample(x = object@cell.names, size = n.cells) )
    subsetString <- paste0("For this heatmap 5000 cells were randomly selected from ", nrow(OsCsub@meta.data)," cells in the experiment. ")
} else {
    OsCsub_HM <- OsCsub
    subsetString <- ""
}

## Scale Data ##
allGenes <- rownames(x = OsCsub_HM@assays$RNA)
OsCsub_HM <- ScaleData(OsCsub_HM, verbose = FALSE, features=allGenes)

## For the moment: resetting heatmap list to keep it short ##
Obio@parameterList[["cat2HMplotList"]] <- list()
Obio@parameterList[["cat2HMplotList"]][["Top5_Cluster_Markers"]] <- as.vector(dfTop5$gene)

## Add top transcription factors for each cluster ##
## Get transcription factor genes ##
if (Obio@parameterList$geneIDcolumn != "mgi_symbol" & Obio@parameterList$geneIDcolumn != "hgnc_symbol") {
    queryGS <- "hgnc_symbol" 
} else {
    queryGS <- Obio@parameterList$geneIDcolumn
}


tempVec <- retrieve.gene.category.from.db(
    cat_id = "ag_lab_categories__10",
    password = db.pwd,
    gene.symbol = queryGS,
    user = Obio@dbDetailList$db.user,
    host = Obio@dbDetailList$host
)

###############################################################################
## If this is fish, translation is non-human or non-mouse, translation is necessary
if (queryGS != Obio@parameterList$geneIDcolumn){
    dfAnno <- Obio@dfGeneAnnotation
    dfAnno <- unique(dfAnno[,c("hgnc_symbol",Obio@parameterList$geneIDcolumn )])
    dfAnno <- dfAnno[dfAnno$hgnc_symbol != "", ]
    dfAnno <- dfAnno[dfAnno$hgnc_symbol %in% tempVec, ]
    tempVec <- unique(dfAnno[,Obio@parameterList$geneIDcolumn])
    tempVec <- tempVec[tempVec != ""]
}

dfHMG <- dfGeneralMarkers[dfGeneralMarkers$gene %in% tempVec, ]
dfHMGsel <- data.frame(dfHMG %>% group_by(cluster) %>% top_n(5, avg_diff))

Obio@parameterList[["cat2HMplotList"]][["Top5_TF_Cluster_Markers"]] <- as.vector(unique(dfHMGsel$gene))

## Done with translation
  



for (i in 1:length(Obio@parameterList[["cat2HMplotList"]])){
    tag <- paste0("HMM_", names(Obio@parameterList$cat2HMplotList)[i])
    textSize <- 5    
    
    plotList[[tag]] <- DoHeatmap(
                object = OsCsub_HM,
                features = Obio@parameterList[["cat2HMplotList"]][[i]],
                #group.by = "hmIdent",
                draw.lines =T,
                label = T,
                group.bar = TRUE,
                slot = "scale.data",
                lines.width = 2 #With of separation lines in 'cells'
                #slim.col.label = TRUE,
                #remove.key = removeKey
            # ) + theme(legend.position = "none"
            ) + theme(text = element_text(size=textSize)
            ) + scale_fill_gradientn(colors = c("blue", "white", "red"))
    
    
    ## Save to file ##
            FNbase <- paste0(tag, VersionPdfExt)
            FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
            FNrel <- paste0("report_figures/", FNbase)

            pdf(FN)
                print(plotList[[tag]])
            dev.off()


            ## Create R markdown chunk ##
            figLegend <- paste0(
                "**Figure ",
                figureCount,
                ":** Heatmap showing the most distinct marker genes in each cluster. " , subsetString,
                "Download a pdf of this figure [here](", FNrel,"). "
            )






            figureCount <- figureCount + 1

            NewChnk <- paste0(
                "#### Heatmap_var_genes",
                "\n```{r Heatmap_", tag,
                ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                figLegend,"'}\n",
                "\n",
                "\n print(plotList[['",tag,"']])",
                "\n cat(  '\n')",
                "\n\n\n```\n"
            )

        
    
    chnkVec <- c(
        chnkVec,
        NewChnk
)
    
}



rm(OsCsub_HM)


## Done making Heatmap                                                       ##
###############################################################################

###############################################################################
## Make dotplot 

Idents(OsCsub) <- "hmIdent2"

levels <- paste0(
    "C",
    sort(unique(OsCsub@meta.data[,Obio@parameterList$singleCellClusterString]))
 )

levels(OsCsub) <- levels


dpGenes <- as.vector(unique(dfTop5$gene))

if (length(dpGenes) >= 50){
  dpGenes <- as.vector(unique(dfTop1$gene))  
}


#dpGenes <- rev(dpGenes[!(duplicated(dpGenes))])

tag <- paste0("Dotplot_", "Var_Genes")
textSize <- 5

plotList[[tag]] <- DotPlotSB(
        object = OsCsub,
        features = dpGenes,
        #cols = cols,
        group.by = NULL,
        split.by = NULL,
        dot.scale = 4,
        col.min = 0,
        col.max = 5
    ) + ggtitle(gsub("_", "", tag)) + coord_fixed() + coord_flip()
    

## Save to file ##
            FNbase <- paste0(tag, VersionPdfExt)
            FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
            FNrel <- paste0("report_figures/", FNbase)

            pdf(FN)
                print(plotList[[tag]])
            dev.off()


            ## Create R markdown chunk ##
            figLegend <- paste0(
                "**Figure ",
                figureCount,
                ":** Dotplot showing showing selected marker genes. ",
                "Download a pdf of this figure [here](", FNrel,"). "
            )






            figureCount <- figureCount + 1

            NewChnk <- paste0(
                "#### Dotplot Markers",
                "\n```{r Dotplot_var_",
                ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                figLegend,"'}\n",
                "\n",
                "\n print(plotList[['",tag,"']])",
                "\n cat(  '\n')",
                "\n\n\n```\n"
            )

        
    
    chnkVec <- c(
        chnkVec,
        NewChnk
)    
    
## Done making dotplot                                                       ##
###############################################################################

############################
## Make cat feature plot

## done making cat feature plot
################################

```

```{r Cluster_overview, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}

cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))

```

### Cluster-defining Genes Table {.tabset .tabset-fade .tabset-pills}

```{r ClusterTable_data_prep, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
library(DT)

###############################################################################
## Calculate percentages of expressed genes                                  ##
DefaultAssay(OsCsub) <- "RNA"
my_genes <- rownames(x = OsCsub@assays$RNA)

exp <- FetchData(OsCsub, my_genes)

ExprMatrix <- round(as.matrix(colMeans(exp  > 0)) *100,1)
colnames(ExprMatrix)[1] <- "count_cut_off"
dfExprMatrix <- data.frame(ExprMatrix)
dfExprMatrix[["gene"]] <- row.names(dfExprMatrix)

Obio@dataTableList[["dfPercCellsExpr"]] <- dfExprMatrix

## Done calculating percentages of expressed gens                            ##
###############################################################################

###############################################################################
## Create one table per cluster                                              ##
## Add expressed in N percent cells ##
dfPercCellsExpr <- Obio@dataTableList$dfPercCellsExpr

dfDat <- Obio@dataTableList$dfGeneralMarkersFilt
dfDat <- dfDat[,c("cluster", "gene", "avg_diff", "pct.1", "pct.2","myAUC", "power")]
dfDat$avg_diff <- round(dfDat$avg_diff,2)
dfDat$cluster <- paste0("Cluster_",dfDat$cluster, "_C")
dfDat$gene <- substr(dfDat$gene,1,50)

dfDat[["uniqueMarker"]] <- as.character(!duplicated(dfDat$gene))
dfDat$uniqueMarker <- substr(dfDat$uniqueMarker, 1,1)

dtList <- list()

tabClusters <- sort(unique(dfDat$cluster))
chnkVec <- as.vector(NULL, mode="character")
    
linkGeneView <- paste0("https://",urlString,"/",Obio@parameterList$project_id,"/gene-view")
linkFeatureView <- paste0("https://",urlString,"/mdata/",Obio@parameterList$project_id,"/html/FeatureView.html")

#for (i in 1:length(tabClusters)){
    #tabLegend = paste0("**Table: ** Positive and negative marker genes for ", tabClusters[i])
    tabLegend = paste0("**Table: ** Positive and negative cluster-defining marker genes. Perc_Cells_Expr: Percentage of total cells expressing gene X. Enr in Cluster: Enrichment of gene X in cluster Y. To collapse the table to one particular cluster, type the name of the cluster in the search box.",
    "Use the [GeneView](",linkGeneView,") or [FeatureView](",linkFeatureView,") functionalities to examine individual genes in more detail. "                   
    )
    #dfTempDat <- dfDat[dfDat$cluster == tabClusters[i],]
    dfTempDat <- dfDat
    
    ## Percent expressed genes 
    dfTempDat <- merge(
        dfTempDat, 
        Obio@dataTableList$dfPercCellsExpr,
        by.x = "gene",
        by.y = "gene"
    )
    
    names(dfTempDat) <- gsub("count_cut_off", "Perc_Cells_Expr",names(dfTempDat))
    names(dfTempDat) <- gsub("myAUC", "AUC", names(dfTempDat))
    names(dfTempDat) <- gsub("[.]", "", names(dfTempDat))
    
    #dtList[[paste0("Table",i)]] <- datatable(dfDat,rownames = FALSE) 
    if (Obio@dbDetailList$host == "10.27.241.234"){
      urlString <- "biologic.thecrick.org"
    } else {
      urlString <- "biologic.crick.ac.uk"
    }
    
    
    dfTempDat$gene <- paste0("<a href='https://",urlString,"/",Obio@parameterList$project_id,"/gene-view?query=",dfTempDat$gene,"&exact=TRUE' target='_blank'>", dfTempDat$gene, "</a>")
    
    NewChnk <- paste0(
        "#### ", names(dtList),
        "\n```{r datatable_",
        i,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
        tabLegend,"'}\n",
        "\n",
        "\n datatable(dfTempDat,rownames = FALSE,  escape = FALSE)",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
    )
    
    chnkVec <- c(
            chnkVec,
            NewChnk
        )
#}

## Done creating one table per cluster                                      ##
##############################################################################
```


```{r render_ClusterTable, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}

cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))

```

### Define categories enriched in individual clusters {.tabset .tabset-fade .tabset-pills}
```{r create-markdown-enrichment-chunks-prepare-data, echo=T, eval=TRUE, warning=FALSE, results=F}
# library(knitr)
# library(ggplot2)
# 
# #save.image("temp.RData")
# 
# DefaultAssay(OsCsub) <- "RNA"
# 
# dfGeneralMarkersFilt <- data.frame(
#     Obio@dataTableList$dfGeneralMarkersFilt,
#     stringsAsFactors = FALSE
# )
# 
# dfGeneralMarkersFilt$cluster <- paste0("C", dfGeneralMarkersFilt$cluster)
# 
# clusterIDs <- as.character(unique(dfGeneralMarkersFilt$cluster))
# 
# plotList <- list()
# chnkVec <- as.vector(NULL, mode = "character")
# 
# for (j in 1:length(clusterIDs)){
#     dfEnrichSel <- dfGeneralMarkersFilt[dfGeneralMarkersFilt$cluster == as.vector(clusterIDs[j]), ]
#     #dfEnrichSel[["gene"]] <- row.names(dfEnrichSel)
#     
#     posTestGeneSet <- as.vector(
#         unique(
#             dfEnrichSel[dfEnrichSel$avg_diff > lgFCthreshold, "gene"]
#         )
#     )
#     
#     
#     negTestGeneSet <- as.vector(
#         unique(
#             dfEnrichSel[dfEnrichSel$avg_diff < -lgFCthreshold, "gene"]
#         )
#     )
#     
#     ## Get background gene set ##
#     #backgroundGeneVec <- row.names(OsCsub[["RNA"]]@counts)
#     if ((length(posTestGeneSet) >= 3) |(length(negTestGeneSet) >= 3)){
#         library(enrichR)
#         topMaxCat <- 10
#         dbs <- listEnrichrDbs()
#         
#         dbs <- c("GO_Biological_Process_2017")
#         
#         
#         PosEnriched <- enrichr(posTestGeneSet, dbs)
#         
#         for (i in 1:length(dbs)){
#             dfTemp <- PosEnriched[[dbs[i]]]
#             
#             if (i ==1){
#                 dfPosEnriched <- dfTemp
#             } else {
#                 dfPosEnriched <- rbind(
#                     dfPosEnriched,
#                     dfTemp
#                 )
#             }
#             
#         }
#         
#         dfPosEnriched[["log10FDR"]] <- -1*log10(dfPosEnriched$Adjusted.P.value)
#         dfPosEnriched <- dfPosEnriched[order(-dfPosEnriched$log10FDR),]
#         dfPosEnriched <- na.omit(dfPosEnriched)
#         
#         ## Negative Side ##
#         NegEnriched <- enrichr(negTestGeneSet, dbs)
#         
#         for (i in 1:length(dbs)){
#             dfTemp <- NegEnriched[[dbs[i]]]
#             
#             if (i ==1){
#                 dfNegEnriched <- dfTemp
#             } else {
#                 dfNegEnriched <- rbind(
#                     dfNegEnriched,
#                     dfTemp
#                 )
#             }
#             
#         }
#         
#         
#         dfNegEnriched[["log10FDR"]] <- -1*log10(dfNegEnriched$Adjusted.P.value)
#         dfNegEnriched <- dfNegEnriched[order(-dfPosEnriched$log10FDR),]
#         dfNegEnriched <- na.omit(dfNegEnriched)
#         
#         dfNegSel <- dfNegEnriched
#         if (nrow(dfNegSel) > topMaxCat){
#             dfNegSel <- dfNegSel[1:topMaxCat,]
#         }
#         
#         dfPosSel <- dfPosEnriched
#         if (nrow(dfPosSel) > topMaxCat){
#             dfPosSel <- dfPosSel[1:topMaxCat,]
#         }
#         
#         if ((nrow(dfNegEnriched) > 0) | (nrow(dfPosEnriched) > 0)){
#             
#             
#             dfNegSel$log10FDR <- -1* dfNegSel$log10FDR
#             
#             dfSel <- rbind(
#                 dfNegSel,
#                 dfPosSel
#             )
#             
#             dfSel <- na.omit(dfSel)
#             dfSel <- dfSel[order(dfSel$log10FDR),]
#             dfSel$log10FDR <- round(dfSel$log10FDR, 2)
#             
#             dfSel[["Category"]] <- ""
#             dfSel[dfSel$log10FDR >= 0, "Category"] <- "Enr."
#             dfSel[dfSel$log10FDR < 0, "Category"] <- "Depl."
#             
#             for (k in 1:nrow(dfSel)){
#                 if (nchar(dfSel[k, "Term"]) > 50 & length(grep("\\(GO", as.vector(dfSel[k, "Term"]))) > 0){
#                     part1 <- unlist(strsplit(as.vector(dfSel[k, "Term"]), "\\(GO"))[1]
#                     part1 <- substr(part1, 1, 45)
#                     part2 <- unlist(strsplit(as.vector(dfSel[k, "Term"]), "\\(GO"))[2]
#                     part2 <- paste0("\\(GO", part2)
#                     
#                     if (nchar(part1) > 40 ){
#                         dfSel[k, "Term"] <- paste0(part1, " \\n", part2)
#                     } else { 
#                         dfSel[k, "Term"] <- paste0(part1, " ", part2)
#                     }
#                 }
#             }
#             
#             
#             #dfSel$Term <- gsub("\\(GO", "\\\n\\(GO", dfSel$Term)
#             
#             dfSel$Term <- factor(dfSel$Term, levels = unique(dfSel$Term))
#             
#             plotList[[paste0("ENR_", j)]] <- ggplot(
#                 data=dfSel, aes(x= Term, y=log10FDR, fill=Category, order=log10FDR)
#             ) + geom_bar(stat="identity", colour="black"
#             ) + coord_flip() +scale_fill_manual(values=c("yellow", "blue"))  +  theme(
#                 axis.text.y   = element_text(size=8),
#                 axis.text.x   = element_text(size=8),
#                 axis.title.y  = element_text(size=8),
#                 axis.title.x  = element_text(size=8),
#                 axis.line = element_line(colour = "black"),
#                 panel.border = element_rect(colour = "black", fill=NA, size=1),
#                 plot.title = element_text(hjust = 0.5, size = 12)
#             )  + labs(title = paste0("Cluster ", clusterIDs[j]," enriched genes") ,y = "-log10(FDR)", x = ""
#             ) + geom_hline(yintercept = c(-log10(0.05), log10(0.05)), color = "red", size=0.5, lty=2
#             ) + geom_hline(yintercept = 0, color = "black", size=0.5
#             )
#             cat("  \n")
#             
#             
#             
#             ## Save to file ##
#             FNbase <- paste0("Cluster_", clusterIDs[j],".enriched.genes", VersionPdfExt)
#             FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
#             FNrel <- paste0("report_figures/", FNbase)
#             
#            
#             pdf(FN)
#             print(plotList[[paste0("ENR_", j)]])
#             dev.off()
#             
#             ## Create R markdown chunk ##
#             link <- paste0(
#                 "https://",urlString,"/",
#                 Obio@parameterList$project_id,
#                 "/category-view?category_type=GO-BP"
#             )
#             
#             
#             figLegend <- paste0(
#                 "**Figure ", 
#                 figureCount, 
#                 "**: GO-BP category enrichment analysis for genes that are <font color = \\'yellow\\'>higher</font> and <font color = \\'blue\\'>lower</font> expressed in Cluster ", 
#                 clusterIDs[j],
#                 " as compared to all other clusters. Download a pdf of this figure [here](", FNrel, "). To view these gene sets in the context of your data, go to [CategoryView > GO-BP](",link,") and find the above categories using the search box."
#             )
#             figureCount <- figureCount + 1 
#             
#             NewChnk <- paste0(
#                 "#### ", clusterIDs[j],
#                 "\n```{r enrchr_cluster_",
#                 clusterIDs[j],", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
#                 figLegend,"'}\n",
#                 "\n",
#                 "\n print(plotList[['",paste0("ENR_", j),"']])",
#                 "\n cat(  '\n')",
#                 "\n\n\n```\n"   
#             )
#         }
#     }
#     chnkVec <- c(
#         chnkVec,
#         NewChnk
#     )
# }

```

```{r create-markdown-enrichment-chunks-dynamically, echo=T, eval=TRUE, warning=FALSE, results='asis'}
###############################################################################
## Do category enrichment on clusters                                        ##
#cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
## Done doing enrichment on clusters                                         ##
###############################################################################


```

### Gene Set Heatmaps {.tabset .tabset-fade .tabset-pills}
```{r MakeHM, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}

## Two options: full heatmap and averaged heatmaps
## https://satijalab.org/seurat/v3.0/interaction_vignette.html

###############################################################################
## Add percentage expressed genes                                            ##
# DefaultAssay(OsCsub) <- "RNA"
# my_genes <- rownames(x = OsCsub@assays$RNA)
# 
# exp <- FetchData(OsCsub, my_genes)
# 
# ExprMatrix <- round(as.matrix(colMeans(exp  > 0)) *100,1)
# colnames(ExprMatrix)[1] <- "count_cut_off"
# dfExprMatrix <- data.frame(ExprMatrix)
# dfExprMatrix[["gene"]] <- row.names(dfExprMatrix)
# 
# Obio@dataTableList[["dfPercCellsExpr"]] <- dfExprMatrix
# 
# hmRelevantGenes <- as.vector(unique(dfExprMatrix[dfExprMatrix$count_cut_off > Obio@parameterList$singleCellPercExpressedMinCutOff, "gene"]))
# 
# 
# 
# ## Done adding percentage expressed                                          ##
# ###############################################################################
# 
# 
# ###############################################################################
# ## Make plot according to reference categories                               ##
# allGenes <- rownames(x = OsCsub@assays$RNA)
# OsCsub <- ScaleData(OsCsub, verbose = FALSE, features=allGenes)
# 
# DefaultAssay(OsCsub) <- "RNA"
# 
# 
# ## Add heatmap identities to meta.data ##
# 
OsCsub@meta.data[["hmIdent"]] <- paste0(
     OsCsub@meta.data[,Obio@parameterList$singleCellClusterString],
     "_",
     substr(OsCsub@meta.data$sampleID,1,10)
     
)
# 
if (length(unique(OsCsub@meta.data$hmIdent)) > 25){
     OsCsub@meta.data[["hmIdent"]] <- OsCsub@meta.data[,Obio@parameterList$singleCellClusterString]
     
}
# 
# Idents(OsCsub) <- "hmIdent"
# 
# ## Done adding heatmap identities to meta.data ##
# 
# printPdf <- TRUE
# referenceList <- Obio@dataTableList$referenceList

# for (i in 1:length(referenceList)){
#     HMname <- names(referenceList)[i]
#     geneVec <- unique(referenceList[[i]][referenceList[[i]] %in% rownames(x = OsCsub@assays$RNA)])
#     
#     if (length(geneVec) > 50){
#         geneVec <- geneVec[geneVec %in% hmRelevantGenes]
#     }
#     
#     
#     ## Do Heatmap ##
#     if (length(geneVec) < 1500 & length(geneVec) > 2){
#         Idents(OsCsub) <- "hmIdent"
#         HMname <- names(referenceList)[i]
#         cat("\n")
#         cat(paste0("**Heatmap ", HMname,"**"))
#         cat("\n")
#         cat("\n")
#         ## Cluster genes ##
#         HMgenes <- referenceList[[i]]
#         #dfCluster <- OsCsub@assays$integrated
#         Mexpr <- GetAssayData(object = OsCsub, assay.type = "integrated", slot = "scale.data")
#         HMgenesSel <- HMgenes[HMgenes %in% row.names(Mexpr)]
#         
#         if (length(HMgenesSel) > 2){
#             Mexpr <- Mexpr[HMgenesSel,]    
#             
#             pdf(paste0("temp", VersionPdfExt))
#             hmRes <- make.hm(
#                 m.df1 = Mexpr, 
#                 filename = "", 
#                 k.number = 1, 
#                 n.colors = 1000, 
#                 hclust.method = "complete", 
#                 dist.method = "euclidean", 
#                 main = "",
#                 Colv = TRUE,
#                 showRowNames = TRUE,
#                 showColNames = F,
#                 plotSeparationLines = FALSE
#             )
#             dev.off()
#             
#             orderedGenes <- as.vector(unique(row.names(hmRes$sorted)))
#             
#             if (length(unique(OsCsub@meta.data$hmIdent)) > 10){
#                 removeKey <- TRUE
#             } else {
#                 removeKey <- FALSE
#             }
#             
#             if (length(orderedGenes) <= 50){
#                 label = TRUE
#             } else {
#                 label = FALSE
#             }
#             
#             p1 <- DoHeatmap(
#                 object = OsCsub, 
#                 features = orderedGenes,
#                 #group.by = "hmIdent",
#                 draw.lines =T,
#                 label = label,
#                 group.bar = TRUE,
#                 slot = "scale.data",
#                 lines.width = 2 #With of separation lines in 'cells'
#                 #slim.col.label = TRUE, 
#                 #remove.key = removeKey
#             ) + theme(legend.position = "none")
#             
#             print(p1)
#             
#             ## Save to file ##
#             FNbase <- paste0("HM", HMname, VersionPdfExt)
#             FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
#             FNrel <- paste0("report_figures/", FNbase)
#             
#             pdf(FN)
#                 print(p1)
#             dev.off()
#             
#             cat("\n")
#             cat(paste0('Download a pdf of this figure [here](', FNrel, '). '))
#             cat("\n")
#             cat("\n")
#             
#         }
#     }
# }


## Done making plots according to gene categories                            ##
###############################################################################
```

```{r calculate-markdown-heatmap-chunks-dynamically, echo=F, eval=TRUE, warning=FALSE, results="asis"}
#setwd(Obio@parameterList$localWorkDir)
#load("temp.RData")
# DefaultAssay(OsCsub) <- "RNA"
# library(knitr)
# library(ggplot2)
# plotList <- list()
# chnkVec <- as.vector(NULL, mode = "character")
# 
# #referenceList <- referenceList[c(1,2)]
# 
# for (i in 1:length(referenceList)) {
#     ## Step 1 create Heatmap object ##
#     HMname <- names(referenceList)[i]
#     geneVec <- unique(referenceList[[i]][referenceList[[i]] %in% rownames(x = OsCsub@assays$RNA)])
#     nGenes <- length(unique(geneVec))
#     
#     if (nGenes > 50){
#         geneVec <- geneVec[geneVec %in% hmRelevantGenes]
#         nGenes <- length(unique(geneVec))
#         
#         legendNote <- paste0(
#             " Note: Only genes that were expressed in more than ",Obio@parameterList$singleCellPercExpressedMinCutOff,"% of all cells are shown in this figure. "
#         )
#         
#     } else {
#         legendNote <- paste0(
#             "Note: All genes in this category are displayed in the heatmap. "
#         )
#     }
#     
#     
#     ## Do Heatmap ##
#     if (length(geneVec) < 1500 & length(geneVec) > 2){
#         Idents(OsCsub) <- "hmIdent"
#         HMname <- names(referenceList)[i]
#         
#         ## Cluster genes ##
#         #HMgenes <- referenceList[[i]]
#         #dfCluster <- OsCsub@assays$integrated
#         Mexpr <- GetAssayData(object = OsCsub, assay.type = "integrated", slot = "scale.data")
#         HMgenesSel <- geneVec[geneVec %in% row.names(Mexpr)]
#         
#         if (length(HMgenesSel) > 2){
#             Mexpr <- Mexpr[HMgenesSel,]    
#             
#             pdf(paste0("temp", VersionPdfExt))
#             hmRes <- make.hm(
#                 m.df1 = Mexpr, 
#                 filename = "", 
#                 k.number = 1, 
#                 n.colors = 1000, 
#                 hclust.method = "complete", 
#                 dist.method = "euclidean", 
#                 main = "",
#                 Colv = TRUE,
#                 showRowNames = TRUE,
#                 showColNames = F,
#                 plotSeparationLines = FALSE
#             )
#             dev.off()
#             
#             orderedGenes <- as.vector(unique(row.names(hmRes$sorted)))
#             
#             if (length(unique(OsCsub@meta.data$hmIdent)) > 20){
#                 removeKey <- TRUE
#             } else {
#                 removeKey <- FALSE
#             }
#             
#             textSize <- 6
#             if (length(orderedGenes) <= 50){
#                 label = TRUE
#                 if (length(orderedGenes) <= 20){
#                     textSize <- 10
#                 } 
#             } else {
#                 label = FALSE
#             }
#             
#             
#             plotList[[paste0("HM_", i)]] <- DoHeatmap(
#                 object = OsCsub, 
#                 features = orderedGenes,
#                 #group.by = "hmIdent",
#                 draw.lines =T,
#                 label = label,
#                 group.bar = TRUE,
#                 slot = "scale.data",
#                 lines.width = 2 #With of separation lines in 'cells'
#                 #slim.col.label = TRUE, 
#                 #remove.key = removeKey
#             # ) + theme(legend.position = "none"
#             ) + theme(text = element_text(size=textSize))
#             
#             ## Save to file ##
#             FNbase <- paste0("HM", gsub(" ", "_", HMname), VersionPdfExt)
#             FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
#             FNrel <- paste0("report_figures/", FNbase)
#             
#             pdf(FN)
#                 print(plotList[[paste0("HM_", i)]])
#             dev.off()
#             
#             
#             ## Create R markdown chunk ##
#             figLegend <- paste0(
#                 "**Figure ", 
#                 figureCount, 
#                 ":** Heatmap showing gene category ", 
#                 HMname, ". ", 
#                 legendNote, 
#                 "Download a pdf of this figure [here](", FNrel,"). "
#             )
#             
#             
#             pos <- grep(names(referenceList)[i], names(Obio@parameterList$HmDisplayCatsFromDb))
#             if (length(pos) ==1){
#                 cat_id <- Obio@parameterList$HmDisplayCatsFromDb[[pos]]
#                 figLegend <- paste0(
#                     figLegend,
#                     "An interactive version of this figure is available [here](",
#                     "https://biologic.thecrick.org/", Obio@parameterList$project_id, "/category-view/", cat_id,
#                     ")."
#                 )
#             }
#             
#             
#             figureCount <- figureCount + 1 
#             
#             NewChnk <- paste0(
#                 "#### Heatmap ",HMname,
#                 "\n```{r Heatmap_",
#                 i,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
#                 figLegend,"'}\n",
#                 "\n",
#                 "\n print(plotList[['",paste0("HM_", i),"']])",
#                 "\n cat(  '\n')",
#                 "\n\n\n```\n"   
#             )
#           
#         }
#     }
#     chnkVec <- c(
#         chnkVec,
#         NewChnk
#     )
#     
# }

```

### Heatmaps for various gene sets {.tabset .tabset-fade .tabset-pills}
```{r create-markdown-heatmap-chunks-dynamically, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
```


### Calculate Average expression
```{r Identify differential expressed genes across conditions, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
## Ensure that hmIdent exists ##

## Want to average by cluster and subset by sampleID ##



OsCsub@meta.data[["clustIdent"]] <- paste0(
    "C_", OsCsub@meta.data[,Obio@parameterList$singleCellClusterString]
)


Idents(OsCsub) <- "clustIdent"
if (length(grep("sampleID", names(OsCsub@meta.data))) > 0){
    cluster.averages <- AverageExpression(OsCsub, return.seurat = TRUE, add.ident = "sampleID")
} else {
    cluster.averages <- AverageExpression(OsCsub, return.seurat = TRUE)
}

Idents(OsCsub) <- "sampleID"

## Retrieved Scaled data ##
dfAvgExpr <- data.frame(cluster.averages[["RNA"]]@data)
dfAvgExpr <- dfAvgExpr[,sort(names(dfAvgExpr))]
dfAvgExpr[["gene"]] <- row.names(dfAvgExpr)

dfAvgScaledData <- data.frame(cluster.averages[["RNA"]]@scale.data)
dfAvgScaledData <- dfAvgScaledData[,sort(names(dfAvgScaledData))]
dfAvgScaledData[["gene"]] <- row.names(dfAvgScaledData)


Obio@dataTableList[["dfAvglg10ExprByClusterBySample"]] <- dfAvgExpr
Obio@dataTableList[["dfAvglg10ExprByClusterBySampleScaled"]] <- dfAvgScaledData
###############################################################################

###############################################################################
## Average by Cluster                                                        ##
Idents(OsCsub) <- "clustIdent"

Idents(OsCsub) <- factor(Idents(OsCsub), levels = sort(levels(OsCsub)))

cluster.averages <- AverageExpression(OsCsub, return.seurat = TRUE)

dfAvgExpr <- data.frame(cluster.averages[["RNA"]]@data)
dfAvgExpr <- dfAvgExpr[,sort(names(dfAvgExpr))]
dfAvgExpr[["gene"]] <- row.names(dfAvgExpr)

dfAvgScaledData <- data.frame(cluster.averages[["RNA"]]@scale.data)
dfAvgScaledData <- dfAvgScaledData[,sort(names(dfAvgScaledData))]
dfAvgScaledData[["gene"]] <- row.names(dfAvgScaledData)


Obio@dataTableList[["dfAvglg10ExprPerCluster"]] <- dfAvgExpr
Obio@dataTableList[["dfAvglg10ExprPerClusterScaled"]] <- dfAvgScaledData

## Done Average by Cluster                                                   ##
###############################################################################

###############################################################################
## Average gene expression by sample                                         ##


Idents(OsCsub) <- "sampleID"
cluster.averages <- AverageExpression(OsCsub, return.seurat = TRUE)

## Retrieved Scaled data ##
dfAvgExpr <- data.frame(cluster.averages[["RNA"]]@data)
selVec <- names(dfAvgExpr)

dfAvgExpr[["gene"]] <- row.names(dfAvgExpr)
selVec <- c("gene", selVec)

dfAvgExpr <- dfAvgExpr[,selVec]


dfAvgScaledData <- data.frame(cluster.averages[["RNA"]]@scale.data)
selVec <- names(dfAvgScaledData)

dfAvgScaledData[["gene"]] <- row.names(dfAvgScaledData)
selVec <- c("gene", selVec)

dfAvgScaledData <- dfAvgScaledData[,selVec]


Obio@dataTableList[["dfAvglg10ExprBySample"]] <- dfAvgExpr
Obio@dataTableList[["dfAvglg10ExprBySampleScaled"]] <- dfAvgScaledData


## Done average gene expression by sample                                    ##
###############################################################################

## Diagnostic plot ##
#DoHeatmap(cluster.averages, features = c("LY75", "PSMB11", "CCL25", "CD274"), size = 3, draw.lines = FALSE, slot = "scale.data")



## Create Plots ##
# Create temp files for plotting #


# for (i in 1:length(clusterIDs)){
#     
#     
#     ## Select Marker genes ##
#     dfTemp <- Obio@dataTableList$dfGeneralMarkersFilt
#     dfTemp <- dfTemp[dfTemp$cluster == clusterIDs[i], ]
#     posGenes <- as.vector(unique(dfTemp[dfTemp$direction == "positive", "gene"]))
#     negGenes <- as.vector(unique(dfTemp[dfTemp$direction == "negative", "gene"]))
#     
#     avgTemp <- dfRes[,c("gene", names(dfRes)[grep(paste0("Cl_", clusterIDs[i]), names(dfRes))])]
#     avgTemp$Selection <- ""
#     avgTemp[avgTemp$gene %in% posGenes, "Selection"] <- "+"
#     avgTemp[avgTemp$gene %in% negGenes, "Selection"] <- "-"
#     names(avgTemp) <- c("gene", "Ctrl", "Prad", "Selection")
#     
#     df_layer_1 <- avgTemp[avgTemp$Selection =="",]
#     df_layer_2 <- avgTemp[avgTemp$Selection =="+",]
#     df_layer_3 <- avgTemp[avgTemp$Selection =="-",]
#     
#     
#     p1 <- ggplot(
#     ) + geom_point(data = df_layer_1, aes(Ctrl, Prad), fill = "black", shape = 21
#     ) + geom_point(data = df_layer_2, aes(Ctrl, Prad), fill = "red", shape = 21
#     ) + geom_point(data = df_layer_3, aes(Ctrl, Prad), fill = "blue", shape = 21
#     ) + ggtitle(paste0("Cluster ", clusterIDs[i])
#     ) +  theme(
#             axis.text.y   = element_text(size=8),
#             axis.text.x   = element_text(size=8),
#             axis.title.y  = element_text(size=8),
#             axis.title.x  = element_text(size=8),
#             axis.line = element_line(colour = "black"),
#             panel.border = element_rect(colour = "black", fill=NA, size=1),
#             plot.title = element_text(hjust = 0.5, size = 12)
#         )
#     
# 
#     #p1 <- LabelPoints(plot = p1, points = genes.to.label, repel = TRUE)
#     cat("\n");cat(paste0("#### Scatterplot Cluster ", clusterIDs[i]));cat("\n")
#     print(p1)
#     
#     ## Save to file ##
#     FNbase <- paste0("scatterplot.",clusterIDs[i],".highlighted.pdf")
#     FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
#     FNrel <- paste0("report_figures/", FNbase)
#     
#     pdf(FN)
#         print(p1)
#     dev.off()
#     
#     
#     cat("Download a pdf of this figure [here](",FNrel,"). ")
#     
# }


```

### Example Feature Plots {.tabset .tabset-fade .tabset-pills}
```{r ExampleFeaturePlot_dataprep, echo=TRUE, eval=TRUE, warning=FALSE,  results="asis"}
DefaultAssay(OsCsub) <- "RNA"

cat(paste0(
    "Feature plots for any gene in this experiment can be viewed [**here**](http://shiny.thecrick.org/babs/boeings/",Obio@parameterList$project_id,"_app/). This link will only work when you are on the Crick network or VPN-connected to the Crick. "
))



plotGenes <- c(
    Obio@dataTableList$referenceList$integrated_top30var[1:10]
)

plotParts <- ceiling(length(plotGenes)/2)

chnkVec <- as.vector(NULL, mode = "character")
plotListF <- list()

for (i in 1:plotParts){
    tag1 <- paste0("Featureplot_",i)
    
    featureGenes <- c(plotGenes[((2*i)-1)], plotGenes[((2*i))])
    
    plotListF[[tag1]] <- FeaturePlot(
        OsCsub,
        features = featureGenes,
        #split.by = "orig.ident",
        reduction = Obio@parameterList$primReduction
    )
    
    
    ## Save to file ##
    FNbase <- paste0("Featureplot.", plotGenes[((2*i)-1)], ".", plotGenes[((2*i))],".", VersionPdfExt)
        FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    pdf(FN)
        print(plotListF[[tag1]])
    dev.off()
    
    linkFeatureView <- paste0("https://",urlString,"/mdata/",Obio@parameterList$project_id,"/html/FeatureView.html")

    figLegend <-  paste0(
        "**Figure ", 
        figureCount, 
        ":** Gene expression plot for genes ", 
        plotGenes[((2*i)-1)], 
        " and ", 
        plotGenes[((2*i))],".",
        " Results for any other gene may be plotted in [FeatureView](",linkFeatureView,")."
    )
      
    NewChnk <- paste0(
        "#### Featureplot ", plotGenes[((2*i)-1)], " and ",plotGenes[((2*i))],
        "\n```{r FeaturePlot_", i,
        ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
        figLegend,"'}\n",
        "\n",
        "\n print(plotListF[['",tag1,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
    )
    
    chnkVec <- c(
        chnkVec,
        NewChnk
    )

}

```

```{r ExampleFeaturePlot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))

```


### PCA Gene and Cell Plots {.tabset .tabset-fade .tabset-pills}

```{r elbow_plot, fig.height=6, fig.width=10, fig.cap=paste0("**Figure, ",figureCount,": Elbowplot** Variance explained per PCA dimension")}
figureCount <- figureCount + 1
# An alternative heuristic method generates an 'Elbow plot': a ranking of principle components based on the percentage of variance explained by each one (`ElbowPlot` function). In this example, we can observe an 'elbow' around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs.

ElbowPlot(object = OsCsub) +  theme(
    axis.text.y   = element_text(size=8),
    axis.text.x   = element_text(size=8),
    axis.title.y  = element_text(size=8),
    axis.title.x  = element_text(size=8),
    axis.line = element_line(colour = "black"),
    panel.border = element_rect(colour = "black", fill=NA, size=1),
    plot.title = element_text(hjust = 0.5, size = 12)
) + ggtitle(paste0("Variance per PCA Dimension"))

## Plot variance per PCA dimension ##

```


```{r PCA_dataprep, echo=TRUE, eval=TRUE, warning=FALSE,  results=F}


## Add PCA coordinates ##
dfTemp <- data.frame(OsCsub@reductions$pca@cell.embeddings)[, 1:20]
OsCsub <- addDf2seuratMetaData(
    obj = OsCsub, 
    dfAdd = dfTemp
)


## Add UMAP coordinates to Metadata ##
dfAdd <- data.frame(OsCsub@reductions$umap@cell.embeddings)

OsCsub <- addDf2seuratMetaData(
    obj = OsCsub, 
    dfAdd = dfAdd
)

## Add tSNE coordinates to Metadata ##
dfAdd <- data.frame(OsCsub@reductions$tsne@cell.embeddings)

OsCsub <- addDf2seuratMetaData(
    obj = OsCsub, 
    dfAdd = dfAdd
)


Obio@dataTableList[["meta.data"]] <- OsCsub@meta.data

xVec <- c("PC_1","PC_3","PC_5","PC_7","PC_9","PC_11","PC_13","PC_15","PC_17","PC_19")
yVec <- c("PC_2","PC_4","PC_6","PC_8","PC_10","PC_12","PC_14","PC_16","PC_18","PC_20")
pcVec <- c("PC_1","PC_2","PC_3","PC_4","PC_5","PC_6","PC_7","PC_8","PC_9","PC_10")
chnkVec <- as.vector(NULL, mode = "character")
plotListCell <- list()
plotListGene <- list()

###############################################################################
## Collect top-enriched genes                                                ##
EnrichedGenesList <- list()
## Done                                                                      ##
###############################################################################

for (i in 1:length(xVec)){
    dfDat <- Obio@dataTableList$meta.data
    dfSel <- dfDat
    selXY <- c(xVec[i], yVec[i])
    colCol <- Obio@parameterList$singleCellClusterString
    
    tag <- paste0(xVec[i], "and", yVec[i])
    tag <- gsub("_", "", tag)
    
    ## Make Cell level PCA
    plotListCell[[tag]] <- ggplot(data=dfDat, aes_string(selXY[1] , selXY[2], col=colCol, shape="sampleID")
    ) + geom_vline(xintercept = 0, color = "grey", size=0.5
    ) + geom_hline(yintercept = 0, color = "grey", size=0.5
    ) + geom_point()+ ggtitle(paste0("PCA - Cell Level")
    ) +  theme(
        axis.text.y   = element_text(size=8),
        axis.text.x   = element_text(size=8),
        axis.title.y  = element_text(size=8),
        axis.title.x  = element_text(size=8),
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1),
        plot.title = element_text(hjust = 0.5, size = 12)
    )    
    
    ## Save to file ##
    FNbase <- paste0("PCA.cell.level.", xVec[i],".", yVec[i], ".", VersionPdfExt)
        FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    pdf(FN)
        print(plotListCell[[tag]])
    dev.off()
    
    link <- paste0("https://",urlString,"/",Obio@parameterList$project_id,"/pca?x_axis=",gsub("_", "", xVec[i]),"&y_axis=",gsub("_", "", yVec[i]))
    
    figCap <- paste0(
        "**Figure, " ,figureCount,"A:** Cell-level PCA plot for dimensions ", xVec[i], " and ", yVec[i],". ", 
        "Download a pdf of this figure [here](", FNrel,"). ",
        "An interactive version of this figure can be found [here](", link, "). "
    )
    
    
    NewChnk <- paste0(
        "#### PCA Cell Level ", xVec[i], " and ",yVec[i],
        "\n```{r PCAcells_", i,
        ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
        figCap,"'}\n",
        "\n",
        "\n print(plotListCell[['",tag,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
    )
    
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
    
    ## Done with cells                                                       ##
    ###########################################################################
    
    ###########################################################################
    ## Start with genes                                                      ##
    dfPCADat <- data.frame(Loadings(OsCsub, reduction = "pca"))
        dfPCADat[["gene"]] <- row.names(dfPCADat)
    
    dfPCADat <- gather(
        dfPCADat, 
        condition, 
        measurement, 1:(ncol(dfPCADat)-1), 
        factor_key=TRUE
    )

    Obio@dataTableList[["dfPCAloadings"]] <- dfPCADat

    ## Make Gene Level PCA ##
    dfPCADat <- data.frame(Loadings(OsCsub, reduction = "pca"))
    dfPCADat[["gene"]] <- row.names(dfPCADat)
    dfPCADat <- gather(
        dfPCADat,
        condition,
        measurement, 1:(ncol(dfPCADat)-1),
        factor_key=TRUE
    )
    
    dfLoad <- dfPCADat
    Obio@dataTableList$dfPCAloadings <- dfLoad
    ## Plot ##
    selXY <- c(xVec[i], yVec[i])
    dfSel <- filter(dfLoad, condition %in% selXY)
    dfSel <- dfSel %>% spread(key=condition, value=measurement)
    row.names(dfSel) <- dfSel$gene
    dfSel[["highlight"]] <- ""
    dfSel <- dfSel[order(dfSel[,selXY[1]], decreasing = FALSE), ]
    dfSel[1:15, "highlight"] <- "+"
    EnrichedGenesList[[paste0(selXY[1], "_neg")]]<- as.vector(dfSel$gene[1:15])
    
    dfSel <- dfSel[order(dfSel[,selXY[1]], decreasing = TRUE), ]
    dfSel[1:15, "highlight"] <- "+"
    EnrichedGenesList[[paste0(selXY[1], "_pos")]]<- as.vector(dfSel$gene[1:15])
    
    dfSel <- dfSel[order(dfSel[,selXY[2]], decreasing = FALSE), ]
    dfSel[1:15, "highlight"] <- "+"
    EnrichedGenesList[[paste0(selXY[2], "_neg")]]<- as.vector(dfSel$gene[1:15])
    
    dfSel <- dfSel[order(dfSel[,selXY[2]], decreasing = TRUE), ]
    dfSel[1:15, "highlight"] <- "+"
    EnrichedGenesList[[paste0(selXY[2], "_pos")]]<- as.vector(dfSel$gene[1:15])
    
    plotListGene[[tag]] <- ggplot(data=dfSel, aes_string(x=selXY[1],y=selXY[2], col="highlight")
    ) + geom_vline(xintercept = 0, color = "grey", size=0.5
    ) + geom_hline(yintercept = 0, color = "grey", size=0.5) + geom_point() + scale_color_manual(values=c("black", "red")) + ggtitle(paste0("PCA - Cell Level")
    ) +  theme(
        axis.text.y   = element_text(size=8),
        axis.text.x   = element_text(size=8),
        axis.title.y  = element_text(size=8),
        axis.title.x  = element_text(size=8),
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1),
        plot.title = element_text(hjust = 0.5, size = 12)
    )  
    
    points <-  as.vector(unique(dfSel[dfSel$highlight=="+", "gene"]))
    plotListGene[[tag]] <- LabelPoints(plot = plotListGene[[tag]], points =points, repel = TRUE, xnudge = 0, ynudge = 0)
    
    ## Save to file ##
    FNbase <- paste0("PCA.cell.level.", xVec[i],".", yVec[i],".", VersionPdfExt)
        FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    pdf(FN)
        print(plotListGene[[tag]])
    dev.off()
    
    dim1 <- gsub("PC_", "", xVec[i])
    dim2 <- gsub("PC_", "", yVec[i])
    link <- paste0(
        " https://",urlString,"/",
        Obio@parameterList$project_id, 
        "/scatterplot?x_axis=add_counts_PCA_Dim_",
        dim1, 
        "_Loadings&y_axis=add_counts_PCA_Dim_",
        dim2,
        "_Loadings&highlight_gene=&cat_id=ag_lab_categories__10"
    )
    
    figCap <- paste0(
        "**Figure, " ,figureCount,"B:**Gene-level PCA plot for dimensions ", xVec[i], " and ", yVec[i], ". An interactive version of this figure can be found [here](", link, "). "
    )
   
    
    NewChnk <- paste0(
        "\n```{r PCA_gene_level_", i,
        ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
        figCap,"'}\n",
        "\n",
        "\n print(plotListGene[['",tag,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
    )
    
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
    
    ## Done with genes                                                       ##
    ###########################################################################
    figureCount <- figureCount + 1
}

```

```{r plot_PCADimPlots, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
```

### Characterize PCA Dimensions {.tabset .tabset-fade .tabset-pills}
```{r pca_boxplot_dataprep, echo=T, eval=TRUE, warning=FALSE, results=F}
selVec <- c(Obio@parameterList$singleCellClusterString, names(OsCsub@meta.data)[grep("^PC", names(OsCsub@meta.data))])

dfPCdat <- OsCsub@meta.data[, selVec]

ymin <- 1.1*min(dfPCdat$PC_1)
ymax <- 1.1*max(dfPCdat$PC_1)

clusterVec <- sort(unique(OsCsub@meta.data[,Obio@parameterList$singleCellClusterString]))

library(scales)
clusterColVec <- hue_pal()(length(clusterVec))

i=1

chnkVec <- as.vector(NULL, mode="character")
plotList <- list()

for (i in 1:length(clusterVec)){
    dfTemp <- dfPCdat[dfPCdat[,Obio@parameterList$singleCellClusterString] == clusterVec[i],]
    dfTemp[,Obio@parameterList$singleCellClusterString] <- NULL
  
  
  
    library(tidyr)
    dfTemp <- gather(dfTemp, PC)
  
    orderVec <- sort(as.numeric(gsub("PC_", "",unique(dfTemp$PC))))
    orderVec <- paste0("PC_", orderVec)
  
    dfTemp$PC <- factor(dfTemp$PC, levels = orderVec)
    Ncolumns <- length(unique(dfTemp$PC))
  
    a <- paste0("Cluster_", clusterVec[i])

    tag <- paste0("PCA_Distributions_", a)

    plotList[[tag]] <-ggplot(
        dfTemp, 
        aes(x=PC, y=value, fill = PC)
        ) + geom_hline(yintercept = 0, color = "black", size=0.5
        )  + geom_jitter(width=0.1,alpha=0.2
        ) + geom_boxplot(
        ) +    theme(
            legend.position = "none",
            axis.text.y   = element_text(size=8),
            axis.text.x   = element_text(size=8, angle = 45,vjust = 1, hjust=1),
            axis.title.y  = element_text(size=8),
            axis.title.x  = element_text(size=8),
            axis.line = element_line(colour = "black"),
            panel.border = element_rect(colour = "black", fill=NA, size=1),
            plot.title = element_text(hjust = 0.5, size = 12)
        ) + ggtitle(paste0("PCA Distribution: ", a)
        ) + ylim(ymin, ymax) + scale_fill_manual(values=rep(clusterColVec[i], Ncolumns))
    
    ## Save to file ##
    FNbase <- paste0(tag, VersionPdfExt)
        FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
    FNrel <- paste0("report_figures/", FNbase)
    
    pdf(FN)
        print(plotList[[tag]])
    dev.off()
    
    
    figCap <- paste0(
        "**Figure, " ,
        figureCount,
        ":** This plot may help you to identify PCA dimensions, in which marker genes for cluster ",
        clusterVec[i],
        " become evident. Download a pdf of this figure [here](", FNrel,"). "
    )
    
    figureCount <- figureCount + 1
    
    NewChnk <- paste0(
        "#### ", tag,
        "\n```{r ", tag,
        ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
        figCap,"'}\n",
        "\n",
        "\n print(plotList[['",tag,"']])",
        "\n cat(  '\n')",
        "\n\n\n```\n"   
    )
    
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
}

```


```{r pca_boxplot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))

```

### Characterize PCA Dimensions {.tabset .tabset-fade .tabset-pills}
```{r create-pca-enrichment-data, echo=T, eval=TRUE, warning=FALSE, results=F}
library(knitr)
library(ggplot2)

#save.image("temp.RData")


PCAdimensions <- paste0("PC_", 1:20)

plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

for (j in 1:length(PCAdimensions)){
    posTestGeneSet <- as.vector(
        unique(
            EnrichedGenesList[[paste0(PCAdimensions[j], "_pos")]]
        )
    )
    
    
    negTestGeneSet <- as.vector(
        unique(
            EnrichedGenesList[[paste0(PCAdimensions[j], "_neg")]]
        )
    )
    
    ## Get background gene set ##
    #backgroundGeneVec <- row.names(OsCsub[["RNA"]]@counts)
    if ((length(posTestGeneSet) >= 3) |(length(negTestGeneSet) >= 3)){
        library(enrichR)
        topMaxCat <- 10
        dbs <- listEnrichrDbs()
        
        dbs <- c("GO_Biological_Process_2017")
        
        
        PosEnriched <- enrichr(posTestGeneSet, dbs)
        
        for (i in 1:length(dbs)){
            dfTemp <- PosEnriched[[dbs[i]]]
            
            if (i ==1){
                dfPosEnriched <- dfTemp
            } else {
                dfPosEnriched <- rbind(
                    dfPosEnriched,
                    dfTemp
                )
            }
            
        }
        
        dfPosEnriched[["log10FDR"]] <- -1*log10(dfPosEnriched$Adjusted.P.value)
        dfPosEnriched <- dfPosEnriched[order(-dfPosEnriched$log10FDR),]
        dfPosEnriched <- na.omit(dfPosEnriched)
        
        ## Negative Side ##
        NegEnriched <- enrichr(negTestGeneSet, dbs)
        
        for (i in 1:length(dbs)){
            dfTemp <- NegEnriched[[dbs[i]]]
            
            if (i ==1){
                dfNegEnriched <- dfTemp
            } else {
                dfNegEnriched <- rbind(
                    dfNegEnriched,
                    dfTemp
                )
            }
            
        }
        
        
        dfNegEnriched[["log10FDR"]] <- -1*log10(dfNegEnriched$Adjusted.P.value)
        dfNegEnriched <- dfNegEnriched[order(-dfPosEnriched$log10FDR),]
        dfNegEnriched <- na.omit(dfNegEnriched)
        
        dfNegSel <- dfNegEnriched
        if (nrow(dfNegSel) > topMaxCat){
            dfNegSel <- dfNegSel[1:topMaxCat,]
        }
        
        dfPosSel <- dfPosEnriched
        if (nrow(dfPosSel) > topMaxCat){
            dfPosSel <- dfPosSel[1:topMaxCat,]
        }
        
        if ((nrow(dfNegEnriched) > 0) | (nrow(dfPosEnriched) > 0)){
            
            
            dfNegSel$log10FDR <- -1* dfNegSel$log10FDR
            
            dfSel <- rbind(
                dfNegSel,
                dfPosSel
            )
            
            dfSel <- na.omit(dfSel)
            dfSel <- dfSel[order(dfSel$log10FDR),]
            dfSel$log10FDR <- round(dfSel$log10FDR, 2)
            
            dfSel[["Category"]] <- ""
            dfSel[dfSel$log10FDR >= 0, "Category"] <- "Enr."
            dfSel[dfSel$log10FDR < 0, "Category"] <- "Depl."
            
            for (k in 1:nrow(dfSel)){
                if (nchar(dfSel[k, "Term"]) > 50 & length(grep("\\(GO", as.vector(dfSel[k, "Term"]))) > 0){
                    part1 <- unlist(strsplit(as.vector(dfSel[k, "Term"]), "\\(GO"))[1]
                    part1 <- substr(part1, 1, 45)
                    part2 <- unlist(strsplit(as.vector(dfSel[k, "Term"]), "\\(GO"))[2]
                    part2 <- paste0("\\(GO", part2)
                    
                    if (nchar(part1) > 40 ){
                        dfSel[k, "Term"] <- paste0(part1, " \\n", part2)
                    } else { 
                        dfSel[k, "Term"] <- paste0(part1, " ", part2)
                    }
                }
            }
            
            
            #dfSel$Term <- gsub("\\(GO", "\\\n\\(GO", dfSel$Term)
            
            dfSel$Term <- factor(dfSel$Term, levels = unique(dfSel$Term))
            
            plotList[[paste0("PCA_ENR_", j)]] <- ggplot(
                data=dfSel, aes(x= Term, y=log10FDR, fill=Category, order=log10FDR)
            ) + geom_bar(stat="identity", colour="black"
            ) + coord_flip() +scale_fill_manual(values=c("yellow", "blue"))  +  theme(
                axis.text.y   = element_text(size=8),
                axis.text.x   = element_text(size=8),
                axis.title.y  = element_text(size=8),
                axis.title.x  = element_text(size=8),
                axis.line = element_line(colour = "black"),
                panel.border = element_rect(colour = "black", fill=NA, size=1),
                plot.title = element_text(hjust = 0.5, size = 12)
            )  + labs(title = paste0("Cluster ", PCAdimensions[j]," enriched genes") ,y = "-log10(FDR)", x = ""
            ) + geom_hline(yintercept = c(-log10(0.05), log10(0.05)), color = "red", size=0.5, lty=2
            ) + geom_hline(yintercept = 0, color = "black", size=0.5
            )
            cat("  \n")
            
            
            
            ## Save to file ##
            FNbase <- paste0("PCA_Cluster_", PCAdimensions[j],".enriched.genes", VersionPdfExt)
            FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
            FNrel <- paste0("report_figures/", FNbase)
            
           
            pdf(FN)
            print(plotList[[paste0("PCA_ENR_", j)]])
            dev.off()
            
            link <- paste0(
                "https://", urlString, "/",
                Obio@parameterList$project_id,
                "/category-view?category_type=GO-BP"
            )
            
            ## Create R markdown chunk ##
            figLegend <- paste0(
                "**Figure ", 
                figureCount, 
                "**: GO-BP category enrichment analysis for the 15 genes that have  <font color = \\'yellow\\'> the most positive </font> and <font color = \\'blue\\'>the most negative</font> PCA loading values in dimension ", 
               PCAdimensions[j],
                " associated with them. Download a pdf of this figure [here](", FNrel, "). To view these gene sets in the context of your data, go to [CategoryView > GO-BP](",link,") and find these categories using the search box."
            )
            figureCount <- figureCount + 1 
            
            NewChnk <- paste0(
                "#### ", PCAdimensions[j],
                "\n```{r enrichr_",
                j,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                figLegend,"'}\n",
                "\n",
                "\n print(plotList[['",paste0("PCA_ENR_", j),"']])",
                "\n cat(  '\n')",
                "\n\n\n```\n"   
            )
        }
    }
    chnkVec <- c(
        chnkVec,
        NewChnk
    )
}

```

```{r create-pca-enrichment-plot, echo=T, eval=TRUE, warning=FALSE, results='asis'}
###############################################################################
## Do category enrichment on clusters                                        ##
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
## Done doing enrichment on clusters                                         ##
###############################################################################


```

### Pseudotime {.tabset .tabset-fade .tabset-pills}
```{r pseudotime-enrichment-data, echo=T, eval=TRUE, warning=FALSE, results=F}

## Add default ##
pos <- grep("addDmaps", names(Obio@parameterList))
if (length(pos) == 0){
    Obio@parameterList[["addDmaps"]] <- FALSE
}

if (Obio@parameterList$addDmaps){
    library(knitr)
    library(ggplot2)
    invertPT <- TRUE
    
    
    plotList <- list()
    chnkVec <- as.vector(NULL, mode = "character")
    
    
    
    library(Seurat)
    library(destiny)
    
    dfPCA <- OsCsub@meta.data
    dfPCA <- dfPCA[,grep("PC_", names(dfPCA))]
    
    dmPCA <- DiffusionMap(dfPCA)
    
    #dpt <- DPT(dm, tips = 268)
    #dpt <- DPT(dm)
    dpt <- DPT(dmPCA)
    #pseudotime <- dpt$dpt
    
    # Plot DC1 vs DC2 and color the cells by their inferred diffusion pseudotime.
    # We can accesss diffusion pseudotime via dpt$dpt.
    df <- data.frame(
      DC1 = eigenvectors(dmPCA)[, 1], 
      DC2 = eigenvectors(dmPCA)[, 2], 
      DC3 = eigenvectors(dmPCA)[, 3], 
      "DM_Pseudotime" = dpt$dpt
    )
    
    df$cellID <- row.names(dfPCA)
    
    ## For this project reverse pseudotime ##
    if (invertPT){
        PTmax <- max(df$DM_Pseudotime)
        df$DM_Pseudotime <- -1* (df$DM_Pseudotime - PTmax)
        ## Invert DC1, 2, 3
        df$DC1 <- -1* df$DC1
        df$DC2 <- -1* df$DC2
        df$DC3 <- -1* df$DC3
    }
    
    ## Add to table ##
    df$cellID <- row.names(dfPCA)
    dfdbTable <- OsCsub@meta.data
    
    dfdbTable$row_names <- NULL
    dfdbTable$DC1 <- NULL
    dfdbTable$DC3 <- NULL
    dfdbTable$DC2 <- NULL
    dfdbTable$DM_Pseudotime <- NULL
    dfdbTable <- unique(dfdbTable)
    
    ###############################################################################
    ## Add pseudotime components                                                 ##
    #dim(dfdbTable)
    
    dfdbTable[["cellID"]] <- row.names(dfdbTable)
    
    dfdbTable <- merge(
      dfdbTable, 
      df, 
      by.x = "cellID",
      by.y = "cellID"
    )
    
    #dim(dfdbTable)
    
    ## Add to Seurat object ##
    OsCsub@meta.data <- dfdbTable
    
    ## Create Pseudotime plot ##
    dfTemp <- dfdbTable
    dotsize <- 0.5
    dotcolor <- "darkblue"
    tag <- "PC1PC2all"
    
    plotList[[tag]] <- ggplot(dfTemp, aes(PC_1, PC_2, color=DM_Pseudotime)
        )+ geom_point( 
            shape = 16,
            size = as.numeric(dotsize)
        ) + xlab("PC1") + ylab("PC2") + scale_color_gradient2(
            low="#ff6600", 
            high=dotcolor #, 
            #limits=c(0,maxExpr)
        ) +  theme(
            axis.text.y   = element_text(size=8),
            axis.text.x   = element_text(size=8),
            axis.title.y  = element_text(size=8),
            axis.title.x  = element_text(size=8),
            axis.line = element_line(colour = "black"),
            panel.border = element_rect(colour = "black", fill=NA, size=1),
            plot.title = element_text(hjust = 0.5, size = 12),
            panel.background = element_rect(fill = "lightgrey")
        ) + ggtitle("PC1, PC2 and DM Pseudotime"
        ) #+ xlim(minX, maxX) + ylim(minY, maxY)  
    
    
    
            ## Save to file ##
            FNbase <- paste0("Pseudotime_overview", VersionPdfExt)
            FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
            FNrel <- paste0("report_figures/", FNbase)
            
           
            pdf(FN)
            print(plotList[[tag]])
            dev.off()
            
            
            
            ## Create R markdown chunk ##
            figLegend <- paste0(
                "**Figure ", 
                figureCount, 
                "**: Figure depicting PCA components 1 and 2 with the diffusion map pseudotime highlighted in color. Download a pdf of this figure [here](", FNrel, "). "
            )
            figureCount <- figureCount + 1 
            
            NewChnk <- paste0(
                "\n#### Pseudotime All Timepoints", 
                "\n```{r ", tag, ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                figLegend,"'}\n",
                "\n",
                "\n print(plotList[['",tag,"']])",
                "\n cat(  '\n')",
                "\n\n\n```\n"   
            )
            
            chnkVec <- c(
                chnkVec,
                NewChnk
            )
} # end dmap




```

```{r pseudotime-plot, echo=T, eval=TRUE, warning=FALSE, results='asis'}
###############################################################################
## Do category enrichment on clusters                                        ##
if (Obio@parameterList$addDmaps){
    cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
}
## Done doing enrichment on clusters                                         ##
###############################################################################


```


```{r Additional tables, echo=F, eval=TRUE, warning=FALSE, results=F} 
dfCounts <- data.frame(OsCsub[["RNA"]]@counts)
dfCounts[["gene"]] <- row.names(dfCounts)
dfCounts <- gather(
    dfCounts, 
    condition, 
    counts, 1:(ncol(dfCounts)-1), 
    factor_key=TRUE
)
dfCounts <- dfCounts[dfCounts$counts != 0,]
Obio@dataTableList[["dfCounts"]] <- dfCounts

# slotNames(OsCsub[["RNA"]])
dfExpr <- data.frame(OsCsub[["RNA"]]@data)
dfExpr[["gene"]] <- row.names(dfExpr)
dfExpr <- gather(
    dfExpr, 
    condition, 
    expr, 1:(ncol(dfExpr)-1), 
    factor_key=TRUE
)
dfExpr <- dfExpr[dfExpr$expr != 0,]
Obio@dataTableList[["dfExpr"]] <- dfExpr
```

```{r saveobject, eval=TRUE, echo=T, results=F}
### Will save Obio object here, so it can be re-used with different parameters
# save(Obio, 
#      file = paste0(
#          Obio@parameterList$localWorkDir,
#          Obio@parameterList$project_id,
#          ".bioLOGIC.Robj"
#      )
# )

# print("Obio Object saved.")
# 
# save(OsCsub,
#     file = paste0(
#          Obio@parameterList$localWorkDir,
#          Obio@parameterList$project_id,
#         ".Seurat.Robj"
#      )
# )

```

## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
sessionInfo()
```                                            