---
author: "Stefan Boeing"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

output: 
    html_document:
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 3
        toc_float: true
    css:
    
always_allow_html: yes
---
    
    
```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```


## Part B Database

module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;R;

```{r populate_meta_data_database, eval=TRUE, echo=F, results=F}
#Create the environment and load a suitable version of R, e.g. so:

myPaths <- .libPaths()
myNewPaths <- c("/camp/stp/babs/working/boeings/Projects/pachnisv/song.chng/330_10xscRNAseq_DIV4_DIV11_DIV20_SC19069/basedata", myPaths)
.libPaths(myNewPaths)
library(glmGamPoi)
.libPaths(myPaths)

library(Seurat)

VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}


FN <- paste0(hpc.mount, "Projects/reference_data/documentation/BC.parameters.txt")
dbTable <- read.delim(
    FN, 
    sep = "\t",
    stringsAsFactors = F
)

db.pwd <- as.vector(dbTable[1,1])


source("assets/SBwebtools.pckg.r")



if (length(.libPaths()) > 2){
    .libPaths(.libPaths()[2:3])
}
## Create biologic Object for visualization ##


ObioFN <- paste0("../",list.files("..")[grep(".bioLOGIC.Robj", list.files(".."))])

load(ObioFN)


checkFile = paste0(
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
)


Obio <- setMountingPoint(Obio)
Obio <- setAnalysisPaths(Obio)
Obio <- setCrickGenomeAndGeneNameTable(Obio)
Obio <- createAnalysisFolders(
    Obio,
    baseDir="/camp/stp/babs/working/boeings/Projects/",
    localBaseDir = paste0(hpc.mount, "Projects/")
)
Obio <- setDataBaseParameters(Obio)

## Upload metadata table > p315_PCA
# Obio@parameterList$host <- "10.152.22.193"
# Obio@parameterList$db.user <- "boeingS"
# db.pwd <- "5+3f4nB04042018"
```


## GSEA

module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/3.6.0-foss-2016b-BABS;R;

```{r doGSEA, eval=TRUE, echo=F, results=F}
dfGSEA <- Obio@dataTableList[["DGE_table"]]

###############################################################################
# Do GSEA                                                                     #
###############################################################################

database.table2 <- Obio@dataTableList[["DGE_table"]]

## Add annotation ##
dfAnno <- unique(Obio@databaseTable[,c("mgi_symbol","hgnc_symbol")])
dfAnno <- dfAnno[dfAnno$mgi_symbol %in% database.table2$gene, ]

database.table2 <- merge(
  dfAnno, 
  database.table2, 
  by.x = "hgnc_symbol",
  by.y = "gene",
  all = TRUE
)
database.table2[is.na(database.table2)] <- 0
rmVec <- c(
     #grep("contrast_2", names(database.table2))
     grep("contrast_P", names(database.table2)),
     grep("contrast_L", names(database.table2)),
     grep("LRT_", names(database.table2)),
     grep("PCA_", names(database.table2)),
     grep("norm_counts_", names(database.table2)),
     grep("intercept_", names(database.table2)),
     grep("r2_P", names(database.table2)),
     grep("DEseq2NormalizedReadCounts", names(database.table2)),
     grep("p_value_P", names(database.table2)),
     grep("lg2_avg_", names(database.table2))
)

if (length(rmVec) > 0){
     database.table2 <- database.table2[,-rmVec]
}
# #
# ## Remove unnecessary columns, if needed ##
# #
# ## Create GSEA rank files ##
create.gsea.rnk.files(
     Obio@parameterList$localWorkDir,
     df.dataTable = database.table2,
     GSEA.colum.type = "_logFC_",
     gene.symbol.column.name = "hgnc_symbol"
 )
# #
# # ## Remove last character from file ##
# # #truncate -s -2 file
# # #sed '$d' file # remove last line
# #
# # ## Remove last character from file ##
# # #truncate -s -2 file
# # #sed '$d' file # remove last line
# #
# ## Function to create gmt file ##
tables <- c(
     "mysigdb_h_hallmarks" ,
     "mysigdb_c5_BP" ,
     "mysigdb_sc_sig",
     "mysigdb_c5_MF",
     Obio@parameterList$lab.categories.table
 )
# #
dfRefGmt <- create.gmt.file.from.ref.data.table(
     host = Obio@parameterList$host,
     dbname = "reference_categories_db_new",
     dataTable = tables,
     pwd = db.pwd,
     user=Obio@parameterList$db.user,
     gene.id.column = "hgnc_symbol"
 )
# #
# # ###############################################################################
# # ## Save gmt file                                                             ##
# # #"/camp/stp/babs/working/boeings/Projects/reference_data/GSEA.gmt.files/20160508.rna.seq.txn.analysis.gmt"
# #
localGmtDir <- paste0(
    Obio@parameterList$localWorkDir,
    "GSEA/"
)
#
gmtDir<- paste0(
    Obio@parameterList$workdir,
    "GSEA/"
)

gmtFileName <- paste0(
    Obio@parameterList$project_id,
    ".",
    "projectGmtFile.gmt"
)

dfRefGmt <- dfRefGmt[!(duplicated(dfRefGmt[,1])),]

write.table(
    dfRefGmt,
    paste0(localGmtDir, gmtFileName),
    col.names = FALSE,
    row.names = FALSE,
    sep="\t",
    quote = F
)
# #
contrasts <- names(database.table2)[grep("logFC",names(database.table2))]
contrasts <- contrasts[contrasts != "logFC_cut_off"]
contrasts

GSEAfn <- paste0(
    Obio@parameterList$localWorkDir,
    "/GSEA/GSEAcommands.sh"
)
sink(GSEAfn)

cat("module load Java/1.8.0_131");cat("\n");cat("\n")
for (i in 1:length(contrasts)){
    gmtFile <- paste0(gmtDir, gmtFileName)
    contrastNo <- unlist(strsplit(contrasts[i], "_"))[2]
    nTopPlots <- 50
    GSEAdir <- paste0(Obio@parameterList$workdir, "GSEA")
    rnkFile <- paste0(GSEAdir, "/",contrasts[i],".rnk")

    gseaCMD <- paste0(
        "java -Xmx2512m -cp /camp/stp/babs/working/boeings/Projects/software/gsea-3.0.jar xtools.gsea.GseaPreranked -gmx ",
        gmtFile,
        " -rnk ",
        rnkFile,
        " -rpt_label ",
        "contrast_",
        contrastNo,
        "_rnaSeqTxnTest",
        " -out ",
        GSEAdir,
        " -collapse false -mode Max_probe -norm meandiv -nperm 1000 -scoring_scheme classic -include_only_symbols true -make_sets true -plot_top_x ",
        nTopPlots,
        " -rnd_seed timestamp -set_max 2500 -set_min 10 -zip_report false -gui false"
    )
    cat(gseaCMD);cat("\n");cat("\n");


}
sink()

```

```{r doGSEA_post_processing, eval=TRUE, echo=F, results=F}
# ###############################################################################
# ## Process GSEA outputs                                                      ##
#
# # # Subfunction to select relevant tables
library(RMySQL)
dbDB = dbConnect(
    drv = RMySQL::MySQL(),
    user = Obio@parameterList$db.user,
    password = db.pwd,
    dbname = "reference_categories_db_new",
    host = Obio@parameterList$host
)

#This will move to the header
availableTables = dbGetQuery(dbDB, "SHOW TABLES in reference_categories_db_new")
availableTables <- as.vector(availableTables$Tables_in_reference_categories_db_new)
dbDisconnect(dbDB)
#
# ## Check if all reference tables are present in the db ##
print(
    paste0(
        sum(!(tables %in% availableTables)),
        " tables NOT present in the reference database."
    )
)
#
#
#
# #temporary change
# #tables <- tables[c(1,2,3,5,6)]
# #tables <- append(tables, "vt_lab_categories")
# # End of parts to be moved to the header
#
gsea.cat.lines <- create.GSEA.table(
    GSEADir = paste0(Obio@parameterList$localWorkDir, "GSEA"),
    logFC.column.name = "logFC",
    host = Obio@parameterList$host,
    refdbname= "reference_categories_db_new",
    refDBTableName = Obio@parameterList$enriched.categories.dbTableName,
    db.user = Obio@parameterList$db.user,
    db.password = db.pwd,
    tables = tables,
    df.dataTable = database.table2,
    outputDir = Obio@parameterList$outputDir
)

## Done processing GSEA outputs                                              ##
###############################################################################
```


## Gene Set Enrichment Analysis (GSEA) {.tabset .tabset-fade .tabset-pills}

Find furter background information on the Gene Set Enrichment Analysis (GSEA) and the interpretation of results can be found [here](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html?_Interpreting_GSEA_Results).

```{r GSEA_section, echo=TRUE, eval=TRUE, warning=FALSE, results=F}
chnkVec <- as.vector(NULL, mode = "character")
plotList <- list()


print(
  paste0(
    "The following gene sets have been used in the GSEA analysis: ",   
    sort(paste0(tables, collapse = ", ")), 
    "."
  )
)

###############################################################################
## Retrieve GSEA Table                                                       ##
gseaDbTable <- paste0(
  Obio@parameterList$project_id,
  "_enriched_categories_table"
)

dfGSEA <- import.db.table.from.db(
    host = Obio@parameterList$host, 
    user = Obio@parameterList$db.user,
    password = db.pwd,
    dbname =  "enriched_categories",
    dbtable =gseaDbTable
)

dfGdat <- dfGSEA
dfGdat[["url"]] <- paste0("https://biologic.crick.ac.uk/",Obio@parameterList$project_id,"/category-view/", dfGdat$cat_id)
dfGdat[["Category"]] <- paste0("<a href='",dfGdat$url,"' >", gsub("_", " ", dfGdat$cat_name), "</a>")

NESvec <- names(dfGdat)[grep("NES_", names(dfGdat))]
padjVec <- names(dfGdat)[grep("padj_", names(dfGdat))]
imgVec <- names(dfGdat)[grep("image_link", names(dfGdat))]
dfImg <- data.frame(NESvec, imgVec)
plotLinkCol <- imgVec
plotLinkCol <- gsub("contrast_", "P", plotLinkCol)
plotLinkCol <- gsub("_image_link", "", plotLinkCol)

repString <- paste0(
    "https://biologic.crick.ac.uk/mdata/",
    Obio@parameterList$project_id,
    "/enrichment_plots"
)

selVec <- as.vector(NULL, mode = "character")
    

for (i in 1:length(plotLinkCol)){
    names(dfGdat) <- gsub(imgVec[i], plotLinkCol[i], names(dfGdat))
    dfGdat[,plotLinkCol[i]] <- gsub(
        "enrichment_plots", 
        repString, 
        dfGdat[,plotLinkCol[i]] 
    )
    dfGdat[dfGdat[,plotLinkCol[i]] != "",plotLinkCol[i]] <- paste0(
        "<a href='", 
        dfGdat[dfGdat[,plotLinkCol[i]] != "",plotLinkCol[i]],
        "'>Plot</a>"
    )
    
    
    selVec <- c(
        selVec,
        NESvec[i],
        plotLinkCol[i],
        padjVec[i]
        
    )
}



colSel <- c(
  "Category",
  "enrichment_type",
  selVec
)

## Done GSEA table                                                           ##
###############################################################################

###############################################################################
## Plot top-scoring categories                                               ##

## Select top 10 categories from each ##

dfTemp <- dfGdat[,colSel]
dfGdatS <- dfGdat
dfGdat <- dfGdat[,colSel]
## Write table as Excel into outputs ##
## Create Excel file ##

###############################################################################
## Creating Excel output table                                               ##
for (i in 1:length(plotLinkCol)){
    names(dfGdat) <- gsub(imgVec[i], plotLinkCol[i], names(dfGdat))
    dfGdat[,plotLinkCol[i]] <- gsub(
        "<a href='", 
        "", 
        dfGdat[,plotLinkCol[i]] 
    )
    names(dfGdat) <- gsub(imgVec[i], plotLinkCol[i], names(dfGdat))
    dfGdat[,plotLinkCol[i]] <- gsub(
        "' download = 'GSEA.pdf'>Plot</a>", 
        "", 
        dfGdat[,plotLinkCol[i]] 
    )
    
}
    
    library(openxlsx)
    fullOutFN <- paste0(Obio@parameterList$outputDir, Obio@parameterList$project_id, "_GSEA.xlsx")
    outFN <- paste0(Obio@parameterList$project_id, "_GSEA.xlsx")
    wb <- createWorkbook()
    addWorksheet(wb, paste0(Obio@parameterList$project_id, "_GSEA"))
    freezePane(wb, paste0(Obio@parameterList$project_id, "_GSEA") ,  firstActiveRow = 2)
    
    ## Filter is inactivated, as it does not appear to be compatible with the current version of Excel
    #addFilter(wb, 1, row = 1, cols = 1:ncol(dfOutput))
    
    ## Style headers ##
    hs1 <- createStyle(
        fontColour = "#ffffff",
        fgFill = "#000000", 
        halign = "CENTER", 
        textDecoration = "Bold"
    )
    
    writeData(wb, 1, dfGdat, startRow = 1, startCol = 1, headerStyle = hs1)
    
    saveWorkbook(
        wb, 
        outFN,
        overwrite = TRUE
    )
    
    ## Done creating Excel output table                                          ##
    ###############################################################################

# library(tidyr)
# row.names(dfTemp) <- NULL
# dfTemp <- unique(dfTemp)
# dfTemp <- gather(dfTemp, key = "DGE_comp", value = c("NES", "padj"), grep("NES", names(dfTemp)))
# dfTemp <- gather(dfTemp, key = "DGE_comp", value = "padj", grep("padj", names(dfTemp)))
# row.names(dfTemp) <- NULL
# dfTemp <- unique(dfTemp)
# 
# dfTemp <- dfTemp[order(-dfTemp$padj, abs(dfTemp$NES), decreasing = T),]
# 
# dfTemp <- dfTemp[1:20,]
# 
# gather(Quarter, Revenue, Qtr.1:Qtr.4)
# 
# ggplot(dfTemp, aes(x = DGE_comp, y = fct_reorder(Description, GeneRatio))) + 
#                geom_point(aes(size = GeneRatio, color = p.adjust)) +
#                theme_bw(base_size = 14) +
#         scale_colour_gradient(limits=c(0, 0.10), low="red") +
#         ylab(NULL) +
#         ggtitle("GO pathway enrichment")

## Done plotting top scoring categories                                      ##
###############################################################################
tableDNlink <- paste0(
    "https://biologic.crick.ac.uk/mdata/",
    Obio@parameterList$project_id,
    "/html/report_tables/",
    outFN
)
    
tableCap <- paste0(
    "**GSEA Result Table:** Find the GSEA normalized enrichment score (NES) and the enrichment p-value in the above table. Plot entries mean that for this category and comparison a GSEA plot is readily available for download. Download the full GSEA result table as Excel file [here](", tableDNlink, ")"
)

dfGdat <- dfGdatS

```

```{r GSEA_table, echo=TRUE, eval=TRUE, warning=FALSE, results=T, fig.cap=tableCap}
chnkVec <- as.vector(NULL, mode = "character")

dfDataTable <- dfGdat[,colSel]

library(DT)
datatable(
    dfDataTable,
    colnames = gsub("_", " ", names(dfDataTable)),
    rownames = FALSE,
    escape = FALSE,
    options = list(
        initComplete = JS(
            "function(settings, json) {",
            "$(this.api().table().header()).css({'background-color': '#000', 'color': '#fff'});",
            "}"
        ),
    order = list(list(3, 'asc'), list(2, 'desc'))
    )
) 




```

## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
sessionInfo()
```

---
title: '`r paste0("GSEA Analysis for Project ",  Obio@parameterList$project_id)`'
---