---

output: 
    html_document:
        highlight: default
        theme: paper
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 3
        toc_float: true
        self_contained: no
        css: src/style/style.css

---
    
    

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```


```{r, include=FALSE}


## Recommended (Crick) R-setup: 
# module purge;source /nemo/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a

```

```{r, eval=TRUE, echo=F, results=F}
###############################################################################
##                                                                           ##
if (!requireNamespace("remotes")){
  install.packages("remotes")
}


if (!requireNamespace("renv")){
  remotes::install_github("rstudio/renv")
}

if (!file.exists("renv.lock")){
    renv::init()
} else {
    renv::restore(prompt = FALSE)
}

## Done                                                                      ##
###############################################################################


```

```{r, eval=TRUE, echo=T, results=F}
###############################################################################
##  Load packages                                                            ##
library(Seurat)
library(biologicSeqTools2)
library(dplyr)

VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")
rasterDpi <- 400
calculateLRTfromScratch <- FALSE

## Load bioLOGIC object
source("load.biologic.robj.R")

## Make sure paths are set correctly in the Obio object
Obio <- Obio %>% 
    biologicSeqTools2::setMountingPoint()  %>% 
    biologicSeqTools2::setAnalysisPaths() %>% 
    biologicSeqTools2::setCrickGenomeAndGeneNameTable() %>% 
    biologicSeqTools2::createAnalysisFolders() %>% 
    biologicSeqTools2::setDataBaseParameters()

## Load Seurat object ##
SeuratFN <- paste0(
    Obio@parameterList$localWorkDir,
    Obio@parameterList$project_id,
    ".Seurat.Robj"
)

load(SeuratFN)

##                                                                           ##
###############################################################################

###############################################################################
## Retrieve database password                                                ##
if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")
    dbTable <- read.delim(
      FN,
      header = F,
      sep = "\t",
      stringsAsFactors = F
    )
#}
db.pwd <- as.vector(dbTable[1,1])

## Done                                                                      ##
###############################################################################

###############################################################################
##  Set variables                                                            ##

# Plotting parameters
figureCount <- 1
legendDotSize <- 5

## Define project ID ##
project_id <- Obio@parameterList$project_id


## Set directory for report tables
reportTableDir <- Obio@parameterList$reportTableDir
reportFigDir <- Obio@parameterList$reportFigDir

localWorkDir <- Obio@parameterList$localWorkDir

## Create url string
if (Obio@parameterList$host == "10.27.241.234"){
    urlString <- "biologic.thecrick.org"
} else {
    urlString <- "biologic.crick.ac.uk"
}

shinyBaseUrl <- "https://bioinformatics.crick.ac.uk/shiny/users/boeings/"

shinyURL <- paste0(
    shinyBaseUrl,
    project_id,
    "_app/"
)            
   
## Set file paths ##
baseFN <- paste0(
   project_id, 
   ".LRT.table.xlsx"
)

FNrel <- paste0("report_tables/", baseFN)

outPutFN <- paste0(
     reportTableDir,
     baseFN
)

tableLink <- paste0(
    'https://',
    urlString,
    '/mdata/',project_id, '/html/', 
    FNrel
)

tableString <- paste0(
    'An Excel table with the LRT results can be downloaded ',
    tableLink
)

tableString <- paste0(
   ' An Excel table with the LRT results can be downloaded <a href="',tableLink,'" target="_blank">here</a>. '
)
  

##  Set variables                                                            ##
###############################################################################



###############################################################################
## Function doLRTsc                                                          ##

#' @title doLRTsc
#' @description Function to do glmgampoi differential gene expression. When using this function, make sure the glmGamPoi R-package is installed on your system: renv::install("renv::install("bioc::glmGamPoi") 
#' @param obj Seurat object for the LRT analysis
#' @param countSlot Seurat object assay to use to retrieve count data.
#' @param LRTselCol Meta data column in the above Seurat object to use for LRT sample group definitions. 
#' @param colName Name for the LRT comparison in the meta data
#' @param contrastTag contrastTag
#' @param LRTsampleList List specifying the LRT comparisons. 
#' @import Seurat
#' @export


setGeneric(
    name="doLRTsc",
    def=function(
        obj,
        countSlot = "RNA",
        LRTselCol = "seurat_clusters",
        colName = "LRT_test",
        contrastTag = "contrast_L1_test",
        LRTsampleList = list(
            "LRT_name" = c(17)
        )
    ) {
    
    ############################################################################
    ## Add LRT cell selection to meta-data                                    ##
    colName <- unlist(colName)
    LRTselCol <- unlist(LRTselCol)

    pos <- grep(colName, names(obj@meta.data))
    if (length(pos) > 0){
        obj@meta.data <- obj@meta.data[,-pos]
    }

    mColName <- substr(paste0("meta_", colName), 1, 63)
    obj@meta.data[[mColName]] <- "Rest" 
    
    if (names(LRTsampleList) == "seurat_clusters"){
      obj@meta.data$seurat_clusters <- as.numeric(obj@meta.data$seurat_clusters )
    }
    
    for (i in 1:length(LRTsampleList)){
        obj@meta.data[obj@meta.data[,names(LRTsampleList)] %in% LRTsampleList[[i]], mColName] <- obj@meta.data[obj@meta.data[,names(LRTsampleList)] %in% LRTsampleList[[i]], LRTselCol]
    }
    
    dfMeta <- obj@meta.data
    dfMeta <- dfMeta[dfMeta[,mColName] != "Rest",]
    #dfMeta[,mColName] <- dfMeta[,LRTselCol]
    
    ##                                                                        ##
    ############################################################################
    
    ############################################################################
    ## Get counts for LRT                                                     ##
    cellVec <- dfMeta$cellID
    group <-  as.vector(dfMeta[,mColName])
    
    dfMatrix <- obj[[countSlot]]@counts
    dfMatrix <- data.matrix(dfMatrix[,cellVec])
    dfMatrix <- dfMatrix[rowSums(dfMatrix) != 0, ]
    
    ## Pseudobulk LRT
    
    ## LRT ##
     fit <- glmGamPoi::glm_gp(dfMatrix, design = group)
     res <- glmGamPoi::test_de(fit, reduced_design = ~ 1)
    
    # sample_labels <- group
    # 
    # sample_labels[sample_labels == names(LRTsampleList)[1]] <- paste0(sample_labels[sample_labels == names(LRTsampleList[1])], "_", 1:length(sample_labels[sample_labels == names(LRTsampleList)[1]]))
    # 
    # sample_labels[sample_labels == names(LRTsampleList)[2]] <- paste0(sample_labels[sample_labels == names(LRTsampleList[2])], "_", 1:length(sample_labels[sample_labels == names(LRTsampleList)[2]]))
    # 
    # cell_type_lables <- group
    # unique(group)
    # 
    # fit <- glmGamPoi::glm_gp(
    #     dfMatrix, 
    #     design = group
    # )
    
    #comparison <- names(LRTsampleList)
    
    # contrastString <- names(LRTsampleList)[1]
    # 
    # res <- glmGamPoi::test_de(fit, contrast = contrastString,
    #         pseudobulk_by = sample_labels, 
    #         #subset_to = cell_type_labels == "T-cells",
    #         #n_max = 4, 
    #         sort_by = pval, 
    #         decreasing = FALSE
    # )
    
    dfRes <- res[order(res$adj_pval, decreasing = F),]
    dfRes[is.na(dfRes)] <- 0
    
    ## Remove extreme outliers ##
    dfRes <- dfRes[dfRes$lfc > -100 & dfRes$lfc < 100, ]
    dfRes[["padj"]] <- dfRes$adj_pval
    minP <- min(dfRes$padj[dfRes$padj != 0])
    dfRes[dfRes$padj == 0, "padj"] <- minP
    dfRes[["lg10p"]] <- -1 * log10(dfRes$padj)
    
    
    comp_1 <- dfRes
    comp_1[["gene"]] <- dfRes$name
    names(comp_1) <- gsub("lfc", paste0(contrastTag, "logFC_" ,colName), names(comp_1))
    
    names(comp_1) <- gsub("padj", paste0(contrastTag, "padj_",colName), names(comp_1))
    
    comp_1 <- comp_1[order(comp_1$lg10p, decreasing = T),]
    
    names(comp_1) <- gsub("lg10p", paste0(contrastTag, "lg10p_",colName), names(comp_1))
    
    selVec <- c(
        "gene",
        names(comp_1)[grep(contrastTag, names(comp_1))]
    )
    
    comp_1 <- unique(comp_1[,selVec])
    
    ## Cut all strings in comp_1 to a maximum of 63 charcters 
    names(comp_1) <- sapply(names(comp_1), function(x) substr(x, 1, 63))
    
    
    
    ##                                                                        ##
    ############################################################################
    
    
    
    returnList <- list(
      "OsC" = obj,
      LRTtable = comp_1
    )
    
    return(returnList)
})


## End doLRTsc Function                                                      ##
###############################################################################

###############################################################################
## Define differential gene expression comparisons in a list                 ##
## One list entry per comparison                                             ##

## Create custom column for this project:
# samplegroup_seurat_clusters


# OsC@meta.data[["samplegroup_seurat_clusters"]] <- paste0(
#     OsC@meta.data$meta_Age_Group,
#     "_C",
#     OsC@meta.data$seurat_clusters
# )
 
dfTemp <- unique(OsC@meta.data[,c("clusterName", "seurat_clusters")])
dfTemp$clusterName <- factor(dfTemp$clusterName, levels= Obio@parameterList$clusterNameOrder)
dfTemp <- dfTemp[order(dfTemp$clusterName),]


comparisonCol <- "seurat_clusters"
groupCol <- "clusterName"

LRToptions <- unique(dfTemp[,groupCol])

compList <- list()

## Make LRT list
for (i in 1:length(LRToptions)){
    tag <- paste0("LRT_", i)
    
    name <- paste0("LRT_", as.vector(dfTemp[i, "clusterName"]), "_Age_Group")
    
    LLRT <- list(
          "seurat_clusters" = as.numeric(dfTemp[i, "seurat_clusters"])
      )
      
      #names(LLRT) <- name
      
      compList[[tag]] <- list(
          "name"    = name,
          "LRTselCol" = "meta_Age_Group",
          "LRTsampleList" =  LLRT
        
      )
      
      names(compList)
     
      
    } 
    
    



LRTtagVec <- as.vector(NULL, mode = "character")


LRTinputList <- compList


# LRTtagVec <- as.vector(NULL, mode = "character")
# 
# LRTinputList <- list(
#   "LRT_1" = list(
#       "name"    = "LRT_Oligo_A",
#       "LRTselCol" = "meta_Age_Group",
#       #"LRTgroupCol" = "meta_Age_Group",
#       "LRTsampleList" = list(
#           "seurat_clusters"   = c(0)
#       )
#   ),
#   "LRT_2" = list(
#       "name"    = "LRT_Astro_A_age_group",
#       "LRTselCol" = "meta_Age_Group",
#       #"LRTgroupCol" = "meta_Age_Group",
#       "LRTsampleList" = list(
#           "seurat_clusters"   = c(1)
#       )
#   )
# )




###############################################################################
## Function do scLRT                                                         ##

doScLRTonList <- function(
    OsC,
    tag,
    LRTinputList,
    LRTtagVec = as.vector(NULL, mode = "character")
    
){
    resList <- doLRTsc(
        obj = OsC,
        LRTselCol = unlist(LRTinputList[[i]]["LRTselCol"]),
        colName = unlist(LRTinputList[[i]]["name"]),
        contrastTag = paste0("contrast_L",i,"_"),
        LRTsampleList = LRTinputList[[i]][["LRTsampleList"]]
    )
    
    OsC <- resList$OsC
    
    ###########################################################################
    ## Add average intensity for MA-Plot to LRT table                        ##
    
    OsC_LRT <- OsC
    OsC_LRT@meta.data[["LRT_sel"]] <- 0
    
    OsC_LRT@meta.data[OsC_LRT@meta.data[,paste0("meta_",unlist(LRTinputList[[i]]["name"]))] != "Rest", "LRT_sel"] <- 1
    
    #row.names(OsC_LRT@meta.data) <- OsC_LRT@meta.data[,"cellID"]
    #unlist(LRTinputList[[i]]["name"])
    
    
   
    
    OsC_LRT <- subset(x = OsC_LRT, subset = LRT_sel == 1)
    
    
    Idents(OsC_LRT) <- "LRT_sel"
    
    cluster.averages <- Seurat::AverageExpression(
        OsC_LRT, 
        return.seurat = TRUE
    )
    
    dfAvgExpr <- data.frame(cluster.averages[["RNA"]]@data)
    dfAvgExpr[["gene"]] <- row.names(dfAvgExpr)
    row.names(dfAvgExpr) <- NULL
    
    nameTag <- paste0("add_MA_cts_MA_Avg_", unlist(LRTinputList[[i]]$name))
    names(dfAvgExpr) <- gsub("all", nameTag, names(dfAvgExpr))
    
    dfDataTable <- resList$LRTtable
    
    dfDataTable <- merge(
        dfDataTable, 
        dfAvgExpr, 
        by.x = "gene",
        by.y = "gene",
        all =TRUE
    )
    
    dfDataTable[is.na(dfDataTable)] <- 0
    names(dfDataTable) <- sapply(names(dfDataTable), function(x) substr(x, 1, 63))
    
    LRTtagVec <- c(
      LRTtagVec,
      tag
    )
    
    resultList <- list(
        "dfDataTable" = dfDataTable,
        "LRTtagVec" = LRTtagVec,
        "OsC" = OsC
    )
    
    return(resultList)
    
}

## Done do scLRT                                                             ##
###############################################################################

###############################################################################
## LRT Analysis                                                              ##

LRTtagVec = as.vector(NULL, mode = "character")
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

## This does the LRT and creates an overview UMAP figure ##
for (i in 1:length(LRTinputList)){
    
    tag <- paste0(unlist(LRTinputList[[i]]["name"]))
    
    if (calculateLRTfromScratch){
      resList <- doScLRTonList(
          OsC = OsC,
          tag = tag,
          LRTinputList = LRTinputList,
          LRTtagVec = LRTtagVec
      )
    
      dfDataTable <- resList[["dfDataTable"]]
      LRTtagVec <- resList[["LRTtagVec"]]
      OsC <- resList[["OsC"]]
      ## Add results to biologic object 
      Obio@dataTableList[[tag]] <- data.frame(NULL)
      Obio@dataTableList[[tag]] <- dfDataTable
    } else {
      dfDataTable <- Obio@dataTableList[[tag]]
      LRTtagVec <- c(
        LRTtagVec, 
        tag
      )
    }
    
    
    ## Done with LRT Analysis                                                ##
    ###########################################################################
    
    ###########################################################################
    ## Create LRT Selection plot                                             ##
    dfPlot <- OsC@meta.data
    
    
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }

    pos <- grep("cellID", names(dfPlot))
      if (length(pos) == 0){
      dfPlot[["cellID"]] <- row.names(dfPlot)
    }

    dfPlot$UMAP_1 <- NULL
    dfPlot$UMAP_2 <- NULL
            
    ## Get UMAP coordinates ##
    coord <- data.frame(OsC@reductions$umap@cell.embeddings)
    coord[["cellID"]] <- row.names(coord)
    coord <-coord[coord$cellID %in% dfPlot$cellID, ]
            
    dfPlot <- merge(dfPlot, coord, by.x = "cellID", by.y="cellID", all=T)
    dfPlot[is.na(dfPlot)] <- 0
    dfPlot <- dfPlot[dfPlot$UMAP_1 != 0 & dfPlot$UMAP_2 != 0,]
            
            
    ## Add cluster colors ##
    dfPlot[["Cluster"]] <- dfPlot[,paste0("meta_", unlist(LRTinputList[[i]]["name"]))]
    dfPlot$Cluster[is.na(dfPlot$Cluster)] <- ""
    dfPlot[dfPlot$Cluster == "", "Cluster"] <- "Rest"
    
    label <- as.vector(gsub("LRT_", "", unlist(LRTinputList[[i]]["name"])))
    
    
    redGroup <- names(LRTinputList[[i]]$LRTsampleList)[1] #as.vector(unlist(strsplit(label, "_vs_"))[1])
    #blueGroup <- names(LRTinputList[[i]]$LRTsampleList)[2] #as.vector(unlist(strsplit(label, "_vs_"))[2])
    greyGroup <- "Rest"
    
    ## Assigning colors ##
    
    
    
    colVec <- c(
      "Young" = "#E6C229", 
      "Middle" = "#ED1D58", 
      "Old"= "#48A6EA",
      "Rest" = "grey"
    ) 
    
    if (sum(unique(dfPlot$Cluster) %in% names(colVec)) == length(colVec)){
      dfCol <- data.frame(name = names(colVec), col = colVec)
    } else {
      dfCol <- data.frame(name = unique(dfPlot$Cluster), col = rep("", length(unique(dfPlot$Cluster))), stringsAsFactors = F)
    
    
   
    
    #dfCol[dfCol$name == redGroup, "col"] <- "#000080"
    cols <- rainbow(nrow(dfCol[dfCol$name != "Rest",]))
    dfCol[dfCol$name != "Rest", "col"] <- cols
    dfCol[dfCol$name == greyGroup, "col"] <- "grey"
    colVec <- dfCol$col
    names(colVec) <- dfCol$name
      
    }
    
    
    dfColExtra <- dfCol[dfCol$col == "",]
    if (nrow(dfColExtra) > 0){
        eColVec <- dfColExtra$name
        library(scales)
        eCols = hue_pal()(length(eColVec))
        for (l in 1:length(eColVec)){
            dfCol[dfCol$name == eColVec[l], "col"] <- eCols[l]
        }
    }
    
    maxX <- 1.1*max(dfPlot$UMAP_1, na.rm = T)
    minX <- 1.1*min(dfPlot$UMAP_1, na.rm = T)
    maxY <- 1.1*max(dfPlot$UMAP_2, na.rm = T)
    minY <- 1.1*min(dfPlot$UMAP_2, na.rm = T)            

    #colVec <- dfPair$sampleColor
    #names(colVec) <- dfPair$sampleName
    levels <- as.vector(dfCol$name)
    levelsRest <- levels[grep("Rest", levels)]
    levels <- levels[levels != "Rest"]
    levels <- c(levels, levelsRest)
    
    dfPlot$Cluster <- factor(dfPlot$Cluster, levels = levels)

    dotsize  = 1
    if (nrow(dfPlot) > 10000){
        dotsize  = 0.5
    } else if (nrow(dfPlot) > 20000){
        dotsize = 0.25
    } else if (nrow(dfPlot) > 50000){
        dotsize = 0.1
    }

    ## make plot ##
    plotList[[tag]] <- ggplot2::ggplot(
        data=dfPlot[dfPlot$included == "+",], 
        ggplot2::aes(UMAP_1, UMAP_2, color=Cluster)) + 
        ggrastr::rasterize(ggplot2::geom_point( shape=16, size = as.numeric(dotsize)), dpi = rasterDpi) + 
        ggplot2::xlab("UMAP1") + 
        ggplot2::ylab("UMAP2") + 
        ggplot2::theme_bw()  +  
        ggplot2::theme(
          axis.text.y   = ggplot2::element_text(size=8),
          axis.text.x   = ggplot2::element_text(size=8),
          axis.title.y  = ggplot2::element_text(size=8),
          axis.title.x  = ggplot2::element_text(size=8),
          axis.line = ggplot2::element_line(colour = "black"),
          panel.border = ggplot2::element_rect(colour = "black", fill=NA, size=1),
          plot.title = ggplot2::element_text(hjust = 0.5, size = 12),
          legend.title = ggplot2::element_blank()
        ) + 
        ggplot2::guides(col = ggplot2::guide_legend(override.aes = list(shape = 16, size = legendDotSize))) +     
        ggplot2::ggtitle(paste0("Cell Selection for ", gsub("_", " ", unlist(LRTinputList[[i]]["name"])))) + 
        ggplot2::xlim(minX, maxX) + 
        ggplot2::ylim(minY, maxY) + 
        ggplot2::coord_fixed(ratio=1) + 
        ggplot2::scale_colour_manual("LRT Groups" ,values = colVec) + 
        ggplot2::guides(col = ggplot2::guide_legend(override.aes = list(shape = 16, size = legendDotSize)))


      ## Add colors if specified ##

            
      ## Make plot figure ##
      FNbase <- paste0(tag, VersionPdfExt)
      FN <- paste0(reportFigDir, FNbase)
      FNrel <- paste0("report_figures/", FNbase)
              
      pdf(FN)
          print(plotList[[tag]])
      dev.off()
              
        # link <- paste0('<a href="https://',urlString,'/',Obio@parameterList$project_id,'/pca?x_axis=UMAP_1&y_axis=UMAP_2" target="_blank">here</a>')  
        
      if (exists("shinyURL") & !is.null(shinyURL)){
          link <- paste0(
                  'An interactive version of this figure with additional viewing options can be found <a href="',shinyURL,'?_inputs_&y_axis=%22UMAP_2%22&x_axis=%22UMAP_1%22&colorBy=%22',paste0("meta_", tag),'%22&splitByColumn=%22all%22" target="_blank">here</a>. '
          )
                
      } else {
          link <- ""
      }
              
      figLegend <- paste0(
      '**Figure ', 
      figureCount, 
                ':** ',
                ' UMAP showing all cells from all samples together. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. ',
                'An interactive version of this figure can be found ', link
            )
            
      figureCount <- figureCount + 1
            
      
      ## Rasterize plot for plotlist to cut down on size
      plotList[[tag]] <-  ggrastr::rasterise(
        plotList[[tag]],  
        layers="Point", 
        dpi=400
      )
      
      NewChnk <- paste0(
          "#### ", tag,
          "\n```{r Sample_LRT_",
          tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
          figLegend,"'}\n",
          "\n",
          "\n print(plotList[['",tag,"']])",
          "\n cat(  '\n')",
          "\n\n\n```\n"   
      )
            
      chnkVec <- c(
          chnkVec,
          NewChnk
      )
  
      ## Done creating LRT selection plot                                      ##
      ###########################################################################
      
      
      print(paste0("Done with ", tag))
}

## Done LRT                                                                  ##
###############################################################################

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
``` 

### Likelihood-ratio test Cell Selections {`r tabVar`}
In this section the cells that were used for the LRT comparisons are highlighted. 

```{r CellTypePlot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))

``` 


```{r data_LRT_summary, echo=F, eval=T, warning=FALSE, results="asis"}

###############################################################################
## Assemble LRT summary table                                                ##
#dfTemp <- Obio@dataTableList[["LRT_table"]] 

wb <- openxlsx::createWorkbook()

hs1 <- openxlsx::createStyle(
    fontColour = "#ffffff",
    fgFill = "#000000",
    halign = "CENTER",
    textDecoration = "Bold"
)

###############################################################################
## Add Create LRT result Excel Spreadsheet                                   ##

for (i in 1:length(LRTtagVec)){
  
  ## Add to Excel workbook ##
  dfOutput <-   Obio@dataTableList[[LRTtagVec[i]]]
  
  pos <- grep("logFC_", names(dfOutput))
  
  if (length(pos) > 0){
    dfOutput[,pos] <- NULL
  }
  
  dfOutput[["gene_FeatureViewLink"]] <- paste0(
      shinyURL,
      "?gene=",
      dfOutput$gene
  )
  
  ## select only significant genes
  pos <- grep("_padj_", names(dfOutput))
  padjCol <- names(dfOutput)[pos]
  
  
  if (nrow(dfOutput[dfOutput[,padjCol] < 0.01,]) > 3){
    cutOff <- 0.01
  } else if (nrow(dfOutput[dfOutput[,padjCol] < 0.05,]) > 3){
    cutOff <- 0.05
  } else {
    cutOff <- 0
  }
  
  dfOutput <- dfOutput[dfOutput[,padjCol] < cutOff,]
  
  pos <- grep("_cts_", names(dfOutput))
  ctsCol <- names(dfOutput)[pos]
  
  if (nrow(dfOutput[dfOutput[,ctsCol] > 0,]) > 0){
    dfOutput <- dfOutput[dfOutput[,ctsCol] > 0,]
  }
  
  ## Order based on intensity * lg10p product
  pos <- grep("_lg10p_", names(dfOutput))
  lg10pCol <- names(dfOutput)[pos]
  
  dfOutput[["intensity_weighted_lg10p"]] <- dfOutput[[lg10pCol]] * dfOutput[[ctsCol]]
  
  dfOutput <- dfOutput[order(dfOutput[,"intensity_weighted_lg10p"], decreasing = T),]
  
  ## Mark transcription factors ##
  if (!is.null(Obio@dataTableList[["referenceList"]]$TFs)){
    cat <- Obio@dataTableList[["referenceList"]]$TFs
    tag <- "Transcription_Factor"
    dfOutput[[tag]] <- ""
    if (sum(dfOutput$gene %in% cat) > 0){
        dfOutput[dfOutput$gene %in% cat, tag] <- "+" 
      } 
  }
  
  if (!is.null(Obio@dataTableList[["referenceList"]]$GO_Secretion)){
    cat <- Obio@dataTableList[["referenceList"]]$GO_Secretion
    
     tag <- "GO_Secretion"
    dfOutput[[tag]] <- ""
    if (sum(dfOutput$gene %in% cat) > 0){
        dfOutput[dfOutput$gene %in% cat, tag] <- "+" 
      }
  }
  
  if (!is.null(Obio@parameterList$catRefFile)){
    dfCat <- read.delim(
      Obio@parameterList$catRefFile,
      header = T, 
      sep = "\t",
      stringsAsFactors = F
    )
    
    pos <- grep("Aging_encoding_probes_Allen_et_al_2023", names(dfCat))
    
    if (length(pos) ==1){
      cat <- dfCat[3:nrow(dfCat),pos]
      cat <- cat[cat != ""]
      tag <- "Aging_Genes"
      dfOutput[[tag]] <- ""
      if (sum(dfOutput$gene %in% cat) > 0){
        dfOutput[dfOutput$gene %in% cat, tag] <- "+" 
      }
    }
  }
    
     
    
  
  orderVec <- names(dfOutput)
  orderVec <- orderVec[orderVec != "gene"]
  orderVec <- c("gene", "gene_FeatureViewLink", orderVec)
  dfOutput <- dfOutput[, orderVec]
  class(dfOutput$gene_FeatureViewLink) <- c(class(dfOutput$gene_FeatureViewLink), "hyperlink")
  
  sheetName <- substr(LRTtagVec[i],1,30)
  
  openxlsx::addWorksheet(
      wb, 
      sheetName = sheetName
  )
  
  
  openxlsx::freezePane(wb, sheetName ,  firstActiveRow = 2)
  openxlsx::addFilter(wb, i, row = 1, 1:ncol(dfOutput))
  openxlsx::writeData(wb, i, dfOutput, startRow = 1, startCol = 1, headerStyle = hs1)
  ## Done adding to Excel workbook ##
  
  if (i ==1){
      dfTemp <- Obio@dataTableList[[LRTtagVec[i]]]
  } else {
      dfTemp <- merge(
        dfTemp, 
        Obio@dataTableList[[LRTtagVec[i]]],
        by.x = "gene",
        by.y = "gene",
        all =TRUE
      )
      dfTemp[is.na(dfTemp)] <- 0
  
  }
}

## Save workbook ##
baseFN <- paste0(
   project_id, 
   ".LRT.table.xlsx"
)


outPutFN <- paste0(
     reportTableDir,
     baseFN
)

if (!dir.exists(reportTableDir)){
    dir.create(reportTableDir, recursive=T)
}

openxlsx::saveWorkbook(
        wb,
        outPutFN ,
        overwrite = TRUE
)

Obio@dataTableList[["LRT_table"]] <- NULL
Obio@dataTableList[["LRT_table"]] <- dfTemp


## Done creating xlsx spreadsheet                                            ##
###############################################################################
```



```{r data_Volcano_plots, echo=F, eval=T, warning=FALSE, results="asis"}
###############################################################################
## Make Volcano plots
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

dfAllPlots <- Obio@dataTableList[["LRT_table"]] 

for (i in 1:length(LRTtagVec)){
    tag <- paste0("LRT_Plot_", LRTtagVec[i])  
    label <- gsub("_", " ", LRTtagVec[i])
    
    selVec <- c(
        "gene",
        names(dfAllPlots)[grep(paste0(LRTtagVec[i], "$"), names(dfAllPlots))]
    )
    
    dfPlot <- dfAllPlots[,selVec]
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }
    
    fourSD <- 4*sd(dfPlot[,grep("_lg10p_", names(dfPlot))])
    
    dfPlot[["LRT_Status"]] <- "Unchanged"
    dfPlot[dfPlot[,grep("_lg10p_", names(dfPlot))] > fourSD & dfPlot[,grep("_cts_", names(dfPlot))] > 0, "LRT_Status"] <- "Variable"
    # dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] < -1 *logFCfourSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "LRT_Status"] <- "Down"
    # 
    colVec <- c("red", "black")
    names(colVec) <- c("Variable",  "Unchanged")
    # 
    
    dfPlot[["X"]] <- dfPlot[,grep("_cts_MA", names(dfPlot))]
    dfPlot[["Y"]] <- dfPlot[,grep("_lg10p_", names(dfPlot))]
    dfPlot[["intensity_weighted_lg10p"]] <- dfPlot[["X"]] * dfPlot[["Y"]]
    
    dotsize  = 1.25
    if (nrow(dfPlot) > 10000){
      dotsize  = 1
    } else if (nrow(dfPlot) > 20000){
      dotsize = 0.75
    } else if (nrow(dfPlot) > 50000){
      dotsize = 0.5
    }

    Ylims <- ceiling(max(abs(dfPlot$X)))
    
    ## Add to-regulated gene names ##
    topNgenes <- 20
    dfPlot <- dfPlot[order(dfPlot$intensity_weighted_lg10p, decreasing = T), ]
    topGenes <- as.vector(dfPlot[1:topNgenes,"gene"])
    
    #dfPlot <- dfPlot[order(dfPlot$Y, decreasing = F), ]
    #bottomGenes <- as.vector(dfPlot[1:topNgenes,"gene"])
    
    dfPlot[["label"]] <- ""
    dfPlot[dfPlot[,"gene"] %in% c(topGenes), "label"] <- dfPlot[dfPlot[,"gene"] %in% c(topGenes), "gene"]
    
    breaks <- ceiling(max(dfPlot$Y) / 20)
    
    roundUp <- function(x, m) {
      m * ceiling(x/m)
    }
    
    ybreak <- roundUp(breaks, m = 5)
    
    plotList[[tag]] <- ggplot2::ggplot(
        data=dfPlot[dfPlot$included == "+",], 
        ggplot2::aes(X, Y, label = label, color= LRT_Status)) + 
        ggplot2::geom_hline(yintercept = fourSD, col = "darkgrey", lty=2) +   
        ggrastr::rasterize(ggplot2::geom_point( shape=16, size = as.numeric(dotsize),alpha = 0.7), dpi=rasterDpi) + 
        ggplot2::xlab(paste0("Average Expression Intensity ", gsub("_", " ",label))) + 
        ggplot2::ylab(paste0("-log10(adj LRT p-value) ", gsub("_", " ",label))) + 
        ggplot2::theme_bw()  +  ggplot2::theme(
          axis.text.y   = ggplot2::element_text(size=8),
          axis.text.x   = ggplot2::element_text(size=8),
          axis.title.y  = ggplot2::element_text(size=8),
          axis.title.x  = ggplot2::element_text(size=8),
          axis.line = ggplot2::element_line(colour = "black"),
          panel.border = ggplot2::element_rect(colour = "black", fill=NA, size=1),
          plot.title = ggplot2::element_text(hjust = 0.5, size = 12)) + 
      ggplot2::guides(col = ggplot2::guide_legend(override.aes = list(shape = 16, size = legendDotSize))) + 
      ggplot2::ggtitle(paste0("LRT Plot ", gsub("_", " ", label))) + 
      ggplot2::ylim(Ylims, -1 * Ylims) + 
      ggplot2::scale_y_continuous(breaks=c(seq(0,400,ybreak))) + 
      ggplot2::scale_x_continuous(breaks=c(seq(0,400,2))) +  
      ggplot2::scale_color_manual(values=colVec
      ) 


      ## Add labels 
      plotList[[tag]] <- plotList[[tag]] + ggrepel::geom_text_repel(size = 3, max.iter = 10000)
      # plotList[[tag]] <- plotList[[tag]] + ggplot2::geom_text(nudge_x = 0.05)      

      FNbase <- paste0(tag, VersionPdfExt)
      FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
      FNrel <- paste0("report_figures/", FNbase)
            
      pdf(FN)
          print(plotList[[tag]])
      dev.off()
      
      xLabel <- names(dfPlot[grep("_logFC_", names(dfPlot))])      
      yLabel <- names(dfPlot[grep("_lg10p_", names(dfPlot))])  
      link <- paste0('<a href="https://',urlString,'/',Obio@parameterList$project_id,'/scatterplot?x_axis=',xLabel,'&y_axis=',yLabel,'" target="_blank">here</a>')  
            
      
      
      figLegend <- paste0(
      '**Figure ', 
      figureCount, 
                ':** ',
                ' Volcano plot for differential gene expression comparison ',gsub("_", " ", LRTtagVec[i]),'. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. ',
                'An interactive version of this figure can be found ', link, '. ',
                tableString
            )
            
      figureCount <- figureCount + 1
            
      NewChnk <- paste0(
          "#### ", tag,
          "\n```{r Volcano_LRT_",
          tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
          figLegend,"'}\n",
          "\n",
          "\n print(plotList[['",tag,"']])",
          "\n cat(  '\n')",
          "\n\n\n```\n"   
      )
            
      chnkVec <- c(
          chnkVec,
          NewChnk
      )

      ## Done creating LRT selection plot                                      ##
      ###########################################################################
    
    
    
    
}


## Done Volcano plots
###############################################################################

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
``` 

### LRT Plots {`r tabVar`}

```{r LRT_volcano_plot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))

``` 


```{r create-lrt-enrichment-data, echo=T, eval=TRUE, warning=FALSE, results=F, error=F}
Obio@dbDetailList$db.user <- "babs"
## Create enriched genes list ##

## Data processing for category enrichment


EnrichedGenesList <- list()
dfAllPlots <- Obio@dataTableList[["LRT_table"]] 

if (Obio@parameterList$geneIDcolumn != "mgi_symbol" & Obio@parameterList$geneIDcolumn != "hgnc_symbol") {
  queryGS <- "hgnc_symbol" 
} else {
  queryGS <- Obio@parameterList$geneIDcolumn
}

for (i in 1:length(LRTtagVec)){
  tag <- paste0("LRT_Enrichment_", LRTtagVec[i])  
  tagGLpos <- paste0(LRTtagVec[i], "_pos") 
  tagGLneg <- paste0(LRTtagVec[i], "_neg") 
    
  selVec <- c(
    "gene",
    names(dfAllPlots)[grep(paste0(LRTtagVec[i], "$"), names(dfAllPlots))]
  )
    
  dfPlot <- dfAllPlots[,selVec]
  pos <- grep("included", names(dfPlot))
  if (length(pos) == 0){
    dfPlot[["included"]] <- "+"
  }
    
  fourSD <- 4*sd(dfPlot[,grep("_lg10p_", names(dfPlot))])
    
  dfPlot[["LRT_Status"]] <- "Unchanged"
  dfPlot[dfPlot[,grep("_cts_", names(dfPlot))] > 0 & dfPlot[,grep("_lg10p_", names(dfPlot))] > fourSD, "LRT_Status"] <- "Variable"
    EnrichedGenesList[[tagGLpos]] <- unique(dfPlot[dfPlot$LRT_Status == "Variable", "gene"])
    
    #dfPlot[dfPlot[,grep("_logFC_", names(dfPlot))] < -1 *logFCfourSD & dfPlot[,grep("_padj_", names(dfPlot))] < 0.05, "LRT_Status"] <- "Down"
    #EnrichedGenesList[[tagGLneg]] <- unique(dfPlot[dfPlot$LRT_Status == "Down", "gene"])
} 

## Done creating gene list                                                   ##
###############################################################################

#library(knitr::)
#library(ggplot2)

#save.image("temp.RData")
#library(clusterProfiler)

gmtList <- list()
pos <- grep("clusterSigEnrichmentList", slotNames(Obio))
    
if (length(pos) > 0){
  if (is.null(Obio@clusterSigEnrichmentList)){
    dbtableList <- list(
      "Cell Type Signatures" = "mysigdb_sc_sig",
      "Cell Type Signatures" = "cibersort_L22",
      "GO-MF" = "mysigdb_c5_MF",
      "Pathways" = "mysigdb_c2_1329_canonical_pathways",
      "Protein Complexes" = "networkcategories"
    )
  } else {
    dbtableList <- Obio@clusterSigEnrichmentList
  }
} else {
  dbtableList <- list(
    "Cell Type Signatures" = "mysigdb_sc_sig",
    "Cell Type Signatures" = "cibersort_L22",
    "GO-MF" = "mysigdb_c5_MF",
    "GO-BP" = "mysigdb_c5_BP",
    "Pathways" = "mysigdb_c2_1329_canonical_pathways",
    "Protein Complexes" = "networkcategories"
  )
}
    
    
rmVec <- grep("Cell Type Signatures", names(dbtableList))
if (length(rmVec) > 0){
  dbtableList <- dbtableList[-rmVec]
}
    
for (i in 1:length(dbtableList)){
  dfTemp <- unique(import.db.table.from.db(
  host = Obio@dbDetailList$host,
  dbname = Obio@dbDetailList$ref.cat.db,
  dbtable = dbtableList [[i]],
  password = db.pwd,
  user = Obio@dbDetailList$db.user
  ))
        
## Remove duplicated entries ##
dfTemp <- dfTemp[!(duplicated(dfTemp$cat_name)),]
        
        rmVec <- grep("temp_", dfTemp$cat_type)
        if (length(rmVec) > 0){
            dfTemp <- dfTemp[-rmVec, ]
        }
        
        dfTemp <- unique(dbcat2gmt(
            dfTemp, # As downloaded from reference_categories_db_new database
            gene.id.column = queryGS
        ))
        
        dfTemp <- unique(dfTemp[!duplicated(as.vector(dfTemp[,1])), ])
        
        write.table(
            dfTemp,
            "temp.gmt.txt",
            row.names = F, 
            sep = "\t",
            col.names = F,
            quote = F
        )
        
        CPgmt <- clusterProfiler::read.gmt("temp.gmt.txt")
        unlink("temp.gmt.txt")
        CPgmt <- unique(CPgmt[CPgmt$gene != "", ])
        
        gmtList[[dbtableList[[i]]]] <- CPgmt
    }
    
    ## Edit collection names for plot
    names(gmtList) <- gsub("mysigdb_", "",names(gmtList))
    names(gmtList) <- gsub("c2_1329_canonical_p", "P",names(gmtList))
    names(gmtList) <- gsub("sc_sig", "CellSig",names(gmtList))
    names(gmtList) <- gsub("cibersort_L22", "CellSig",names(gmtList))
    names(gmtList) <- gsub("c5_", "GO_",names(gmtList))
    names(gmtList) <- gsub("networkcategories", "Complexes",names(gmtList))
    ## Done creating gmt list
    ###########################
    
    ## Select colors ##
    #library(scales)
    enrCols <- scales::hue_pal()(length(gmtList))
    names(enrCols) <- substr(names(gmtList),1,10)



plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")



for (j in 1:length(LRTtagVec)){
  posTestGeneSet <- as.vector(
      unique(
          EnrichedGenesList[[paste0(LRTtagVec[j], "_pos")]]
      )
  )
  
  if (Obio@parameterList$geneIDcolumn != "mgi_symbol" & Obio@parameterList$geneIDcolumn != "hgnc_symbol") {
    queryGS <- "hgnc_symbol" 
  } else {
            queryGS <- Obio@parameterList$geneIDcolumn
  }
        
  if (Obio@parameterList$host == "10.27.241.234"){
    urlString <- "biologic.thecrick.org"
  } else {
    urlString <- "biologic.crick.ac.uk"
  }
    
  colVec <- c("#009900")
  pvalueCutoff <- 0.5
  topMaxCat <- 20
    
  ## Get background gene set ##
  #backgroundGeneVec <- row.names(OsC[["RNA"]]@counts)
  if ((length(posTestGeneSet) >= 3) ){
    ## Do enrichment ##
    first <- TRUE
    if (length(posTestGeneSet) >= 3){
      for (k in 1:length(gmtList)){
        egmt <- data.frame(
          clusterProfiler::enricher(
            posTestGeneSet, 
            TERM2GENE=gmtList[[k]],
            pvalueCutoff = pvalueCutoff
          )
        )
        if (!is.null(egmt)){
          if (nrow(egmt) > 0){
            egmt[["Collection"]] <- substr(names(gmtList)[k], 1,10)
          }
          if (first){
            dfTempEnriched <- egmt    
            first <- FALSE
          } else {
            dfTempEnriched <- rbind(
              dfTempEnriched, 
              egmt
            )    
          }
                        
        }
      }
           
      if (exists("dfTempEnriched") && nrow(dfTempEnriched) > 0){
        dfTempEnriched[["direction"]] <- "positive"
        dfTempEnriched[["log10FDR"]] <- log10(dfTempEnriched$p.adjust)
        
        dfTempEnriched[["log10FDR"]] <- -1*log10(dfTempEnriched$p.adjust)
        
        dfTempEnriched <- dfTempEnriched[order(dfTempEnriched$log10FDR, decreasing = F),]
        dfTempEnriched <- na.omit(dfTempEnriched)
        dfEnriched <- unique(dfTempEnriched)
                
        if (nrow(dfTempEnriched) > topMaxCat){
          dfTempEnriched <- dfTempEnriched[1:topMaxCat, ]
        }
      }
          
            
    } # end positive
            
            ## Now the negative side ##
            # if (length(negTestGeneSet) >= 3){
            # first <- TRUE
            # for (k in 1:length(gmtList)){
            #         egmt <- data.frame(
            #             clusterProfiler::enricher(
            #                 posTestGeneSet, 
            #                 TERM2GENE=gmtList[[k]],
            #                 pvalueCutoff = pvalueCutoff
            #             )
            #         )
            #         if (!is.null(egmt)){
            #             if (nrow(egmt) > 0){
            #                 egmt[["Collection"]] <- substr(names(gmtList)[k], 1,10)
            #             }
            #             if (first){
            #                 dfTempEnrichedNeg <- egmt    
            #                 first <- FALSE
            #             } else {
            #                 dfTempEnrichedNeg <- rbind(
            #                     dfTempEnrichedNeg, 
            #                     egmt
            #                 )    
            #             }
            #             
            #         } 
            # }
            # if (exists("dfTempEnrichedNeg") && nrow(dfTempEnrichedNeg) > 0){
            #     dfTempEnrichedNeg[["direction"]] <- "negative"
            #     dfTempEnrichedNeg[["log10FDR"]] <- -1*log10(dfTempEnrichedNeg$p.adjust)
            #     dfTempEnrichedNeg <- dfTempEnrichedNeg[order(dfTempEnrichedNeg$log10FDR, decreasing = T),]
            #     dfTempEnrichedNeg <- na.omit(dfTempEnrichedNeg)
            #     dfEnrichedNeg <- dfTempEnrichedNeg
            #     
            #     if (nrow(dfTempEnrichedNeg) > topMaxCat){
            #         dfTempEnrichedNeg <- dfTempEnrichedNeg[1:topMaxCat, ]
            #     }
            # }
            # } # end negative
        
            
            
            ## Make plot 
            if (nrow(dfTempEnriched) > 0){
            
            
            
            
            dfSel <- rbind(
                dfTempEnriched
                #dfTempEnrichedNeg
            )
            
            dfSel <- na.omit(dfSel)
            dfSel <- dfSel[order(dfSel$log10FDR),]
            dfSel$log10FDR <- round(dfSel$log10FDR, 2)
            
            dfSel[["Category"]] <- ""
            dfSel[dfSel$log10FDR >= 0, "Category"] <- "Enr."
            #dfSel[dfSel$log10FDR < 0, "Category"] <- "Depl."
            
            for (l in 1:nrow(dfSel)){
                if (nchar(dfSel[l, "ID"]) > 30){
                    part1 <- substr(dfSel[l, "ID"], 1, 30)
                    part2 <- substr(dfSel[l, "ID"], 31, 60)
                    dfSel[l, "ID"] <- paste0(part1, " \n ", part2)
                  
                }
            }
            
            
            #dfSel$Term <- gsub("\\(GO", "\\\n\\(GO", dfSel$Term)
            
            dfSel$ID <- factor(dfSel$ID, levels = unique(dfSel$ID))
            
            
            
            plotList[[paste0("LRT_ENR_", j)]] <- ggplot2::ggplot(
                data=dfSel, ggplot2::aes(x= ID, y=log10FDR, fill=Collection, order=log10FDR)
            ) + ggplot2::geom_bar(stat="identity", colour="black"
            ) + ggplot2::coord_flip() + ggplot2::scale_fill_manual(values=enrCols
            ) + ggplot2::theme_bw(
            )  +  ggplot2::theme(
                axis.text.y   = ggplot2::element_text(size=8),
                axis.text.x   = ggplot2::element_text(size=8),
                axis.title.y  = ggplot2::element_text(size=8),
                axis.title.x  = ggplot2::element_text(size=8),
                axis.line = ggplot2::element_line(colour = "black"),
                panel.border = ggplot2::element_rect(colour = "black", fill=NA, size=1),
                plot.title = ggplot2::element_text(hjust = 0.5, size = 12)
            )  + ggplot2::labs(title = paste0("Comparison ", LRTtagVec[j]," enriched genes") ,y = "-log10(FDR)", x = ""
            ) + ggplot2::geom_hline(yintercept = c(-1 * log10(0.05)), color = "grey", size=0.5, lty=2
            ) + ggplot2::geom_hline(yintercept = 0, color = "black", size=0.5
            ) 
            cat("  \n")
            
            
            
            ## Save to file ##
            FNbase <- paste0("LRT_comparison_", LRTtagVec[j],".enriched.genes", VersionPdfExt)
            FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
            FNrel <- paste0("report_figures/", FNbase)
            
           
            pdf(FN)
            print(plotList[[paste0("PCA_ENR_", j)]])
            dev.off()
            
            link <- paste0(
                '<a href="https://', urlString, '/',
                Obio@parameterList$project_id,
                '/category-view?category_type=GO-BP" target="_blank">CategoryView</a>'
            )
            
            ## Create R markdown chunk ##
            figLegend <- paste0(
                '**Figure ', 
                figureCount, 
                '**: Category enrichment analysis for the top genes that have  <font color = "',colVec[2],'"> the most positive </font> and <font color = "',colVec[1],'">the most negative</font> PCA loading values in dimension ', 
               LRTtagVec[j],
                ' associated with them. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. To view these gene sets in the context of your data, go to ',link,' and find these categories using the search box.'
            )
            figureCount <- figureCount + 1 
            
            NewChnk <- paste0(
                "#### ", LRTtagVec[j],
                "\n```{r enrichr_",
                j,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                figLegend,"'}\n",
                "\n",
                "\n print(plotList[['",paste0("LRT_ENR_", j),"']])",
                "\n cat(  '\n')",
                "\n\n\n```\n"   
            )
            
            chnkVec <- c(
                chnkVec,
                NewChnk
            )
            
            
            ####################################################################
            ## Create full output table                                       ##
            dfResTemp <- rbind(
                dfEnriched
                #dfEnrichedNeg
            )
            
            dfResTemp[["comparison"]] <- tag
            
            if (j==1){
                dfResTable <- dfResTemp
            } else {
                dfResTable <- rbind(
                    dfResTable, 
                    dfResTemp
                )
            }
            ## Done                                                           ##
            ####################################################################
        }
            
            
            ## done with plot 
            
    } ## Done with per dimension loops
}        
     
 
 

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
```

### Category Enrichments Differential Gene Expression Results {`r tabVar`} 
```{r create-cat-enrichment-plot, echo=T, eval=TRUE, warning=FALSE, results='asis'}
###############################################################################
## Do category enrichment on clusters                                        ##
cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))
## Done doing enrichment on clusters                                         ##
###############################################################################


```

```{r process_heatmap_data, echo=T, eval=TRUE, warning=FALSE, results=F, error=F}

## Get transcription factor genes ##
if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/nemo/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}


FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")
    dbTable <- read.delim(
      FN,
      header = F,
      sep = "\t",
      stringsAsFactors = F
    )
db.pwd <- as.vector(dbTable[1,1])

if (Obio@parameterList$geneIDcolumn != "mgi_symbol" & Obio@parameterList$geneIDcolumn != "hgnc_symbol") {
    queryGS <- "hgnc_symbol" 
} else {
    queryGS <- Obio@parameterList$geneIDcolumn
}


TFs <- retrieve.gene.category.from.db(
    cat_id = "ag_lab_categories__10",
    password = db.pwd,
    gene.symbol = queryGS,
    user = Obio@parameterList$db.user,
    host = Obio@parameterList$host
)

GO_Secretion <-  retrieve.gene.category.from.db(
                          cat_id = "mysigdb_c5_BP__1347",
                          password = db.pwd,
                          gene.symbol = queryGS,
                          user = Obio@dbDetailList$db.user,
                          host = Obio@dbDetailList$host
                    )


## Interesting genes for each cluster are stored in this list:
# EnrichedGenesList

## Create a heatmap list for each element of the above list

hmVec <- names(EnrichedGenesList)
HMgeneSelections <- list()

for (i in 1: length(hmVec)){
  tag <- paste0("HM_", hmVec[i])
  
  geneVec <- EnrichedGenesList[[hmVec[i]]]
  TFgeneVec <- geneVec[geneVec %in% TFs]
  secGeneVec <- geneVec[geneVec %in% GO_Secretion]
  RestGeneVec <- geneVec[!(geneVec %in% unique(c(TFgeneVec, secGeneVec)))]
  
  catList <- list()
  
  if (length(TFgeneVec) > 0){
    catList[["TFs"]] <- TFgeneVec
  }
  
  if (length(secGeneVec) > 0){
    catList[["Secretion"]] <- secGeneVec
  }
  
  if (length( RestGeneVec) > 0){
     catList[["Other"]] <- RestGeneVec
  }
  
  
  HMgeneSelections[[tag]] <- catList
  
  
}

## Done creation of gene categories
##############################

###############################################################################
## Function to create heatmaps                                               ##

addHeatmap2List <- function(
    dataMatrix,
    geneGroupList,
    cluster_column_slices   = FALSE,
    cluster_columns         = TRUE,
    relativeExpression = FALSE,
    cmdVec = NULL,
    tag = NULL,
    showAllLegends = TRUE,
    figureCount = 1,
    chnkVec = NULL, 
    plotList = NULL,
    maxNcharCatName = 10,
    dfMeta,
    group1ColorCol = "clusterColor", 
    group1NameCol = "clusterName",
    group2ColorCol = "sampleColor", 
    group2NameCol = "sampleName",
    columnSplitCol = "clusterName"
    ){
  
  if (is.null(chnkVec)){
    chnkVec <- as.vector(NULL, mode="character")
  }
  
  if (is.null(plotList)){
    plotList <- list()
  }
  
    allGenes <- unique(as.vector(unlist(geneGroupList)))
    
    
    if (is.null(cmdVec)){
      cmdVec <- as.vector(NULL, mode="character")
    }
    
    if (is.null(tag)){
      tag <- paste("Heatmap_", (length(plotList) + 1))
    }
    
    if (relativeExpression){
      legendString <- "log2 row mean"
    } else {
      legendString <- "lg10 Expr"
    }

# library("RColorBrewer")
# library("Seurat")
# library("ggplot2")
# library("viridis")
# #library("scMCA")
# library("circlize")
# library("ComplexHeatmap")



## Complex heatmap of cell/celltype correlations using just the top annotation per cell.  It's split by seurat cluster.
numbers_plot <- 1
# cors       <-  data.frame(OsC[["RNA"]]@data)
#cors_index <- apply(cors, 2, gettissue, numbers_plot)
#cors_index <- sort(unique(as.integer(cors_index)))
allGenes <- allGenes[allGenes %in% row.names(dataMatrix)]

heat.dat   <- dataMatrix[allGenes, ]
#heat.dat <- heat.dat
#heat.dat   <- t(apply(heat.dat,1,function(x){(x-mean(x))/sd(x)}))

# Comment out for absolute expression

heat.dat <- heat.dat[apply(heat.dat, 1, mean) != 0, ]

if (relativeExpression){
  heat.dat   <- t(apply(heat.dat,1,function(x){log2((x/mean(x)))}))  
} 

heat.dat[heat.dat == -Inf] <- 0

## make sure cat names aren't too long
nMax <- max(nchar(names(geneGroupList)))

if (nMax > maxNcharCatName){
  shortNames <- sapply(
    names(geneGroupList),
    function(x){
      substr(x, 1, maxNcharCatName)
    }
  )
  
  if (sum(duplicated(shortNames)) > 0){
    indexVec <- 1:length(shortNames)
    shortNames[duplicated(shortNames)] <- paste0(shortNames[duplicated(shortNames)],"_", indexVec[duplicated(shortNames)])
  }
  
  names(geneGroupList) <- shortNames
  
}



rowSplitVec <- row.names(heat.dat)   
for (i in 1:length(geneGroupList)){
  rowSplitVec[rowSplitVec %in% geneGroupList[[i]]] <- names(geneGroupList)[i]
}



dfCol <- unique(dfMeta[,c(group1ColorCol, group1NameCol)])
colVec <- dfCol[,group1ColorCol]
names(colVec) <- dfCol[,group1NameCol]

dfCol2 <- unique(dfMeta[,c(group2ColorCol, group2NameCol)])
colVec2 <- dfCol2[,group2ColorCol]
names(colVec2) <- dfCol2[,group2NameCol]
# 
# colList <- list(
#   dfCol, 
#   dfCol2
# )


ht.anno <- ComplexHeatmap::HeatmapAnnotation(
    Age =  dfMeta[,group2NameCol],
    Cluster =  dfMeta[,group1NameCol],
    col = list(
        Age = colVec2,
        Cluster = colVec
    ),
    show_legend = showAllLegends
)


#bulkLogFC <- dfSel[allGenes, "contrast_2_logFC_p27NULL_E185OC_vs_WT_E185OC"]
#xcol <- ifelse(bulkLogFC < 0, "blue", "red")



#row_ha = rowAnnotation(Bulk = anno_barplot(bulkLogFC, gp = gpar(fill=xcol)),
#        show_legend = showAllLegends
#)   

#lrow_ha = rowAnnotation(foo = anno_block(gp = gpar(fill = c("blue", "red")),
#        labels = c("Down", "Up"), 
#        labels_gp = gpar(col = "white", fontsize = 10)),
#        show_legend = showAllLegends
#)
#     df = data.frame(
#         cluster = OsC_HM@meta.data$clusterName),
#     # col = list(
#     #     gp = gpar(col = "white")
#     # ),
#     annotation_legend_param = list(direction = "horizontal")
# )

if (relativeExpression){
  colRamp <- circlize::colorRamp2(c(-3, 0, 3),c("#3060cf", "#fffbbc","#c4463a"))
} else {
  colRamp <- circlize::colorRamp2(c(0, 6),c("#d3d3d3","#3060cf"))
}

if (nrow(heat.dat) > 100){
    rowFont <- 1
} else if (nrow(heat.dat) > 50){
    rowFont <- 1
} else if (nrow(heat.dat) > 25){
    rowFont <- 6
} else {
    rowFont <- 8
}

plotList[[tag]] <- ComplexHeatmap::Heatmap(
    heat.dat,
    row_split               = rowSplitVec,
    name                    = legendString,
    column_title_gp         = grid::gpar(fontsize = 6),
    row_title_rot           = 0, # Must be 0 or 90
    column_title_rot        = 90,
    column_split            = dfMeta[,columnSplitCol],
    #column_split            = column_split,
    cluster_column_slices   = cluster_column_slices,
    cluster_columns         = cluster_columns,
    cluster_rows            = TRUE,
    show_row_names          = TRUE,
    show_column_names       = FALSE,
    column_names_side       = "bottom",
    show_column_dend        = TRUE,
    row_dend_width          = grid::unit(20, "mm"),
    show_heatmap_legend     = showAllLegends,
    column_names_max_height = grid::unit(8, "cm"),
    row_names_gp            = grid::gpar(fontsize = 6),
    top_annotation          = ht.anno,
    #right_annotation = row_ha,
    #left_annotation = lrow_ha,
    #col                     = colorRamp2(c(-2, 0, 2),magma(3)),
    col                     = colRamp,
    column_names_rot        = 90,
    border = TRUE
)

  FNbase <- paste0(tag, VersionPdfExt)

  FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
  FNrel <- paste0("report_figures/", FNbase)
            
  pdf(FN)
    print(plotList[[tag]])
  dev.off()

  figLegend <- paste0(
    '**Figure ', 
    figureCount, 
    ':** ',
    'Heatmap split by cluster for gene set ',gsub("_", " ", tag),'. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>.'
  )
            
  figureCount <- figureCount + 1

  NewChnk <- paste0(
    "#### ", tag,
    "\n```{r hm_",
    tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
    figLegend,"'}\n",
    "\n",
    "\n print(plotList[['",tag,"']])",
    "\n cat(  '\n')",
    "\n\n\n```\n"   
)

  chnkVec <- c(
    chnkVec,
    NewChnk
  )
  
  returnList <- list(
    "plotList" = plotList,
    "chnkVec"  = chnkVec,
    "figureCount" = figureCount
  )

return(returnList)
}

## Done function to create heatmaps
###############################################################################

dfData <- data.frame(OsC[["RNA"]]@data)
dataMatrix <- data.matrix(dfData)
dfMeta <- OsC@meta.data

####################
## Create datamatrix and dfMeta


## Need cluster_stage column
selClust <- "cluster_age"
OsC_HM <- OsC
OsC_HM@meta.data[[selClust]] <- paste0(
  OsC_HM@meta.data$clusterName, 
  "_",
  OsC_HM@meta.data$meta_Age_Group
)



###############################################################################
## Calculataing average expression across all samples                        ##
Seurat::Idents(OsC_HM) <- selClust
OsCsel <- Seurat::AverageExpression(OsC_HM, return.seurat = TRUE)


# 

dfAdd <- unique(OsC_HM@meta.data[,c("clusterName", "clusterColor",  "meta_Age_Group", selClust)])

dfAdd[["ageGroupColor"]] <- dfAdd$meta_Age_Group
dfAdd$ageGroupColor <- gsub("Young", "#E6C229", dfAdd$ageGroupColor)
dfAdd$ageGroupColor <- gsub("Middle", "#ED1D58", dfAdd$ageGroupColor)
dfAdd$ageGroupColor <- gsub("Old", "#48A6EA", dfAdd$ageGroupColor)

row.names(dfAdd) <- dfAdd[,selClust]

OsCsel <- biologicToolsSC::addDf2seuratMetaData(
  obj = OsCsel, 
  dfAdd = dfAdd
)

OsCsel@meta.data$meta_Age_Group <- factor(
  OsCsel@meta.data$meta_Age_Group,
  levels = c("Young", "Middle", "Old")
)

OsCsel@meta.data$clusterName <- factor(
  OsCsel@meta.data$clusterName,
  levels=Obio@parameterList$clusterNameOrder
)


OsCsel@meta.data <- OsCsel@meta.data[order(OsCsel@meta.data$clusterName,OsCsel@meta.data$meta_Age_Group),]

dfData <- data.frame(OsCsel[["RNA"]]@data)
dataMatrix <- data.matrix(dfData)
dfMeta <- OsCsel@meta.data
 
dataMatrix <- dataMatrix[,dfMeta[,selClust]]

#library(ComplexHeatmap)
plotList <- list()

chnkVec <- as.vector(NULL, mode="character")

if (!exists("figureCount")){
  figureCount <- 1  
}


for (i in 1:length(HMgeneSelections)){
  if (length(unlist(HMgeneSelections[[i]])) > 1){
    returnList <- addHeatmap2List(
        dataMatrix = dataMatrix,
        cluster_column_slices   = FALSE,
        cluster_columns         = FALSE,
        geneGroupList = HMgeneSelections[[i]],
        relativeExpression = FALSE,
        cmdVec = NULL,
        tag = paste0(names(HMgeneSelections)[i], "_Abs"),
        showAllLegends = TRUE,
        figureCount = figureCount,
        plotList = plotList,
        maxNcharCatName = 10,
        dfMeta = dfMeta,
        group1ColorCol = "clusterColor", 
        group1NameCol = "clusterName",
        group2ColorCol = "ageGroupColor", 
        group2NameCol = "meta_Age_Group",
        chnkVec = chnkVec,
        columnSplitCol = "clusterName"
        
    )
    
    plotList <- returnList[["plotList"]]
    chnkVec <- returnList[["chnkVec"]]
    figureCount <- returnList[["figureCount"]]
  }
}

# plotList <- addHeatmap2List(
#     OsC = OsC,
#     geneGroupList = geneGroupList,
#     relativeExpression = TRUE,
#     plotList = plotList,
#     cmdVec = NULL,
#     tag = "Channels_Rel",
#     showAllLegends = TRUE
#     )




            
if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}

## To plot all heatmaps in one file
# FNbase <- paste0("Summary.heatmaps", VersionPdfExt)

#FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
#FNrel <- paste0("report_figures/", FNbase)
            
#pdf(FN)
#    print(plotList)
#dev.off()

```


### Heatmaps Absolute {`r tabVar`}
The heatmaps depict the genes shown in red in the above plots. 

```{r plot-heatmaps, echo=T, eval=TRUE, warning=FALSE, results='asis'}
###############################################################################
## Do category enrichment on clusters                                        ##
cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))
## Done doing enrichment on clusters                                         ##
###############################################################################

```

```{r process_heatmap_data_rel, echo=T, eval=TRUE, warning=FALSE, results=F, error=F}

plotList <- list()

chnkVec <- as.vector(NULL, mode="character")

if (!exists("figureCount")){
  figureCount <- 1  
}


for (i in 1:length(HMgeneSelections)){
  
  # tag <- names(HMgeneSelections)[i]
  # tag <- unlist(strsplit(tag, "_LRT_"))[2]
  # tag <- unlist(strsplit(tag,"_Age_Group_pos"))[1]
  # 
  # selCols <- colnames(dataMatrix)[grep(paste0(tag, "_"), colnames(dataMatrix))]
  # dataMatrix <- dataMatrix[,selCols]
  # dataMatrix <- dataMatrix[rowSums(dataMatrix) != 0, ]
  
  if (length(unlist(HMgeneSelections[[i]])) > 1){
    returnList <- addHeatmap2List(
        dataMatrix = dataMatrix,
        cluster_column_slices   = FALSE,
        cluster_columns         = FALSE,
        geneGroupList = HMgeneSelections[[i]],
        relativeExpression = TRUE,
        cmdVec = NULL,
        tag = paste0(names(HMgeneSelections)[i], "_Rel"),
        showAllLegends = TRUE,
        figureCount = figureCount,
        plotList = plotList,
        maxNcharCatName = 10,
        dfMeta = dfMeta,
        group1ColorCol = "clusterColor", 
        group1NameCol = "clusterName",
        group2ColorCol = "ageGroupColor", 
        group2NameCol = "meta_Age_Group",
        chnkVec = chnkVec ,
        columnSplitCol = "clusterName"
        
    )
    
    plotList <- returnList[["plotList"]]
    chnkVec <- returnList[["chnkVec"]]
    figureCount <- returnList[["figureCount"]]
  }
}

# plotList <- addHeatmap2List(
#     OsC = OsC,
#     geneGroupList = geneGroupList,
#     relativeExpression = TRUE,
#     plotList = plotList,
#     cmdVec = NULL,
#     tag = "Channels_Rel",
#     showAllLegends = TRUE
#     )




            
if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}

## To plot all heatmaps in one file
# FNbase <- paste0("Summary.heatmaps", VersionPdfExt)

#FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
#FNrel <- paste0("report_figures/", FNbase)
            
#pdf(FN)
#    print(plotList)
#dev.off()

```


### Heatmaps Relative {`r tabVar`}
The heatmaps depict the genes shown in red in the above plots. 

```{r plot-heatmaps-rel, echo=T, eval=TRUE, warning=FALSE, results='asis'}
###############################################################################
## Do category enrichment on clusters                                        ##
cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))
## Done doing enrichment on clusters                                         ##
###############################################################################

```

```{r documentation_saving, echo=F, eval=TRUE, warning=FALSE, results=F, error = FALSE} 

source("save.biologic.robj.R")

print("Obio Object saved.")

save(OsC,
    file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
        ".Seurat.Robj"
     )
)

###############################################################################
## Rerunning of part C is required                                           ##

###############################################################################
## Create Excel output files                                                 ##

## Add ref-gene column ##
#dfTemp[["ref_gene"]] <- ""
#dfTemp[dfTemp$gene %in% refGenes, "ref_gene"] <- "+"

library(openxlsx)

dfTemp <- Obio@dataTableList[["LRT_table"]] 

if (!dir.exists( paste0(
   Obio@parameterList$html_local,
   "report_tables/"))){
  dir.create(paste0(
   Obio@parameterList$html_local,
   "report_tables/"))
}

baseFN <- paste0(
   Obio@parameterList$project_id, 
   ".LRT.table.xlsx"
)


outPutFN <- paste0(
     Obio@parameterList$html_local,
      "report_tables/",
     baseFN
)
  
 
FNrel <- paste0("report_table/", baseFN)
    
sheetName <- substr(paste0(Obio@parameterList$project_id, "_subcluster_LRT"),1,30)

wb <- createWorkbook()
    addWorksheet(wb, sheetName)
    freezePane(wb, sheetName ,  firstActiveRow = 2)
    
    ## Filter is inactivated, as it does not appear to be compatible with the current version of Excel
    #addFilter(wb, 1, row = 1, cols = 1:ncol(dfOutput))
    
## Style headers ##
hs1 <- createStyle(
  fontColour = "#ffffff",
  fgFill = "#000000", 
  halign = "CENTER", 
  textDecoration = "Bold"
)
    
writeData(wb, 1,dfTemp, startRow = 1, startCol = 1, headerStyle = hs1)
    
saveWorkbook(
  wb, 
  outPutFN , 
  overwrite = TRUE
)

## Done creating Excel output files                                          ##
###############################################################################

###############################################################################
## Create cat enrichment table                                               ##

## Add ref-gene column ##
#dfTemp[["ref_gene"]] <- ""
#dfTemp[dfTemp$gene %in% refGenes, "ref_gene"] <- "+"

library(openxlsx)

## Enrichment result table is called dfResTable

if (!dir.exists( paste0(
   Obio@parameterList$html_local,
   "report_tables/"))){
  dir.create(paste0(
   Obio@parameterList$html_local,
   "report_tables/"))
}

baseFN <- paste0(
   Obio@parameterList$project_id, 
   ".LRT.cat.enrichment.table.xlsx"
)


outPutFN <- paste0(
     Obio@parameterList$html_local,
      "report_tables/",
     baseFN
)
  
 
FNrel <- paste0("report_table/", baseFN)
    
sheetName <- substr(paste0(Obio@parameterList$project_id, "_cat_enrichment_LRT"),1,30)

wb <- openxlsx::createWorkbook()
    openxlsx::addWorksheet(wb, sheetName)
    openxlsx::freezePane(wb, sheetName ,  firstActiveRow = 2)
    
    ## Filter is inactivated, as it does not appear to be compatible with the current version of Excel
    #addFilter(wb, 1, row = 1, cols = 1:ncol(dfOutput))
    
## Style headers ##
hs1 <- openxlsx::createStyle(
  fontColour = "#ffffff",
  fgFill = "#000000", 
  halign = "CENTER", 
  textDecoration = "Bold"
)
    
openxlsx::writeData(wb, 1,dfTemp, startRow = 1, startCol = 1, headerStyle = hs1)
    
saveWorkbook(
  wb, 
  outPutFN , 
  overwrite = TRUE
)

## Done creating Excel output files                                          ##
###############################################################################

```




```{r create_report_params, eval=T, results="asis"}
documentationParams <- list(

    "title" = "Pseudobulk Differential Gene Expression Analysis",
    "subtitle" =  "",
    "abstract" = ""

)

Obio@parameterList$lims.id <- "SC19235"

## Try to retrieve project data from db ##
library(RMySQL)
db.pwd2 <- "_asf_"
db.user2 <- "asf"
host2 <- "ms1.thecrick.org"
projectParams <- documentationParams

tryCatch({
    dbDB = dbConnect(drv = RMySQL::MySQL(), user = db.user2, password = db.pwd2, host = host2, dbname = "asf");
dfProposal =  dbGetQuery(dbDB, paste0("SELECT * FROM asf_proposals WHERE project_name ='",Obio@parameterList$lims.id,"'"));
dbDisconnect(dbDB)
  }, error = function(x) {
    message("Project Database could not be reached or has no entry in Obio@parameterList$lims.id for this analysis.")
   
})

if (exists("dfProposal")){
  if (nrow(dfProposal) == 1){
      if (!is.na(dfProposal[1,"ProjectAlias"]) & dfProposal[1,"ProjectAlias"] != ""){
          projectParams[["title"]] = paste0(dfProposal[1,"ProjectAlias"], " - ", dfProposal[1,"project_name"])
      }
      
      if (!is.na(dfProposal[1,"project_user"]) & dfProposal[1,"project_user"] != ""){
          projectParams[["subtitle"]] = paste0(dfProposal[1,"user_lab"], " Lab - ", dfProposal[1,"project_user"])
          projectParams[["subtitle"]] <- gsub("^ Lab - ", "", projectParams[["subtitle"]])
          
      }
      
      if (!is.na(dfProposal[1,"proposal_text"]) & dfProposal[1,"proposal_text"] != ""){
          projectParams[["abstract"]] = dfProposal[1,"proposal_text"]
         
          
      }
  }
}
   
## Escape all special characters
projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("([.|()/\\^{}+$*?]|\\[|\\])", " ", x)
) 

projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("\\\n", " ", x)
) 


#projectParams$title <- "Title"
# projectParams$abstract <- "This is the QC section."
#projectParams$subtitle <- "Abstract"

```



## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
print(paste0("Analysisdirectory: ", getwd()))

biologic_active_object_dir <- paste0(
    Obio@parameterList$folder,
    "data/biologic_active_object/"
)

bFN <- paste0(
    biologic_active_object_dir,
    Obio@parameterList$project_id,
    ".bioLOGIC.Robj"
)
print(paste0("bioLOGIC data object: ", bFN))

sessionInfo()
```

---
title: "`r projectParams$title`"
subtitle:  Pseudobulk Differential Gene Expression Analyses on Single-cell Results
author:
    - Bioinformatics: Stefan Boeing^[The Francis Crick Institute, stefan.boeing@crick.ac.uk]
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

abstract: |
    In this section pseudo-bulk differential gene expression analyses for this project are presented. The method used here was developed recently by the Huber lab and is desribed in detail [here](https://www.biorxiv.org/content/10.1101/2020.08.13.249623v2).

---
