---
output: 
    html_document:
        highlight: default
        theme: paper
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 3
        toc_float: true
        self_contained: yes
        css: src/style/style.css
---
    
    

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```

<!-- Set PArameters Module -->
<!-- Set the chnkPrefix to make all chunks unique in the overall folder -->
```{r partB_doublet_estimates, echo=TRUE, eval=TRUE, warning=FALSE, include=F}
chnkPrefix <- "scType."
VersionPdfExt <- VersionPdfExt <- paste0(".",chnkPrefix,"V", gsub("-", "", Sys.Date()), ".pdf")

```

```{r, include=FALSE}


## Recommended (Crick) R-setup: 
# module purge;source /nemo/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a

```

```{r, eval=TRUE, echo=F, results=F}
###############################################################################
##                                                                           ##
if (!requireNamespace("remotes")){
  install.packages("remotes")
}


if (!requireNamespace("renv")){
  remotes::install_github("rstudio/renv")
}

if (!file.exists("renv.lock")){
    renv::init()
} else {
    renv::restore(prompt = FALSE)
}

## Done                                                                      ##
###############################################################################


```

```{r, eval=TRUE, echo=T, results=F}
###############################################################################
##  Load packages                                                            ##
library(Seurat)
library(biologicSeqTools2)
library(dplyr)

VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")
rasterDpi <- 400
calculateLRTfromScratch <- TRUE

## Load bioLOGIC object
source("load.biologic.robj.R")

## Make sure paths are set correctly in the Obio object
Obio <- Obio %>% 
    biologicSeqTools2::setMountingPoint()  %>% 
    biologicSeqTools2::setAnalysisPaths() %>% 
    biologicSeqTools2::setCrickGenomeAndGeneNameTable() %>% 
    biologicSeqTools2::createAnalysisFolders() %>% 
    biologicSeqTools2::setDataBaseParameters()

## Load Seurat object ##
SeuratFN <- paste0(
    Obio@parameterList$localWorkDir,
    Obio@parameterList$project_id,
    ".Seurat.Robj"
)

load(SeuratFN)

##                                                                           ##
###############################################################################

###############################################################################
## Retrieve database password                                                ##
if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")
    dbTable <- read.delim(
      FN,
      header = F,
      sep = "\t",
      stringsAsFactors = F
    )
#}
db.pwd <- as.vector(dbTable[1,1])

## Done                                                                      ##
###############################################################################

###############################################################################
##  Set variables                                                            ##

# Plotting parameters
figureCount <- 1
legendDotSize <- 5

## Define project ID ##
project_id <- Obio@parameterList$project_id


## Set directory for report tables
reportTableDir <- Obio@parameterList$reportTableDir
reportFigDir <- Obio@parameterList$reportFigDir

localWorkDir <- Obio@parameterList$localWorkDir

## Create url string
if (Obio@parameterList$host == "10.27.241.234"){
    urlString <- "biologic.thecrick.org"
} else {
    urlString <- "biologic.crick.ac.uk"
}

shinyBaseUrl <- "https://bioinformatics.crick.ac.uk/shiny/users/boeings/"

shinyURL <- paste0(
    shinyBaseUrl,
    project_id,
    "_app/"
)            
   
## Set file paths ##
baseFN <- paste0(
   project_id, 
   ".LRT.table.xlsx"
)

FNrel <- paste0("report_tables/", baseFN)

outPutFN <- paste0(
     reportTableDir,
     baseFN
)

tableLink <- paste0(
    'https://',
    urlString,
    '/mdata/',project_id, '/html/', 
    FNrel
)

tableString <- paste0(
    'An Excel table with the LRT results can be downloaded ',
    tableLink
)

tableString <- paste0(
   ' An Excel table with the LRT results can be downloaded <a href="',tableLink,'" target="_blank">here</a>. '
)
  

##  Set variables                                                            ##
###############################################################################




 
plotList <- list() 

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
```

```{r scType-1, echo=T, eval=TRUE, warning=FALSE, results='asis'}

#renv::install("HGNChelper")
lapply(c("dplyr","Seurat","HGNChelper","openxlsx"), library, character.only = T)
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R"); source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R")

# get cell-type-specific gene sets from our in-built database (DB)
gs_list = gene_sets_prepare("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_short.xlsx", "Immune system") # e.g. Immune system, Liver, Pancreas, Kidney, Eye, Brain

# assign cell types
scRNAseqData = readRDS(gzcon(url('https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/exampleData.RDS'))); #load example scRNA-seq matrix
es.max = sctype_score(scRNAseqData = scRNAseqData, scaled = TRUE, gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)

# View results, cell-type by cell matrix. See the complete example below
# View(es.max)

# load gene set preparation function
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R")
# load cell type annotation function
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R")

# DB file
db_ = "https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_full.xlsx";
tissue = "Immune system" # e.g. Immune system,Pancreas,Liver,Eye,Kidney,Brain,Lung,Adrenal,Heart,Intestine,Muscle,Placenta,Spleen,Stomach,Thymus 

# prepare gene sets
gs_list = gene_sets_prepare(db_, tissue)

# get cell-type by cell matrix
es.max = sctype_score(scRNAseqData = OsC[["RNA"]]@scale.data, scaled = TRUE, 
                      gs = gs_list$gs_positive, gs2 = gs_list$gs_negative) 

# NOTE: scRNAseqData parameter should correspond to your input scRNA-seq matrix. 
# In case Seurat is used, it is either OsC[["RNA"]]@scale.data (default), OsC[["SCT"]]@scale.data, in case sctransform is used for normalization,
# or OsC[["integrated"]]@scale.data, in case a joint analysis of multiple single-cell datasets is performed.

# merge by cluster
cL_resutls = do.call("rbind", lapply(unique(OsC@meta.data$seurat_clusters), function(cl){
    es.max.cl = sort(rowSums(es.max[ ,rownames(OsC@meta.data[OsC@meta.data$seurat_clusters==cl, ])]), decreasing = !0)
    head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(OsC@meta.data$seurat_clusters==cl)), 10)
}))
sctype_scores = cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  

# set low-confident (low ScType score) clusters to "unknown"
sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] = "Unknown"
print(sctype_scores[,1:3])

OsC@meta.data$customclassif = ""
for(j in unique(sctype_scores$cluster)){
  cl_type = sctype_scores[sctype_scores$cluster==j,]; 
  OsC@meta.data$customclassif[OsC@meta.data$seurat_clusters == j] = as.character(cl_type$type[1])
}





# Seurat::DimPlot(
#   OsC, reduction = "umap", 
#   label = TRUE, 
#   repel = TRUE, 
#   group.by = 'customclassif'
# )  


# load libraries
lapply(c("ggraph","igraph","tidyverse", "data.tree"), library, character.only = T)

# prepare edges
cL_resutls=cL_resutls[order(cL_resutls$cluster),]; edges = cL_resutls; edges$type = paste0(edges$type,"_",edges$cluster); edges$cluster = paste0("cluster ", edges$cluster); edges = edges[,c("cluster", "type")]; colnames(edges) = c("from", "to"); rownames(edges) <- NULL

# prepare nodes
nodes_lvl1 = sctype_scores[,c("cluster", "ncells")]; nodes_lvl1$cluster = paste0("cluster ", nodes_lvl1$cluster); nodes_lvl1$Colour = "#f1f1ef"; nodes_lvl1$ord = 1; nodes_lvl1$realname = nodes_lvl1$cluster; nodes_lvl1 = as.data.frame(nodes_lvl1); nodes_lvl2 = c(); 
ccolss= c("#5f75ae","#92bbb8","#64a841","#e5486e","#de8e06","#eccf5a","#b5aa0f","#e4b680","#7ba39d","#b15928","#ffff99", "#6a3d9a","#cab2d6","#ff7f00","#fdbf6f","#e31a1c","#fb9a99","#33a02c","#b2df8a","#1f78b4","#a6cee3")
for (i in 1:length(unique(cL_resutls$cluster))){
  dt_tmp = cL_resutls[cL_resutls$cluster == unique(cL_resutls$cluster)[i], ]; nodes_lvl2 = rbind(nodes_lvl2, data.frame(cluster = paste0(dt_tmp$type,"_",dt_tmp$cluster), ncells = dt_tmp$scores, Colour = ccolss[i], ord = 2, realname = dt_tmp$type))
}
nodes = rbind(nodes_lvl1, nodes_lvl2); nodes$ncells[nodes$ncells<1] = 1;
files_db = openxlsx::read.xlsx(db_)[,c("cellName","shortName")]; files_db = unique(files_db); nodes = merge(nodes, files_db, all.x = T, all.y = F, by.x = "realname", by.y = "cellName", sort = F)
nodes$shortName[is.na(nodes$shortName)] = nodes$realname[is.na(nodes$shortName)]; nodes = nodes[,c("cluster", "ncells", "Colour", "ord", "shortName", "realname")]

mygraph <- graph_from_data_frame(edges, vertices=nodes)

# Make the graph
gggr <- ggraph(mygraph, layout = 'circlepack', weight=I(ncells)) + 
  geom_node_circle(aes(filter=ord==1,fill=I("#F5F5F5"), colour=I("#D3D3D3")), alpha=0.9) + geom_node_circle(aes(filter=ord==2,fill=I(Colour), colour=I("#D3D3D3")), alpha=0.9) +
  theme_void() + geom_node_text(aes(filter=ord==2, label=shortName, colour=I("#ffffff"), fill="white", repel = !1, parse = T, size = I(log(ncells,25)*1.5)))+ geom_node_label(aes(filter=ord==1,  label=shortName, colour=I("#000000"), size = I(3), fill="white", parse = T), repel = !0, segment.linetype="dotted")
  
#scater::multiplot(DimPlot(OsC, reduction = "umap", label = TRUE, repel = TRUE, cols = ccolss), gggr, cols = 2)



plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

tag <- "Dimplot"
plotList[[tag]] <- Seurat::DimPlot(
  OsC, reduction = "umap", 
  label = TRUE, 
  repel = TRUE, 
  group.by = 'customclassif'
)  

## Save to file ##
FNbase <- paste0("scType", VersionPdfExt)
FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
FNrel <- paste0("report_figures/", FNbase)


pdf(FN)
print(plotList[[tag]])
dev.off()



## Create R markdown chunk ##
figLegend <- paste0(
    '**Figure ', 
    figureCount, 
    '**: Celltype assignment UMAP. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. '
)

figureCount <- figureCount + 1 

NewChnk <- paste0(
    "#### Automatic celltype assignment for all clusters", 
    "\n```{r ", tag, ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
    figLegend,"'}\n",
    "\n",
    "\n print(plotList[['",tag,"']])",
    "\n cat(  '\n')",
    "\n\n\n```\n"   
)

chnkVec <- c(
    chnkVec,
    NewChnk
)


tag <- "Bubblechart"
plotList[[tag]] <- gggr

## Save to file ##
FNbase <- paste0("scType.bubblechart", VersionPdfExt)
FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
FNrel <- paste0("report_figures/", FNbase)


pdf(FN)
print(plotList[[tag]])
dev.off()



## Create R markdown chunk ##
figLegend <- paste0(
    '**Figure ', 
    figureCount, 
    '**: Bubblechart. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. '
)

figureCount <- figureCount + 1 

NewChnk <- paste0(
    "#### Bubblechart", 
    "\n```{r ", tag, ", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
    figLegend,"'}\n",
    "\n",
    "\n print(plotList[['",tag,"']])",
    "\n cat(  '\n')",
    "\n\n\n```\n"   
)

chnkVec <- c(
    chnkVec,
    NewChnk
)

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}

```

### Cell Type Assignments {`r tabVar`} 
```{r create-cat-enrichment-plot, echo=T, eval=TRUE, warning=FALSE, results='asis'}
###############################################################################
## Do category enrichment on clusters                                        ##
cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))
## Done doing enrichment on clusters                                         ##
###############################################################################


```





```{r create_report_params, eval=T, results="asis"}
documentationParams <- list(

    "title" = "Celltype assignment with scType",
    "subtitle" =  "",
    "abstract" = ""

)


## Try to retrieve project data from db ##
library(RMySQL)
db.pwd2 <- "_asf_"
db.user2 <- "asf"
host2 <- "ms1.thecrick.org"
projectParams <- documentationParams

tryCatch({
    dbDB = dbConnect(drv = RMySQL::MySQL(), user = db.user2, password = db.pwd2, host = host2, dbname = "asf");
dfProposal =  dbGetQuery(dbDB, paste0("SELECT * FROM asf_proposals WHERE project_name ='",Obio@parameterList$lims.id,"'"));
dbDisconnect(dbDB)
  }, error = function(x) {
    message("Project Database could not be reached or has no entry in Obio@parameterList$lims.id for this analysis.")
   
})

if (exists("dfProposal")){
  if (nrow(dfProposal) == 1){
      if (!is.na(dfProposal[1,"ProjectAlias"]) & dfProposal[1,"ProjectAlias"] != ""){
          projectParams[["title"]] = paste0(dfProposal[1,"ProjectAlias"], " - ", dfProposal[1,"project_name"])
      }
      
      if (!is.na(dfProposal[1,"project_user"]) & dfProposal[1,"project_user"] != ""){
          projectParams[["subtitle"]] = paste0(dfProposal[1,"user_lab"], " Lab - ", dfProposal[1,"project_user"])
          projectParams[["subtitle"]] <- gsub("^ Lab - ", "", projectParams[["subtitle"]])
          
      }
      
      if (!is.na(dfProposal[1,"proposal_text"]) & dfProposal[1,"proposal_text"] != ""){
          projectParams[["abstract"]] = dfProposal[1,"proposal_text"]
         
          
      }
  }
}
   
## Escape all special characters
projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("([.|()/\\^{}+$*?]|\\[|\\])", " ", x)
) 

projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("\\\n", " ", x)
) 


#projectParams$title <- "Title"
# projectParams$abstract <- "This is the QC section."
#projectParams$subtitle <- "Abstract"

```



## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
print(paste0("Analysisdirectory: ", getwd()))

biologic_active_object_dir <- paste0(
    Obio@parameterList$folder,
    "data/biologic_active_object/"
)

bFN <- paste0(
    biologic_active_object_dir,
    Obio@parameterList$project_id,
    ".bioLOGIC.Robj"
)
print(paste0("bioLOGIC data object: ", bFN))

sessionInfo()
```

---
title: "`r projectParams$title`"
subtitle:  Celltype Assignment using the scType tool
author:
    - Bioinformatics: Stefan Boeing^[The Francis Crick Institute, stefan.boeing@crick.ac.uk]
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

abstract: |
   In this section cell type inferences are made using the scType package. More details on that [here](https://www.nature.com/articles/s41467-022-28803-w).
---
