---

output: 
    html_document:
        highlight: default
        theme: paper
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 3
        toc_float: true
        css: src/style/style.css

---
    
    

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```


```{r hpc_notes, include=FALSE}
# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a

###############################################################################
##                                                                           ##
if (!requireNamespace("remotes")){
  install.packages("remotes")
}


if (!requireNamespace("renv")){
  remotes::install_github("rstudio/renv")
}

if (!file.exists("renv.lock")){
    renv::init()
} else {
    renv::restore(prompt = FALSE)
}

## Done                                                                      ##
###############################################################################

```


## Load Required Datasets
```{r , eval=TRUE, echo=F, results=F}
## Add required packages if needed

# Sys.unsetenv("GITHUB_PAT")
# renv::install("immunogenomics/harmony")
# renv::install("powellgenomicslab/scPred")


## Load required packages
library("scPred")
library("Seurat")
library("magrittr")


## Example
# reference <- scPred::pbmc_1
# query <- scPred::pbmc_2


## Load reference dataset
refFN <- "/camp/stp/babs/working/boeings/Projects/bonfantip/roberta.ragazzini/446_scRNAseq_subClustering_pbl359B_12clusters/workdir/pbl359BsubCluster.Seurat.Robj"
load(refFN)
reference <- OsC
DefaultAssay(reference) <- "integrated"

## Query dataset: New patient 1 + 2 dataset, similar to the above
queryFN <- "/camp/stp/babs/working/boeings/Projects/bonfantip/roberta.ragazzini/476_scRNAseq_A1A6_SC21201/workdir/A123456.Seurat.Robj"
load(queryFN)
testQuery <- OsC
DefaultAssay(testQuery) <- "integrated"

## Load in vitro query dataset
queryFN <- "/camp/stp/babs/working/boeings/Projects/bonfantip/roberta.ragazzini/484B_SC22096subset_scRNAseq_Characterisation_EpCAM_CD90_thymic_epithelial_cells_SC22096/workdir/SC22096subset.Seurat.Robj"

load(queryFN)
query <- OsC
DefaultAssay(query) <- "integrated"


rm(OsC)

```



```{r, echo=F, eval=T, warning=FALSE, results="asis"}
## Create plotting list ##
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")


tag <- "Original_in_vivo_experiment"
plotList[[tag]] <- DimPlot(reference, group.by = "clusterName", label = TRUE, repel = TRUE)

figLegend <- ""

NewChnk <- paste0(
    "### ", tag,
    "\n```{r ",
    tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
    figLegend,"'}\n",
    "\n",
    "\n print(plotList[['",tag,"']])",
    "\n cat(  '\n')",
    "\n\n\n```\n"   
)
            
chnkVec <- c(
    chnkVec,
    NewChnk
)

tag <- "Second_in_vivo_experiment"
plotList[[tag]] <- DimPlot(testQuery, group.by = "clusterName", label = TRUE, repel = TRUE)

figLegend <- ""

NewChnk <- paste0(
    "### ", tag,
    "\n```{r ",
    tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
    figLegend,"'}\n",
    "\n",
    "\n print(plotList[['",tag,"']])",
    "\n cat(  '\n')",
    "\n\n\n```\n"   
)
            
chnkVec <- c(
    chnkVec,
    NewChnk
)

tag <- "New_in_vitro_dataset"
plotList[[tag]] <- DimPlot(query, group.by = "clusterName", label = TRUE, repel = TRUE)

figLegend <- ""

NewChnk <- paste0(
    "### ", tag,
    "\n```{r ",
    tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
    figLegend,"'}\n",
    "\n",
    "\n print(plotList[['",tag,"']])",
    "\n cat(  '\n')",
    "\n\n\n```\n"   
)
            
chnkVec <- c(
    chnkVec,
    NewChnk
)

if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
``` 

## Explore Datasets {`r tabVar`}


```{r, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"} 

cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))

``` 

## First Test: Predict Second In Vivo Experiment Using First
```{r, echo=T, eval=T, warning=FALSE}

# , results="asis"

reference <- getFeatureSpace(reference, "clusterName")

reference <- trainModel(reference)

get_probabilities(reference) %>% head()

get_scpred(reference)

plot_probabilities(reference)



get_classifiers(reference)

# Optional
# reference <- trainModel(reference, model = "mda", reclassify = c("cMono", "ncMono"))

DimPlot(reference, group.by = "clusterName", label = TRUE, repel = TRUE)
DimPlot(query, group.by = "clusterName", label = TRUE, repel = TRUE)

testQuery <- scPredict(testQuery, reference)

crossTab(testQuery, "clusterName", "scpred_prediction")

crossTab(testQuery, "clusterName", "scpred_prediction", output = "prop")

DimPlot(testQuery, group.by = "scpred_prediction", reduction = "umap")
DimPlot(testQuery, group.by = "clusterName", reduction = "umap")


```

## Second Test: Predict Second In Vitro Experiment
```{r, echo=T, eval=T, warning=FALSE}

#, results="asis"

reference <- getFeatureSpace(reference, "clusterName")

reference <- trainModel(reference)

get_probabilities(reference) %>% head()

get_scpred(reference)

plot_probabilities(reference)



get_classifiers(reference)

# Optional
# reference <- trainModel(reference, model = "mda", reclassify = c("cMono", "ncMono"))

DimPlot(reference, group.by = "clusterName", label = TRUE, repel = TRUE)
DimPlot(query, group.by = "clusterName", label = TRUE, repel = TRUE)

query <- scPredict(query, reference)

crossTab(query, "clusterName", "scpred_prediction")

crossTab(query, "clusterName", "scpred_prediction", output = "prop")

DimPlot(query, group.by = "scpred_prediction", reduction = "umap")
DimPlot(query, group.by = "clusterName", reduction = "umap")


```

## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
print(paste0("Analysisdirectory: ", getwd()))

sessionInfo()
```

---
title: scPred Classifier Tests
subtitle:  Cell type assignment using classifiers
author:
    - Bioinformatics: Stefan Boeing^[The Francis Crick Institute, stefan.boeing@crick.ac.uk]
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

abstract: |
    In this section we explore the scPred R-package. It is desribed in detail [here](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1862-5).

---


