---
output: 
    html_document:
        code_folding: hide
        df_print: tibble
        highlight: default
        theme: paper
        toc: true
        toc_depth: 5
        toc_float: true
        css: src/style/style.css

always_allow_html: yes
---


```{r setup, include=FALSE}
###############################################################################
## Load conda environment                                                    ##

# To get started, we set up the environment for monocle3 - which is 
# different from the analysis used in the previous analysis. 

# The conda environment requires to source the .bashrc when run on the cluster. 

## When running on the cluster, it's necessary to source the .bashrc to 
## initiate the conda environment
# !! Contents within this block are managed by 'conda init' !!
# __conda_setup="$('/camp/apps/eb/software/Anaconda2/2019.03/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
# if [ $? -eq 0 ]; then
#     eval "$__conda_setup"
# else
#     if [ -f "/camp/apps/eb/software/Anaconda2/2019.03/etc/profile.d/conda.sh" ]; then
#         . "/camp/apps/eb/software/Anaconda2/2019.03/etc/profile.d/conda.sh"
#     else
#         export PATH="/camp/apps/eb/software/Anaconda2/2019.03/bin:$PATH"
#     fi
# fi
# unset __conda_setup

# The critical bit is to do an conda initiate

# Make sure that no renv.lock file is present and make sure that 
# the renv session is deactivated. 

# module purge
# conda deactivate
# ml cairo/1.16.0-GCCcore-10.3.0;ml rgdal/1.5-16-foss-2020a-R-4.0.0
# source .bashrc
# conda activate R-monocle3;R

# Check for presence of monocle3
# library(monocle3)

# renv::snapshot() - reinitialise with a new library
# Activate renv session

# check again for presence of monocle
# library(monocle3)

# R version 4.0.2

# in the data analysis = false mode, the following R is used:
# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a;R

###############################################################################
## alternative 20220630                                                      ##
## start by deactivating renv renv::deactivate()
## Quit R
# conda deactivate
# module purge
# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a;
# ml cairo/1.16.0-GCCcore-10.3.0;ml rgdal/1.5-16-foss-2020a-R-4.0.0
# source ~/.bashrc
# conda activate R-monocle3;R
## Check - can monocle3 still be loaded? library(monocle3) - not present
## renv::init() - option 2
## Check - can monocle3 still be loaded? library(monocle3) 

##
###############################################################################

###############################################################################
## Note: 2022 06 30                                                          ##

## For the time being, run monocle3 ALWAYS without renv. 

## Make sure to do renv::deactivate before running monocle3 Rmds.

## Done                                                                      ##
###############################################################################

###############################################################################
## Running monocle3 envs                                                     ##
## start by deactivating renv renv::deactivate()
## Quit R
# conda deactivate
# module purge
# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a;
# ml cairo/1.16.0-GCCcore-10.3.0;ml rgdal/1.5-16-foss-2020a-R-4.0.0
# source ~/.bashrc
# conda activate R-monocle3;R
## Check - can monocle3 still be loaded? library(monocle3) 

## Run on the cluster:
## Critical: source .bashrc
##
# sbatch --time=02:00:00 --wrap "module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;ml cairo/1.16.0-GCCcore-10.3.0;ml pandoc/2.2.3.2-foss-2016b;ml Anaconda3/2019.07;source ~/.bashrc;conda activate R-monocle3;Rscript runRmd.r sc_Trajectory_Analysis_Main_Template.Rmd " --job-name=$project  --mem=200G -o ../rT.$project.slurm >> commands.txt

##
###############################################################################


knitr::opts_chunk$set(
    tidy = F,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)


if (!requireNamespace("remotes")){
  install.packages("remotes")
}


if (!requireNamespace("renv")){
  remotes::install_github("rstudio/renv")
}

if (!file.exists("renv.lock")){
    #renv::init()
    # Use renv snapshot to capture the R libraries already installed in the Conda 
    # environment
  
    # !!!Don't activate before snapshot!!!
    renv::snapshot()
} else {
    renv::restore(prompt = FALSE)
}

## If data analysis mode is used, no renv is used in order to prevent 
## interference with bespoke packages and environments
## In the data analysis mode is false, we expect the monocle3 outputs to be pre-generated.

data.analysis.mode <- TRUE



#Create the environment and load a suitable version of R, e.g. so:
# if (!require("remotes")){
#   install.packages("remotes")
# }

#remotes::install_github("rstudio/renv")
# if (!data.analysis.mode){
#     if (!file.exists("renv.lock")){
#       renv::init(prompt=F)
#     } else {
#       renv::restore()
#     }
# }


# renv::update("vctrs")


#devtools::install_github("decusInLabore/biologicSeqTools2")
# skip updates - just enter on prompt


#devtools::install_github("decusInLabore/biologicToolsSC")

# renv::install("bioc::ComplexHeatmap")
# renv::install("decusInLabore/biologicSeqTools2")
# renv::install("decusInLabore/biologicToolsSC")
# renv::install("gam")
# renv::install('circlize')




source("trajectory.helper.functions.R")         

#remotes::install_github("decusInLabore/biologicSeqTools@main")

library(Seurat)
library(biologicSeqTools2)
library(SingleCellExperiment)
library(dplyr)
library(ggplot2)
#library(ouija)
# library(biologicToolsSC)

# library(tidyverse)
# library(dplyr)
# library(Seurat)
# library(ggplot2)
# library(tidyr)
# library(knitr)

cwd <- paste0(here::here(),"/")
tempWorkDir <- paste0(cwd, "../")

projectDir <- gsub("scripts/scRNAseq/analyses/Monocle_Trajectory_Analysis", "", getwd())
workdir <- paste0(projectDir, "workdir/")


#library(biologicToolsSC)

upload.results.to.database <- TRUE
save.chunk.intermediates <- TRUE


###############################################################################
##                                                                           ##
figureCount <- 1

## Load R module load R/3.5.1-foss-2018b ##
#setwd(Obio@parameterList$localWorkDir)




##                                                                           ##
###############################################################################

###############################################################################
## Set db access credentials                                                 ##
if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

## Loading the BABS password ##
FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")
dbTable <- read.delim(
  FN,
  header = F,
  sep = "\t",
  stringsAsFactors = F
)

db.pwd <- as.vector(dbTable[1,1])


## Done                                                                      ##
###############################################################################


```


<!-- Running Individual Modules in R -->
<!-- For example rmarkdown::render("src/modules/settings/partB.set.parameters.Rmd", output_dir = "..") -->


<!-- Essential 1: Load data (output required later) -->
```{r child = '01-load.data.Rmd', eval=TRUE}

```

<!-- Optional 2: Retrieve Reference Genes from Database or Gmt file -->
```{r child = '02-retrieve.reference.gene.sets.Rmd', eval=TRUE}
```


<!-- Optional 5: monocle3 -->
```{r child = '03-monocle.trajectory.module.overview.Rmd', eval=TRUE}
```

<!-- Optional 5: monocle3 Branch 1 Subtrajectory -->
<!-- Add as many sub-trajectory analyses as necessary as banch.1, branch.2, branch.3... -->
```{r child = '04-monocle.trajectory.module.branch.1.Rmd', eval=TRUE}
```



```{r saveobject, eval=TRUE, echo=T, results=F}
### Will save Obio object here, so it can be re-used with different parameters
save(Obio, 
     file = paste0(
         Obio@parameterList$localWorkDir,
         Obio@parameterList$project_id,
         ".bioLOGIC.Robj"
     )
)

print("Obio Object saved.")

## OsC saving occurs within the modules already
# save(OsC,
#     file = paste0(
#          Obio@parameterList$localWorkDir,
#          Obio@parameterList$project_id,
#         ".Seurat.Robj"
#      )
# )



```



```{r create_report_params, eval=T, results="asis"}

## Try to retrieve project data from db ##
library(RMySQL)
db.pwd2 <- "zU3ufd9L"
db.user2 <- "reader"
host2 <- "clvd1-db-u-p-17.thecrick.org"
projectParams <- Obio@documentationParams

tryCatch({
    dbDB = DBI::dbConnect(
        drv = RMySQL::MySQL(), 
        user = db.user2, 
        password = db.pwd2, 
        host = host2, 
        dbname = "clarity_shadow"
    )
dfProposal =  DBI::dbGetQuery(dbDB, paste0("SELECT * FROM clarify_asf_proposals WHERE project_name ='",Obio@parameterList$lims.id,"'"))
dbDisconnect(dbDB)
  }, error = function(x) {
    message("Project Database could not be reached or has no entry in Obio@parameterList$lims.id for this analysis.")
   
})

if (exists("dfProposal")){
  if (nrow(dfProposal) == 1){
      if (!is.na(dfProposal[1,"ProjectAlias"]) & dfProposal[1,"ProjectAlias"] != ""){
          projectParams[["title"]] = paste0(dfProposal[1,"ProjectAlias"], " - ", dfProposal[1,"project_name"])
      }
      
      if (!is.na(dfProposal[1,"project_user"]) & dfProposal[1,"project_user"] != ""){
          projectParams[["subtitle"]] = paste0(dfProposal[1,"user_lab"], " Lab - ", dfProposal[1,"project_user"])
          projectParams[["subtitle"]] <- gsub("^ Lab - ", "", projectParams[["subtitle"]])
          
      }
      
      if (!is.na(dfProposal[1,"proposal_text"]) & dfProposal[1,"proposal_text"] != ""){
          projectParams[["abstract"]] = dfProposal[1,"proposal_text"]
         
          
      }
  }
}
   
## Escape all special characters
projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("([.|()/\\^{}+$*?]|\\[|\\])", " ", x)
) 

projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("\\\n", " ", x)
) 


#projectParams$title <- "Title"
# projectParams$abstract <- "This is the QC section."
#projectParams$subtitle <- "Abstract"

```



## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
print(paste0("Analysisdirectory: ", getwd()))

biologic_active_object_dir <- paste0(
    Obio@parameterList$folder,
    "data/biologic_active_object/"
)

bFN <- paste0(
    biologic_active_object_dir,
    Obio@parameterList$project_id,
    ".bioLOGIC.Robj"
)
print(paste0("bioLOGIC data object: ", bFN))

sessionInfo()
```

---
title: "`r projectParams$title`"
subtitle:  "Trajectory Analysis Module"
author:
    - Bioinformatics: Stefan Boeing^[The Francis Crick Institute, stefan.boeing@crick.ac.uk]
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

abstract: |
    "This is the trajectory analysis module"


---