---
title: "LINCS Drug Pertubation Analysis"
author: "Stefan Boeing stefan.boeing@crick.ac.uk"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output: 
    html_document:
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 5
        toc_float: true
        css:

always_allow_html: yes

---

```{css setup_css, echo=FALSE}


.table{
  width:auto;
  font-size: 10px;
}

```

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)
```


```{r methods, eval=T}

```

```{r set_directories, eval=T}
## Setup plot collection object
library(knitr)
library(ggplot2)
library(ggpubr)
library(DT)


if (length(.libPaths()) > 2){
    .libPaths(.libPaths()[2:3])
}

addCorCatsToLabDb <- FALSE
figureCount <- 1
chnkVec <- as.vector(NULL, mode = "character")

VersionPdfExt <- paste0(".V", gsub("-", "", Sys.Date()), ".pdf")

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

source("assets/SBwebtools.pckg.r")


if (length(.libPaths()) > 2){
    .libPaths(.libPaths()[2:3])
}


## Heatmap setup ##
## Select Heatmap samples ##
FN <- paste0(hpc.mount, "Projects/reference_data/documentation/BC.parameters.txt")
dbTable <- read.delim(
    FN, 
    sep = "\t",
    stringsAsFactors = F
)

db.pwd <- as.vector(dbTable[1,1])
#setwd(Obio@parameterList$localWorkDir)

ObioFN <- paste0("../",list.files("..")[grep(".bioLOGIC.Robj", list.files(".."))])

load(ObioFN)

# checkFile = paste0(
#          Obio@parameterList$project_id,
#          ".bioLOGIC.Robj"
# )
# 
# if (ObioFN != checkFile){
#     exit()
# }


Obio <- setMountingPoint(Obio)
Obio <- setAnalysisPaths(Obio)
Obio <- setCrickGenomeAndGeneNameTable(Obio)
Obio <- createAnalysisFolders(
    Obio,
    baseDir="/camp/stp/babs/working/boeings/Projects/",
    localBaseDir = paste0(hpc.mount, "Projects/")
)
Obio <- setDataBaseParameters(Obio)


Obio@parameterList[["reportFigDir"]] <- paste0(Obio@parameterList$localWorkDir,Obio@parameterList$project_id, "/report_figures/")

## Create outputfolders ##
if (!dir.exists(paste0(Obio@parameterList$localWorkDir,Obio@parameterList$project_id))){
    dir.create(paste0(Obio@parameterList$localWorkDir,Obio@parameterList$project_id))
}

if (!dir.exists(Obio@parameterList$reportFigDir)){
    dir.create(Obio@parameterList$reportFigDir)
}

figureCount <- 1

## Load R module load R/3.5.1-foss-2018b ##
#setwd(Obio@parameterList$localWorkDir)

###########################
## Set default 
Obio@parameterList$HmDisplayCatsFromDb <- list(
  "Transcription Factors" = "ag_lab_categories__10"
  )

```

## Do LINCS GSEA enrichment
The correlation heatmap is based on the 500 most time-variable genes in this experiment. 

First, as KS test
```{r template, eval=TRUE, collapse=TRUE, echo=TRUE, warning=FALSE}

```

```{r GSEA_enrichment, echo=F, eval=TRUE, warning=FALSE, result="asis", include = FALSE}

###############################################################################
## Load data                                                                 ##



setGeneric(
    name="doLINCSenrichment",
    def=function(
      lgFCvec = NULL,
      topN = 5,
      dfData = "dfData",
      password = "db.pwd",
      db.user = "Obio@dbDetailList$db.user",
      host = "Obio@dbDetailList$host",
      project_id = "Obio@parameterList$project_id"
    ){


if (is.null(lgFCvec)){
lgFCvec <- names(dfData)[grep("_logFC_", names(dfData))]
lgFCvec <- lgFCvec[grep("contrast_", lgFCvec)]
rmVec <- c(
  grep("contrast_L", lgFCvec),
  grep("contrast_P", lgFCvec),
  grep("Expressed_in_N_Percent_Cells", lgFCvec)
)
if (length(rmVec) > 0){
  lgFCvec <- lgFCvec[-rmVec]
}
}

print(paste0("Using the following DGE comparisons: ", paste0(lgFCvec, collapse = ", ")))      
      
for (k in 1:length(lgFCvec)){
    selCol <- lgFCvec[k]
    scatterX <- selCol
    scatterY <- gsub("logFC", "lg10p", selCol)
    
    selVec <- c(
      "hgnc_symbol",
      selCol
    )
    
    
    
    dfDat <- unique(dfData[,selVec])
    dfDat <- dfDat[dfDat[,selCol] != 0, ]
    dfDat <- dfDat[dfDat[,"hgnc_symbol"] != "", ]
    dfDat <- dfDat[order(dfDat[,selCol], decreasing = T) , ]
    
    ranks <- as.numeric(dfDat[,2])  #rnorm(n = 12000, mean = 0, sd=5)
    names(ranks) <- as.character(dfDat[,1])
    
    ## retrieve path list
    dfCat <- import.db.table.from.db(
        dbname = "reference_categories_db_new",
        dbtable = "LINCS_L1000_Chem_Pert_down",
        password = db.pwd,
        user = db.user, 
        host = host
    )
    dfCat[["type"]] <- "LINCS_L1000_Chem_Pert_down"
    
    dfCat2 <- import.db.table.from.db(
        dbname = "reference_categories_db_new",
        dbtable = "LINCS_L1000_Chem_Pert_up",
        password = db.pwd,
        user = db.user, 
        host = host
    )
    dfCat2[["type"]] <- "LINCS_L1000_Chem_Pert_up"
    
    dfCat <-rbind(dfCat, dfCat2)
    rm(dfCat2)
    
    #############
    ## KS Test
    l1 <- sapply(dfCat[,"hgnc_symbol"], function(x) list(x) )
    names(l1) <- as.vector(dfCat[,"cat_id"])
    l2 <- lapply(l1, function(x) unlist(strsplit(x, ";")))
    l3 <- lapply(l2, function(x) x[x != ""])
    l4 <- lapply(l3, function(x) x[x %in% dfDat$hgnc_symbol])
    l5 <- l4[lapply(l4, function(x) length(x)) > 3]
    l6 <- lapply(l5, function(x) dfDat[dfDat[,"hgnc_symbol"] %in% x,2])
    l7 <- l6
    
    catValueMedian <- unlist(lapply(l7, function(x) median(x)))
    cat_id <- names(l7)
    
    allValues <- dfDat[,2]
    KsTestRes <- lapply(l7, function(x) ks.test(x, allValues))
    
    pValueKsTest <- unlist(lapply(KsTestRes, function (x) x$p.value))
    statisticKsTest <- unlist(lapply(KsTestRes, function (x) x$statistic))
    
    dfKS <- data.frame(catValueMedian, pValueKsTest, cat_id, statisticKsTest)
    
    dfKS <- merge(
        dfKS, 
        dfCat,
        by.x = "cat_id",
        by.y = "cat_id"
    )
    
    dfKS$mgi_symbol <- NULL
    
    dfKS <- dfKS[order(dfKS$pValueKsTest, decreasing = FALSE),]
    
    
    ## Split in up- and down-regulated
    dfResUp <- dfKS[dfKS$type == "LINCS_L1000_Chem_Pert_up",]
    dfResDown <- dfKS[dfKS$type == "LINCS_L1000_Chem_Pert_down", c("catValueMedian", "pValueKsTest", "cat_name", "statisticKsTest")]
    names(dfResDown) <- paste0("Down_", names(dfResDown))
    names(dfResDown) <- gsub("Down_cat_name", "cat_name", names(dfResDown))
    
    names(dfResUp) <- gsub("^catValueMedian", "Up_catValueMedian", names(dfResUp))
    names(dfResUp) <- gsub("^pValueKsTest", "Up_pValueKsTest", names(dfResUp))
    names(dfResUp) <- gsub("^statisticKsTest", "Up_statisticKsTest", names(dfResUp))
    
    dfKS2 <- merge(
        dfResUp, 
        dfResDown, 
        by.x = "cat_name",
        by.y = "cat_name"
    )
    
    
    dfKS2[["connectivity"]] <- ifelse(
        ((dfKS2$Up_catValueMedian > 0) & (dfKS2$Down_catValueMedian < 0)) |
            ((dfKS2$Up_catValueMedian < 0) & (dfKS2$Down_catValueMedian > 0)),
        "yes",
        "no"
    )
    
    #Add url#
    dfKS2[["heatmapURL_Drug_Repressed"]] <- dfKS2$cat_id
    dfKS2$heatmapURL_Drug_Repressed <- gsub("_up_", "_down_", dfKS2$heatmapURL_Drug_Repressed)
    
    dfKS2[["heatmapURL_Drug_Induced"]] <- dfKS2$cat_id
    dfKS2$heatmapURL_Drug_Induced <- gsub("_down_", "_up_", dfKS2$heatmapURL_Drug_Repressed)
    
    
    
    dfKS2$heatmapURL_Drug_Repressed <- paste0(
        "https://biologic.crick.ac.uk/",project_id,"/category-view/", 
        dfKS2$heatmapURL_Drug_Repressed
    )
    
    dfKS2$heatmapURL_Drug_Induced <- paste0(
        "https://biologic.crick.ac.uk/",project_id,"/category-view/", 
        dfKS2$heatmapURL_Drug_Induced
    )
    
    
    
    #Add Scatter#
    dfKS2[["scatterplotURL_Drug_Repressed"]] <- dfKS2$cat_id
    dfKS2$scatterplotURL_Drug_Repressed <- gsub("_up_", "_down_", dfKS2$scatterplotURL_Drug_Repressed)
    
    dfKS2$scatterplotURL_Drug_Repressed <- paste0(
      "https://biologic.crick.ac.uk/",
      Obio@parameterList$project_id,
      "/scatterplot?x_axis=",
      scatterX, 
      "&y_axis=",
      scatterY,
      "&highlight_gene=&cat_id=",
      dfKS2$scatterplotURL_Drug_Repressed
    )
    
    dfKS2[["scatterplotURL_Drug_Induced"]] <- dfKS2$cat_id
    dfKS2$scatterplotURL_Drug_Induced <- gsub("_down_", "_up_", dfKS2$scatterplotURL_Drug_Induced)
    dfKS2$scatterplotURL_Drug_Induced <- paste0(
      "https://biologic.crick.ac.uk/",
      Obio@parameterList$project_id,
      "/scatterplot?x_axis=",
      scatterX, 
      "&y_axis=",
      scatterY,
      "&highlight_gene=&cat_id=",
      dfKS2$scatterplotURL_Drug_Induced
    )
    
    dfKS2$cat_id <- gsub("_down_", "", dfKS2$cat_id)
    dfKS2$cat_id <- gsub("_up_", "", dfKS2$cat_id)
    
    ## Reorder ##
    dfKS2 <- dfKS2[,c(
        'cat_name', 
        "scatterplotURL_Drug_Induced",
        "scatterplotURL_Drug_Repressed",
        'heatmapURL_Drug_Induced',
        'heatmapURL_Drug_Repressed',
        'cat_id','connectivity','Down_catValueMedian','Up_catValueMedian','Down_pValueKsTest','Up_pValueKsTest','Down_statisticKsTest','Up_statisticKsTest','comments_1','cat_item_size')]
    
    ## Add drug lable ##
    dfKS2[["temp"]] <- dfKS2$cat_name
    
    dfKS2[["Code"]] <- sapply(dfKS2$cat_name, function(x) unlist(strsplit(x, "_"))[1])
    dfKS2[,"temp"] <- apply(dfKS2[,c("temp", "Code")], 1, function(x) gsub(paste0(x[2], "_"), "", x[1]))
    
    dfKS2[["Conc"]] <- sapply(dfKS2$cat_name, function(x) unlist(strsplit(x, "-"))[length(unlist(strsplit(x, "-")))])
    dfKS2[,"temp"] <- apply(dfKS2[,c("temp", "Conc")], 1, function(x) gsub(paste0("-", x[2]), "", x[1]))
    
    dfKS2[["CellLine"]] <- sapply(dfKS2$temp, function(x) unlist(strsplit(x, "_"))[1])
    dfKS2[,"temp"] <- apply(dfKS2[,c("temp", "CellLine")], 1, function(x) gsub(paste0(x[2], "_"), "", x[1]))
    
    dfKS2[["time"]] <- sapply(dfKS2$temp, function(x) unlist(strsplit(x, "-"))[1])
    dfKS2[,"temp"] <- apply(dfKS2[,c("temp", "time")], 1, function(x) gsub(paste0(x[2], "-"), "", x[1]))
    
    names(dfKS2) <- gsub("temp", "Drug", names(dfKS2))
    ## Done KS Test
    ################
    
    
    dfCatTemp <- dfCat
    l1 <- sapply(dfCatTemp[,"hgnc_symbol"], function(x) list(x) )
    names(l1) <- as.vector(dfCatTemp[,"cat_id"])
    l1 <- lapply(l1, function(x) unlist(strsplit(x, ";")))
    l1 <- lapply(l1, function(x) x[x != ""])
    l1 <- lapply(l1, function(x) x[x %in% dfDat$hgnc_symbol])
    pathList <- l1[lapply(l1, function(x) length(x)) > 3]
       
    library(fgsea) 
    dfgseaTemp <- fgsea(
        pathways = pathList, 
        stats = ranks, ##exampleRanks,  
        minSize=15,
        maxSize=500,
        nperm=10000
    )
    dfgseaTemp <- dfgseaTemp[,c("pathway","padj", "ES", "NES", "nMoreExtreme","size")]
    
    dfgseaTempUp <- dfgseaTemp[grep("_up_", dfgseaTemp$pathway),] 
    names(dfgseaTempUp) <- paste0("Up_",names(dfgseaTempUp))
    
    dfgseaTempDown <- dfgseaTemp[grep("_down_", dfgseaTemp$pathway),] 
    names(dfgseaTempDown) <- paste0("Down_",names(dfgseaTempDown))
    
    dfgseaTempUp[["id"]] <- gsub("_up_","", dfgseaTempUp$Up_pathway)
    dfgseaTempDown[["id"]] <- gsub("_down_","", dfgseaTempDown$Down_pathway)
    
    dfES <- merge(
        dfgseaTempUp, 
        dfgseaTempDown,
        by.x = "id",
        by.y = "id"
    )
    
    
    ##                                                                           ##
    ###############################################################################
    
    dfResTable <- data.frame(merge(
        dfES,
        dfKS2,
        by.x = "id",
        by.y = "cat_id"
    ))
    
    
    dfResTable[["connectivity"]] <- ifelse(
        (dfResTable$Up_ES > 0 & dfResTable$Down_ES > 0) |
            (dfResTable$Up_ES < 0 & dfResTable$Down_ES < 0),
        0,
        dfResTable$Up_ES - dfResTable$Down_ES
    )
    
    ## Add interesting column ##
    dfResTable[["interesting"]] <- ifelse(
        (dfResTable$Up_padj < 0.05 &
            dfResTable$Down_padj < 0.05 &
            dfResTable$connectivity != 0 &
             dfResTable$Down_size > 50 & 
             dfResTable$Up_size > 50 
        ),
        "+",
        ""
    )
    
    
    ## Reorder columns ##
    colOrder <- names(dfResTable)
    colOrder <- c(
        'id',
        'cat_name',
        'heatmapURL_Drug_Induced',
        'heatmapURL_Drug_Repressed',
        "scatterplotURL_Drug_Induced",
        "scatterplotURL_Drug_Repressed",
        'Drug',
        'Code',
        'Conc',
        'CellLine',
        'time',
        'Up_ES',
        'Down_ES',
        'Up_padj',
        'Down_padj',
        'connectivity',
        'interesting',
        'Up_NES',
        'Down_NES',
        #'Up_nMoreExtreme',
        #'Down_nMoreExtreme',
        'Up_size',
        'Down_size',
        'Down_catValueMedian',
        'Up_catValueMedian',
        'Down_pValueKsTest',
        'Up_pValueKsTest',
        'Down_statisticKsTest',
        'Up_statisticKsTest',
        'comments_1'
    )
    
    dfResTable <- dfResTable[,colOrder]
    
    if (nrow(dfResTable) > 0){
      dfResTable[["comparison"]] <- selCol
    } else {
      dfResTable[1,1] <- "No results"
      dfResTable[["comparison"]] <- ""
    }
    
    dfResTable <- dfResTable[dfResTable$connectivity != 0,]
    dfResTable <- dfResTable[order(dfResTable$connectivity, decreasing=F),]
    dfResTable[["TopN"]] <- ""
    
    dfResTable[1:topN,"TopN"] <- "+"
    dfResTable <- dfResTable[order(dfResTable$connectivity, decreasing=T),]
    dfResTable[1:topN,"TopN"] <- "+"
    
    if (k ==1){
      dfFullRes <- dfResTable
    } else {
      dfFullRes <- rbind(
        dfResTable, 
        dfFullRes
      )
      
      string <- paste0(selCol, " completed.")
      print(string)
    }
    
}
return(dfFullRes)
})

## End do LINCS enrichment function                                          ##
###############################################################################


#Obio@dbDetailList$primDataDB <- "agl_data"
#Obio@dbDetailList$db.user <- Obio@parameterList$db.user
#Obio@dbDetailList$host <- Obio@parameterList$host

dfData <- import.db.table.from.db(
    dbname =  "csl_data", #Obio@dbDetailList$primDataDB,
    dbtable = Obio@parameterList$rnaseqdbTableName,
    password = db.pwd,
    user = Obio@parameterList$db.user, 
    host = Obio@parameterList$host
)

dfRes <- doLINCSenrichment(
      lgFCvec = NULL,
      dfData = dfData,
      topN = 5,
      password = db.pwd,
      db.user =  Obio@parameterList$db.user,
      host = Obio@parameterList$host,
      project_id = Obio@parameterList$project_id
)

#############################
## Create Excel output table 

## Create Excel file ##
    
  library(openxlsx)

  sheets <- unique(dfRes$comparison)
  wb <- createWorkbook()
  
  ## Style headers ##
    hs1 <- createStyle(
        fontColour = "#ffffff",
        fgFill = "#000000",
        halign = "CENTER",
        textDecoration = "Bold"
    )
  
  for (i in 1:length(sheets)){
    tabName <- sheets[i]
    tabName <- gsub("contrast_", "", tabName)
    tabName <- gsub("logFC_", "", tabName)
    tabName <- substr(paste0(tabName, "_LINCs_results"),1,30)
    addWorksheet(wb, tabName)
    freezePane(wb, tabName ,  firstActiveRow = 2)
    dfTab <- dfRes[dfRes$comparison == sheets[i],]
    writeData(wb, i, dfTab, startRow = 1, startCol = 1, headerStyle = hs1)
  }
    
    

    ## Filter is inactivated, as it does not appear to be compatible with the current version of Excel
    #addFilter(wb, 1, row = 1, cols = 1:ncol(dfOutput))

    

    FNout <- paste0(
      Obio@parameterList$outputDir,
      "LINCS.result.table.xlsx"
    )

    saveWorkbook(
        wb,
        FNout ,
        overwrite = TRUE
    )

## Done creating Excel output table                                          ##
###############################################################################

```



## Volcano Plots for topN Categories {.tabset .tabset-fade .tabset-pills}
```{r make_enrichment_plots, echo=F, eval=TRUE, warning=FALSE, result="asis",include = FALSE}



plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

for (j in 1:nrow(dfENRgmt)){
    
    tag <- paste0("ENR_", j)
    Cname <- gsub("_", " ", dfENRgmt[j,1])
    
    geneVec <- as.vector(unique(dfENRgmt[j, 3:ncol(dfENRgmt)]))
    geneVec <- geneVec[geneVec != ""]
    geneVec <- na.omit(geneVec)
    
    ## Translate 
    # dfAnnoSel <- unique(
    #     dfAnno[dfAnno[,Obio@parameterList$primaryAlignmentGeneID] %in% geneVec, ]
    # )
    # 
    # posTestGeneSet <- unique(dfAnnoSel[,Obio@parameterList$geneIDcolumn])
    # 
    posTestGeneSet <- geneVec
    
    negTestGeneSet <- NULL
    
    ## Get background gene set ##
    #backgroundGeneVec <- row.names(OsC[["RNA"]]@counts)
    if ((length(posTestGeneSet) >= 3) |(length(negTestGeneSet) >= 3)){
        library(enrichR)
        topMaxCat <- 20
        dbs <- listEnrichrDbs()
        
        dbs <- c("GO_Biological_Process_2017")
        
        
        PosEnriched <- enrichr(posTestGeneSet, dbs)
        
        for (i in 1:length(dbs)){
            dfTemp <- PosEnriched[[dbs[i]]]
            
            if (i ==1){
                dfPosEnriched <- dfTemp
            } else {
                dfPosEnriched <- rbind(
                    dfPosEnriched,
                    dfTemp
                )
            }
            
        }
        
        dfPosEnriched[["log10FDR"]] <- -1*log10(dfPosEnriched$Adjusted.P.value)
        dfPosEnriched <- dfPosEnriched[order(-dfPosEnriched$log10FDR),]
        dfPosEnriched <- na.omit(dfPosEnriched)
        
        ## Negative Side ##
        # NegEnriched <- enrichr(negTestGeneSet, dbs)
        # 
        # for (i in 1:length(dbs)){
        #     dfTemp <- NegEnriched[[dbs[i]]]
        #     
        #     if (i ==1){
        #         dfNegEnriched <- dfTemp
        #     } else {
        #         dfNegEnriched <- rbind(
        #             dfNegEnriched,
        #             dfTemp
        #         )
        #     }
        #     
        # }
        # 
        # 
        # dfNegEnriched[["log10FDR"]] <- -1*log10(dfNegEnriched$Adjusted.P.value)
        # dfNegEnriched <- dfNegEnriched[order(-dfPosEnriched$log10FDR),]
        # dfNegEnriched <- na.omit(dfNegEnriched)
        # 
        # dfNegSel <- dfNegEnriched
        # if (nrow(dfNegSel) > topMaxCat){
        #     dfNegSel <- dfNegSel[1:topMaxCat,]
        # }
        
        dfPosSel <- dfPosEnriched
        if (nrow(dfPosSel) > topMaxCat){
            dfPosSel <- dfPosSel[1:topMaxCat,]
        }
        
        if ((nrow(dfPosEnriched) > 0)){
            
            
            #dfNegSel$log10FDR <- -1* dfNegSel$log10FDR
            
            dfSel <- dfPosSel
            
            dfSel <- na.omit(dfSel)
            dfSel <- dfSel[order(dfSel$log10FDR),]
            dfSel$log10FDR <- round(dfSel$log10FDR, 2)
            
            dfSel[["Category"]] <- ""
            dfSel[dfSel$log10FDR >= 0, "Category"] <- "Enr."
            #dfSel[dfSel$log10FDR < 0, "Category"] <- "Depl."
            
            for (k in 1:nrow(dfSel)){
                if (nchar(dfSel[k, "Term"]) > 50 & length(grep("\\(GO", as.vector(dfSel[k, "Term"]))) > 0){
                    part1 <- unlist(strsplit(as.vector(dfSel[k, "Term"]), "\\(GO"))[1]
                    part1 <- substr(part1, 1, 45)
                    part2 <- unlist(strsplit(as.vector(dfSel[k, "Term"]), "\\(GO"))[2]
                    part2 <- paste0("\\(GO", part2)
                    
                    if (nchar(part1) > 40 ){
                        dfSel[k, "Term"] <- paste0(part1, " \\n", part2)
                    } else { 
                        dfSel[k, "Term"] <- paste0(part1, " ", part2)
                    }
                }
            }
            
            
            #dfSel$Term <- gsub("\\(GO", "\\\n\\(GO", dfSel$Term)
            
            dfSel$Term <- factor(dfSel$Term, levels = unique(dfSel$Term))
            
            #tag <- paste0("ENR_", j)
            #Cname <- gsub("_", " ", dfENRgmt[j,1])
            
            plotList[[tag]] <- ggplot(
                data=dfSel, aes(x= Term, y=log10FDR, fill=Category, order=log10FDR)
            ) + geom_bar(stat="identity", colour="black"
            ) + coord_flip() +scale_fill_manual(values=c("grey"))  +  theme(
                axis.text.y   = element_text(size=8),
                axis.text.x   = element_text(size=8),
                axis.title.y  = element_text(size=8),
                axis.title.x  = element_text(size=8),
                axis.line = element_line(colour = "black"),
                panel.border = element_rect(colour = "black", fill=NA, size=1),
                plot.title = element_text(hjust = 0.5, size = 12)
            )  + labs(title = paste0("Cluster ", Cname," enriched genes") ,y = "-log10(FDR)", x = ""
            ) + geom_hline(yintercept = c(-log10(0.05), log10(0.05)), color = "red", size=0.5, lty=2
            ) + geom_hline(yintercept = 0, color = "black", size=0.5
            )
            cat("  \n")
            
            
            
            ## Save to file ##
            FNbase <- paste0("Cluster_", Cname,".enriched.genes", VersionPdfExt)
            FN <- paste0(Obio@parameterList$reportFigDir, FNbase)
            FNrel <- paste0("report_figures/", FNbase)
            
           
            pdf(FN)
            print(plotList[[paste0("ENR_", j)]])
            dev.off()
            
            ## Create R markdown chunk ##
            # link <- paste0(
            #     "https://",urlString,"/",
            #     Obio@parameterList$project_id,
            #     "/category-view?category_type=GO-BP"
            # )
            
            
            figLegend <- paste0(
                "**Figure ", 
                figureCount, 
                "**: GO-BP category enrichment analysis for genes that are <font color = \\'yellow\\'>higher</font> and <font color = \\'blue\\'>lower</font> expressed in Cluster ", 
                Cname,
                " as compared to all other clusters. Download a pdf of this figure [here](", FNrel, "). "

#To view these gene sets in the context of your data, go to [CategoryView > GO-BP](",link,") and find the above categories using the search box."
            )
            figureCount <- figureCount + 1 
            
            NewChnk <- paste0(
                "### ", Cname,
                "\n```{r enrichr_cluster_",
                j,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
                figLegend,"'}\n",
                "\n",
                "\n print(plotList[['",tag,"']])",
                "\n cat(  '\n')",
                "\n\n\n```\n"   
            )
            
            chnkVec <- c(
                chnkVec,
                NewChnk
            )
        }
        
    }
    
}

```

```{r create-markdown-enrichment-chunks-dynamically, echo=T, eval=TRUE, warning=FALSE, results='asis'}
###############################################################################
## Do category enrichment on clusters                                        ##
cat(paste(knit(text = chnkVec, quiet = T), collapse = '\n'))
## Done doing enrichment on clusters                                         ##
###############################################################################


```
